<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Python API &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="_static/vrna_32.png"/>
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
        <script defer="defer" src="_static/js/mathjax/tex-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]]}, "options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="News" href="news.html" />
    <link rel="prev" title="SWIG Wrappers" href="wrappers.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="manpages.html">Manpages</a></li>
<li class="toctree-l1"><a class="reference internal" href="using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1"><a class="reference internal" href="io.html">I/O Formats</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Python API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#global-variables">Global Variables</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pythonic-interface">Pythonic interface</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#object-orientation">Object orientation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#lists-and-tuples">Lists and Tuples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#energy-parameters">Energy Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-RNA">The <code class="docutils literal notranslate"><span class="pre">RNA</span></code> Python module</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Python API</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="python-api">
<h1>Python API<a class="headerlink" href="#python-api" title="Permalink to this headline"></a></h1>
<p>Almost all symbols of the API available in our <em>RNAlib</em> C-library is wrapped for
use in Python using <code class="docutils literal notranslate"><span class="pre">swig</span></code>. That makes our fast and efficient algorithms and
tools available for third-party Python programs and scripting languages.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Our Python API is automatically generated and translated from our C-library documentation.
If you find anything problematic or want to to help us improve the documentation, do
not hesitate to contact us or make a PR at our <a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA">official github repository</a>.</p>
</div>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline"></a></h2>
<p>The Python interface is usually part of the installation of the ViennaRNA Package,
see also <a class="reference internal" href="install.html"><span class="doc">Installation</span></a> and <a class="reference internal" href="configuration.html#scripting-language-interfaces"><span class="std std-ref">Scripting Language Interfaces</span></a>.</p>
<p>If for any reason your installation does not provide our Python interface or
in cases where you don’t want to install the full ViennaRNA Package but <em>only</em>
the Python bindings to <em>RNAlib</em>, you may also install them via Pythons <code class="docutils literal notranslate"><span class="pre">pip</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python -m pip install viennarna
</pre></div>
</div>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline"></a></h2>
<p>To use our Python bindings simply <code class="docutils literal notranslate"><span class="pre">import</span></code> the <code class="docutils literal notranslate"><span class="pre">RNA</span></code> or <code class="docutils literal notranslate"><span class="pre">ViennaRNA</span></code> package
like</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">RNA</span>
</pre></div>
</div>
<p>or</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">ViennaRNA</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">RNA</span></code> module that provides access to our <em>RNAlib</em> C-library can also be imported
directly using</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">RNA</span> <span class="kn">import</span> <span class="n">RNA</span>
</pre></div>
</div>
<p>or</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ViennaRNA</span> <span class="kn">import</span> <span class="n">RNA</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In previous release of the ViennaRNA Packge, only the <code class="docutils literal notranslate"><span class="pre">RNA</span></code> package/module has
been available.
Since version 2.6.2 we maintain the <a class="reference external" href="https://pypi.org/project/ViennaRNA/">ViennaRNA</a>
project at <a class="reference external" href="https://pypi.org">https://pypi.org</a>. The former maintainer additionally introduced the
<code class="docutils literal notranslate"><span class="pre">ViennaRNA</span></code> package which we intend to keep and extend in future releases.</p>
</div>
</section>
<section id="global-variables">
<h2>Global Variables<a class="headerlink" href="#global-variables" title="Permalink to this headline"></a></h2>
<p>For the Python interface(s) SWIG places global variables of the C-library
into an additional namespace <code class="docutils literal notranslate"><span class="pre">cvar</span></code>. For instance, changing the global temperature
variable thus becomes</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">RNA</span><span class="o">.</span><span class="n">cvar</span><span class="o">.</span><span class="n">temperature</span> <span class="o">=</span> <span class="mi">25</span>
</pre></div>
</div>
</section>
<section id="pythonic-interface">
<h2>Pythonic interface<a class="headerlink" href="#pythonic-interface" title="Permalink to this headline"></a></h2>
<p>Since our library is written in <code class="docutils literal notranslate"><span class="pre">C</span></code> the functions we provide in our API might
seem awkward for users more familiar with Pythons object oriented fashion. Therefore,
we spend some effort on creating a more <em>pythonic</em> interface here. In particular, we
tried to group together particular data structures and functions operating on them to
derive classes and objects with corresponding methods attached.</p>
<p>If you browse through our <a class="reference internal" href="api.html"><span class="doc">reference manual</span></a>, many C-functions have additional
<em>SWIG Wrapper Notes</em> in their description. These descriptions should give an idea how
the function is available in the Python interface. Usually, our <code class="docutils literal notranslate"><span class="pre">C</span></code> functions,
data structures, typedefs, and enumerations use the <code class="docutils literal notranslate"><span class="pre">vrna_</span></code> prefixes and <code class="docutils literal notranslate"><span class="pre">_s</span></code>,
<code class="docutils literal notranslate"><span class="pre">_t</span></code>, <code class="docutils literal notranslate"><span class="pre">_e</span></code> suffixes. Those decorators are useful in <code class="docutils literal notranslate"><span class="pre">C</span></code> but of less use in the
context of Python packages or modules. Therefore, these prefixes and suffixes are
<em>dropped</em> from the Python interface.</p>
<section id="object-orientation">
<h3>Object orientation<a class="headerlink" href="#object-orientation" title="Permalink to this headline"></a></h3>
<p>Consider the C-function <a class="reference internal" href="utils/datastructures/fold_compound.html#c.vrna_fold_compound" title="vrna_fold_compound"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_fold_compound()</span></code></a>. This creates a
<a class="reference internal" href="utils/datastructures/fold_compound.html#c.vrna_fold_compound_t" title="vrna_fold_compound_t"><code class="xref c c-type docutils literal notranslate"><span class="pre">vrna_fold_compound_t</span></code></a> data structure that is then passed around to various
functions, e.g. to <a class="reference internal" href="files.html#c.vrna_mfe" title="vrna_mfe"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_mfe()</span></code></a> to compute the <a class="reference internal" href="mfe/global.html"><span class="doc">MFE structure</span></a>.
A corresponding C-code may look like this:</p>
<div class="highlight-c notranslate"><div class="highlight"><pre><span></span><span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;stdio.h&gt;</span><span class="cp"></span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;stdlib.h&gt;</span><span class="cp"></span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;string.h&gt;</span><span class="cp"></span>

<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;ViennaRNA/utils/basic.h&gt;</span><span class="cp"></span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;ViennaRNA/fold_compound.h&gt;</span><span class="cp"></span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;ViennaRNA/mfe.h&gt;</span><span class="cp"></span>

<span class="kt">int</span><span class="w"></span>
<span class="nf">main</span><span class="p">(</span><span class="kt">int</span><span class="w">  </span><span class="n">argc</span><span class="p">,</span><span class="w"></span>
<span class="w">     </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">argv</span><span class="p">[])</span><span class="w"></span>
<span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="kt">char</span><span class="w"> </span><span class="o">*</span><span class="n">seq</span><span class="p">,</span><span class="w"> </span><span class="o">*</span><span class="n">ss</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="kt">float</span><span class="w"> </span><span class="n">mfe</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="n">vrna_fold_compound_t</span><span class="w"> </span><span class="o">*</span><span class="n">fc</span><span class="p">;</span><span class="w"></span>

<span class="w">  </span><span class="n">seq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;AGACGACAAGGUUGAAUCGCACCCACAGUCUAUGAGUCGGUG&quot;</span><span class="p">;</span><span class="w"></span>
<span class="w">  </span><span class="n">ss</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="n">vrna_alloc</span><span class="p">(</span><span class="k">sizeof</span><span class="p">(</span><span class="kt">char</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="n">strlen</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="p">));</span><span class="w"></span>
<span class="w">  </span><span class="n">fc</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="n">vrna_fold_compound</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span><span class="w"> </span><span class="nb">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">VRNA_OPTION_DEFAULT</span><span class="p">);</span><span class="w"></span>
<span class="w">  </span><span class="n">mfe</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">vrna_mfe</span><span class="p">(</span><span class="n">fc</span><span class="p">,</span><span class="w"> </span><span class="n">ss</span><span class="p">);</span><span class="w"></span>

<span class="w">  </span><span class="n">printf</span><span class="p">(</span><span class="s">&quot;%s</span><span class="se">\n</span><span class="s">%s (%6.2f)</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">seq</span><span class="p">,</span><span class="w"> </span><span class="n">ss</span><span class="p">,</span><span class="w"> </span><span class="n">mfe</span><span class="p">);</span><span class="w"></span>

<span class="w">  </span><span class="k">return</span><span class="w"> </span><span class="n">EXIT_SUCCESS</span><span class="p">;</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
</pre></div>
</div>
<p>In our Python interface, the <a class="reference internal" href="utils/datastructures/fold_compound.html#c.vrna_fold_compound_t" title="vrna_fold_compound_t"><code class="xref c c-type docutils literal notranslate"><span class="pre">vrna_fold_compound_t</span></code></a> data structure becomes the
<a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a> class, the <a class="reference internal" href="utils/datastructures/fold_compound.html#c.vrna_fold_compound" title="vrna_fold_compound"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_fold_compound()</span></code></a> becomes one of
its constructors and the <a class="reference internal" href="files.html#c.vrna_mfe" title="vrna_mfe"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_mfe()</span></code></a> function becomes the method
<a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe()</span></code></a>. So, the Python code would probably translate
to something like</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">RNA</span>

<span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;AGACGACAAGGUUGAAUCGCACCCACAGUCUAUGAGUCGGUG&quot;</span>
<span class="n">fc</span>  <span class="o">=</span> <span class="n">RNA</span><span class="o">.</span><span class="n">fold_compound</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="p">(</span><span class="n">ss</span><span class="p">,</span> <span class="n">mfe</span><span class="p">)</span> <span class="o">=</span> <span class="n">fc</span><span class="o">.</span><span class="n">mfe</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">ss</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">mfe</span><span class="si">:</span><span class="s2">6.2f</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The C-function <a class="reference internal" href="files.html#c.vrna_mfe" title="vrna_mfe"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_mfe()</span></code></a> actually returns two values, the MFE in units
of <span class="math notranslate nohighlight">\(\text{kcal} \cdot \text{mol}^{-1}\)</span> and the corresponding MFE structure.
The latter is written to the <code class="docutils literal notranslate"><span class="pre">ss</span></code> pointer. This is necessary since <code class="docutils literal notranslate"><span class="pre">C</span></code> functions
can at most return one single value. In Python, function and methods may return
arbitrarily many values instead, and in addition, passing parameters to a function
or method such that it changes its content is generally discouraged. Therefore,
our functions that return values through function parameters usually return them
<em>regularly</em> in the Python interface.</p>
</div>
</section>
<section id="lists-and-tuples">
<h3>Lists and Tuples<a class="headerlink" href="#lists-and-tuples" title="Permalink to this headline"></a></h3>
<p>C-functions in our API that return or receive list-like data usually utilize <em>pointers</em>.
Since there are no such things in Python, they would be wrapped as particular kind of
objects that would then be tedious to work with. For the Python interface, we therefore
tried to wrap the majority of these instances to <em>native</em> Python types, such as <code class="docutils literal notranslate"><span class="pre">list</span></code>
or <code class="docutils literal notranslate"><span class="pre">tuple</span></code>. Therefore, one can usually pass a <code class="docutils literal notranslate"><span class="pre">list</span></code> to a function that uses pointers
to array in <code class="docutils literal notranslate"><span class="pre">C</span></code>, and expect to receive a <code class="docutils literal notranslate"><span class="pre">list</span></code> or <code class="docutils literal notranslate"><span class="pre">tuple</span></code> from functions that return
pointers to arrays.</p>
</section>
</section>
<section id="energy-parameters">
<h2>Energy Parameters<a class="headerlink" href="#energy-parameters" title="Permalink to this headline"></a></h2>
<p>Energy parameters are compiled into our library, so there is usually no necessity
to load them from a file. All parameter files shipped with the ViennaRNA Package
can be loaded by simply calling any of the dedicated functions:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004()</span></code></a> (default RNA parameters)</p></li>
<li><p><a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004()</span></code></a> (default DNA parameters)</p></li>
<li><p><a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999()</span></code></a> (old DNA parameters)</p></li>
<li><p><a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999()</span></code></a> (old RNA parameters)</p></li>
<li><p><a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007()</span></code></a> (trained RNA parameters)</p></li>
<li><p><a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018()</span></code></a> (trained RNA parameters)</p></li>
<li><p><a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins()</span></code></a> (special hairpin loop parameters)</p></li>
</ul>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h2>
<p>A few more Python code examples can be found <a class="reference internal" href="examples/python.html"><span class="doc">here</span></a>.</p>
</section>
<section id="module-RNA">
<span id="the-rna-python-module"></span><h2>The <code class="docutils literal notranslate"><span class="pre">RNA</span></code> Python module<a class="headerlink" href="#module-RNA" title="Permalink to this headline"></a></h2>
<p>A library for the prediction and comparison of RNA secondary structures.</p>
<p>Amongst other things, our implementations allow you to:</p>
<ul class="simple">
<li><p>predict minimum free energy secondary structures</p></li>
<li><p>calculate the partition function for the ensemble of structures</p></li>
<li><p>compute various equilibrium probabilities</p></li>
<li><p>calculate suboptimal structures in a given energy range</p></li>
<li><p>compute local structures in long sequences</p></li>
<li><p>predict consensus secondary structures from a multiple sequence alignment</p></li>
<li><p>predict melting curves</p></li>
<li><p>search for sequences folding into a given structure</p></li>
<li><p>compare two secondary structures</p></li>
<li><p>predict interactions between multiple RNA molecules</p></li>
</ul>
<dl class="py class">
<dt class="sig sig-object py" id="RNA.COORDINATE">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">COORDINATE</span></span><a class="headerlink" href="#RNA.COORDINATE" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>this is a workarround for the SWIG Perl Wrapper RNA plot function that returns an array of type
COORDINATE</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.COORDINATE.X">
<span class="sig-name descname"><span class="pre">X</span></span><a class="headerlink" href="#RNA.COORDINATE.X" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.COORDINATE.Y">
<span class="sig-name descname"><span class="pre">Y</span></span><a class="headerlink" href="#RNA.COORDINATE.Y" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/plotting/layouts.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id0">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">X</span></span><a class="headerlink" href="#id0" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Y</span></span><a class="headerlink" href="#id1" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.COORDINATE.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.COORDINATE.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.COORDINATE.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.COORDINATE.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.ConstCharVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ConstCharVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ConstCharVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ConstCharVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.ConstCharVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.ConstCharVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.CoordinateVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">CoordinateVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.CoordinateVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.CoordinateVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.CoordinateVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.CoordinateVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">DoubleDoubleVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleDoubleVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.DoubleDoubleVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.DoubleDoubleVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.DoublePair">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">DoublePair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoublePair" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.DoublePair.first">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">first</span></span><a class="headerlink" href="#RNA.DoublePair.first" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.DoublePair.second">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">second</span></span><a class="headerlink" href="#RNA.DoublePair.second" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.DoublePair.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.DoublePair.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.DoubleVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">DoubleVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DoubleVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DoubleVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.DoubleVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.DoubleVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.DuplexVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">DuplexVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.DuplexVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.DuplexVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.DuplexVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.DuplexVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_ExtLoop">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_ExtLoop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_ExtLoop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_Hairpin">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_Hairpin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_Hairpin" title="Permalink to this definition"></a></dt>
<dd><p>Compute the Energy of a hairpin-loop.</p>
<p>To evaluate the free energy of a hairpin-loop, several parameters have to be known. A general
hairpin-loop has this structure:</p>
<blockquote>
<div><blockquote>
<div><blockquote>
<div><p>a3 a4</p>
</div></blockquote>
<p>a2     a5
a1     a6</p>
<blockquote>
<div><p>X - Y
|   |
5’  3’</p>
</div></blockquote>
</div></blockquote>
<p>where X-Y marks the closing pair [e.g. a <strong>(G,C)</strong> pair]. The length of this loop is 6 as there are</p>
</div></blockquote>
<p>six unpaired nucleotides (a1-a6) enclosed by (X,Y). The 5’ mismatching nucleotide is a1 while the 3’
mismatch is a6. The nucleotide sequence of this loop is “a1.a2.a3.a4.a5.a6”</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the loop (number of unpaired nucleotides)</p></li>
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The pair type of the base pair closing the hairpin</p></li>
<li><p><strong>si1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 5’-mismatching nucleotide</p></li>
<li><p><strong>sj1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 3’-mismatching nucleotide</p></li>
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The sequence of the loop (May be <cite>NULL</cite>, otherwise mst be at least <span class="math notranslate nohighlight">\(size + 2\)</span> long)</p></li>
<li><p><strong>P</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.param()</span> <span class="pre">*</span></code>) – The datastructure containing scaled energy parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Free energy of the Hairpin-loop in dcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Not (really) thread safe! A threadsafe implementation will replace this function in a future
release!</p>
<blockquote>
<div><p>Energy evaluation may change due to updates in global variable “tetra_loop”</p>
</div></blockquote>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">scale_parameters</span></code>, <a class="reference internal" href="#RNA.param" title="RNA.param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.param</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The parameter sequence should contain the sequence of the loop in capital letters of the nucleic
acid alphabet if the loop size is below 7. This is useful for unusually stable tri-, tetra- and
hexa-loops which are treated differently (based on experimental data) if they are tabulated.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_IntLoop">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_IntLoop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sp1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_IntLoop" title="Permalink to this definition"></a></dt>
<dd><p>This function computes the free energy <span class="math notranslate nohighlight">\(\Delta G\)</span> of an interior-loop with the following
structure:</p>
<blockquote>
<div><blockquote>
<div><blockquote>
<div><p>3’  5’
|   |
U - V</p>
</div></blockquote>
<dl class="simple">
<dt>a_n       b_1</dt><dd><p>.        .
.        .
.        .</p>
</dd>
<dt>a_1       b_m</dt><dd><p>X - Y
|   |
5’  3’</p>
</dd>
</dl>
</div></blockquote>
<p>This general structure depicts an interior-loop that is closed by the base pair (X,Y). The enclosed</p>
</div></blockquote>
<p>base pair is (V,U) which leaves the unpaired bases a_1-a_n and b_1-b_n that constitute the loop. In
this example, the length of the interior-loop is <span class="math notranslate nohighlight">\((n+m)\)</span> where n or m may be 0 resulting in a
bulge-loop or base pair stack. The mismatching nucleotides for the closing pair (X,Y) are:</p>
<blockquote>
<div><p>5’-mismatch: a_1
3’-mismatch: b_m
and for the enclosed base pair (V,U):
5’-mismatch: b_1
3’-mismatch: a_n</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the ‘left’-loop (number of unpaired nucleotides)</p></li>
<li><p><strong>n2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the ‘right’-loop (number of unpaired nucleotides)</p></li>
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The pair type of the base pair closing the interior loop</p></li>
<li><p><strong>type_2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The pair type of the enclosed base pair</p></li>
<li><p><strong>si1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 5’-mismatching nucleotide of the closing pair</p></li>
<li><p><strong>sj1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 3’-mismatching nucleotide of the closing pair</p></li>
<li><p><strong>sp1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 3’-mismatching nucleotide of the enclosed pair</p></li>
<li><p><strong>sq1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 5’-mismatching nucleotide of the enclosed pair</p></li>
<li><p><strong>P</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.param()</span> <span class="pre">*</span></code>) – The datastructure containing scaled energy parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Free energy of the Interior-loop in dcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">scale_parameters</span></code>, <a class="reference internal" href="#RNA.param" title="RNA.param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.param</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Base pairs are always denoted in 5’-&gt;3’ direction. Thus the enclosed base pair must be ‘turned
arround’ when evaluating the free energy of the interior-loop</p>
<blockquote>
<div><p>This function is threadsafe</p>
</div></blockquote>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_IntLoop_Co">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_IntLoop_Co</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutpoint</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sp1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dangles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_IntLoop_Co" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_MLstem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_MLstem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_MLstem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_Stem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_Stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extLoop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_Stem" title="Permalink to this definition"></a></dt>
<dd><p>Compute the energy contribution of a stem branching off a loop-region.</p>
<p>This function computes the energy contribution of a stem that branches off a loop region. This can
be the case in multiloops, when a stem branching off increases the degree of the loop but also
<em>immediately interior base pairs</em> of an exterior loop contribute free energy. To switch the behavior
of the function according to the evaluation of a multiloop- or exterior-loop-stem, you pass the flag
‘extLoop’. The returned energy contribution consists of a TerminalAU penalty if the pair type is
greater than 2, dangling end contributions of mismatching nucleotides adjacent to the stem if only
one of the si1, sj1 parameters is greater than 0 and mismatch energies if both mismatching
nucleotides are positive values. Thus, to avoid incorporating dangling end or mismatch energies just
pass a negative number, e.g. -1 to the mismatch argument.</p>
<dl class="simple">
<dt>This is an illustration of how the energy contribution is assembled:</dt><dd><p>3’  5’
|   |
X - Y</p>
</dd>
</dl>
<p>5’-si1     sj1-3’</p>
<p>Here, (X,Y) is the base pair that closes the stem that branches off a loop region. The nucleotides
si1 and sj1 are the 5’- and 3’- mismatches, respectively. If the base pair type of (X,Y) is greater
than 2 (i.e. an A-U or G-U pair, the TerminalAU penalty will be included in the energy contribution
returned. If si1 and sj1 are both nonnegative numbers, mismatch energies will also be included. If
one of si1 or sj1 is a negative value, only 5’ or 3’ dangling end contributions are taken into
account. To prohibit any of these mismatch contributions to be incorporated, just pass a negative
number to both, si1 and sj1. In case the argument extLoop is 0, the returned energy contribution
also includes the <em>internal-loop-penalty</em> of a multiloop stem with closing pair type.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Please use one of the functions RNA.E_ext_stem() and E_MLstem() instead! Use the former for
cases where <cite>extLoop</cite> != 0 and the latter otherwise.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.E_MLstem" title="RNA.E_MLstem"><code class="xref py py-obj docutils literal notranslate"><span class="pre">E_MLstem</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">_ExtLoop</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is threadsafe</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The pair type of the first base pair un the stem</p></li>
<li><p><strong>si1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 5’-mismatching nucleotide</p></li>
<li><p><strong>sj1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 3’-mismatching nucleotide</p></li>
<li><p><strong>extLoop</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – A flag that indicates whether the contribution reflects the one of an exterior loop or not</p></li>
<li><p><strong>P</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.param()</span> <span class="pre">*</span></code>) – The data structure containing scaled energy parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Free energy of the branch off the loop in dcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_ext_stem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_ext_stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n5d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n3d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_ext_stem" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate a stem branching off the exterior loop.</p>
<p>Given a base pair <span class="math notranslate nohighlight">\((i,j)\)</span> encoded by <em>type</em>, compute the energy contribution including
dangling-end/terminal-mismatch contributions. Instead of returning the energy contribution per-se,
this function returns the corresponding Boltzmann factor. If either of the adjacent nucleotides
<span class="math notranslate nohighlight">\((i - 1)\)</span> and <span class="math notranslate nohighlight">\((j+1)\)</span> must not contribute stacking energy, the corresponding encoding
must be <span class="math notranslate nohighlight">\(-1\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The base pair encoding</p></li>
<li><p><strong>n5d</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The encoded nucleotide directly adjacent at the 5’ side of the base pair (may be -1)</p></li>
<li><p><strong>n3d</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The encoded nucleotide directly adjacent at the 3’ side of the base pair (may be -1)</p></li>
<li><p><strong>p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.param()</span> <span class="pre">*</span></code>) – The pre-computed energy parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The energy contribution of the introduced exterior-loop stem</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.E_exp_stem</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.E_ml_rightmost_stem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">E_ml_rightmost_stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.E_ml_rightmost_stem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.ElemProbVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ElemProbVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.ElemProbVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ElemProbVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.ElemProbVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.ElemProbVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">HeatCapacityVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HeatCapacityVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.HeatCapacityVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.HeatCapacityVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.HelixVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">HelixVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.HelixVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.HelixVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.HelixVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.HelixVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.IntIntVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">IntIntVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntIntVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntIntVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.IntIntVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.IntIntVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.IntVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">IntVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.IntVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.IntVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.IntVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.IntVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.Lfold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">Lfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nullfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.Lfold" title="Permalink to this definition"></a></dt>
<dd><p>Local MFE prediction using a sliding window approach (simplified interface)</p>
<p>This simplified interface to RNA.fold_compound.mfe_window() computes the MFE and locally optimal secondary
structure using default options. Structures are predicted using a sliding window approach, where
base pairs may not span outside the window. Memory required for dynamic programming (DP) matrices
will be allocated and free’d on-the-fly. Hence, after return of this function, the recursively
filled matrices are not available any more for any post-processing.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>Lfold()</cite> in the global namespace. The
parameter <cite>file</cite> defaults to <cite>NULL</cite> and may be omitted. See e.g. <a class="reference internal" href="#RNA.Lfold" title="RNA.Lfold"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.Lfold()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The nucleic acid sequence</p></li>
<li><p><strong>window_size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The window size for locally optimal structures</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – The output file handle where predictions are written to (if NULL, output is written to stdout)</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.mfe_window" title="RNA.fold_compound.mfe_window"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window</span></code></a>, <a class="reference internal" href="#RNA.Lfoldz" title="RNA.Lfoldz"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.Lfoldz</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe_window_zscore" title="RNA.fold_compound.mfe_window_zscore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window_zscore</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case you want to use the filled DP matrices for any subsequent post-processing step, or you
require other conditions than specified by the default model details, use
RNA.fold_compound.mfe_window(), and the
data structure RNA.fold_compound() instead.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.Lfold_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">Lfold_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">*</span> <span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#RNA.Lfold_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.Lfoldz">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">Lfoldz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nullfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.Lfoldz" title="Permalink to this definition"></a></dt>
<dd><p>Local MFE prediction using a sliding window approach with z-score cut-off (simplified interface)</p>
<p>This simplified interface to RNA.fold_compound.mfe_window_zscore() computes the MFE and locally optimal secondary
structure using default options. Structures are predicted using a sliding window approach, where
base pairs may not span outside the window. Memory required for dynamic programming (DP) matrices
will be allocated and free’d on-the-fly. Hence, after return of this function, the recursively
filled matrices are not available any more for any post-processing. This function is the z-score
version of RNA.Lfold(), i.e. only predictions above a certain z-score cut-off value are printed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The nucleic acid sequence</p></li>
<li><p><strong>window_size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The window size for locally optimal structures</p></li>
<li><p><strong>min_z</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The minimal z-score for a predicted structure to appear in the output</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – The output file handle where predictions are written to (if NULL, output is written to stdout)</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.mfe_window_zscore" title="RNA.fold_compound.mfe_window_zscore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window_zscore</span></code></a>, <a class="reference internal" href="#RNA.Lfold" title="RNA.Lfold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.Lfold</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe_window" title="RNA.fold_compound.mfe_window"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case you want to use the filled DP matrices for any subsequent post-processing step, or you
require other conditions than specified by the default model details, use
RNA.fold_compound.mfe_window(), and the
data structure RNA.fold_compound() instead.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.Lfoldz_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">Lfoldz_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">*</span> <span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">min_z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#RNA.Lfoldz_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.MEA_from_plist">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">MEA_from_plist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MEA_from_plist" title="Permalink to this definition"></a></dt>
<dd><p>Compute a MEA (maximum expected accuracy) structure from a list of probabilities.</p>
<p>The algorithm maximizes the expected accuracy</p>
<div class="math notranslate nohighlight">
\[A(S) = \sum_{(i,j) \in S} 2 \gamma p_{ij} + \sum_{i \notin S} p^u_{i}\]</div>
<p>Higher values of <span class="math notranslate nohighlight">\(\gamma\)</span> result in more base pairs of lower probability and thus higher
sensitivity. Low values of <span class="math notranslate nohighlight">\(\gamma\)</span> result in structures containing only highly likely pairs
(high specificity). The code of the MEA function also demonstrates the use of sparse dynamic
programming scheme to reduce the time and memory complexity of folding.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>MEA_from_plist`(gamma = 1., md = NULL). Note,
that it returns the MEA structure and MEA value as a tuple (MEA_structure, MEA). See, e.g.
:py:func:`RNA.MEA_from_plist()</cite> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>plist</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code>) – A list of base pair probabilities the MEA structure is computed from</p></li>
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence that corresponds to the list of probability values</p></li>
<li><p><strong>gamma</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The weighting factor for base pairs vs. unpaired nucleotides</p></li>
<li><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – A model details data structure (maybe NULL)</p></li>
<li><p><strong>mea</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – A pointer to a variable where the MEA value will be written to</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An MEA structure (or NULL on any error)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The unpaired probabilities <span class="math notranslate nohighlight">\(p^u_{i} = 1 - \sum_{j \neq i} p_{ij}\)</span> are usually computed from
the supplied pairing probabilities <span class="math notranslate nohighlight">\(p_{ij}\)</span> as stored in <cite>plist</cite> entries of type
RNA.PLIST_TYPE_BASEPAIR. To overwrite individual <span class="math notranslate nohighlight">\(p^u_{o}\)</span> values simply add entries with
type RNA.PLIST_TYPE_UNPAIRED</p>
<blockquote>
<div><p>To include G-Quadruplex support, the corresponding field in <cite>md</cite> must be set.</p>
</div></blockquote>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.Make_bp_profile">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">Make_bp_profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.Make_bp_profile" title="Permalink to this definition"></a></dt>
<dd><div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is deprecated and will be removed soon! See Make_bp_profile_bppm() for a
replacement</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.Make_bp_profile_bppm" title="RNA.Make_bp_profile_bppm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Make_bp_profile_bppm</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is NOT threadsafe</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.Make_bp_profile_bppm">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">Make_bp_profile_bppm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bppm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.Make_bp_profile_bppm" title="Permalink to this definition"></a></dt>
<dd><p>condense pair probability matrix into a vector containing probabilities for unpaired, upstream
paired and downstream paired.</p>
<p>This resulting probability profile is used as input for profile_edit_distance</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bppm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – A pointer to the base pair probability matrix</p></li>
<li><p><strong>length</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The length of the sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The bp profile</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.Make_swString">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">Make_swString</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.Make_swString" title="Permalink to this definition"></a></dt>
<dd><p>Convert a structure into a format suitable for string_edit_distance().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">swString</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.MoveVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">MoveVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.MoveVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.MoveVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.MoveVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.MoveVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_color_dot_plot">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_color_dot_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_color_dot_plot" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_color_dot_plot_turn">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_color_dot_plot_turn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">winSize</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_color_dot_plot_turn" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_dot_plot">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_dot_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_dot_plot" title="Permalink to this definition"></a></dt>
<dd><p>Produce postscript dot-plot.</p>
<p>Wrapper to PS_dot_plot_list</p>
<p>Reads base pair probabilities produced by pf_fold() from the global array pr and the pair list
base_pair produced by fold() and produces a postscript “dot plot” that is written to ‘filename’.
The “dot plot” represents each base pairing probability by a square of corresponding area in a
upper triangle matrix. The lower part of the matrix contains the minimum free energy</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is deprecated and will be removed soon! Use PS_dot_plot_list() instead!</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>DO NOT USE THIS FUNCTION ANYMORE SINCE IT IS NOT THREADSAFE</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_dot_plot_list">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_dot_plot_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pl</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_dot_plot_list" title="Permalink to this definition"></a></dt>
<dd><p>Produce a postscript dot-plot from two pair lists.</p>
<p>This function reads two plist structures (e.g. base pair probabilities and a secondary structure) as
produced by assign_plist_from_pr() and assign_plist_from_db() and produces a postscript “dot plot”
that is written to ‘filename’.
Using base pair probabilities in the first and mfe structure in the second plist, the resulting
“dot plot” represents each base pairing probability by a square of corresponding area in a upper
triangle matrix. The lower part of the matrix contains the minimum free energy structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p></li>
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A filename for the postscript output</p></li>
<li><p><strong>pl</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code>) – The base pair probability pairlist</p></li>
<li><p><strong>mf</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code>) – The mfe secondary structure pairlist</p></li>
<li><p><strong>comment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A comment</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 if postscript was successfully written, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">assign_plist_from_pr</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">assign_plist_from_db</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_dot_plot_turn">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_dot_plot_turn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pl</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">winSize</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_dot_plot_turn" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_rna_plot">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_rna_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_rna_plot" title="Permalink to this definition"></a></dt>
<dd><p>Produce a secondary structure graph in PostScript and write it to ‘filename’.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.file_PS_rnaplot() instead!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_rna_plot_a">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_rna_plot_a</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">post</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_rna_plot_a" title="Permalink to this definition"></a></dt>
<dd><p>Produce a secondary structure graph in PostScript including additional annotation macros and write
it to ‘filename’.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.file_PS_rnaplot_a() instead!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.PS_rna_plot_a_gquad">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PS_rna_plot_a_gquad</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ssfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">post</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PS_rna_plot_a_gquad" title="Permalink to this definition"></a></dt>
<dd><p>Produce a secondary structure graph in PostScript including additional annotation macros and write
it to ‘filename’ (detect and draw g-quadruplexes)</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.file_PS_rnaplot_a() instead!</p>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.PathVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">PathVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.PathVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.PathVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.PathVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.PathVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.SOLUTION">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">SOLUTION</span></span><a class="headerlink" href="#RNA.SOLUTION" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.SOLUTION.energy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">energy</span></span><a class="headerlink" href="#RNA.SOLUTION.energy" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTION.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTION.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTION.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTION.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.SOLUTION.structure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">structure</span></span><a class="headerlink" href="#RNA.SOLUTION.structure" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.SOLUTION.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.SOLUTION.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">SOLUTIONVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SOLUTIONVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.SOLUTIONVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.SOLUTIONVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.StringVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">StringVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.StringVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.StringVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.StringVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.StringVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.SuboptVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">SuboptVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SuboptVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SuboptVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.SuboptVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.SuboptVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.SwigPyIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">SwigPyIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.advance">
<span class="sig-name descname"><span class="pre">advance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.advance" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.copy" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.decr">
<span class="sig-name descname"><span class="pre">decr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.decr" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.distance">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.distance" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.equal">
<span class="sig-name descname"><span class="pre">equal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.equal" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.incr">
<span class="sig-name descname"><span class="pre">incr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.incr" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.next">
<span class="sig-name descname"><span class="pre">next</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.next" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.previous">
<span class="sig-name descname"><span class="pre">previous</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.previous" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.SwigPyIterator.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.SwigPyIterator.value">
<span class="sig-name descname"><span class="pre">value</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.SwigPyIterator.value" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.UIntVector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">UIntVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.append" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.assign" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.back">
<span class="sig-name descname"><span class="pre">back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.begin">
<span class="sig-name descname"><span class="pre">begin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.begin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.capacity">
<span class="sig-name descname"><span class="pre">capacity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.clear" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.empty">
<span class="sig-name descname"><span class="pre">empty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.empty" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.end">
<span class="sig-name descname"><span class="pre">end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.erase">
<span class="sig-name descname"><span class="pre">erase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.erase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.front">
<span class="sig-name descname"><span class="pre">front</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.front" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.get_allocator">
<span class="sig-name descname"><span class="pre">get_allocator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.get_allocator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.insert" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.iterator">
<span class="sig-name descname"><span class="pre">iterator</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.iterator" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.pop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.pop_back">
<span class="sig-name descname"><span class="pre">pop_back</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.pop_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.push_back">
<span class="sig-name descname"><span class="pre">push_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.push_back" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.rbegin">
<span class="sig-name descname"><span class="pre">rbegin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.rbegin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.rend">
<span class="sig-name descname"><span class="pre">rend</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.rend" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.reserve">
<span class="sig-name descname"><span class="pre">reserve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.reserve" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.resize">
<span class="sig-name descname"><span class="pre">resize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.resize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.UIntVector.swap">
<span class="sig-name descname"><span class="pre">swap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.UIntVector.swap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.UIntVector.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.UIntVector.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.abstract_shapes">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">abstract_shapes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">level=5</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span><a class="headerlink" href="#RNA.abstract_shapes" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">abstract_shapes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">level=5</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">abstract_shapes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">level=5</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span></dt>
<dd><p>Convert a secondary structure in dot-bracket notation to its abstract shapes representation.</p>
<p>This function converts a secondary structure into its abstract shapes representation as presented by
<span id="id2">Giegerich <em>et al.</em> [<a class="reference internal" href="bibliography.html#id42" title="Robert Giegerich, Björn Voß, and Marc Rehmsmeier. Abstract shapes of RNA. Nucleic Acids Research, 32(16):4843–4851, 2004. doi:10.1093/nar/gkh779.">2004</a>]</span> .</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as an overloaded function <cite>abstract_shapes()</cite> where the optional
second parameter <cite>level</cite> defaults to 5. See, e.g.  <a class="reference internal" href="#RNA.abstract_shapes" title="RNA.abstract_shapes"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.abstract_shapes()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A secondary structure in dot-bracket notation</p></li>
<li><p><strong>level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The abstraction level (integer in the range of 0 to 5)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The secondary structure in abstract shapes notation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.abstract_shapes_pt</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.add_root">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">add_root</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.add_root" title="Permalink to this definition"></a></dt>
<dd><p>Adds a root to an un-rooted tree in any except bracket notation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aliLfold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aliLfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nullfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.aliLfold" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>aliLfold()</cite> in the global namespace. The
parameter <cite>fp</cite> defaults to <cite>NULL</cite> and may be omitted. See e.g. <a class="reference internal" href="#RNA.aliLfold" title="RNA.aliLfold"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aliLfold()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aliLfold_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aliLfold_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#RNA.aliLfold_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aliduplex_subopt">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aliduplex_subopt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">delta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">w</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.DuplexVector" title="RNA.DuplexVector"><span class="pre">DuplexVector</span></a></span></span><a class="headerlink" href="#RNA.aliduplex_subopt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aliduplexfold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aliduplexfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.duplex_list_t" title="RNA.duplex_list_t"><span class="pre">duplex_list_t</span></a></span></span><a class="headerlink" href="#RNA.aliduplexfold" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.alifold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">alifold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.alifold" title="Permalink to this definition"></a></dt>
<dd><p>Compute Minimum Free Energy (MFE), and a corresponding consensus secondary structure for an RNA
sequence alignment using a comparative method.</p>
<p>This simplified interface to RNA.fold_compound.mfe() computes the MFE and, if required, a consensus secondary
structure for an RNA sequence alignment using default options. Memory required for dynamic
programming (DP) matrices will be allocated and free’d on-the-fly. Hence, after return of this
function, the recursively filled matrices are not available any more for any post-processing, e.g.
suboptimal backtracking, etc.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>alifold()</cite> in the global namespace. The parameter
<cite>structure</cite> is returned along with the MFE und must not be provided. See e.g.
<a class="reference internal" href="#RNA.alifold" title="RNA.alifold"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.alifold()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequences</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – RNA sequence alignment</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the secondary structure in dot-bracket notation will be
written to</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the minimum free energy (MFE) in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.circalifold" title="RNA.circalifold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.circalifold</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case you want to use the filled DP matrices for any subsequent post-processing step, or you
require other conditions than specified by the default model details, use RNA.fold_compound.mfe(),
and the data
structure RNA.fold_compound() instead.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aln_consensus_mis">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aln_consensus_mis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span> <span class="pre">md_p=None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span><a class="headerlink" href="#RNA.aln_consensus_mis" title="Permalink to this definition"></a></dt>
<dd><p>Compute the Most Informative Sequence (MIS) for a given multiple sequence alignment.</p>
<p>The most informative sequence (MIS)  <span id="id3">[<a class="reference internal" href="bibliography.html#id39" title="Eva Freyhult, Vincent Moulton, and Paul Gardner. Predicting RNA structure using mutual information. Applied bioinformatics, 4(1):53–59, 2005. doi:10.2165/00822942-200504010-00006.">Freyhult <em>et al.</em>, 2005</a>]</span>  displays for each alignment column the
nucleotides with frequency greater than the background frequency, projected into IUPAC notation.
Columns where gaps are over-represented are in lower case.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>aln_consensus_mis()</cite> where the last parameter
may be omitted, indicating <cite>md</cite> = <cite>NULL</cite>. See e.g.  <a class="reference internal" href="#RNA.aln_consensus_mis" title="RNA.aln_consensus_mis"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_consensus_mis()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The input sequence alignment (last entry must be <em>NULL</em> terminated)</p></li>
<li><p><strong>md_p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – Model details that specify known nucleotides (Maybe <em>NULL</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The most informative sequence for the alignment</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aln_consensus_sequence">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aln_consensus_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span> <span class="pre">md_p=None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span><a class="headerlink" href="#RNA.aln_consensus_sequence" title="Permalink to this definition"></a></dt>
<dd><p>Compute the consensus sequence for a given multiple sequence alignment.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>aln_consensus_sequence()</cite> where the last
parameter may be omitted, indicating <cite>md</cite> = <cite>NULL</cite>. See e.g.
<a class="reference internal" href="#RNA.aln_consensus_sequence" title="RNA.aln_consensus_sequence"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_consensus_sequence()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The input sequence alignment (last entry must be <em>NULL</em> terminated)</p></li>
<li><p><strong>md_p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – Model details that specify known nucleotides (Maybe <em>NULL</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The consensus sequence of the alignment, i.e. the most frequent nucleotide for each alignment
column</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aln_conservation_col">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aln_conservation_col</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span> <span class="pre">md=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.DoubleVector" title="RNA.DoubleVector"><span class="pre">DoubleVector</span></a></span></span><a class="headerlink" href="#RNA.aln_conservation_col" title="Permalink to this definition"></a></dt>
<dd><p>Compute nucleotide conservation in an alignment.</p>
<p>This function computes the conservation of nucleotides in alignment columns. The simples measure is
Shannon Entropy and can be selected by passing the RNA.MEASURE_SHANNON_ENTROPY flag in the
<cite>options</cite> parameter.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>aln_conservation_col()</cite> where the last two
parameters may be omitted, indicating <cite>md</cite> = <cite>NULL</cite>, and <cite>options</cite> =
RNA.MEASURE_SHANNON_ENTROPY, respectively. See e.g. <a class="reference internal" href="#RNA.aln_conservation_col" title="RNA.aln_conservation_col"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_conservation_col()</span></code></a> in
the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The input sequence alignment (last entry must be <em>NULL</em> terminated)</p></li>
<li><p><strong>md</strong> – Model details that specify known nucleotides (Maybe <em>NULL</em>)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A flag indicating which measure of conservation should be applied</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 1-based vector of column conservations</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MEASURE_SHANNON_ENTROPY</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently, only RNA.MEASURE_SHANNON_ENTROPY is supported as conservation measure.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aln_conservation_struct">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aln_conservation_struct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span> <span class="pre">md=None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.DoubleVector" title="RNA.DoubleVector"><span class="pre">DoubleVector</span></a></span></span><a class="headerlink" href="#RNA.aln_conservation_struct" title="Permalink to this definition"></a></dt>
<dd><p>Compute base pair conservation of a consensus structure.</p>
<p>This function computes the base pair conservation (fraction of canonical base pairs) of a consensus
structure given a multiple sequence alignment. The base pair types that are considered canonical may
be specified using the RNA.md().pair array. Passing <em>NULL</em> as parameter <cite>md</cite> results in default
pairing rules, i.e. canonical Watson-Crick and GU Wobble pairs.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>aln_conservation_struct()</cite> where the last
parameter <cite>md</cite> may be omitted, indicating <cite>md</cite> = <cite>NULL</cite>. See, e.g.
<a class="reference internal" href="#RNA.aln_conservation_struct" title="RNA.aln_conservation_struct"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_conservation_struct()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The input sequence alignment (last entry must be <em>NULL</em> terminated)</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The consensus structure in dot-bracket notation</p></li>
<li><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – Model details that specify compatible base pairs (Maybe <em>NULL</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 1-based vector of base pair conservations</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aln_mpi">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aln_mpi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.aln_mpi" title="Permalink to this definition"></a></dt>
<dd><p>Get the mean pairwise identity in steps from ?to?(ident)</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>aln_mpi()</cite>. See e.g.  <a class="reference internal" href="#RNA.aln_mpi" title="RNA.aln_mpi"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_mpi()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alignment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – Aligned sequences</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The mean pairwise identity</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.aln_pscore">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">aln_pscore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span> <span class="pre">md=None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.IntIntVector" title="RNA.IntIntVector"><span class="pre">IntIntVector</span></a></span></span><a class="headerlink" href="#RNA.aln_pscore" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>aln_pscore()</cite> where the last parameter may be
omitted, indicating <cite>md</cite> = <cite>NULL</cite>. See e.g.  <a class="reference internal" href="#RNA.aln_pscore" title="RNA.aln_pscore"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_pscore()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.b2C">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">b2C</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.b2C" title="Permalink to this definition"></a></dt>
<dd><p>Converts the full structure from bracket notation to the a coarse grained notation using the ‘H’ ‘B’
‘I’ ‘M’ and ‘R’ identifiers.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>See RNA.db_to_tree_string() and RNA.STRUCTURE_TREE_SHAPIRO_SHORT for a replacement</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.b2HIT">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">b2HIT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.b2HIT" title="Permalink to this definition"></a></dt>
<dd><p>Converts the full structure from bracket notation to the HIT notation including root.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>See RNA.db_to_tree_string() and RNA.STRUCTURE_TREE_HIT for a replacement</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.b2Shapiro">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">b2Shapiro</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.b2Shapiro" title="Permalink to this definition"></a></dt>
<dd><p>Converts the full structure from bracket notation to the <em>weighted</em> coarse grained notation using
the ‘H’ ‘B’ ‘I’ ‘M’ ‘S’ ‘E’ and ‘R’ identifiers.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>See RNA.db_to_tree_string() and RNA.STRUCTURE_TREE_SHAPIRO_WEIGHT for a replacement</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.basepair">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">basepair</span></span><a class="headerlink" href="#RNA.basepair" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.basepair.i">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">i</span></span><a class="headerlink" href="#RNA.basepair.i" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.basepair.j">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">j</span></span><a class="headerlink" href="#RNA.basepair.j" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.basepair.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.basepair.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.boustrophedon">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">boustrophedon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.boustrophedon" title="Permalink to this definition"></a></dt>
<dd><p>Generate a sequence of Boustrophedon distributed numbers.</p>
<p>This function generates a sequence of positive natural numbers within the interval <span class="math notranslate nohighlight">\([start,
end]\)</span> in a Boustrophedon fashion. That is, the numbers <span class="math notranslate nohighlight">\(start, \ldots, end\)</span> in the resulting
list are alternating between left and right ends of the interval while progressing to the inside,
i.e. the list consists of a sequence of natural numbers of the form:</p>
<div class="math notranslate nohighlight">
\[start, end, start + 1, end - 1, start + 2, end - 2, \ldots\]</div>
<p>The resulting list is 1-based and contains the length of the sequence of numbers at it’s 0-th
position.</p>
<p>Upon failure, the function returns <strong>NULL</strong></p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded global function <cite>boustrophedon()</cite>. See, e.g.
<a class="reference internal" href="#RNA.boustrophedon" title="RNA.boustrophedon"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.boustrophedon()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">size()</span></code>) – The first number of the list (left side of the interval)</p></li>
<li><p><strong>end</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">size()</span></code>) – The last number of the list (right side of the interval)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of alternating numbers from the interval <span class="math notranslate nohighlight">\([start, end]\)</span> (or <strong>NULL</strong> on error)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.boustrophedon_pos</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.bp_distance">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">bp_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">str1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">str2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.bp_distance" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">bp_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">bp_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span></dt>
<dd><p>Compute the “base pair” distance between two secondary structures s1 and s2.</p>
<p>This is a wrapper around <strong>RNA.bp_distance_pt()</strong>. The sequences should have the same length. dist
= number of base pairs in one structure but not in the other same as edit distance with open-pair
close-pair as move-set</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as an overloaded method <strong>bp_distance()</strong>. Note that the SWIG wrapper
takes two structure in dot-bracket notation and converts them into pair tables using
RNA.ptable_from_string(). The resulting pair tables are then internally passed to
RNA.bp_distance_pt(). To control which kind of matching brackets will be used during
conversion, the optional argument <cite>options</cite> can be used. See also the description of
RNA.ptable_from_string() for available options. (default: <strong>RNA.BRACKETS_RND</strong>). See, e.g.
<a class="reference internal" href="#RNA.bp_distance" title="RNA.bp_distance"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.bp_distance()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>str1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – First structure in dot-bracket notation</p></li>
<li><p><strong>str2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Second structure in dot-bracket notation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The base pair distance between str1 and str2</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.bp_distance_pt</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.cdata">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">cdata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nelements</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.cdata" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.centroid">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.centroid" title="Permalink to this definition"></a></dt>
<dd><div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is deprecated and should not be used anymore as it is not threadsafe!</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.get_centroid_struct_pl" title="RNA.get_centroid_struct_pl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_centroid_struct_pl</span></code></a>, <a class="reference internal" href="#RNA.get_centroid_struct_pr" title="RNA.get_centroid_struct_pr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_centroid_struct_pr</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.circalifold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">circalifold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.circalifold" title="Permalink to this definition"></a></dt>
<dd><p>Compute MFE and according structure of an alignment of sequences assuming the sequences are circular
instead of linear.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Usage of this function is discouraged! Use RNA.alicircfold(), and RNA.fold_compound.mfe()</p>
</div>
<p>instead!</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strings</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – A pointer to a NULL terminated array of character arrays</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to a character array that may contain a constraining consensus structure (will be
overwritten by a consensus structure that exhibits the MFE)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy score in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.alicircfold</span></code>, <a class="reference internal" href="#RNA.alifold" title="RNA.alifold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.alifold</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.circfold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">circfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.circfold" title="Permalink to this definition"></a></dt>
<dd><p>Compute Minimum Free Energy (MFE), and a corresponding secondary structure for a circular RNA
sequence.</p>
<p>This simplified interface to RNA.fold_compound.mfe() computes the MFE and, if required, a secondary structure for
a circular RNA sequence using default options. Memory required for dynamic programming (DP) matrices
will be allocated and free’d on-the-fly. Hence, after return of this function, the recursively
filled matrices are not available any more for any post-processing, e.g. suboptimal backtracking,
etc.</p>
<p>Folding of circular RNA sequences is handled as a post-processing step of the forward recursions.
See  <span id="id4">Hofacker and Stadler [<a class="reference internal" href="bibliography.html#id21" title="I.L. Hofacker and P.F. Stadler. Memory efficient folding algorithms for circular RNA secondary structures. Bioinformatics, 22(10):1172–1176, 2006. doi:10.1093/bioinformatics/btl023.">2006</a>]</span>  for further details.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>circfold()</cite> in the global namespace. The parameter
<cite>structure</cite> is returned along with the MFE und must not be provided. See e.g.
<a class="reference internal" href="#RNA.circfold" title="RNA.circfold"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.circfold()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the secondary structure in dot-bracket notation will be
written to</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the minimum free energy (MFE) in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold" title="RNA.fold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case you want to use the filled DP matrices for any subsequent post-processing step, or you
require other conditions than specified by the default model details, use RNA.fold_compound.mfe(),
and the data
structure RNA.fold_compound() instead.</p>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.cmd">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">cmd</span></span><a class="headerlink" href="#RNA.cmd" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.cmd.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.cmd.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.co_pf_fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">co_pf_fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.co_pf_fold" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.cofold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">cofold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.cofold" title="Permalink to this definition"></a></dt>
<dd><p>Compute Minimum Free Energy (MFE), and a corresponding secondary structure for two dimerized RNA
sequences.</p>
<p>This simplified interface to RNA.fold_compound.mfe() computes the MFE and, if required, a secondary structure for
two RNA sequences upon dimerization using default options. Memory required for dynamic programming
(DP) matrices will be allocated and free’d on-the-fly. Hence, after return of this function, the
recursively filled matrices are not available any more for any post-processing, e.g. suboptimal
backtracking, etc.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is obsolete since RNA.mfe()/RNA.fold() can handle complexes multiple sequences
since v2.5.0. Use RNA.mfe()/RNA.fold() for connected component MFE instead and compute MFEs of
unconnected states separately.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case you want to use the filled DP matrices for any subsequent post-processing step, or you
require other conditions than specified by the default model details, use RNA.fold_compound.mfe(),
and the data
structure RNA.fold_compound() instead.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>cofold()</cite> in the global namespace. The parameter
<cite>structure</cite> is returned along with the MFE und must not be provided. See e.g.
<a class="reference internal" href="#RNA.cofold" title="RNA.cofold"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.cofold()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – two RNA sequences separated by the ‘&amp;’ character</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the secondary structure in dot-bracket notation will be
written to</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the minimum free energy (MFE) in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold" title="RNA.fold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.cut_point_insert</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.consens_mis">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">consens_mis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md_p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.consens_mis" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_flatten">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_flatten</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.db_flatten" title="Permalink to this definition"></a></dt>
<dd><p>Substitute pairs of brackets in a string with parenthesis.</p>
<p>This function can be used to replace brackets of unusual types, such as angular brackets <cite>&lt;&gt;</cite> , to
dot-bracket format. The <cite>options</cite> parameter is used tpo specify which types of brackets will be
replaced by round parenthesis <a href="#id5"><span class="problematic" id="id6">``</span></a>() .</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function flattens an input structure string in-place! The second parameter is optional and
defaults to RNA.BRACKETS_DEFAULT.</p>
<p>An overloaded version of this function exists, where an additional second parameter can be
passed to specify the target brackets, i.e. the type of matching pair characters all brackets
will be flattened to. Therefore, in the scripting language interface this function is a
replacement for RNA.db_flatten_to(). See, e.g.  <a class="reference internal" href="#RNA.db_flatten" title="RNA.db_flatten"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.db_flatten()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The structure string where brackets are flattened in-place</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitmask to specify which types of brackets should be flattened out</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_flatten_to</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_RND</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_ANG</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_CLY</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_SQR</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_from_WUSS">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_from_WUSS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wuss</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.db_from_WUSS" title="Permalink to this definition"></a></dt>
<dd><p>Convert a WUSS annotation string to dot-bracket format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wuss</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The input string in WUSS notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dot-bracket notation of the input secondary structure</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function flattens all brackets, and treats pseudo-knots annotated by matching pairs of
upper/lowercase letters as unpaired nucleotides</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_from_plist">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_from_plist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ElemProbVector</span> <span class="pre">elem_probs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">length</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span><a class="headerlink" href="#RNA.db_from_plist" title="Permalink to this definition"></a></dt>
<dd><p>Convert a list of base pairs into dot-bracket notation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pairs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code>) – A RNA.ep() containing the pairs to be included in the dot-bracket string</p></li>
<li><p><strong>n</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The length of the structure (number of nucleotides)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The dot-bracket string containing the provided base pairs</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.plist" title="RNA.plist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plist</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_from_ptable">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_from_ptable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span></span></span><a class="headerlink" href="#RNA.db_from_ptable" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_from_ptable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span></dt>
<dd><p>Convert a pair table into dot-parenthesis notation.</p>
<p>This function also converts pair table formatted structures that contain pseudoknots. Non-nested
base pairs result in additional pairs of parenthesis and brackets within the resulting dot-
parenthesis string. The following pairs are awailable: (), []. {}. &lt;&gt;, as well as pairs of matching
upper-/lower-case characters from the alphabet A-Z.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – The pair table to be copied</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A char pointer to the dot-bracket string</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In cases where the level of non-nested base pairs exceeds the maximum number of 30 different base
pair indicators (4 parenthesis/brackets, 26 matching characters), a warning is printed and the
remaining base pairs are left out from the conversion.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_pack">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_pack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">struc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.db_pack" title="Permalink to this definition"></a></dt>
<dd><p>Pack secondary secondary structure, 5:1 compression using base 3 encoding.</p>
<p>Returns a binary string encoding of the secondary structure using a 5:1 compression scheme. The
string is NULL terminated and can therefore be used with standard string functions such as strcmp().
Useful for programs that need to keep many structures in memory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>struc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The binary encoded structure</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.db_unpack" title="RNA.db_unpack"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_unpack</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_pk_remove">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_pk_remove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span><a class="headerlink" href="#RNA.db_pk_remove" title="Permalink to this definition"></a></dt>
<dd><p>Remove pseudo-knots from an input structure.</p>
<p>This function removes pseudo-knots from an input structure by determining the minimum number of base
pairs that need to be removed to make the structure pseudo-knot free.</p>
<p>To accomplish that, we use a dynamic programming algorithm similar to the Nussinov maxmimum matching
approach.</p>
<p>The input structure must be in a dot-bracket string like form where crossing base pairs are denoted
by the use of additional types of matching brackets, e.g. <cite>&lt;&gt;</cite>, <cite>{}</cite>, <a href="#id7"><span class="problematic" id="id8">``</span></a>[], <cite>{}</cite>. Furthermore,
crossing pairs may be annotated by matching uppercase/lowercase letters from the alphabet <cite>A-Z</cite>. For
the latter, the uppercase letter must be the 5’ and the lowercase letter the 3’ nucleotide of the
base pair. The actual type of brackets to be recognized by this function must be specifed through
the <cite>options</cite> parameter.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as an overloaded function <cite>db_pk_remove()</cite> where the optional second
parameter <cite>options</cite> defaults to RNA.BRACKETS_ANY. See, e.g.  <a class="reference internal" href="#RNA.db_pk_remove" title="RNA.db_pk_remove"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.db_pk_remove()</span></code></a> in
the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Input structure in dot-bracket format that may include pseudo-knots</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitmask to specify which types of brackets should be processed</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The input structure devoid of pseudo-knots in dot-bracket notation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.pt_pk_remove" title="RNA.pt_pk_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pt_pk_remove</span></code></a>, <a class="reference internal" href="#RNA.db_flatten" title="RNA.db_flatten"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_flatten</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_RND</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_ANG</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_CLY</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_SQR</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_ALPHA</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_ANY</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Brackets in the input structure string that are not covered by the <cite>options</cite> bitmask will be
silently ignored!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_to_element_string">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_to_element_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.db_to_element_string" title="Permalink to this definition"></a></dt>
<dd><p>Convert a secondary structure in dot-bracket notation to a nucleotide annotation of loop contexts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A string annotating each nucleotide according to it’s structural context</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_to_tree_string">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_to_tree_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">type</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">std::string</span></span></span><a class="headerlink" href="#RNA.db_to_tree_string" title="Permalink to this definition"></a></dt>
<dd><p>Convert a Dot-Bracket structure string into tree string representation.</p>
<p>This function allows one to convert a secondary structure in dot-bracket notation into one of the
various tree representations for secondary structures. The resulting tree is then represented as a
string of parenthesis and node symbols, similar to to the Newick format.</p>
<p>Currently we support conversion into the following formats, denoted by the value of parameter
<cite>type:</cite></p>
<ul class="simple">
<li><p>RNA.STRUCTURE_TREE_HIT - Homeomorphically Irreducible Tree (HIT) representation of a secondary
structure. (See also  <span id="id9">Fontana <em>et al.</em> [<a class="reference internal" href="bibliography.html#id25" title="W. Fontana, P.F. Stadler, E.G. Bornberg-Bauer, T. Griesmacher, I.L. Hofacker, M. Tacker, P. Tarazona, E.D. Weinberger, and P. Schuster. RNA folding and combinatory landscapes. Physical review E, 47(3):2083, 1993. doi:10.1103/PhysRevE.47.2083.">1993</a>]</span> )</p></li>
<li><p>RNA.STRUCTURE_TREE_SHAPIRO_SHORT - (short) Coarse Grained representation of a secondary
structure (same as  <span id="id10">Shapiro [<a class="reference internal" href="bibliography.html#id27" title="B.A. Shapiro. An algorithm for comparing multiple RNA secondary structures. Computer applications in the biosciences: CABIOS, 4(3):387–393, 1988. doi:10.1093/bioinformatics/4.3.387.">1988</a>]</span> , but with root node <cite>R</cite> and without <cite>S</cite> nodes for
the stems)</p></li>
<li><p>RNA.STRUCTURE_TREE_SHAPIRO - (full) Coarse Grained representation of a secondary structure (See
also  <span id="id11">Shapiro [<a class="reference internal" href="bibliography.html#id27" title="B.A. Shapiro. An algorithm for comparing multiple RNA secondary structures. Computer applications in the biosciences: CABIOS, 4(3):387–393, 1988. doi:10.1093/bioinformatics/4.3.387.">1988</a>]</span> )</p></li>
<li><p>RNA.STRUCTURE_TREE_SHAPIRO_EXT - (extended) Coarse Grained representation of a secondary
structure (same as  <span id="id12">Shapiro [<a class="reference internal" href="bibliography.html#id27" title="B.A. Shapiro. An algorithm for comparing multiple RNA secondary structures. Computer applications in the biosciences: CABIOS, 4(3):387–393, 1988. doi:10.1093/bioinformatics/4.3.387.">1988</a>]</span> , but external nodes denoted as <cite>E</cite> )</p></li>
<li><p>RNA.STRUCTURE_TREE_SHAPIRO_WEIGHT - (weighted) Coarse Grained representation of a secondary
structure (same as RNA.STRUCTURE_TREE_SHAPIRO_EXT but with additional weights for number of
unpaired nucleotides in loop, and number of pairs in stems)</p></li>
<li><p>RNA.STRUCTURE_TREE_EXPANDED - Expanded Tree representation of a secondary structure.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The null-terminated dot-bracket structure string</p></li>
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A switch to determine the type of tree string representation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tree representation of the input <cite>structure</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sec_structure_representations_tree</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.db_unpack">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">db_unpack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">packed</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.db_unpack" title="Permalink to this definition"></a></dt>
<dd><p>Unpack secondary structure previously packed with RNA.db_pack()</p>
<p>Translate a compressed binary string produced by RNA.db_pack() back into the familiar dot-bracket
notation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>packed</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The binary encoded packed secondary structure</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The unpacked secondary structure in dot-bracket notation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.db_pack" title="RNA.db_pack"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_pack</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.delete_doubleP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">delete_doubleP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.delete_doubleP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.delete_floatP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">delete_floatP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.delete_floatP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.delete_intP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">delete_intP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.delete_intP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.delete_shortP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">delete_shortP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.delete_shortP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.delete_ushortP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">delete_ushortP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.delete_ushortP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.deref_any">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">deref_any</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.deref_any" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.dist_mountain">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">dist_mountain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">str1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">str2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.dist_mountain" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.doubleArray">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">doubleArray</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.doubleArray" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.doubleArray.cast">
<span class="sig-name descname"><span class="pre">cast</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.doubleArray.cast" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.doubleArray.frompointer">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">frompointer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.doubleArray.frompointer" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.doubleArray.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.doubleArray.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.doubleArray_frompointer">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">doubleArray_frompointer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.doubleArray_frompointer" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.doubleP_getitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">doubleP_getitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.doubleP_getitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.doubleP_setitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">doubleP_setitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.doubleP_setitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.duplex_list_t">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">duplex_list_t</span></span><a class="headerlink" href="#RNA.duplex_list_t" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.duplex_list_t.energy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">energy</span></span><a class="headerlink" href="#RNA.duplex_list_t.energy" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.duplex_list_t.i">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">i</span></span><a class="headerlink" href="#RNA.duplex_list_t.i" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.duplex_list_t.j">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">j</span></span><a class="headerlink" href="#RNA.duplex_list_t.j" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.duplex_list_t.structure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">structure</span></span><a class="headerlink" href="#RNA.duplex_list_t.structure" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.duplex_list_t.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.duplex_list_t.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.duplex_subopt">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">duplex_subopt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">delta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">w</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.DuplexVector" title="RNA.DuplexVector"><span class="pre">DuplexVector</span></a></span></span><a class="headerlink" href="#RNA.duplex_subopt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.duplexfold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">duplexfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.duplex_list_t" title="RNA.duplex_list_t"><span class="pre">duplex_list_t</span></a></span></span><a class="headerlink" href="#RNA.duplexfold" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.encode_seq">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">encode_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.encode_seq" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_circ_struct">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_circ_struct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_circ_struct" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded circular RNA</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is deprecated and should not be used in future programs Use
energy_of_circ_structure() instead!</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is not entirely threadsafe! Depending on the state of the global variable eos_debug it
prints energy information to stdout or not…</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – secondary structure in dot-bracket notation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.energy_of_circ_structure" title="RNA.energy_of_circ_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_of_circ_structure</span></code></a>, <a class="reference internal" href="#RNA.energy_of_struct" title="RNA.energy_of_struct"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_of_struct</span></code></a>, <a class="reference internal" href="#RNA.energy_of_struct_pt" title="RNA.energy_of_struct_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_of_struct_pt</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_circ_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_circ_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity_level</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_circ_structure" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded circular RNA.</p>
<p>If verbosity level is set to a value &gt;0, energies of structure elements are printed to stdout</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>OpenMP: This function relies on several global model settings variables and thus is not to be
considered threadsafe. See energy_of_circ_struct_par() for a completely threadsafe implementation.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.eval_structure() or RNA.fold_compound.eval_structure_verbose() instead!</p>
</div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket notation</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – A flag to turn verbose output on/off</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_structure" title="RNA.fold_compound.eval_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_gquad_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_gquad_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity_level</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_gquad_structure" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_move">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_move</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_move" title="Permalink to this definition"></a></dt>
<dd><p>Calculate energy of a move (closing or opening of a base pair)</p>
<p>If the parameters m1 and m2 are negative, it is deletion (opening) of a base pair, otherwise it is
insertion (opening).</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.eval_move() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – secondary structure in dot-bracket notation</p></li>
<li><p><strong>m1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – first coordinate of base pair</p></li>
<li><p><strong>m2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – second coordinate of base pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>energy change of the move in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_move" title="RNA.fold_compound.eval_move"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_move</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_move_pt">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_move_pt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_move_pt" title="Permalink to this definition"></a></dt>
<dd><p>Calculate energy of a move (closing or opening of a base pair)</p>
<p>If the parameters m1 and m2 are negative, it is deletion (opening) of a base pair, otherwise it is
insertion (opening).</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.eval_move_pt() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – the pair table of the secondary structure</p></li>
<li><p><strong>s</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
<li><p><strong>m1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – first coordinate of base pair</p></li>
<li><p><strong>m2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – second coordinate of base pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>energy change of the move in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_move_pt" title="RNA.fold_compound.eval_move_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_move_pt</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_struct">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_struct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_struct" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is deprecated and should not be used in future programs! Use energy_of_structure()
instead!</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is not entirely threadsafe! Depending on the state of the global variable eos_debug it
prints energy information to stdout or not…</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – secondary structure in dot-bracket notation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.energy_of_structure" title="RNA.energy_of_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_of_structure</span></code></a>, <a class="reference internal" href="#RNA.energy_of_circ_struct" title="RNA.energy_of_circ_struct"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_of_circ_struct</span></code></a>, <a class="reference internal" href="#RNA.energy_of_struct_pt" title="RNA.energy_of_struct_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_of_struct_pt</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_struct_pt">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_struct_pt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_struct_pt" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is deprecated and should not be used in future programs! Use
energy_of_structure_pt() instead!</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is not entirely threadsafe! Depending on the state of the global variable eos_debug it
prints energy information to stdout or not…</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>ptable</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – the pair table of the secondary structure</p></li>
<li><p><strong>s</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the free energy of the input structure given the input sequence in 10kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_pair_table</span></code>, <a class="reference internal" href="#RNA.energy_of_structure" title="RNA.energy_of_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_of_structure</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity_level</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_structure" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA using global model detail settings.</p>
<p>If verbosity level is set to a value &gt;0, energies of structure elements are printed to stdout</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.eval_structure() or RNA.fold_compound.eval_structure_verbose() instead!</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>OpenMP: This function relies on several global model settings variables and thus is not to be
considered threadsafe. See energy_of_struct_par() for a completely threadsafe implementation.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – secondary structure in dot-bracket notation</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – a flag to turn verbose output on/off</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_structure" title="RNA.fold_compound.eval_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.energy_of_structure_pt">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">energy_of_structure_pt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity_level</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.energy_of_structure_pt" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA.</p>
<p>If verbosity level is set to a value &gt;0, energies of structure elements are printed to stdout</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.eval_structure_pt() or RNA.fold_compound.eval_structure_pt_verbose()</p>
</div>
<p>instead!</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>OpenMP: This function relies on several global model settings variables and thus is not to be
considered threadsafe. See energy_of_struct_pt_par() for a completely threadsafe implementation.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>ptable</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – the pair table of the secondary structure</p></li>
<li><p><strong>s</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – a flag to turn verbose output on/off</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the free energy of the input structure given the input sequence in 10kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_structure_pt" title="RNA.fold_compound.eval_structure_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.enumerate_necklaces">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">enumerate_necklaces</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_counts</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.enumerate_necklaces" title="Permalink to this definition"></a></dt>
<dd><p>Enumerate all necklaces with fixed content.</p>
<p>This function implements <em>A fast algorithm to generate necklaces with fixed content</em> as published by
<span id="id13">Sawada [<a class="reference internal" href="bibliography.html#id36" title="Joe Sawada. A fast algorithm to generate necklaces with fixed content. Theoretical Computer Science, 301(1):477–489, 2003. doi:10.1016/S0304-3975(03)00049-5.">2003</a>]</span> .</p>
<p>The function receives a list of counts (the elements on the necklace) for each type of object within
a necklace. The list starts at index 0 and ends with an entry that has a count of 0. The algorithm
then enumerates all non-cyclic permutations of the content, returned as a list of necklaces. This
list, again, is zero-terminated, i.e. the last entry of the list is a <cite>NULL</cite> pointer.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as global function <cite>enumerate_necklaces()</cite> which accepts lists input,
an produces list of lists output. See, e.g.   <a class="reference internal" href="#RNA.enumerate_necklaces" title="RNA.enumerate_necklaces"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.enumerate_necklaces()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>type_counts</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A 0-terminated list of entity counts</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of all non-cyclic permutations of the entities</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(list-like(unsigned</span> <span class="pre">int))</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.ep">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ep</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ep" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Data structure representing a single entry of an element probability list (e.g. list of pair
probabilities)</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.plist" title="RNA.plist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plist</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.plist_from_probs" title="RNA.fold_compound.plist_from_probs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.plist_from_probs</span></code></a>, <a class="reference internal" href="#RNA.db_from_plist" title="RNA.db_from_plist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_from_plist</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PLIST_TYPE_BASEPAIR</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PLIST_TYPE_GQUAD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PLIST_TYPE_H_MOTIF</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PLIST_TYPE_I_MOTIF</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PLIST_TYPE_UD_MOTIF</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PLIST_TYPE_STACK</span></code></p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.ep.i">
<span class="sig-name descname"><span class="pre">i</span></span><a class="headerlink" href="#RNA.ep.i" title="Permalink to this definition"></a></dt>
<dd><p>Start position (usually 5’ nucleotide that starts the element, e.g. base pair)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.ep.j">
<span class="sig-name descname"><span class="pre">j</span></span><a class="headerlink" href="#RNA.ep.j" title="Permalink to this definition"></a></dt>
<dd><p>End position (usually 3’ nucleotide that ends the element, e.g. base pair)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.ep.p">
<span class="sig-name descname"><span class="pre">p</span></span><a class="headerlink" href="#RNA.ep.p" title="Permalink to this definition"></a></dt>
<dd><p>Probability of the element.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.ep.type">
<span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#RNA.ep.type" title="Permalink to this definition"></a></dt>
<dd><p>Type of the element.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/utils/structures.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id14">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">i</span></span><a class="headerlink" href="#id14" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id15">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">j</span></span><a class="headerlink" href="#id15" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id16">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">p</span></span><a class="headerlink" href="#id16" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.ep.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.ep.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id17">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#id17" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.eval_circ_gquad_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">eval_circ_gquad_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.eval_circ_gquad_structure" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate free energy of a sequence/structure pair, assume sequence to be circular, allow for
G-Quadruplexes in the structure, and print contributions per loop.</p>
<p>This function is the same as RNA.eval_structure_simple_v() but assumes the input sequence to be
circular and allows for annotated G-Quadruplexes in the dot-bracket structure input.</p>
<p>G-Quadruplexes are annotated as plus signs (‘+’) for each G involved in the motif. Linker sequences
must be denoted by dots (‘.’) as they are considered unpaired. Below is an example of a 2-layer
G-quadruplex:</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available through an overloaded version of RNA.eval_circ_gquad_structure().
The last two arguments for this function are optional and default to RNA.VERBOSITY_QUIET and
<cite>NULL</cite>, respectively. See, e.g.   <a class="reference internal" href="#RNA.eval_circ_gquad_structure" title="RNA.eval_circ_gquad_structure"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.eval_circ_gquad_structure()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence in uppercase letters</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket notation</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The level of verbosity of this function</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle where this function should print to (may be NULL).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.eval_circ_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">eval_circ_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.eval_circ_structure" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate free energy of a sequence/structure pair, assume sequence to be circular and print
contributions per loop.</p>
<p>This function is the same as RNA.eval_structure_simple_v() but assumes the input sequence to be
circularized.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available through an overloaded version of RNA.eval_circ_structure(). The last
two arguments for this function are optional and default to RNA.VERBOSITY_QUIET and <cite>NULL</cite>,
respectively. See, e.g.   <a class="reference internal" href="#RNA.eval_circ_structure" title="RNA.eval_circ_structure"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.eval_circ_structure()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence in uppercase letters</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket notation</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The level of verbosity of this function</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle where this function should print to (may be NULL).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.eval_structure_simple_v</span></code>, <a class="reference internal" href="#RNA.eval_circ_structure" title="RNA.eval_circ_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.eval_circ_structure</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_verbose" title="RNA.fold_compound.eval_structure_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_verbose</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.eval_gquad_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">eval_gquad_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.eval_gquad_structure" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate free energy of a sequence/structure pair, allow for G-Quadruplexes in the structure and
print contributions per loop.</p>
<p>This function is the same as RNA.eval_structure_simple_v() but allows for annotated G-Quadruplexes
in the dot-bracket structure input.</p>
<p>G-Quadruplexes are annotated as plus signs (‘+’) for each G involved in the motif. Linker sequences
must be denoted by dots (‘.’) as they are considered unpaired. Below is an example of a 2-layer
G-quadruplex:</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available through an overloaded version of RNA.eval_gquad_structure(). The
last two arguments for this function are optional and default to RNA.VERBOSITY_QUIET and
<cite>NULL</cite>, respectively. See, e.g.   <a class="reference internal" href="#RNA.eval_gquad_structure" title="RNA.eval_gquad_structure"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.eval_gquad_structure()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence in uppercase letters</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket notation</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The level of verbosity of this function</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle where this function should print to (may be NULL).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.eval_structure_simple_v</span></code>, <a class="reference internal" href="#RNA.eval_gquad_structure" title="RNA.eval_gquad_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.eval_gquad_structure</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_verbose" title="RNA.fold_compound.eval_structure_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_verbose</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.eval_structure_pt_simple">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">eval_structure_pt_simple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.eval_structure_pt_simple" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA.</p>
<p>This function allows for energy evaluation of a given sequence/structure pair where the structure is
provided in pair_table format as obtained from RNA.ptable(). Model details, energy parameters, and
possibly soft constraints are used as provided via the parameter ‘fc’. The fold_compound does not
need to contain any DP matrices, but all the most basic init values as one would get from a call
like this:  In contrast to RNA.fold_compound.eval_structure_pt_verbose() this function assumes default model
details and default energy parameters in order to evaluate the free energy of the secondary
structure. Threefore, it serves as a simple interface function for energy evaluation for situations
where no changes on the energy model are required.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence in uppercase letters</p></li>
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – Secondary structure as pair_table</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The level of verbosity of this function</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle where this function should print to (may be NULL).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.eval_structure_pt_v</span></code>, <a class="reference internal" href="#RNA.eval_structure_simple" title="RNA.eval_structure_simple"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.eval_structure_simple</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.eval_structure_simple">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">eval_structure_simple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.eval_structure_simple" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA and print contributions per loop.</p>
<p>This function allows for detailed energy evaluation of a given sequence/structure pair. In contrast
to RNA.fold_compound.eval_structure() this function prints detailed energy contributions based on individual
loops to a file handle. If NULL is passed as file handle, this function defaults to print to stdout.
Any positive <cite>verbosity_level</cite> activates potential warning message of the energy evaluting
functions, while values <span class="math notranslate nohighlight">\(\ge 1\)</span> allow for detailed control of what data is printed. A
negative parameter <cite>verbosity_level</cite> turns off printing all together.</p>
<p>In contrast to RNA.fold_compound.eval_structure_verbose() this function assumes default model details and default
energy parameters in order to evaluate the free energy of the secondary structure. Threefore, it
serves as a simple interface function for energy evaluation for situations where no changes on the
energy model are required.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available through an overloaded version of RNA.eval_structure_simple(). The
last two arguments for this function are optional and default to RNA.VERBOSITY_QUIET and
<cite>NULL</cite>, respectively. See, e.g.   <a class="reference internal" href="#RNA.eval_structure_simple" title="RNA.eval_structure_simple"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.eval_structure_simple()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence in uppercase letters</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket notation</p></li>
<li><p><strong>verbosity_level</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The level of verbosity of this function</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle where this function should print to (may be NULL).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_structure_verbose" title="RNA.fold_compound.eval_structure_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_verbose</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt" title="RNA.fold_compound.eval_structure_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt_verbose" title="RNA.fold_compound.eval_structure_pt_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt_verbose</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.exp_E_ExtLoop">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">exp_E_ExtLoop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.exp_E_ExtLoop" title="Permalink to this definition"></a></dt>
<dd><p>This is the partition function variant of E_ExtLoop()</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.exp_E_ext_stem() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The Boltzmann weighted energy contribution of the introduced exterior-loop stem</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.E_ExtLoop" title="RNA.E_ExtLoop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">E_ExtLoop</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.exp_E_Hairpin">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">exp_E_Hairpin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.exp_E_Hairpin" title="Permalink to this definition"></a></dt>
<dd><p>Compute Boltzmann weight <span class="math notranslate nohighlight">\(e^{-\Delta G/kT}\)</span> of a hairpin loop.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the loop (number of unpaired nucleotides)</p></li>
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The pair type of the base pair closing the hairpin</p></li>
<li><p><strong>si1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">short</span></code>) – The 5’-mismatching nucleotide</p></li>
<li><p><strong>sj1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">short</span></code>) – The 3’-mismatching nucleotide</p></li>
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The sequence of the loop (May be <cite>NULL</cite>, otherwise mst be at least <span class="math notranslate nohighlight">\(size + 2\)</span> long)</p></li>
<li><p><strong>P</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.exp_param()</span> <span class="pre">*</span></code>) – The datastructure containing scaled Boltzmann weights of the energy parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Boltzmann weight of the Hairpin-loop</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Not (really) thread safe! A threadsafe implementation will replace this function in a future
release!</p>
<blockquote>
<div><p>Energy evaluation may change due to updates in global variable “tetra_loop”</p>
</div></blockquote>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_scaled_pf_parameters</span></code>, <a class="reference internal" href="#RNA.exp_param" title="RNA.exp_param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.exp_param</span></code></a>, <a class="reference internal" href="#RNA.E_Hairpin" title="RNA.E_Hairpin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">E_Hairpin</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>multiply by scale[u+2]</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.exp_E_IntLoop">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">exp_E_IntLoop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">u2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sp1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.exp_E_IntLoop" title="Permalink to this definition"></a></dt>
<dd><p>multiply by scale[u1+u2+2] for scaling</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the ‘left’-loop (number of unpaired nucleotides)</p></li>
<li><p><strong>u2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the ‘right’-loop (number of unpaired nucleotides)</p></li>
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The pair type of the base pair closing the interior loop</p></li>
<li><p><strong>type2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The pair type of the enclosed base pair</p></li>
<li><p><strong>si1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">short</span></code>) – The 5’-mismatching nucleotide of the closing pair</p></li>
<li><p><strong>sj1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">short</span></code>) – The 3’-mismatching nucleotide of the closing pair</p></li>
<li><p><strong>sp1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">short</span></code>) – The 3’-mismatching nucleotide of the enclosed pair</p></li>
<li><p><strong>sq1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">short</span></code>) – The 5’-mismatching nucleotide of the enclosed pair</p></li>
<li><p><strong>P</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.exp_param()</span> <span class="pre">*</span></code>) – The datastructure containing scaled Boltzmann weights of the energy parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Boltzmann weight of the Interior-loop</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_scaled_pf_parameters</span></code>, <a class="reference internal" href="#RNA.exp_param" title="RNA.exp_param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.exp_param</span></code></a>, <a class="reference internal" href="#RNA.E_IntLoop" title="RNA.E_IntLoop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">E_IntLoop</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is threadsafe</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.exp_E_MLstem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">exp_E_MLstem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.exp_E_MLstem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.exp_E_Stem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">exp_E_Stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extLoop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.exp_E_Stem" title="Permalink to this definition"></a></dt>
<dd><p>This is the partition function variant of E_Stem()</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The Boltzmann weighted energy contribution of the branch off the loop</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.E_Stem" title="RNA.E_Stem"><code class="xref py py-obj docutils literal notranslate"><span class="pre">E_Stem</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is threadsafe</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.exp_E_ext_stem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">exp_E_ext_stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n5d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n3d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.exp_E_ext_stem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.exp_param">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">exp_param</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_details</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.exp_param" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The data structure that contains temperature scaled Boltzmann weights of the energy parameters.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#RNA.exp_param.id" title="Permalink to this definition"></a></dt>
<dd><p>An identifier for the data structure.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This attribute will be removed in version 3</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expstack">
<span class="sig-name descname"><span class="pre">expstack</span></span><a class="headerlink" href="#RNA.exp_param.expstack" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.exphairpin">
<span class="sig-name descname"><span class="pre">exphairpin</span></span><a class="headerlink" href="#RNA.exp_param.exphairpin" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expbulge">
<span class="sig-name descname"><span class="pre">expbulge</span></span><a class="headerlink" href="#RNA.exp_param.expbulge" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expinternal">
<span class="sig-name descname"><span class="pre">expinternal</span></span><a class="headerlink" href="#RNA.exp_param.expinternal" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expmismatchExt">
<span class="sig-name descname"><span class="pre">expmismatchExt</span></span><a class="headerlink" href="#RNA.exp_param.expmismatchExt" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expmismatchI">
<span class="sig-name descname"><span class="pre">expmismatchI</span></span><a class="headerlink" href="#RNA.exp_param.expmismatchI" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expmismatch23I">
<span class="sig-name descname"><span class="pre">expmismatch23I</span></span><a class="headerlink" href="#RNA.exp_param.expmismatch23I" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expmismatch1nI">
<span class="sig-name descname"><span class="pre">expmismatch1nI</span></span><a class="headerlink" href="#RNA.exp_param.expmismatch1nI" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expmismatchH">
<span class="sig-name descname"><span class="pre">expmismatchH</span></span><a class="headerlink" href="#RNA.exp_param.expmismatchH" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expmismatchM">
<span class="sig-name descname"><span class="pre">expmismatchM</span></span><a class="headerlink" href="#RNA.exp_param.expmismatchM" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expdangle5">
<span class="sig-name descname"><span class="pre">expdangle5</span></span><a class="headerlink" href="#RNA.exp_param.expdangle5" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expdangle3">
<span class="sig-name descname"><span class="pre">expdangle3</span></span><a class="headerlink" href="#RNA.exp_param.expdangle3" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expint11">
<span class="sig-name descname"><span class="pre">expint11</span></span><a class="headerlink" href="#RNA.exp_param.expint11" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expint21">
<span class="sig-name descname"><span class="pre">expint21</span></span><a class="headerlink" href="#RNA.exp_param.expint21" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expint22">
<span class="sig-name descname"><span class="pre">expint22</span></span><a class="headerlink" href="#RNA.exp_param.expint22" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expninio">
<span class="sig-name descname"><span class="pre">expninio</span></span><a class="headerlink" href="#RNA.exp_param.expninio" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.lxc">
<span class="sig-name descname"><span class="pre">lxc</span></span><a class="headerlink" href="#RNA.exp_param.lxc" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expMLbase">
<span class="sig-name descname"><span class="pre">expMLbase</span></span><a class="headerlink" href="#RNA.exp_param.expMLbase" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expMLintern">
<span class="sig-name descname"><span class="pre">expMLintern</span></span><a class="headerlink" href="#RNA.exp_param.expMLintern" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expMLclosing">
<span class="sig-name descname"><span class="pre">expMLclosing</span></span><a class="headerlink" href="#RNA.exp_param.expMLclosing" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expTermAU">
<span class="sig-name descname"><span class="pre">expTermAU</span></span><a class="headerlink" href="#RNA.exp_param.expTermAU" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expDuplexInit">
<span class="sig-name descname"><span class="pre">expDuplexInit</span></span><a class="headerlink" href="#RNA.exp_param.expDuplexInit" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.exptetra">
<span class="sig-name descname"><span class="pre">exptetra</span></span><a class="headerlink" href="#RNA.exp_param.exptetra" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.exptri">
<span class="sig-name descname"><span class="pre">exptri</span></span><a class="headerlink" href="#RNA.exp_param.exptri" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.exphex">
<span class="sig-name descname"><span class="pre">exphex</span></span><a class="headerlink" href="#RNA.exp_param.exphex" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.Tetraloops">
<span class="sig-name descname"><span class="pre">Tetraloops</span></span><a class="headerlink" href="#RNA.exp_param.Tetraloops" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expTriloop">
<span class="sig-name descname"><span class="pre">expTriloop</span></span><a class="headerlink" href="#RNA.exp_param.expTriloop" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.Triloops">
<span class="sig-name descname"><span class="pre">Triloops</span></span><a class="headerlink" href="#RNA.exp_param.Triloops" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.Hexaloops">
<span class="sig-name descname"><span class="pre">Hexaloops</span></span><a class="headerlink" href="#RNA.exp_param.Hexaloops" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expTripleC">
<span class="sig-name descname"><span class="pre">expTripleC</span></span><a class="headerlink" href="#RNA.exp_param.expTripleC" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expMultipleCA">
<span class="sig-name descname"><span class="pre">expMultipleCA</span></span><a class="headerlink" href="#RNA.exp_param.expMultipleCA" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expMultipleCB">
<span class="sig-name descname"><span class="pre">expMultipleCB</span></span><a class="headerlink" href="#RNA.exp_param.expMultipleCB" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expgquad">
<span class="sig-name descname"><span class="pre">expgquad</span></span><a class="headerlink" href="#RNA.exp_param.expgquad" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expgquadLayerMismatch">
<span class="sig-name descname"><span class="pre">expgquadLayerMismatch</span></span><a class="headerlink" href="#RNA.exp_param.expgquadLayerMismatch" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.gquadLayerMismatchMax">
<span class="sig-name descname"><span class="pre">gquadLayerMismatchMax</span></span><a class="headerlink" href="#RNA.exp_param.gquadLayerMismatchMax" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.kT">
<span class="sig-name descname"><span class="pre">kT</span></span><a class="headerlink" href="#RNA.exp_param.kT" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.pf_scale">
<span class="sig-name descname"><span class="pre">pf_scale</span></span><a class="headerlink" href="#RNA.exp_param.pf_scale" title="Permalink to this definition"></a></dt>
<dd><p>Scaling factor to avoid over-/underflows.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.temperature">
<span class="sig-name descname"><span class="pre">temperature</span></span><a class="headerlink" href="#RNA.exp_param.temperature" title="Permalink to this definition"></a></dt>
<dd><p>Temperature used for loop contribution scaling.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.alpha">
<span class="sig-name descname"><span class="pre">alpha</span></span><a class="headerlink" href="#RNA.exp_param.alpha" title="Permalink to this definition"></a></dt>
<dd><p>Scaling factor for the thermodynamic temperature.</p>
<p>This allows for temperature scaling in Boltzmann factors independently from the energy
contributions. The resulting Boltzmann factors are then computed by <span class="math notranslate nohighlight">\(e^{-E/(\alpha \cdot
K \cdot T)}\)</span></p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.model_details">
<span class="sig-name descname"><span class="pre">model_details</span></span><a class="headerlink" href="#RNA.exp_param.model_details" title="Permalink to this definition"></a></dt>
<dd><p>Model details to be used in the recursions.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_md_t</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.param_file">
<span class="sig-name descname"><span class="pre">param_file</span></span><a class="headerlink" href="#RNA.exp_param.param_file" title="Permalink to this definition"></a></dt>
<dd><p>The filename the parameters were derived from, or empty string if they represent the default.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expSaltStack">
<span class="sig-name descname"><span class="pre">expSaltStack</span></span><a class="headerlink" href="#RNA.exp_param.expSaltStack" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.expSaltLoop">
<span class="sig-name descname"><span class="pre">expSaltLoop</span></span><a class="headerlink" href="#RNA.exp_param.expSaltLoop" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.SaltLoopDbl">
<span class="sig-name descname"><span class="pre">SaltLoopDbl</span></span><a class="headerlink" href="#RNA.exp_param.SaltLoopDbl" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.SaltMLbase">
<span class="sig-name descname"><span class="pre">SaltMLbase</span></span><a class="headerlink" href="#RNA.exp_param.SaltMLbase" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.SaltMLintern">
<span class="sig-name descname"><span class="pre">SaltMLintern</span></span><a class="headerlink" href="#RNA.exp_param.SaltMLintern" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.SaltMLclosing">
<span class="sig-name descname"><span class="pre">SaltMLclosing</span></span><a class="headerlink" href="#RNA.exp_param.SaltMLclosing" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.exp_param.SaltDPXInit">
<span class="sig-name descname"><span class="pre">SaltDPXInit</span></span><a class="headerlink" href="#RNA.exp_param.SaltDPXInit" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/params/basic.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id18">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Hexaloops</span></span><a class="headerlink" href="#id18" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id19">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltDPXInit</span></span><a class="headerlink" href="#id19" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id20">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltLoopDbl</span></span><a class="headerlink" href="#id20" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id21">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltMLbase</span></span><a class="headerlink" href="#id21" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id22">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltMLclosing</span></span><a class="headerlink" href="#id22" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id23">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltMLintern</span></span><a class="headerlink" href="#id23" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id24">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Tetraloops</span></span><a class="headerlink" href="#id24" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id25">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Triloops</span></span><a class="headerlink" href="#id25" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id26">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alpha</span></span><a class="headerlink" href="#id26" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id27">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expDuplexInit</span></span><a class="headerlink" href="#id27" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id28">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expMLbase</span></span><a class="headerlink" href="#id28" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id29">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expMLclosing</span></span><a class="headerlink" href="#id29" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id30">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expMLintern</span></span><a class="headerlink" href="#id30" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id31">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expMultipleCA</span></span><a class="headerlink" href="#id31" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id32">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expMultipleCB</span></span><a class="headerlink" href="#id32" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id33">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expSaltLoop</span></span><a class="headerlink" href="#id33" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id34">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expSaltStack</span></span><a class="headerlink" href="#id34" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id35">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expTermAU</span></span><a class="headerlink" href="#id35" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id36">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expTriloop</span></span><a class="headerlink" href="#id36" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id37">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expTripleC</span></span><a class="headerlink" href="#id37" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id38">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expbulge</span></span><a class="headerlink" href="#id38" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id39">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expdangle3</span></span><a class="headerlink" href="#id39" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id40">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expdangle5</span></span><a class="headerlink" href="#id40" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id41">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expgquad</span></span><a class="headerlink" href="#id41" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id42">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expgquadLayerMismatch</span></span><a class="headerlink" href="#id42" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id43">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exphairpin</span></span><a class="headerlink" href="#id43" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id44">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exphex</span></span><a class="headerlink" href="#id44" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id45">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expint11</span></span><a class="headerlink" href="#id45" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id46">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expint21</span></span><a class="headerlink" href="#id46" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id47">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expint22</span></span><a class="headerlink" href="#id47" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id48">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expinternal</span></span><a class="headerlink" href="#id48" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id49">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expmismatch1nI</span></span><a class="headerlink" href="#id49" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id50">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expmismatch23I</span></span><a class="headerlink" href="#id50" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id51">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expmismatchExt</span></span><a class="headerlink" href="#id51" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id52">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expmismatchH</span></span><a class="headerlink" href="#id52" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id53">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expmismatchI</span></span><a class="headerlink" href="#id53" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id54">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expmismatchM</span></span><a class="headerlink" href="#id54" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id55">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expninio</span></span><a class="headerlink" href="#id55" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id56">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expstack</span></span><a class="headerlink" href="#id56" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id57">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exptetra</span></span><a class="headerlink" href="#id57" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id58">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exptri</span></span><a class="headerlink" href="#id58" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id59">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gquadLayerMismatchMax</span></span><a class="headerlink" href="#id59" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id60">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#id60" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id61">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">kT</span></span><a class="headerlink" href="#id61" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id62">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lxc</span></span><a class="headerlink" href="#id62" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id63">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">model_details</span></span><a class="headerlink" href="#id63" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id64">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">param_file</span></span><a class="headerlink" href="#id64" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id65">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pf_scale</span></span><a class="headerlink" href="#id65" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id66">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">temperature</span></span><a class="headerlink" href="#id66" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.exp_param.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.exp_param.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.expand_Full">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">expand_Full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.expand_Full" title="Permalink to this definition"></a></dt>
<dd><p>Convert the full structure from bracket notation to the expanded notation including root.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.expand_Shapiro">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">expand_Shapiro</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coarse</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.expand_Shapiro" title="Permalink to this definition"></a></dt>
<dd><p>Inserts missing ‘S’ identifiers in unweighted coarse grained structures as obtained from b2C().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>coarse</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.extract_record_rest_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">extract_record_rest_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lines</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">option</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.extract_record_rest_structure" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.fc_add_pycallback">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">fc_add_pycallback</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyFunc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fc_add_pycallback" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.fc_add_pydata">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">fc_add_pydata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyFuncOrNone</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fc_add_pydata" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_PS_aln">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_PS_aln</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">identifiers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">start=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">end=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">offset=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">columns=60</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.file_PS_aln" title="Permalink to this definition"></a></dt>
<dd><p>Create an annotated PostScript alignment plot.</p>
<p>Similar to RNA.file_PS_aln() but allows the user to print a particular slice of the alignment by
specifying a <cite>start</cite> and <cite>end</cite> position. The additional <cite>offset</cite> parameter allows for adjusting the
alignment position ruler value.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>file_PS_aln()</cite> where the last four parameter
may be omitted, indicating <cite>start</cite> = <em>0</em>, <cite>end</cite> = <em>0</em>, <cite>offset</cite> = <em>0</em>, and <cite>columns</cite> = <em>60</em>.
See, e.g.  <a class="reference internal" href="#RNA.file_PS_aln" title="RNA.file_PS_aln"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.file_PS_aln()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The output file name</p></li>
<li><p><strong>seqs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The aligned sequences</p></li>
<li><p><strong>names</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The names of the sequences</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The consensus structure in dot-bracket notation</p></li>
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The start of the alignment slice (a value of 0 indicates the first position of the alignment,
i.e. no slicing at 5’ side)</p></li>
<li><p><strong>end</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The end of the alignment slice (a value of 0 indicates the last position of the alignment, i.e.
no slicing at 3’ side)</p></li>
<li><p><strong>offset</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The alignment coordinate offset for the position ruler.</p></li>
<li><p><strong>columns</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The number of columns before the alignment is wrapped as a new block (a value of 0 indicates no
wrapping)</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_PS_aln_slice</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_PS_rnaplot">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_PS_rnaplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.file_PS_rnaplot" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_PS_rnaplot_a">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_PS_rnaplot_a</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.file_PS_rnaplot_a" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_RNAstrand_db_read_record">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_RNAstrand_db_read_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.file_RNAstrand_db_read_record" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_SHAPE_read">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_SHAPE_read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.file_SHAPE_read" title="Permalink to this definition"></a></dt>
<dd><p>Read data from a given SHAPE reactivity input file.</p>
<p>This function parses the informations from a given file and stores the result in the preallocated
string sequence and the double array values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Path to the constraints file</p></li>
<li><p><strong>length</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Length of the sequence (file entries exceeding this limit will cause an error)</p></li>
<li><p><strong>default_value</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – Value for missing indices</p></li>
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Pointer to an array used for storing the sequence obtained from the SHAPE reactivity file</p></li>
<li><p><strong>values</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – Pointer to an array used for storing the values obtained from the SHAPE reactivity file</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_commands_read">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_commands_read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.cmd" title="RNA.cmd"><span class="pre">cmd</span></a></span></span><a class="headerlink" href="#RNA.file_commands_read" title="Permalink to this definition"></a></dt>
<dd><p>Extract a list of commands from a command file.</p>
<p>Read a list of commands specified in the input file and return them as list of abstract commands</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as global function <cite>file_commands_read()</cite>. See, e.g.
<a class="reference internal" href="#RNA.file_commands_read" title="RNA.file_commands_read"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.file_commands_read()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The filename</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to limit the type of commands read from the file</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of abstract commands</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#RNA.cmd" title="RNA.cmd"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.cmd()</span></code></a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.commands_apply" title="RNA.fold_compound.commands_apply"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.commands_apply</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_commands_apply</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.commands_free</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_connect_read_record">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_connect_read_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remainder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.file_connect_read_record" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_fasta_read">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_fasta_read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">FILE</span> <span class="pre">*</span> <span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.file_fasta_read" title="Permalink to this definition"></a></dt>
<dd><p>Get a (fasta) data set from a file or stdin.</p>
<p>This function may be used to obtain complete datasets from a filehandle or stdin. A dataset is
always defined to contain at least a sequence. If data starts with a fasta header, i.e. a line like</p>
<blockquote>
<div><p>&gt;some header info  then RNA.file_fasta_read_record() will assume that the sequence that follows</p>
</div></blockquote>
<p>the header may span over several lines. To disable this behavior and to assign a single line to the
argument ‘sequence’ one can pass RNA.INPUT_NO_SPAN in the ‘options’ argument. If no fasta header is
read in the beginning of a data block, a sequence must not span over multiple lines!</p>
<p>Unless the options RNA.INPUT_NOSKIP_COMMENTS or RNA.INPUT_NOSKIP_BLANK_LINES are passed, a
sequence may be interrupted by lines starting with a comment character or empty lines.</p>
<blockquote>
<div><p>A sequence is regarded as completely read if it was either assumed to not span over multiple lines,</p>
</div></blockquote>
<p>a secondary structure or structure constraint follows the sequence on the next line, or a new header
marks the beginning of a new sequence…</p>
<p>All lines following the sequence (this includes comments) that do not initiate a new dataset
according to the above definition are available through the line-array ‘rest’. Here one can usually
find the structure constraint or other information belonging to the current dataset. Filling of
‘rest’ may be prevented by passing RNA.INPUT_NO_REST to the options argument.</p>
<p>The main purpose of this function is to be able to easily parse blocks of data in the header of a
loop where all calculations for the appropriate data is done inside the loop. The loop may be then
left on certain return values, e.g.:</p>
<p>In the example above, the while loop will be terminated when RNA.file_fasta_read_record() returns
either an error, EOF, or a user initiated quit request.</p>
<p>As long as data is read from stdin (we are passing NULL as the file pointer), the id is printed if
it is available for the current block of data. The sequence will be printed in any case and if some
more lines belong to the current block of data each line will be printed as well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>header</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code>) – A pointer which will be set such that it points to the header of the record</p></li>
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code>) – A pointer which will be set such that it points to the sequence of the record</p></li>
<li><p><strong>rest</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">***</span></code>) – A pointer which will be set such that it points to an array of lines which also belong to the
record</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle to read from (if NULL, this function reads from stdin)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Some options which may be passed to alter the behavior of the function, use 0 for no options</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A flag with information about what the function actually did read</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<dl class="simple">
<dt>This function will exit any program with an error message if no sequence could be read!</dt><dd><p>This function is NOT threadsafe! It uses a global variable to store information about the next data</p>
</dd>
</dl>
<p>block. Do not forget to free the memory occupied by header, sequence and rest!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_msa_detect_format">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_msa_detect_format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">unsigned</span> <span class="pre">int</span></span></span><a class="headerlink" href="#RNA.file_msa_detect_format" title="Permalink to this definition"></a></dt>
<dd><p>Detect the format of a multiple sequence alignment file.</p>
<p>This function attempts to determine the format of a file that supposedly contains a multiple
sequence alignment (MSA). This is useful in cases where a MSA file contains more than a single
record and therefore RNA.file_msa_read() can not be applied, since it only retrieves the first.
Here, one can try to guess the correct file format using this function and then loop over the file,
record by record using one of the low-level record retrieval functions for the corresponding MSA
file format.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function exists as an overloaded version where the <cite>options</cite> parameter may be omitted! In
that case, the <cite>options</cite> parameter defaults to RNA.FILE_FORMAT_MSA_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.file_msa_detect_format" title="RNA.file_msa_detect_format"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.file_msa_detect_format()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The name of input file that contains the alignment</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to manipulate the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The MSA file format, or RNA.FILE_FORMAT_MSA_UNKNOWN</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.file_msa_read" title="RNA.file_msa_read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_msa_read</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_stockholm_read_record</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_clustal_read_record</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_fasta_read_record</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function parses the entire first record within the specified file. As a result, it returns
RNA.FILE_FORMAT_MSA_UNKNOWN not only if it can’t detect the file’s format, but also in cases where
the file doesn’t contain sequences!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_msa_read">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_msa_read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.file_msa_read" title="Permalink to this definition"></a></dt>
<dd><p>Read a multiple sequence alignment from file.</p>
<p>This function reads the (first) multiple sequence alignment from an input file. The read alignment
is split into the sequence id/name part and the actual sequence information and stored in memory as
arrays of ids/names and sequences. If the alignment file format allows for additional information,
such as an ID of the entire alignment or consensus structure information, this data is retrieved as
well and made available. The <cite>options</cite> parameter allows to specify the set of alignment file formats
that should be used to retrieve the data. If 0 is passed as option, the list of alignment file
formats defaults to RNA.FILE_FORMAT_MSA_DEFAULT.</p>
<p>Currently, the list of parsable multiple sequence alignment file formats consists of:</p>
<ul class="simple">
<li><p>msa-formats-clustal</p></li>
<li><p>msa-formats-stockholm</p></li>
<li><p>msa-formats-fasta</p></li>
<li><p>msa-formats-maf</p></li>
</ul>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>In the target scripting language, only the first and last argument, <cite>filename</cite> and <cite>options</cite>,
are passed to the corresponding function. The other arguments, which serve as output in the
C-library, are available as additional return values. This function exists as an overloaded
version where the <cite>options</cite> parameter may be omitted! In that case, the <cite>options</cite> parameter
defaults to RNA.FILE_FORMAT_MSA_STOCKHOLM. See, e.g.   <a class="reference internal" href="#RNA.file_msa_read" title="RNA.file_msa_read"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.file_msa_read()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>  and   <a class="reference internal" href="examples/python.html#parsing-alignments"><span class="std std-ref">Parsing Alignments</span></a>  in the Python examples.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The name of input file that contains the alignment</p></li>
<li><p><strong>names</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">***</span></code>) – An address to the pointer where sequence identifiers should be written to</p></li>
<li><p><strong>aln</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">***</span></code>) – An address to the pointer where aligned sequences should be written to</p></li>
<li><p><strong>id</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code>) – An address to the pointer where the alignment ID should be written to (Maybe NULL)</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code>) – An address to the pointer where consensus structure information should be written to (Maybe
NULL)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to manipulate the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of sequences in the alignment, or -1 if no alignment record could be found</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.file_msa_read_record" title="RNA.file_msa_read_record"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_msa_read_record</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_CLUSTAL</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_STOCKHOLM</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_FASTA</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_MAF</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_NOCHECK</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>After successfully reading an alignment, this function performs a validation of the data that
includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be
deactivated by passing RNA.FILE_FORMAT_MSA_NOCHECK in the <cite>options</cite> parameter.</p>
<blockquote>
<div><p>It is the users responsibility to free any memory occupied by the output arguments <cite>names</cite>, <cite>aln</cite>,</p>
</div></blockquote>
<p><cite>id</cite>, and <cite>structure</cite> after calling this function. The function automatically sets the latter two
arguments to <cite>NULL</cite> in case no corresponding data could be retrieved from the input alignment.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_msa_read_record">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_msa_read_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">FILE</span> <span class="pre">*</span> <span class="pre">filehandle</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.file_msa_read_record" title="Permalink to this definition"></a></dt>
<dd><p>Read a multiple sequence alignment from file handle.</p>
<p>Similar to RNA.file_msa_read(), this function reads a multiple sequence alignment from an input
file handle. Since using a file handle, this function is not limited to the first alignment record,
but allows for looping over all alignments within the input.</p>
<p>The read alignment is split into the sequence id/name part and the actual sequence information and
stored in memory as arrays of ids/names and sequences. If the alignment file format allows for
additional information, such as an ID of the entire alignment or consensus structure information,
this data is retrieved as well and made available. The <cite>options</cite> parameter allows to specify the
alignment file format used to retrieve the data. A single format must be specified here, see
RNA.file_msa_detect_format() for helping to determine the correct MSA file format.</p>
<p>Currently, the list of parsable multiple sequence alignment file formats consists of:</p>
<ul class="simple">
<li><p>msa-formats-clustal</p></li>
<li><p>msa-formats-stockholm</p></li>
<li><p>msa-formats-fasta</p></li>
<li><p>msa-formats-maf</p></li>
</ul>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>In the target scripting language, only the first and last argument, <cite>fp</cite> and <cite>options</cite>, are
passed to the corresponding function. The other arguments, which serve as output in the
C-library, are available as additional return values. This function exists as an overloaded
version where the <cite>options</cite> parameter may be omitted! In that case, the <cite>options</cite> parameter
defaults to RNA.FILE_FORMAT_MSA_STOCKHOLM. See, e.g.   <a class="reference internal" href="#RNA.file_msa_read_record" title="RNA.file_msa_read_record"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.file_msa_read_record()</span></code></a> in
the <a class="reference internal" href="#"><span class="doc">Python API</span></a>  and   <a class="reference internal" href="examples/python.html#parsing-alignments"><span class="std std-ref">Parsing Alignments</span></a>  in the Python examples.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – The file pointer the data will be retrieved from</p></li>
<li><p><strong>names</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">***</span></code>) – An address to the pointer where sequence identifiers should be written to</p></li>
<li><p><strong>aln</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">***</span></code>) – An address to the pointer where aligned sequences should be written to</p></li>
<li><p><strong>id</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code>) – An address to the pointer where the alignment ID should be written to (Maybe NULL)</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code>) – An address to the pointer where consensus structure information should be written to (Maybe
NULL)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to manipulate the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of sequences in the alignment, or -1 if no alignment record could be found</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.file_msa_read" title="RNA.file_msa_read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_msa_read</span></code></a>, <a class="reference internal" href="#RNA.file_msa_detect_format" title="RNA.file_msa_detect_format"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_msa_detect_format</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_CLUSTAL</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_STOCKHOLM</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_FASTA</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_MAF</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_NOCHECK</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>After successfully reading an alignment, this function performs a validation of the data that
includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be
deactivated by passing RNA.FILE_FORMAT_MSA_NOCHECK in the <cite>options</cite> parameter.</p>
<blockquote>
<div><p>It is the users responsibility to free any memory occupied by the output arguments <cite>names</cite>, <cite>aln</cite>,</p>
</div></blockquote>
<p><cite>id</cite>, and <cite>structure</cite> after calling this function. The function automatically sets the latter two
arguments to <cite>NULL</cite> in case no corresponding data could be retrieved from the input alignment.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.file_msa_write">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">file_msa_write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">id=&quot;&quot;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure=&quot;&quot;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">source=&quot;&quot;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=VRNA_FILE_FORMAT_MSA_STOCKHOLM|VRNA_FILE_FORMAT_MSA_APPEND</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.file_msa_write" title="Permalink to this definition"></a></dt>
<dd><p>Write multiple sequence alignment file.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>In the target scripting language, this function exists as a set of overloaded versions, where
the last four parameters may be omitted. If the <cite>options</cite> parameter is missing the options
default to (RNA.FILE_FORMAT_MSA_STOCKHOLM | RNA.FILE_FORMAT_MSA_APPEND). See, e.g.
<a class="reference internal" href="#RNA.file_msa_write" title="RNA.file_msa_write"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.file_msa_write()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The output filename</p></li>
<li><p><strong>names</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The array of sequence names / identifies</p></li>
<li><p><strong>aln</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – The array of aligned sequences</p></li>
<li><p><strong>id</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – An optional ID for the alignment</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – An optional consensus structure</p></li>
<li><p><strong>source</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A string describing the source of the alignment</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to manipulate the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-null upon successfully writing the alignment to file</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_STOCKHOLM</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_APPEND</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.FILE_FORMAT_MSA_MIS</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently, we only support msa-formats-stockholm output</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.filename_sanitize">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">filename_sanitize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.filename_sanitize" title="Permalink to this definition"></a></dt>
<dd><p>Sanitize a file name.</p>
<p>Returns a new file name where all invalid characters are substituted by a replacement character. If
no replacement character is supplied, invalid characters are simply removed from the filename. File
names may also never exceed a length of 255 characters. Longer file names will undergo a ‘smart’
truncation process, where the filenames suffix, i.e. everything after the last dot .’, is attempted
to be kept intact. Hence, only the filename part before the suffix is reduced in such a way that the
total filename complies to the length restriction of 255 characters. If no suffix is present or the
suffix itself already exceeds the maximum length, the filename is simply truncated from the back of
the string.</p>
<p>For now we consider the following characters invalid:</p>
<ul class="simple">
<li><p>backslash ‘'</p></li>
<li><p>slash ‘/’</p></li>
<li><p>question mark ‘?’</p></li>
<li><p>percent sign ‘’</p></li>
<li><p>asterisk ‘*’</p></li>
<li><p>colon ‘:’</p></li>
<li><p>pipe symbol ‘|’</p></li>
<li><p>double quote ‘”’</p></li>
<li><p>triangular brackets ‘&lt;’ and ‘&gt;’</p></li>
</ul>
<p>Furthermore, the (resulting) file name must not be a reserved file name, such as:</p>
<ul class="simple">
<li><p>‘.’</p></li>
<li><p>‘..’</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The input file name</p></li>
<li><p><strong>replacement</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The replacement character, or NULL</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sanitized file name, or NULL</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function allocates a new block of memory for the sanitized string. It also may return (a) NULL
if the input is pointing to NULL, or (b) an empty string if the input only consists of invalid
characters which are simply removed!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.find_saddle">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">find_saddle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.find_saddle" title="Permalink to this definition"></a></dt>
<dd><p>Find energy of a saddle point between 2 structures (search only direct path)</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.path_findpath_saddle() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the first secondary structure in dot-bracket notation
will be written to</p></li>
<li><p><strong>s2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the second secondary structure in dot-bracket notation
will be written to</p></li>
<li><p><strong>width</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – integer how many strutures are being kept during the search</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the saddle energy in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.floatArray">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">floatArray</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.floatArray" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.floatArray.cast">
<span class="sig-name descname"><span class="pre">cast</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.floatArray.cast" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.floatArray.frompointer">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">frompointer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.floatArray.frompointer" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.floatArray.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.floatArray.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.floatArray_frompointer">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">floatArray_frompointer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.floatArray_frompointer" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.floatP_getitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">floatP_getitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.floatP_getitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.floatP_setitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">floatP_setitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.floatP_setitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string)</span> <span class="pre">-&gt;</span> <span class="pre">(structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mfe)fold(string)</span> <span class="pre">-&gt;</span> <span class="pre">(structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mfe</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold" title="Permalink to this definition"></a></dt>
<dd><p>Compute Minimum Free Energy (MFE), and a corresponding secondary structure for an RNA sequence.</p>
<p>This simplified interface to RNA.fold_compound.mfe() computes the MFE and, if required, a secondary structure for
an RNA sequence using default options. Memory required for dynamic programming (DP) matrices will be
allocated and free’d on-the-fly. Hence, after return of this function, the recursively filled
matrices are not available any more for any post-processing, e.g. suboptimal backtracking, etc.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>fold()</cite> in the global namespace. The parameter
<cite>structure</cite> is returned along with the MFE und must not be provided. See e.g.
<a class="reference internal" href="#RNA.fold" title="RNA.fold"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.fold()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the secondary structure in dot-bracket notation will be
written to</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the minimum free energy (MFE) in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.circfold" title="RNA.circfold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.circfold</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case you want to use the filled DP matrices for any subsequent post-processing step, or you
require other conditions than specified by the default model details, use RNA.fold_compound.mfe(),
and the data
structure RNA.fold_compound() instead.</p>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.fold_compound">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">fold_compound</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The most basic data structure required by many functions throughout the RNAlib.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Please read the documentation of this data structure carefully! Some attributes are only available
for specific types this data structure can adopt.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Reading/Writing from/to attributes that are not within the scope of the current type usually result
in undefined behavior!</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>RNA.fold_compound().type, RNA.fold_compound(), RNA.fold_compound_comparative(),
RNA.fold_compound_free(), RNA.FC_TYPE_SINGLE, RNA.FC_TYPE_COMPARATIVE</p>
<p><strong>SWIG Wrapper Notes</strong></p>
<blockquote>
<div><p>This data structure is wrapped as class <cite>fold_compound</cite> with several related functions attached
as methods.</p>
<p>A new <cite>fold_compound</cite> can be obtained by calling one of its constructors:</p>
<ul class="simple">
<li><p><cite>fold_compound(seq)</cite> - Initialize with a single sequence, or two concatenated sequences
separated by an ampersand character <cite>&amp;</cite> (for cofolding)</p></li>
<li><p><cite>fold_compound(aln)</cite> - Initialize with a sequence alignment <em>aln</em> stored as a list of
sequences (with gap characters).</p></li>
</ul>
<p>The resulting object has a list of attached methods which in most cases directly correspond to
functions that mainly operate on the corresponding <cite>C</cite> data structure:</p>
<ul class="simple">
<li><p><cite>type()</cite> - Get the type of the <em>fold_compound</em> (See RNA.fc_type)</p></li>
<li><p><cite>length()</cite> - Get the length of the sequence(s) or alignment stored within the
<cite>fold_compound</cite>.</p></li>
</ul>
<p>See, e.g.  <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</div></blockquote>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.type">
<span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#RNA.fold_compound.type" title="Permalink to this definition"></a></dt>
<dd><p>The type of the RNA.fold_compound().</p>
<p>Currently possible values are RNA.FC_TYPE_SINGLE, and RNA.FC_TYPE_COMPARATIVE</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Do not edit this attribute, it will be automagically set by the corresponding get() methods for
the RNA.fold_compound(). The value specified in this attribute dictates the set of other
attributes to use within this data structure.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">vrna_fc_type_e</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.length">
<span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#RNA.fold_compound.length" title="Permalink to this definition"></a></dt>
<dd><p>The length of the sequence (or sequence alignment)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.cutpoint">
<span class="sig-name descname"><span class="pre">cutpoint</span></span><a class="headerlink" href="#RNA.fold_compound.cutpoint" title="Permalink to this definition"></a></dt>
<dd><p>The position of the (cofold) cutpoint within the provided sequence. If there is no cutpoint,
this field will be set to -1.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.strand_number">
<span class="sig-name descname"><span class="pre">strand_number</span></span><a class="headerlink" href="#RNA.fold_compound.strand_number" title="Permalink to this definition"></a></dt>
<dd><p>The strand number a particular nucleotide is associated with.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.strand_order">
<span class="sig-name descname"><span class="pre">strand_order</span></span><a class="headerlink" href="#RNA.fold_compound.strand_order" title="Permalink to this definition"></a></dt>
<dd><p>The strand order, i.e. permutation of current concatenated sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.strand_order_uniq">
<span class="sig-name descname"><span class="pre">strand_order_uniq</span></span><a class="headerlink" href="#RNA.fold_compound.strand_order_uniq" title="Permalink to this definition"></a></dt>
<dd><p>The strand order array where identical sequences have the same ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.strand_start">
<span class="sig-name descname"><span class="pre">strand_start</span></span><a class="headerlink" href="#RNA.fold_compound.strand_start" title="Permalink to this definition"></a></dt>
<dd><p>The start position of a particular strand within the current concatenated sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.strand_end">
<span class="sig-name descname"><span class="pre">strand_end</span></span><a class="headerlink" href="#RNA.fold_compound.strand_end" title="Permalink to this definition"></a></dt>
<dd><p>The end (last) position of a particular strand within the current concatenated sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.strands">
<span class="sig-name descname"><span class="pre">strands</span></span><a class="headerlink" href="#RNA.fold_compound.strands" title="Permalink to this definition"></a></dt>
<dd><p>Number of interacting strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.nucleotides">
<span class="sig-name descname"><span class="pre">nucleotides</span></span><a class="headerlink" href="#RNA.fold_compound.nucleotides" title="Permalink to this definition"></a></dt>
<dd><p>Set of nucleotide sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_seq_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.alignment">
<span class="sig-name descname"><span class="pre">alignment</span></span><a class="headerlink" href="#RNA.fold_compound.alignment" title="Permalink to this definition"></a></dt>
<dd><p>Set of alignments.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_msa_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.hc">
<span class="sig-name descname"><span class="pre">hc</span></span><a class="headerlink" href="#RNA.fold_compound.hc" title="Permalink to this definition"></a></dt>
<dd><p>The hard constraints data structure used for structure prediction.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_hc_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.matrices">
<span class="sig-name descname"><span class="pre">matrices</span></span><a class="headerlink" href="#RNA.fold_compound.matrices" title="Permalink to this definition"></a></dt>
<dd><p>The MFE DP matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_mx_mfe_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_matrices">
<span class="sig-name descname"><span class="pre">exp_matrices</span></span><a class="headerlink" href="#RNA.fold_compound.exp_matrices" title="Permalink to this definition"></a></dt>
<dd><p>The PF DP matrices</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_mx_pf_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.params">
<span class="sig-name descname"><span class="pre">params</span></span><a class="headerlink" href="#RNA.fold_compound.params" title="Permalink to this definition"></a></dt>
<dd><p>The precomputed free energy contributions for each type of loop.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#RNA.param" title="RNA.param"><code class="xref py py-class docutils literal notranslate"><span class="pre">param</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_params">
<span class="sig-name descname"><span class="pre">exp_params</span></span><a class="headerlink" href="#RNA.fold_compound.exp_params" title="Permalink to this definition"></a></dt>
<dd><p>The precomputed free energy contributions as Boltzmann factors</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#RNA.exp_param" title="RNA.exp_param"><code class="xref py py-class docutils literal notranslate"><span class="pre">exp_param</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.iindx">
<span class="sig-name descname"><span class="pre">iindx</span></span><a class="headerlink" href="#RNA.fold_compound.iindx" title="Permalink to this definition"></a></dt>
<dd><p>DP matrix accessor</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.jindx">
<span class="sig-name descname"><span class="pre">jindx</span></span><a class="headerlink" href="#RNA.fold_compound.jindx" title="Permalink to this definition"></a></dt>
<dd><p>DP matrix accessor</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.stat_cb">
<span class="sig-name descname"><span class="pre">stat_cb</span></span><a class="headerlink" href="#RNA.fold_compound.stat_cb" title="Permalink to this definition"></a></dt>
<dd><p>Recursion status callback (usually called just before, and after recursive computations in the
library.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.recursion_status</span></code>, <a class="reference internal" href="#RNA.fold_compound.add_callback" title="RNA.fold_compound.add_callback"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.add_callback</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_recursion_status_f</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.auxdata">
<span class="sig-name descname"><span class="pre">auxdata</span></span><a class="headerlink" href="#RNA.fold_compound.auxdata" title="Permalink to this definition"></a></dt>
<dd><p>A pointer to auxiliary, user-defined data.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.add_auxdata" title="RNA.fold_compound.add_auxdata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.add_auxdata</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.free_auxdata">
<span class="sig-name descname"><span class="pre">free_auxdata</span></span><a class="headerlink" href="#RNA.fold_compound.free_auxdata" title="Permalink to this definition"></a></dt>
<dd><p>A callback to free auxiliary user data whenever the fold_compound itself is free’d.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.auxdata_free</span></code></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_auxdata_free_f</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.domains_struc">
<span class="sig-name descname"><span class="pre">domains_struc</span></span><a class="headerlink" href="#RNA.fold_compound.domains_struc" title="Permalink to this definition"></a></dt>
<dd><p>Additional structured domains.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_sd_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.domains_up">
<span class="sig-name descname"><span class="pre">domains_up</span></span><a class="headerlink" href="#RNA.fold_compound.domains_up" title="Permalink to this definition"></a></dt>
<dd><p>Additional unstructured domains.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_ud_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.aux_grammar">
<span class="sig-name descname"><span class="pre">aux_grammar</span></span><a class="headerlink" href="#RNA.fold_compound.aux_grammar" title="Permalink to this definition"></a></dt>
<dd><p>Additional decomposition grammar rules.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_gr_aux_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.sequence">
<span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#RNA.fold_compound.sequence" title="Permalink to this definition"></a></dt>
<dd><p>The input sequence string.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_SINGLE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.sequence_encoding">
<span class="sig-name descname"><span class="pre">sequence_encoding</span></span><a class="headerlink" href="#RNA.fold_compound.sequence_encoding" title="Permalink to this definition"></a></dt>
<dd><p>Numerical encoding of the input sequence.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sequence_encode</span></code></p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_SINGLE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.encoding5">
<span class="sig-name descname"><span class="pre">encoding5</span></span><a class="headerlink" href="#RNA.fold_compound.encoding5" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.encoding3">
<span class="sig-name descname"><span class="pre">encoding3</span></span><a class="headerlink" href="#RNA.fold_compound.encoding3" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.sequence_encoding2">
<span class="sig-name descname"><span class="pre">sequence_encoding2</span></span><a class="headerlink" href="#RNA.fold_compound.sequence_encoding2" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.ptype">
<span class="sig-name descname"><span class="pre">ptype</span></span><a class="headerlink" href="#RNA.fold_compound.ptype" title="Permalink to this definition"></a></dt>
<dd><p>Pair type array.</p>
<p>Contains the numerical encoding of the pair type for each pair (i,j) used in MFE, Partition
function and Evaluation computations.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This array is always indexed via jindx, in contrast to previously different indexing between mfe
and pf variants!</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_SINGLE</p>
</div></blockquote>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.idx_col_wise</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptypes</span></code></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.ptype_pf_compat">
<span class="sig-name descname"><span class="pre">ptype_pf_compat</span></span><a class="headerlink" href="#RNA.fold_compound.ptype_pf_compat" title="Permalink to this definition"></a></dt>
<dd><p>ptype array indexed via iindx</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This attribute will vanish in the future! It’s meant for backward compatibility only!</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_SINGLE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.sc">
<span class="sig-name descname"><span class="pre">sc</span></span><a class="headerlink" href="#RNA.fold_compound.sc" title="Permalink to this definition"></a></dt>
<dd><p>The soft constraints for usage in structure prediction and evaluation.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_SINGLE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_sc_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.sequences">
<span class="sig-name descname"><span class="pre">sequences</span></span><a class="headerlink" href="#RNA.fold_compound.sequences" title="Permalink to this definition"></a></dt>
<dd><p>The aligned sequences.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The end of the alignment is indicated by a NULL pointer in the second dimension</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.n_seq">
<span class="sig-name descname"><span class="pre">n_seq</span></span><a class="headerlink" href="#RNA.fold_compound.n_seq" title="Permalink to this definition"></a></dt>
<dd><p>The number of sequences in the alignment.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.cons_seq">
<span class="sig-name descname"><span class="pre">cons_seq</span></span><a class="headerlink" href="#RNA.fold_compound.cons_seq" title="Permalink to this definition"></a></dt>
<dd><p>The consensus sequence of the aligned sequences.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.S_cons">
<span class="sig-name descname"><span class="pre">S_cons</span></span><a class="headerlink" href="#RNA.fold_compound.S_cons" title="Permalink to this definition"></a></dt>
<dd><p>Numerical encoding of the consensus sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.S">
<span class="sig-name descname"><span class="pre">S</span></span><a class="headerlink" href="#RNA.fold_compound.S" title="Permalink to this definition"></a></dt>
<dd><p>Numerical encoding of the sequences in the alignment.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">short</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.S5">
<span class="sig-name descname"><span class="pre">S5</span></span><a class="headerlink" href="#RNA.fold_compound.S5" title="Permalink to this definition"></a></dt>
<dd><p>S5[s][i] holds next base 5’ of i in sequence s.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">short</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.S3">
<span class="sig-name descname"><span class="pre">S3</span></span><a class="headerlink" href="#RNA.fold_compound.S3" title="Permalink to this definition"></a></dt>
<dd><p>Sl[s][i] holds next base 3’ of i in sequence s.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">short</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.Ss">
<span class="sig-name descname"><span class="pre">Ss</span></span><a class="headerlink" href="#RNA.fold_compound.Ss" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.a2s">
<span class="sig-name descname"><span class="pre">a2s</span></span><a class="headerlink" href="#RNA.fold_compound.a2s" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(list-like(unsigned</span> <span class="pre">int))</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.pscore">
<span class="sig-name descname"><span class="pre">pscore</span></span><a class="headerlink" href="#RNA.fold_compound.pscore" title="Permalink to this definition"></a></dt>
<dd><p>Precomputed array of pair types expressed as pairing scores.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.pscore_local">
<span class="sig-name descname"><span class="pre">pscore_local</span></span><a class="headerlink" href="#RNA.fold_compound.pscore_local" title="Permalink to this definition"></a></dt>
<dd><p>Precomputed array of pair types expressed as pairing scores.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.pscore_pf_compat">
<span class="sig-name descname"><span class="pre">pscore_pf_compat</span></span><a class="headerlink" href="#RNA.fold_compound.pscore_pf_compat" title="Permalink to this definition"></a></dt>
<dd><p>Precomputed array of pair types expressed as pairing scores indexed via iindx.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This attribute will vanish in the future!</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.scs">
<span class="sig-name descname"><span class="pre">scs</span></span><a class="headerlink" href="#RNA.fold_compound.scs" title="Permalink to this definition"></a></dt>
<dd><p>A set of soft constraints (for each sequence in the alignment)</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only available if</p>
<blockquote>
<div><p>type==RNA.FC_TYPE_COMPARATIVE</p>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_sc_t</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.oldAliEn">
<span class="sig-name descname"><span class="pre">oldAliEn</span></span><a class="headerlink" href="#RNA.fold_compound.oldAliEn" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.maxD1">
<span class="sig-name descname"><span class="pre">maxD1</span></span><a class="headerlink" href="#RNA.fold_compound.maxD1" title="Permalink to this definition"></a></dt>
<dd><p>Maximum allowed base pair distance to first reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.maxD2">
<span class="sig-name descname"><span class="pre">maxD2</span></span><a class="headerlink" href="#RNA.fold_compound.maxD2" title="Permalink to this definition"></a></dt>
<dd><p>Maximum allowed base pair distance to second reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.reference_pt1">
<span class="sig-name descname"><span class="pre">reference_pt1</span></span><a class="headerlink" href="#RNA.fold_compound.reference_pt1" title="Permalink to this definition"></a></dt>
<dd><p>A pairtable of the first reference structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.reference_pt2">
<span class="sig-name descname"><span class="pre">reference_pt2</span></span><a class="headerlink" href="#RNA.fold_compound.reference_pt2" title="Permalink to this definition"></a></dt>
<dd><p>A pairtable of the second reference structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.referenceBPs1">
<span class="sig-name descname"><span class="pre">referenceBPs1</span></span><a class="headerlink" href="#RNA.fold_compound.referenceBPs1" title="Permalink to this definition"></a></dt>
<dd><p>Matrix containing number of basepairs of reference structure1 in interval [i,j].</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.referenceBPs2">
<span class="sig-name descname"><span class="pre">referenceBPs2</span></span><a class="headerlink" href="#RNA.fold_compound.referenceBPs2" title="Permalink to this definition"></a></dt>
<dd><p>Matrix containing number of basepairs of reference structure2 in interval [i,j].</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.bpdist">
<span class="sig-name descname"><span class="pre">bpdist</span></span><a class="headerlink" href="#RNA.fold_compound.bpdist" title="Permalink to this definition"></a></dt>
<dd><p>Matrix containing base pair distance of reference structure 1 and 2 on interval [i,j].</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.mm1">
<span class="sig-name descname"><span class="pre">mm1</span></span><a class="headerlink" href="#RNA.fold_compound.mm1" title="Permalink to this definition"></a></dt>
<dd><p>Maximum matching matrix, reference struct 1 disallowed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.mm2">
<span class="sig-name descname"><span class="pre">mm2</span></span><a class="headerlink" href="#RNA.fold_compound.mm2" title="Permalink to this definition"></a></dt>
<dd><p>Maximum matching matrix, reference struct 2 disallowed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(unsigned</span> <span class="pre">int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.window_size">
<span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#RNA.fold_compound.window_size" title="Permalink to this definition"></a></dt>
<dd><p>window size for local folding sliding window approach</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.ptype_local">
<span class="sig-name descname"><span class="pre">ptype_local</span></span><a class="headerlink" href="#RNA.fold_compound.ptype_local" title="Permalink to this definition"></a></dt>
<dd><p>Pair type array (for local folding)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.fold_compound.zscore_data">
<span class="sig-name descname"><span class="pre">zscore_data</span></span><a class="headerlink" href="#RNA.fold_compound.zscore_data" title="Permalink to this definition"></a></dt>
<dd><p>Data structure with settings for z-score computations.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_zsc_dat_t</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&#64;17</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">union</span> <span class="pre">vrna_fc_s::&#64;16</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/fold_compound.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.E_ext_hp_loop">
<span class="sig-name descname"><span class="pre">E_ext_hp_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.E_ext_hp_loop" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate the free energy of an exterior hairpin loop and consider possible hard constraints.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic! The provided RNA.fold_compound() may be of type RNA.FC_TYPE_SINGLE
or RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.E_ext_int_loop">
<span class="sig-name descname"><span class="pre">E_ext_int_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.E_ext_int_loop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.E_hp_loop">
<span class="sig-name descname"><span class="pre">E_hp_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.E_hp_loop" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate the free energy of a hairpin loop and consider hard constraints if they apply.</p>
<p>This function evaluates the free energy of a hairpin loop</p>
<p>In case the base pair is not allowed due to a constraint conflict, this function returns INF.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 5’ nucleotide of the base pair (3’ to evaluate the pair as exterior hairpin loop)</p></li>
<li><p><strong>j</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 3’ nucleotide of the base pair (5’ to evaluate the pair as exterior hairpin loop)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the hairpin loop in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic! The provided RNA.fold_compound() may be of type RNA.FC_TYPE_SINGLE
or RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.E_int_loop">
<span class="sig-name descname"><span class="pre">E_int_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.E_int_loop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.E_stack">
<span class="sig-name descname"><span class="pre">E_stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.E_stack" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.MEA">
<span class="sig-name descname"><span class="pre">MEA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span></span></span><a class="headerlink" href="#RNA.fold_compound.MEA" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">MEA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">gamma</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span></dt>
<dd><p>Compute a MEA (maximum expected accuracy) structure.</p>
<p>The algorithm maximizes the expected accuracy</p>
<div class="math notranslate nohighlight">
\[A(S) = \sum_{(i,j) \in S} 2 \gamma p_{ij} + \sum_{i \notin S} p^u_{i}\]</div>
<p>Higher values of <span class="math notranslate nohighlight">\(\gamma\)</span> result in more base pairs of lower probability and thus higher
sensitivity. Low values of <span class="math notranslate nohighlight">\(\gamma\)</span> result in structures containing only highly likely pairs
(high specificity). The code of the MEA function also demonstrates the use of sparse dynamic
programming scheme to reduce the time and memory complexity of folding.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>RNA.fold_compound.pf() must be executed on input parameter <cite>fc</cite></p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>MEA`(gamma = 1.) to objects of type
`fold_compound</cite>. Note, that it returns the MEA structure and MEA value as a tuple
(MEA_structure, MEA). See, e.g.  <a class="reference internal" href="#RNA.fold_compound.MEA" title="RNA.fold_compound.MEA"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.MEA()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gamma</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The weighting factor for base pairs vs. unpaired nucleotides</p></li>
<li><p><strong>mea</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – A pointer to a variable where the MEA value will be written to</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An MEA structure (or NULL on any error)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.add_auxdata">
<span class="sig-name descname"><span class="pre">add_auxdata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFuncOrNone=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.add_auxdata" title="Permalink to this definition"></a></dt>
<dd><p>Add auxiliary data to the RNA.fold_compound().</p>
<p>This function allows one to bind arbitrary data to a RNA.fold_compound() which may later on be used
by one of the callback functions, e.g. RNA.recursion_status(). To allow for proper cleanup of the
memory occupied by this auxiliary data, the user may also provide a pointer to a cleanup function
that free’s the corresponding memory. This function will be called automatically when the
RNA.fold_compound() is free’d with RNA.fold_compound_free().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code>) – A pointer to an arbitrary data structure</p></li>
<li><p><strong>f</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.auxdata_free</span></code>) – A pointer to function that free’s memory occupied by the arbitrary data (May be NULL)</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.auxdata_free</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before attaching the arbitrary data pointer, this function will call the RNA.auxdata_free() on
any pre-existing data that is already attached.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.add_callback">
<span class="sig-name descname"><span class="pre">add_callback</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.add_callback" title="Permalink to this definition"></a></dt>
<dd><p>Add a recursion status callback to the RNA.fold_compound().</p>
<p>Binding a recursion status callback function to a RNA.fold_compound() allows one to perform
arbitrary operations just before, or after an actual recursive computations, e.g. MFE prediction, is
performed by the RNAlib. The callback function will be provided with a pointer to its
RNA.fold_compound(), and a status message. Hence, it has complete access to all variables that
incluence the recursive computations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>f</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.recursion_status</span></code>) – The pointer to the recursion status callback function</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.recursion_status</span></code>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.STATUS_MFE_PRE</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.STATUS_MFE_POST</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.STATUS_PF_PRE</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.STATUS_PF_POST</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.backtrack">
<span class="sig-name descname"><span class="pre">backtrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">length</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span></span></span><a class="headerlink" href="#RNA.fold_compound.backtrack" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">backtrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span></dt>
<dd><p>Backtrack an MFE (sub)structure.</p>
<p>This function allows one to backtrack the MFE structure for a (sub)sequence</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>Requires pre-filled MFE dynamic programming matrices, i.e. one has to call</p>
</dd>
<dt>RNA.fold_compound.mfe() prior to</dt><dd><p>calling this function</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>backtrack()</cite> to objects of type <cite>fold_compound</cite>.
The parameter <cite>length</cite> defaults to the total length of the RNA sequence and may be omitted. The
parameter <cite>structure</cite> is returned along with the MFE und must not be provided. See e.g.
<a class="reference internal" href="#RNA.fold_compound.backtrack" title="RNA.fold_compound.backtrack"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.backtrack()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>length</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The length of the subsequence, starting from the 5’ end</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the secondary structure in dot-bracket notation will be
written to. (Must have size of at least $p length + 1)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The minimum free energy (MFE) for the specified <cite>length</cite> in kcal/mol and a corresponding
secondary structure in dot-bracket notation (stored in <cite>structure</cite>)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.pbacktrack5" title="RNA.fold_compound.pbacktrack5"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.pbacktrack5</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>On error, the function returns INF / 100. and stores the empty string in <cite>structure</cite>.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.bpp">
<span class="sig-name descname"><span class="pre">bpp</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.bpp" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.centroid">
<span class="sig-name descname"><span class="pre">centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.centroid" title="Permalink to this definition"></a></dt>
<dd><p>Get the centroid structure of the ensemble.</p>
<p>The centroid is the structure with the minimal average distance to all other structures
<span class="math notranslate nohighlight">\(&lt;d(S)&gt; = \sum_{(i,j) \in S} (1-p_{ij}) + \sum_{(i,j) \notin S} p_{ij}\)</span>
Thus, the centroid is simply the structure containing all pairs with <span class="math notranslate nohighlight">\(p_{i}j&gt;0.5\)</span> The distance
of the centroid to the ensemble is written to the memory adressed by <em>dist</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dist</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – A pointer to the distance variable where the centroid distance will be written to</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The centroid structure of the ensemble in dot-bracket notation (<cite>NULL</cite> on error)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.commands_apply">
<span class="sig-name descname"><span class="pre">commands_apply</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span> <span class="pre">commands</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.commands_apply" title="Permalink to this definition"></a></dt>
<dd><p>Apply a list of commands to a RNA.fold_compound().</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>commands_apply()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.commands_apply" title="RNA.fold_compound.commands_apply"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.commands_apply()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>commands</strong> (<a class="reference internal" href="#RNA.cmd" title="RNA.cmd"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.cmd()</span></code></a>) – The commands to apply</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to limit the type of commands read from the file</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of commands successfully applied</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.constraints_add">
<span class="sig-name descname"><span class="pre">constraints_add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">constraint</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.constraints_add" title="Permalink to this definition"></a></dt>
<dd><p>Add constraints to a RNA.fold_compound() data structure.</p>
<p>Use this function to add/update the hard/soft constraints The function allows for passing a string
‘constraint’ that can either be a filename that points to a constraints definition file or it may be
a pseudo dot-bracket notation indicating hard constraints. For the latter, the user has to pass the
RNA.CONSTRAINT_DB option. Also, the user has to specify, which characters are allowed to be
interpreted as constraints by passing the corresponding options via the third parameter.</p>
<p>The following is an example for adding hard constraints given in pseudo dot-bracket notation. Here,
<cite>fc</cite> is the RNA.fold_compound() object, <cite>structure</cite> is a char array with the hard constraint in
dot-bracket notation, and <cite>enforceConstraints</cite> is a flag indicating whether or not constraints for
base pairs should be enforced instead of just doing a removal of base pair that conflict with the
constraint.</p>
<blockquote>
<div><p>In constrat to the above, constraints may also be read from file:</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>constraint</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A string with either the filename of the constraint definitions or a pseudo dot-bracket notation
of the hard constraint. May be NULL.</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The option flags</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.hc_add_from_db" title="RNA.fold_compound.hc_add_from_db"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_from_db</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_add_up" title="RNA.fold_compound.hc_add_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_up</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.hc_add_up_batch</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.hc_add_bp_unspecific</span></code>, <a class="reference internal" href="#RNA.fold_compound.hc_add_bp" title="RNA.fold_compound.hc_add_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_init" title="RNA.fold_compound.hc_init"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_init</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_set_up" title="RNA.fold_compound.sc_set_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_set_bp" title="RNA.fold_compound.sc_set_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_deigan" title="RNA.fold_compound.sc_add_SHAPE_deigan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_deigan</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_zarringhalam" title="RNA.fold_compound.sc_add_SHAPE_zarringhalam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_zarringhalam</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.hc_free</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_free</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_PIPE</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_DOT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_X</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_ANG_BRACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_RND_BRACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_INTRAMOL</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_INTERMOL</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_GQUAD</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.db_from_probs">
<span class="sig-name descname"><span class="pre">db_from_probs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.db_from_probs" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.ensemble_defect">
<span class="sig-name descname"><span class="pre">ensemble_defect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.ensemble_defect" title="Permalink to this definition"></a></dt>
<dd><p>Compute the Ensemble Defect for a given target structure.</p>
<p>This is a wrapper around <strong>RNA.ensemble_defect_pt()</strong>. Given a target structure <span class="math notranslate nohighlight">\(s\)</span>, compute
the average dissimilarity of a randomly drawn structure from the ensemble, i.e.:</p>
<div class="math notranslate nohighlight">
\[ED(s) = 1 - \frac{1}{n} \sum_{ij, (i,j) \in s} p_{ij} - \frac{1}{n} \sum_{i}(1 - s_{i})q_{i}\]</div>
<p>with sequence length <span class="math notranslate nohighlight">\(n\)</span>, the probability <span class="math notranslate nohighlight">\(p_{ij}\)</span> of a base pair <span class="math notranslate nohighlight">\((i,j)\)</span>, the
probability <span class="math notranslate nohighlight">\(q_{i} = 1 - \sum_{j} p_{ij}\)</span> of nucleotide <span class="math notranslate nohighlight">\(i\)</span> being unpaired, and the
indicator variable <span class="math notranslate nohighlight">\(s_{i} = 1\)</span> if <span class="math notranslate nohighlight">\(\exists (i,j) \in s\)</span>, and <span class="math notranslate nohighlight">\(s_{i} = 0\)</span>
otherwise.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>The RNA.fold_compound() input parameter <cite>fc</cite> must contain a valid base pair probability matrix.
This means that partition function and base pair probabilities must have been computed using
<cite>fc</cite> before execution of this function!</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>ensemble_defect()</cite> to objects of type <cite>fold_compound</cite>. Note
that the SWIG wrapper takes a structure in dot-bracket notation and converts it into a pair
table using RNA.ptable_from_string(). The resulting pair table is then internally passed to
RNA.ensemble_defect_pt(). To control which kind of matching brackets will be used during
conversion, the optional argument <cite>options</cite> can be used. See also the description of
RNA.ptable_from_string() for available options. (default: <cite>RNA.BRACKETS_RND</cite>). See, e.g.
<a class="reference internal" href="#RNA.fold_compound.ensemble_defect" title="RNA.fold_compound.ensemble_defect"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.ensemble_defect()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A target structure in dot-bracket notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ensemble defect with respect to the target structure, or -1. upon failure, e.g. pre-
conditions are not met</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.pf" title="RNA.fold_compound.pf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.pf</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pairing_probs</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ensemble_defect_pt</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_covar_structure">
<span class="sig-name descname"><span class="pre">eval_covar_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_covar_structure" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.</p>
<p>Consensus structure prediction is driven by covariance scores of base pairs in rows of the provided
alignment. This function allows one to retrieve the total amount of this covariance pseudo energy
scores. The RNA.fold_compound() does not need to contain any DP matrices, but requires all most
basic init values as one would get from a call like this:</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_covar_structure()</cite> to objects of type <cite>fold_compound</cite>.
See, e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_covar_structure" title="RNA.fold_compound.eval_covar_structure"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_covar_structure()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary (consensus) structure in dot-bracket notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The covariance pseudo energy score of the input structure given the input sequence alignment in
kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound_comparative</span></code>, <a class="reference internal" href="#RNA.fold_compound.eval_structure" title="RNA.fold_compound.eval_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Accepts RNA.fold_compound() of type RNA.FC_TYPE_COMPARATIVE only!</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_ext_hp_loop">
<span class="sig-name descname"><span class="pre">eval_ext_hp_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_ext_hp_loop" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate free energy of an exterior hairpin loop.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_ext_stem">
<span class="sig-name descname"><span class="pre">eval_ext_stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_ext_stem" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate the free energy of a base pair in the exterior loop.</p>
<p>Evalue the free energy of a base pair connecting two nucleotides in the exterior loop and take hard
constraints into account.</p>
<p>Typically, this is simply dangling end contributions of the adjacent nucleotides, potentially a
terminal A-U mismatch penalty, and maybe some generic soft constraint contribution for that
decomposition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – 5’ position of the base pair</p></li>
<li><p><strong>j</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – 3’ position of the base pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Free energy contribution that arises when this pair is formed in the exterior loop</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For dangles == 1 || 3 this function also evaluates the three additional pairs (i + 1, j), (i, j -
1), and (i + 1, j - 1) and returns the minimum for all four possibilities in total.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_hp_loop">
<span class="sig-name descname"><span class="pre">eval_hp_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">j</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.eval_hp_loop" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate free energy of a hairpin loop.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_hp_loop()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_hp_loop" title="RNA.fold_compound.eval_hp_loop"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_hp_loop()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – 5’-position of the base pair</p></li>
<li><p><strong>j</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – 3’-position of the base pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Free energy of the hairpin loop closed by <span class="math notranslate nohighlight">\((i,j)\)</span> in deka-kal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic! The provided RNA.fold_compound() may be of type RNA.FC_TYPE_SINGLE
or RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_int_loop">
<span class="sig-name descname"><span class="pre">eval_int_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">k</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">l</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.eval_int_loop" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate the free energy contribution of an interior loop with delimiting base pairs <span class="math notranslate nohighlight">\((i,j)\)</span>
and <span class="math notranslate nohighlight">\((k,l)\)</span>.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_int_loop()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_int_loop" title="RNA.fold_compound.eval_int_loop"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_int_loop()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic, i.e. it accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE as
well as RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_loop_pt">
<span class="sig-name descname"><span class="pre">eval_loop_pt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_loop_pt" title="Permalink to this definition"></a></dt>
<dd><p>Calculate energy of a loop.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_loop_pt()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_loop_pt" title="RNA.fold_compound.eval_loop_pt"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_loop_pt()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – position of covering base pair</p></li>
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – the pair table of the secondary structure</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>free energy of the loop in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_move">
<span class="sig-name descname"><span class="pre">eval_move</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_move" title="Permalink to this definition"></a></dt>
<dd><p>Calculate energy of a move (closing or opening of a base pair)</p>
<p>If the parameters m1 and m2 are negative, it is deletion (opening) of a base pair, otherwise it is
insertion (opening).</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_move()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.eval_move" title="RNA.fold_compound.eval_move"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_move()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – secondary structure in dot-bracket notation</p></li>
<li><p><strong>m1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – first coordinate of base pair</p></li>
<li><p><strong>m2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – second coordinate of base pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>energy change of the move in kcal/mol (INF / 100. upon any error)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_move_pt" title="RNA.fold_compound.eval_move_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_move_pt</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_move_pt">
<span class="sig-name descname"><span class="pre">eval_move_pt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_move_pt" title="Permalink to this definition"></a></dt>
<dd><p>Calculate energy of a move (closing or opening of a base pair)</p>
<p>If the parameters m1 and m2 are negative, it is deletion (opening) of a base pair, otherwise it is
insertion (opening).</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_move_pt()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_move_pt" title="RNA.fold_compound.eval_move_pt"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_move_pt()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – the pair table of the secondary structure</p></li>
<li><p><strong>m1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – first coordinate of base pair</p></li>
<li><p><strong>m2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – second coordinate of base pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>energy change of the move in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_move" title="RNA.fold_compound.eval_move"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_move</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_structure">
<span class="sig-name descname"><span class="pre">eval_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_structure" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA.</p>
<p>This function allows for energy evaluation of a given pair of structure and sequence (alignment).
Model details, energy parameters, and possibly soft constraints are used as provided via the
parameter ‘fc’. The RNA.fold_compound() does not need to contain any DP matrices, but requires all
most basic init values as one would get from a call like this:</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_structure()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_structure" title="RNA.fold_compound.eval_structure"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_structure_pt" title="RNA.fold_compound.eval_structure_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_verbose" title="RNA.fold_compound.eval_structure_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_verbose</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt_verbose" title="RNA.fold_compound.eval_structure_pt_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt_verbose</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound_comparative</span></code>, <a class="reference internal" href="#RNA.fold_compound.eval_covar_structure" title="RNA.fold_compound.eval_covar_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_covar_structure</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE and RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_structure_pt">
<span class="sig-name descname"><span class="pre">eval_structure_pt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_structure_pt" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA.</p>
<p>This function allows for energy evaluation of a given sequence/structure pair where the structure is
provided in pair_table format as obtained from RNA.ptable(). Model details, energy parameters, and
possibly soft constraints are used as provided via the parameter ‘fc’. The fold_compound does not
need to contain any DP matrices, but all the most basic init values as one would get from a call
like this:</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_structure_pt()</cite> to objects of type <cite>fold_compound</cite>.
See, e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt" title="RNA.fold_compound.eval_structure_pt"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – Secondary structure as pair_table</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure" title="RNA.fold_compound.eval_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt_verbose" title="RNA.fold_compound.eval_structure_pt_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt_verbose</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_structure_pt_verbose">
<span class="sig-name descname"><span class="pre">eval_structure_pt_verbose</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_structure_pt_verbose" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA.</p>
<p>This function is a simplyfied version of RNA.eval_structure_simple_v() that uses the <em>default</em>
verbosity level.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_structure_pt_verbose()</cite> to objects of type
<cite>fold_compound</cite>. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt_verbose" title="RNA.fold_compound.eval_structure_pt_verbose"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt_verbose()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – Secondary structure as pair_table</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle where this function should print to (may be NULL).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.eval_structure_pt_v</span></code>, <a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt" title="RNA.fold_compound.eval_structure_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_verbose" title="RNA.fold_compound.eval_structure_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_verbose</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.eval_structure_verbose">
<span class="sig-name descname"><span class="pre">eval_structure_verbose</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nullfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.eval_structure_verbose" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the free energy of an already folded RNA and print contributions on a per-loop base.</p>
<p>This function is a simplyfied version of RNA.eval_structure_v() that uses the <em>default</em> verbosity
level.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>eval_structure_verbose()</cite> to objects of type
<cite>fold_compound</cite>. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.eval_structure_verbose" title="RNA.fold_compound.eval_structure_verbose"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_verbose()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket notation</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – A file handle where this function should print to (may be NULL).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The free energy of the input structure given the input sequence in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_structure_pt" title="RNA.fold_compound.eval_structure_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_verbose" title="RNA.fold_compound.eval_structure_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_verbose</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.eval_structure_pt_verbose" title="RNA.fold_compound.eval_structure_pt_verbose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_structure_pt_verbose</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_E_ext_stem">
<span class="sig-name descname"><span class="pre">exp_E_ext_stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.exp_E_ext_stem" title="Permalink to this definition"></a></dt>
<dd><p>Evaluate a stem branching off the exterior loop (Boltzmann factor version)</p>
<p>Given a base pair <span class="math notranslate nohighlight">\((i,j)\)</span> encoded by <em>type</em>, compute the energy contribution including
dangling-end/terminal-mismatch contributions. Instead of returning the energy contribution per-se,
this function returns the corresponding Boltzmann factor. If either of the adjacent nucleotides
<span class="math notranslate nohighlight">\((i - 1)\)</span> and <span class="math notranslate nohighlight">\((j+1)\)</span> must not contribute stacking energy, the corresponding encoding
must be <span class="math notranslate nohighlight">\(-1\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The base pair encoding</p></li>
<li><p><strong>n5d</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The encoded nucleotide directly adjacent at the 5’ side of the base pair (may be -1)</p></li>
<li><p><strong>n3d</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The encoded nucleotide directly adjacent at the 3’ side of the base pair (may be -1)</p></li>
<li><p><strong>p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.exp_param()</span> <span class="pre">*</span></code>) – The pre-computed energy parameters (Boltzmann factor version)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The Boltzmann weighted energy contribution of the introduced exterior-loop stem</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.E_ext_stem" title="RNA.E_ext_stem"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.E_ext_stem</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_E_hp_loop">
<span class="sig-name descname"><span class="pre">exp_E_hp_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.exp_E_hp_loop" title="Permalink to this definition"></a></dt>
<dd><p>High-Level function for hairpin loop energy evaluation (partition function variant)</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.E_hp_loop" title="RNA.fold_compound.E_hp_loop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.E_hp_loop</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic! The provided RNA.fold_compound() may be of type RNA.FC_TYPE_SINGLE
or RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_E_int_loop">
<span class="sig-name descname"><span class="pre">exp_E_int_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.exp_E_int_loop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_E_interior_loop">
<span class="sig-name descname"><span class="pre">exp_E_interior_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.exp_E_interior_loop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id67">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exp_matrices</span></span><a class="headerlink" href="#id67" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id68">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">exp_params</span></span><a class="headerlink" href="#id68" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_params_rescale">
<span class="sig-name descname"><span class="pre">exp_params_rescale</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.exp_params_rescale" title="Permalink to this definition"></a></dt>
<dd><p>Rescale Boltzmann factors for partition function computations.</p>
<p>This function may be used to (automatically) rescale the Boltzmann factors used in partition
function computations. Since partition functions over subsequences can easily become extremely
large, the RNAlib internally rescales them to avoid numerical over- and/or underflow. Therefore, a
proper scaling factor <span class="math notranslate nohighlight">\(s\)</span> needs to be chosen that in turn is then used to normalize the
corresponding partition functions <span class="math notranslate nohighlight">\(\hat{q}[i,j] = q[i,j] / s^{(j-i+1)}\)</span>.</p>
<p>This function provides two ways to automatically adjust the scaling factor.</p>
<ol class="arabic simple">
<li><p>Automatic guess</p></li>
<li><p>Automatic adjustment according to MFE</p></li>
</ol>
<p>Passing <cite>NULL</cite> as second parameter activates the <em>automatic guess mode</em>. Here, the scaling factor is
recomputed according to a mean free energy of <cite>184.3*length</cite> cal for random sequences.
On the other hand, if the MFE for a sequence is known, it can be used to recompute a more robust
scaling factor, since it represents the lowest free energy of the entire ensemble of structures,
i.e. the highest Boltzmann factor. To activate this second mode of <em>automatic adjustment according
to MFE</em>, a pointer to the MFE value needs to be passed as second argument. This value is then taken
to compute the scaling factor as <span class="math notranslate nohighlight">\(s = exp((sfact * MFE) / kT / length )\)</span>, where sfact is an
additional scaling weight located in the RNA.md() data structure of <cite>exp_params</cite> in <cite>fc</cite>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This recomputation only takes place if the <cite>pf_scale</cite> attribute of the <cite>exp_params</cite> data structure
contained in <cite>fc</cite> has a value below <cite>1.0</cite>.</p>
<p>The computed scaling factor <span class="math notranslate nohighlight">\(s\)</span> will be stored as <cite>pf_scale</cite> attribute of the <cite>exp_params</cite>
data structure in <cite>fc</cite>.</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached to RNA.fc() objects as overloaded <cite>exp_params_rescale()</cite> method.</p>
<p>When no parameter is passed to this method, the resulting action is the same as passing <cite>NULL</cite>
as second parameter to RNA.fold_compound.exp_params_rescale(), i.e. default scaling of the
partition
function. Passing an energy in kcal/mol, e.g. as retrieved by a previous call to the <cite>mfe()</cite>
method, instructs all subsequent calls to scale the partition function accordingly. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.exp_params_rescale" title="RNA.fold_compound.exp_params_rescale"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_rescale()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mfe</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – A pointer to the MFE (in kcal/mol) or NULL</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.exp_params_subst" title="RNA.fold_compound.exp_params_subst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_subst</span></code></a>, <a class="reference internal" href="#RNA.md" title="RNA.md"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.md</span></code></a>, <a class="reference internal" href="#RNA.exp_param" title="RNA.exp_param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.exp_param</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_params_reset">
<span class="sig-name descname"><span class="pre">exp_params_reset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">md</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.exp_params_reset" title="Permalink to this definition"></a></dt>
<dd><p>Reset Boltzmann factors for partition function computations within a RNA.fold_compound() according
to provided, or default model details.</p>
<p>This function allows one to rescale Boltzmann factors for subsequent partition function computations
according to a set of model details, e.g. temperature values. To do so, the caller provides either a
pointer to a set of model details to be used for rescaling, or NULL if global default setting should
be used.</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached to RNA.fc() objects as overloaded <cite>exp_params_reset()</cite> method.</p>
<p>When no parameter is passed to this method, the resulting action is the same as passing <cite>NULL</cite>
as second parameter to RNA.fold_compound.exp_params_reset(), i.e. global default model settings
are used.
Passing an object of type RNA.md() resets the fold compound according to the specifications
stored within the RNA.md() object. See, e.g.  <a class="reference internal" href="#RNA.fold_compound.exp_params_reset" title="RNA.fold_compound.exp_params_reset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_reset()</span></code></a>
in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – A pointer to the new model details (or NULL for reset to defaults)</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.params_reset" title="RNA.fold_compound.params_reset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.params_reset</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.exp_params_subst" title="RNA.fold_compound.exp_params_subst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_subst</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.exp_params_rescale" title="RNA.fold_compound.exp_params_rescale"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_rescale</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.exp_params_subst">
<span class="sig-name descname"><span class="pre">exp_params_subst</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">par</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.exp_params_subst" title="Permalink to this definition"></a></dt>
<dd><p>Update the energy parameters for subsequent partition function computations.</p>
<p>This function can be used to properly assign new energy parameters for partition function
computations to a RNA.fold_compound(). For this purpose, the data of the provided pointer <cite>params</cite>
will be copied into <cite>fc</cite> and a recomputation of the partition function scaling factor is issued, if
the <cite>pf_scale</cite> attribute of <cite>params</cite> is less than <cite>1.0</cite>.</p>
<p>Passing NULL as second argument leads to a reset of the energy parameters within fc to their default
values</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached to RNA.fc() objects as overloaded <cite>exp_params_subst()</cite> method.</p>
<p>When no parameter is passed, the resulting action is the same as passing <cite>NULL</cite> as second
parameter to RNA.fold_compound.exp_params_subst(), i.e. resetting the parameters to the global
defaults. See,
e.g.  <a class="reference internal" href="#RNA.fold_compound.exp_params_subst" title="RNA.fold_compound.exp_params_subst"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_subst()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>params</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.exp_param()</span> <span class="pre">*</span></code>) – A pointer to the new energy parameters</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.exp_params_reset" title="RNA.fold_compound.exp_params_reset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_reset</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.exp_params_rescale" title="RNA.fold_compound.exp_params_rescale"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_rescale</span></code></a>, <a class="reference internal" href="#RNA.exp_param" title="RNA.exp_param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.exp_param</span></code></a>, <a class="reference internal" href="#RNA.md" title="RNA.md"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.md</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.exp_params</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.file_commands_apply">
<span class="sig-name descname"><span class="pre">file_commands_apply</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.file_commands_apply" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id69">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">hc</span></span><a class="headerlink" href="#id69" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.hc_add_bp">
<span class="sig-name descname"><span class="pre">hc_add_bp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">option=VRNA_CONSTRAINT_CONTEXT_ALL_LOOPS</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.hc_add_bp" title="Permalink to this definition"></a></dt>
<dd><p>Favorize/Enforce a certain base pair (i,j)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 5’ located nucleotide position of the base pair (1-based)</p></li>
<li><p><strong>j</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 3’ located nucleotide position of the base pair (1-based)</p></li>
<li><p><strong>option</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">char</span></code>) – The options flag indicating how/where to store the hard constraints</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.hc_add_bp_nonspecific" title="RNA.fold_compound.hc_add_bp_nonspecific"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_bp_nonspecific</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_add_up" title="RNA.fold_compound.hc_add_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_init" title="RNA.fold_compound.hc_init"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_init</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_EXT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_HP_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_INT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_INT_LOOP_ENC</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_MB_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_MB_LOOP_ENC</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_ENFORCE</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_ALL_LOOPS</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.hc_add_bp_nonspecific">
<span class="sig-name descname"><span class="pre">hc_add_bp_nonspecific</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">option=VRNA_CONSTRAINT_CONTEXT_ALL_LOOPS</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.hc_add_bp_nonspecific" title="Permalink to this definition"></a></dt>
<dd><p>Enforce a nucleotide to be paired (upstream/downstream)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The position that needs to stay unpaired (1-based)</p></li>
<li><p><strong>d</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The direction of base pairing ( <span class="math notranslate nohighlight">\(d &lt; 0\)</span>: pairs upstream, <span class="math notranslate nohighlight">\(d &gt; 0\)</span>: pairs downstream,
<span class="math notranslate nohighlight">\(d == 0\)</span>: no direction)</p></li>
<li><p><strong>option</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">char</span></code>) – The options flag indicating in which loop type context the pairs may appear</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.hc_add_bp" title="RNA.fold_compound.hc_add_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_add_up" title="RNA.fold_compound.hc_add_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_init" title="RNA.fold_compound.hc_init"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_init</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_EXT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_HP_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_INT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_INT_LOOP_ENC</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_MB_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_MB_LOOP_ENC</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_ALL_LOOPS</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.hc_add_from_db">
<span class="sig-name descname"><span class="pre">hc_add_from_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">constraint</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.hc_add_from_db" title="Permalink to this definition"></a></dt>
<dd><p>Add hard constraints from pseudo dot-bracket notation.</p>
<p>This function allows one to apply hard constraints from a pseudo dot-bracket notation. The <cite>options</cite>
parameter controls, which characters are recognized by the parser. Use the
RNA.CONSTRAINT_DB_DEFAULT convenience macro, if you want to allow all known characters</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>hc_add_from_db()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.hc_add_from_db" title="RNA.fold_compound.hc_add_from_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_from_db()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>constraint</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pseudo dot-bracket notation of the hard constraint.</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The option flags</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_PIPE</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_DOT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_X</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_ANG_BRACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_RND_BRACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_INTRAMOL</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_INTERMOL</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_DB_GQUAD</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.hc_add_up">
<span class="sig-name descname"><span class="pre">hc_add_up</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">option=VRNA_CONSTRAINT_CONTEXT_ALL_LOOPS</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.hc_add_up" title="Permalink to this definition"></a></dt>
<dd><p>Make a certain nucleotide unpaired.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The position that needs to stay unpaired (1-based)</p></li>
<li><p><strong>option</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">char</span></code>) – The options flag indicating how/where to store the hard constraints</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.hc_add_bp" title="RNA.fold_compound.hc_add_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_add_bp_nonspecific" title="RNA.fold_compound.hc_add_bp_nonspecific"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_bp_nonspecific</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_init" title="RNA.fold_compound.hc_init"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_init</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_EXT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_HP_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_INT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_MB_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_ALL_LOOPS</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.hc_init">
<span class="sig-name descname"><span class="pre">hc_init</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.hc_init" title="Permalink to this definition"></a></dt>
<dd><p>Initialize/Reset hard constraints to default values.</p>
<p>This function resets the hard constraints to their default values, i.e. all positions may be
unpaired in all contexts, and base pairs are allowed in all contexts, if they resemble canonical
pairs. Previously set hard constraints will be removed before initialization.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>hc_init()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.hc_init" title="RNA.fold_compound.hc_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_init()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.hc_add_bp" title="RNA.fold_compound.hc_add_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_add_bp_nonspecific" title="RNA.fold_compound.hc_add_bp_nonspecific"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_bp_nonspecific</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.hc_add_up" title="RNA.fold_compound.hc_add_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_add_up</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.heat_capacity">
<span class="sig-name descname"><span class="pre">heat_capacity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">float</span> <span class="pre">T_min=0.</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">float</span> <span class="pre">T_max=100.</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">float</span> <span class="pre">T_increment=1.</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">mpoints=2</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.HeatCapacityVector" title="RNA.HeatCapacityVector"><span class="pre">HeatCapacityVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.heat_capacity" title="Permalink to this definition"></a></dt>
<dd><p>Compute the specific heat for an RNA.</p>
<p>This function computes an RNAs specific heat in a given temperature range from the partition
function by numeric differentiation. The result is returned as a list of pairs of temperature in C
and specific heat in Kcal/(Mol*K).</p>
<p>Users can specify the temperature range for the computation from <cite>T_min</cite> to <cite>T_max</cite>, as well as the
increment step size <cite>T_increment</cite>. The latter also determines how many times the partition function
is computed. Finally, the parameter <cite>mpoints</cite> determines how smooth the curve should be. The
algorithm itself fits a parabola to <span class="math notranslate nohighlight">\(2 \cdot mpoints + 1\)</span> data points to calculate 2nd
derivatives. Increasing this parameter produces a smoother curve.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>heat_capacity()</cite> to objects of type
<cite>fold_compound</cite>. If the optional function arguments <cite>T_min</cite>, <cite>T_max</cite>, <cite>T_increment</cite>, and
<cite>mpoints</cite> are omitted, they default to 0.0, 100.0, 1.0 and 2, respectively. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.heat_capacity" title="RNA.fold_compound.heat_capacity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.heat_capacity()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>T_min</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Lowest temperature in C</p></li>
<li><p><strong>T_max</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Highest temperature in C</p></li>
<li><p><strong>T_increment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Stepsize for temperature incrementation in C (a reasonable choice might be 1C)</p></li>
<li><p><strong>mpoints</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The number of interpolation points to calculate 2nd derivative (a reasonable choice might be 2,
min: 1, max: 100)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of pairs of temperatures and corresponding heat capacity or <em>NULL</em> upon any failure. The
last entry of the list is indicated by a <strong>temperature</strong> field set to a value smaller than
<cite>T_min</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.heat_capacity()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.heat_capacity_cb" title="RNA.fold_compound.heat_capacity_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.heat_capacity_cb</span></code></a>, <a class="reference internal" href="#RNA.heat_capacity" title="RNA.heat_capacity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.heat_capacity</span></code></a>, <a class="reference internal" href="#RNA.heat_capacity" title="RNA.heat_capacity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.heat_capacity</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.heat_capacity_cb">
<span class="sig-name descname"><span class="pre">heat_capacity_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">float</span> <span class="pre">T_min</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">float</span> <span class="pre">T_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">float</span> <span class="pre">T_increment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">mpoints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.heat_capacity_cb" title="Permalink to this definition"></a></dt>
<dd><p>Compute the specific heat for an RNA (callback variant)</p>
<p>Similar to RNA.fold_compound.heat_capacity(), this function computes an RNAs specific heat in a given temperature
range from the partition function by numeric differentiation. Instead of returning a list of
temperature/specific heat pairs, however, this function returns the individual results through a
callback mechanism. The provided function will be called for each result and passed the
corresponding temperature and specific heat values along with the arbitrary data as provided through
the <cite>data</cite> pointer argument.</p>
<p>Users can specify the temperature range for the computation from <cite>T_min</cite> to <cite>T_max</cite>, as well as the
increment step size <cite>T_increment</cite>. The latter also determines how many times the partition function
is computed. Finally, the parameter <cite>mpoints</cite> determines how smooth the curve should be. The
algorithm itself fits a parabola to <span class="math notranslate nohighlight">\(2 \cdot mpoints + 1\)</span> data points to calculate 2nd
derivatives. Increasing this parameter produces a smoother curve.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>heat_capacity_cb()</cite> to objects of type <cite>fold_compound</cite>.
See, e.g.  <a class="reference internal" href="#RNA.fold_compound.heat_capacity_cb" title="RNA.fold_compound.heat_capacity_cb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.heat_capacity_cb()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>T_min</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Lowest temperature in C</p></li>
<li><p><strong>T_max</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Highest temperature in C</p></li>
<li><p><strong>T_increment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Stepsize for temperature incrementation in C (a reasonable choice might be 1C)</p></li>
<li><p><strong>mpoints</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The number of interpolation points to calculate 2nd derivative (a reasonable choice might be 2,
min: 1, max: 100)</p></li>
<li><p><strong>cb</strong> (<a class="reference internal" href="#RNA.heat_capacity" title="RNA.heat_capacity"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.heat_capacity</span></code></a>) – The user-defined callback function that receives the individual results</p></li>
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code>) – An arbitrary data structure that will be passed to the callback in conjunction with the results</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns 0 upon failure, and non-zero otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.heat_capacity" title="RNA.fold_compound.heat_capacity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.heat_capacity</span></code></a>, <a class="reference internal" href="#RNA.heat_capacity" title="RNA.heat_capacity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.heat_capacity</span></code></a></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id70">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">iindx</span></span><a class="headerlink" href="#id70" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id71">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">jindx</span></span><a class="headerlink" href="#id71" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id72">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#id72" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id73">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrices</span></span><a class="headerlink" href="#id73" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.maxmimum_matching">
<span class="sig-name descname"><span class="pre">maxmimum_matching</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.maxmimum_matching" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.mean_bp_distance">
<span class="sig-name descname"><span class="pre">mean_bp_distance</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.mean_bp_distance" title="Permalink to this definition"></a></dt>
<dd><p>Get the mean base pair distance in the thermodynamic ensemble.</p>
<div class="math notranslate nohighlight">
\[&lt;d&gt; = \sum_{a,b} p_{a} p_{b} d(S_{a},S_{b})\]</div>
<p>this can be computed from the pair probs <span class="math notranslate nohighlight">\(p_{ij}\)</span> as</p>
<div class="math notranslate nohighlight">
\[&lt;d&gt; = \sum_{ij} p_{ij}(1-p_{ij})\]</div>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>mean_bp_distance()</cite> to objects of type <cite>fold_compound</cite>.
See, e.g.  <a class="reference internal" href="#RNA.fold_compound.mean_bp_distance" title="RNA.fold_compound.mean_bp_distance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.mean_bp_distance()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The mean pair distance of the structure ensemble</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.mfe">
<span class="sig-name descname"><span class="pre">mfe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.mfe" title="Permalink to this definition"></a></dt>
<dd><p>Compute minimum free energy and an appropriate secondary structure of an RNA sequence, or RNA
sequence alignment.</p>
<p>Depending on the type of the provided RNA.fold_compound(), this function predicts the MFE for a
single sequence (or connected component of multiple sequences), or an averaged MFE for a sequence
alignment. If backtracking is activated, it also constructs the corresponding secondary structure,
or consensus structure. Therefore, the second parameter, <em>structure</em>, has to point to an allocated
block of memory with a size of at least <span class="math notranslate nohighlight">\(\mathrm{strlen}(\mathrm{sequence})+1\)</span> to store the
backtracked MFE structure. (For consensus structures, this is the length of the alignment + 1. If
<cite>NULL</cite> is passed, no backtracking will be performed.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>mfe()</cite> to objects of type <cite>fold_compound</cite>. The parameter
<cite>structure</cite> is returned along with the MFE und must not be provided. See e.g.
<a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where the secondary structure in dot-bracket notation will be
written to (Maybe NULL)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the minimum free energy (MFE) in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.fold" title="RNA.fold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold</span></code></a>, <a class="reference internal" href="#RNA.circfold" title="RNA.circfold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.circfold</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound_comparative</span></code>, <a class="reference internal" href="#RNA.alifold" title="RNA.alifold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.alifold</span></code></a>, <a class="reference internal" href="#RNA.circalifold" title="RNA.circalifold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.circalifold</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic. It accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE, and
RNA.FC_TYPE_COMPARATIVE.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.mfe_dimer">
<span class="sig-name descname"><span class="pre">mfe_dimer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.mfe_dimer" title="Permalink to this definition"></a></dt>
<dd><p>Compute the minimum free energy of two interacting RNA molecules.</p>
<p>The code is analog to the RNA.fold_compound.mfe() function.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is obsolete since RNA.fold_compound.mfe() can handle complexes multiple sequences</p>
</div>
<dl class="simple">
<dt>since v2.5.0.</dt><dd><p>Use RNA.fold_compound.mfe() for connected component MFE instead and compute MFEs of unconnected</p>
</dd>
<dt>states</dt><dd><p>separately.</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>mfe_dimer()</cite> to objects of type <cite>fold_compound</cite>. The
parameter <cite>structure</cite> is returned along with the MFE und must not be provided. See e.g.
<a class="reference internal" href="#RNA.fold_compound.mfe_dimer" title="RNA.fold_compound.mfe_dimer"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_dimer()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Will hold the barcket dot structure of the dimer molecule</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>minimum free energy of the structure</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.mfe_window">
<span class="sig-name descname"><span class="pre">mfe_window</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nullfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.mfe_window" title="Permalink to this definition"></a></dt>
<dd><p>Local MFE prediction using a sliding window approach.</p>
<p>Computes minimum free energy structures using a sliding window approach, where base pairs may not
span outside the window. In contrast to RNA.fold_compound.mfe(), where a maximum base pair span may be set using
the RNA.md().max_bp_span attribute and one globally optimal structure is predicted, this function
uses a sliding window to retrieve all locally optimal structures within each window. The size of the
sliding window is set in the RNA.md().window_size attribute, prior to the retrieval of the
RNA.fold_compound() using RNA.fold_compound() with option RNA.OPTION_WINDOW</p>
<p>The predicted structures are written on-the-fly, either to stdout, if a NULL pointer is passed as
file parameter, or to the corresponding filehandle.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>mfe_window()</cite> to objects of type
<cite>fold_compound</cite>. The parameter <cite>FILE</cite> has default value of <cite>NULL</cite> and can be omitted. See e.g.
<a class="reference internal" href="#RNA.fold_compound.mfe_window" title="RNA.fold_compound.mfe_window"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – The output file handle where predictions are written to (maybe NULL)</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe_window_zscore" title="RNA.fold_compound.mfe_window_zscore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window_zscore</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a>, <a class="reference internal" href="#RNA.Lfold" title="RNA.Lfold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.Lfold</span></code></a>, <a class="reference internal" href="#RNA.Lfoldz" title="RNA.Lfoldz"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.Lfoldz</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.OPTION_WINDOW</span></code>, <a class="reference internal" href="#RNA.md" title="RNA.md"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.md</span></code></a>, <a class="reference internal" href="#RNA.md" title="RNA.md"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.md</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.mfe_window_cb">
<span class="sig-name descname"><span class="pre">mfe_window_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#RNA.fold_compound.mfe_window_cb" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>mfe_window_cb()</cite> to objects of type
<cite>fold_compound</cite>. The parameter <cite>data</cite> has default value of <cite>NULL</cite> and can be omitted. See e.g.
<a class="reference internal" href="#RNA.fold_compound.mfe_window_cb" title="RNA.fold_compound.mfe_window_cb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window_cb()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.mfe_window_zscore">
<span class="sig-name descname"><span class="pre">mfe_window_zscore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nullfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.mfe_window_zscore" title="Permalink to this definition"></a></dt>
<dd><p>Local MFE prediction using a sliding window approach (with z-score cut-off)</p>
<p>Computes minimum free energy structures using a sliding window approach, where base pairs may not
span outside the window. This function is the z-score version of RNA.fold_compound.mfe_window(), i.e. only
predictions above a certain z-score cut-off value are printed. As for RNA.fold_compound.mfe_window(), the size of
the sliding window is set in the RNA.md().window_size attribute, prior to the retrieval of the
RNA.fold_compound() using RNA.fold_compound() with option RNA.OPTION_WINDOW.</p>
<p>The predicted structures are written on-the-fly, either to stdout, if a NULL pointer is passed as
file parameter, or to the corresponding filehandle.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>mfe_window_zscore()</cite> to objects of type
<cite>fold_compound</cite>. The parameter <cite>FILE</cite> has default value of <cite>NULL</cite> and can be omitted. See e.g.
<a class="reference internal" href="#RNA.fold_compound.mfe_window_zscore" title="RNA.fold_compound.mfe_window_zscore"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window_zscore()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_z</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The minimal z-score for a predicted structure to appear in the output</p></li>
<li><p><strong>file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – The output file handle where predictions are written to (maybe NULL)</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe_window_zscore" title="RNA.fold_compound.mfe_window_zscore"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_window_zscore</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.mfe" title="RNA.fold_compound.mfe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe</span></code></a>, <a class="reference internal" href="#RNA.Lfold" title="RNA.Lfold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.Lfold</span></code></a>, <a class="reference internal" href="#RNA.Lfoldz" title="RNA.Lfoldz"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.Lfoldz</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.OPTION_WINDOW</span></code>, <a class="reference internal" href="#RNA.md" title="RNA.md"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.md</span></code></a>, <a class="reference internal" href="#RNA.md" title="RNA.md"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.md</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.mfe_window_zscore_cb">
<span class="sig-name descname"><span class="pre">mfe_window_zscore_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">min_z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#RNA.fold_compound.mfe_window_zscore_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.move_neighbor_diff">
<span class="sig-name descname"><span class="pre">move_neighbor_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span> <span class="pre">|</span> <span class="pre">8</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.varArrayMove" title="RNA.varArrayMove"><span class="pre">varArrayMove</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.move_neighbor_diff" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">move_neighbor_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">move</span> <span class="pre">move</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=(4|8)</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span></dt>
<dd><p>Apply a move to a secondary structure and indicate which neighbors have changed consequentially.</p>
<p>Similar to RNA.move_neighbor_diff_cb(), this function applies a move to a secondary structure and
reports back the neighbors of the current structure become affected by this move. Instead of
executing a callback for each of the affected neighbors, this function compiles two lists of
neighbor moves, one that is returned and consists of all moves that are novel or may have changed in
energy, and a second, <cite>invalid_moves</cite>, that consists of all the neighbor moves that become invalid,
respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ptable</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – The current structure as pair table</p></li>
<li><p><strong>move</strong> (<a class="reference internal" href="#RNA.move" title="RNA.move"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.move()</span></code></a>) – The move to apply</p></li>
<li><p><strong>invalid_moves</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.move()</span> <span class="pre">**</span></code>) – The address of a move list where the function stores those moves that become invalid</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to modify the behavior of this function, .e.g available move set</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of moves that might have changed in energy or are novel compared to the structure before
application of the move</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.move()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.neighbors">
<span class="sig-name descname"><span class="pre">neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=(4|8)</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.varArrayMove" title="RNA.varArrayMove"><span class="pre">varArrayMove</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.neighbors" title="Permalink to this definition"></a></dt>
<dd><p>Generate neighbors of a secondary structure.</p>
<p>This function allows one to generate all structural neighbors (according to a particular move set)
of an RNA secondary structure. The neighborhood is then returned as a list of transitions / moves
required to transform the current structure into the actual neighbor.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as an overloaded method <cite>neighbors()</cite> to objects of type
<cite>fold_compound</cite>. The optional parameter <cite>options</cite> defaults to RNA.MOVESET_DEFAULT if it is
omitted. See, e.g.  <a class="reference internal" href="#RNA.fold_compound.neighbors" title="RNA.fold_compound.neighbors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.neighbors()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – The pair table representation of the structure</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to modify the behavior of this function, e.g. available move set</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Neighbors as a list of moves / transitions (the last element in the list has both of its fields
set to 0)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.move()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.neighbors_successive</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.move_apply</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_INSERTION</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_DELETION</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_SHIFT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id74">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">params</span></span><a class="headerlink" href="#id74" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.params_reset">
<span class="sig-name descname"><span class="pre">params_reset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">md</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.params_reset" title="Permalink to this definition"></a></dt>
<dd><p>Reset free energy parameters within a RNA.fold_compound() according to provided, or default model
details.</p>
<p>This function allows one to rescale free energy parameters for subsequent structure prediction or
evaluation according to a set of model details, e.g. temperature values. To do so, the caller
provides either a pointer to a set of model details to be used for rescaling, or NULL if global
default setting should be used.</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached to RNA.fc() objects as overloaded <cite>params_reset()</cite> method.</p>
<p>When no parameter is passed to this method, the resulting action is the same as passing <cite>NULL</cite>
as second parameter to RNA.fold_compound.params_reset(), i.e. global default model settings are
used. Passing
an object of type RNA.md() resets the fold compound according to the specifications stored
within the RNA.md() object. See, e.g.  <a class="reference internal" href="#RNA.fold_compound.params_reset" title="RNA.fold_compound.params_reset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.params_reset()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – A pointer to the new model details (or NULL for reset to defaults)</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.exp_params_reset" title="RNA.fold_compound.exp_params_reset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.exp_params_reset</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_subs</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.params_subst">
<span class="sig-name descname"><span class="pre">params_subst</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">par</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.params_subst" title="Permalink to this definition"></a></dt>
<dd><p>Update/Reset energy parameters data structure within a RNA.fold_compound().</p>
<p>Passing NULL as second argument leads to a reset of the energy parameters within fc to their default
values. Otherwise, the energy parameters provided will be copied over into fc.</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached to RNA.fc() objects as overloaded <cite>params_subst()</cite> method.</p>
<p>When no parameter is passed, the resulting action is the same as passing <cite>NULL</cite> as second
parameter to RNA.fold_compound.params_subst(), i.e. resetting the parameters to the global
defaults. See,
e.g.  <a class="reference internal" href="#RNA.fold_compound.params_subst" title="RNA.fold_compound.params_subst"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.params_subst()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>par</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.param()</span> <span class="pre">*</span></code>) – The energy parameters used to substitute those within fc (Maybe NULL)</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.params_reset" title="RNA.fold_compound.params_reset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.params_reset</span></code></a>, <a class="reference internal" href="#RNA.param" title="RNA.param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.param</span></code></a>, <a class="reference internal" href="#RNA.md" title="RNA.md"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.md</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.path">
<span class="sig-name descname"><span class="pre">path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">steps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.MoveVector" title="RNA.MoveVector"><span class="pre">MoveVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.path" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">steps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.MoveVector" title="RNA.MoveVector"><span class="pre">MoveVector</span></a></span></span></dt>
<dd><p>Compute a path, store the final structure, and return a list of transition moves from the start to
the final structure.</p>
<p>This function computes, given a start structure in pair table format, a transition path, updates the
pair table to the final structure of the path. Finally, if not requested otherwise by using the
RNA.PATH_NO_TRANSITION_OUTPUT flag in the <cite>options</cite> field, this function returns a list of
individual transitions that lead from the start to the final structure if requested.</p>
<p>The currently available transition paths are</p>
<ul class="simple">
<li><p>Steepest Descent / Gradient walk (flag: RNA.PATH_STEEPEST_DESCENT)</p></li>
<li><p>Random walk (flag: RNA.PATH_RANDOM)</p></li>
</ul>
<p>The type of transitions must be set through the <cite>options</cite> parameter</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as an overloaded method <cite>path()</cite> to objects of type <cite>fold_compound</cite>.
The optional parameter <cite>options</cite> defaults to RNA.PATH_DEFAULT if it is omitted. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.path" title="RNA.fold_compound.path"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.path()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – The pair table containing the start structure. Used to update to the final structure after
execution of this function</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to modify the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of transition moves (default), or NULL (if options &amp; RNA.PATH_NO_TRANSITION_OUTPUT)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.move()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.path_gradient" title="RNA.fold_compound.path_gradient"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path_gradient</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.path_random" title="RNA.fold_compound.path_random"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path_random</span></code></a>, <a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable_copy</span></code>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PATH_RANDOM</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_SHIFT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PATH_NO_TRANSITION_OUTPUT</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Since the result is written to the input structure you may want to use RNA.ptable_copy() before
calling this function to keep the initial structure</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.path_direct">
<span class="sig-name descname"><span class="pre">path_direct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">maxE=INT_MAX-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_options</span> <span class="pre">options=None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.PathVector" title="RNA.PathVector"><span class="pre">PathVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.path_direct" title="Permalink to this definition"></a></dt>
<dd><p>Determine an optimal direct (re-)folding path between two secondary structures.</p>
<p>This function is similar to RNA.path_direct(), but allows to specify an <em>upper-bound</em> for the
saddle point energy. The underlying algorithms will stop determining an (optimal) (re-)folding path,
if none can be found that has a saddle point below the specified upper-bound threshold <cite>maxE</cite>.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as an overloaded method <cite>path_direct()</cite> to objects of type
<cite>fold_compound</cite>. The optional parameter <cite>maxE</cite> defaults to #INT_MAX - 1 if it is omitted, while
the optional parameter <cite>options</cite> defaults to <cite>NULL</cite>. In case the function did not find a path
with <span class="math notranslate nohighlight">\(E_{saddle} &lt; E_{max}\)</span> it returns an empty list. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.path_direct" title="RNA.fold_compound.path_direct"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.path_direct()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The start structure in dot-bracket notation</p></li>
<li><p><strong>s2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The target structure in dot-bracket notation</p></li>
<li><p><strong>maxE</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Upper bound for the saddle point along the (re-)folding path</p></li>
<li><p><strong>options</strong> (<a class="reference internal" href="#RNA.path_options" title="RNA.path_options"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.path_options()</span></code></a>) – An options data structure that specifies the path heuristic and corresponding settings (maybe
<em>NULL</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An optimal (re-)folding path between the two input structures</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.path()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The argument <cite>maxE</cite> enables one to specify an upper bound, or maximum free energy for the saddle
point between the two input structures. If no path with <span class="math notranslate nohighlight">\(E_{saddle} &lt; E_{max}\)</span> is found, the
function simply returns <em>NULL</em></p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.path_direct" title="RNA.fold_compound.path_direct"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path_direct</span></code></a>, <a class="reference internal" href="#RNA.path_options_findpath" title="RNA.path_options_findpath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.path_options_findpath</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.path_options_free</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.path_free</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.path_findpath">
<span class="sig-name descname"><span class="pre">path_findpath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">width=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">maxE=INT_MAX-1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.PathVector" title="RNA.PathVector"><span class="pre">PathVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.path_findpath" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.path_findpath_saddle">
<span class="sig-name descname"><span class="pre">path_findpath_saddle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">width=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">maxE=INT_MAX</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.path_findpath_saddle" title="Permalink to this definition"></a></dt>
<dd><p>Find energy of a saddle point between 2 structures (search only direct path)</p>
<p>This function uses an inplementation of the <em>findpath</em> algorithm  <span id="id75">[<a class="reference internal" href="bibliography.html#id31" title="Christoph Flamm, Ivo L Hofacker, Sebastian Maurer-Stroh, Peter F Stadler, and Martin Zehl. Design of multistable RNA molecules. RNA, 7(02):254–265, 2001. doi:10.1017/s1355838201000863.">Flamm <em>et al.</em>, 2001</a>]</span>  for near-
optimal direct refolding path prediction.</p>
<p>Model details, and energy parameters are used as provided via the parameter ‘fc’. The
RNA.fold_compound() does not require memory for any DP matrices, but requires all most basic init
values as one would get from a call like this:</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as an overloaded method <cite>path_findpath_saddle()</cite> to objects of type
<cite>fold_compound</cite>. The optional parameter <cite>width</cite> defaults to 1 if it is omitted, while the
optional parameter <cite>maxE</cite> defaults to INF. In case the function did not find a path with
<span class="math notranslate nohighlight">\(E_{saddle} &lt; E_{max}\)</span> the function returns a <cite>NULL</cite> object, i.e. <cite>undef</cite> for Perl and
<cite>None</cite> for Python. See, e.g.  <a class="reference internal" href="#RNA.fold_compound.path_findpath_saddle" title="RNA.fold_compound.path_findpath_saddle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.path_findpath_saddle()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The start structure in dot-bracket notation</p></li>
<li><p><strong>s2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The target structure in dot-bracket notation</p></li>
<li><p><strong>width</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – A number specifying how many strutures are being kept at each step during the search</p></li>
<li><p><strong>maxE</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – An upper bound for the saddle point energy in 10cal/mol</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The saddle energy in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The argument <cite>maxE</cite> ( <span class="math notranslate nohighlight">\(E_{max}\)</span>) enables one to specify an upper bound, or maximum free energy
for the saddle point between the two input structures. If no path with <span class="math notranslate nohighlight">\(E_{saddle} &lt; E_{max}\)</span>
is found, the function simply returns <cite>maxE</cite></p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.path_findpath_saddle</span></code>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.path_findpath</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.path_gradient">
<span class="sig-name descname"><span class="pre">path_gradient</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.MoveVector" title="RNA.MoveVector"><span class="pre">MoveVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.path_gradient" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">path_gradient</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.MoveVector" title="RNA.MoveVector"><span class="pre">MoveVector</span></a></span></span></dt>
<dd><p>Compute a steepest descent / gradient path, store the final structure, and return a list of
transition moves from the start to the final structure.</p>
<p>This function computes, given a start structure in pair table format, a steepest descent path,
updates the pair table to the final structure of the path. Finally, if not requested otherwise by
using the RNA.PATH_NO_TRANSITION_OUTPUT flag in the <cite>options</cite> field, this function returns a list
of individual transitions that lead from the start to the final structure if requested.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as an overloaded method <cite>path_gradient()</cite> to objects of type
<cite>fold_compound</cite>. The optional parameter <cite>options</cite> defaults to RNA.PATH_DEFAULT if it is
omitted. See, e.g.  <a class="reference internal" href="#RNA.fold_compound.path_gradient" title="RNA.fold_compound.path_gradient"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.path_gradient()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – The pair table containing the start structure. Used to update to the final structure after
execution of this function</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to modify the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of transition moves (default), or NULL (if options &amp; RNA.PATH_NO_TRANSITION_OUTPUT)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.move()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.path_random" title="RNA.fold_compound.path_random"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path_random</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.path" title="RNA.fold_compound.path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path</span></code></a>, <a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable_copy</span></code>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_SHIFT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PATH_NO_TRANSITION_OUTPUT</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Since the result is written to the input structure you may want to use RNA.ptable_copy() before
calling this function to keep the initial structure</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.path_random">
<span class="sig-name descname"><span class="pre">path_random</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">steps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.MoveVector" title="RNA.MoveVector"><span class="pre">MoveVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.path_random" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">path_random</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">steps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.MoveVector" title="RNA.MoveVector"><span class="pre">MoveVector</span></a></span></span></dt>
<dd><p>Generate a random walk / path of a given length, store the final structure, and return a list of
transition moves from the start to the final structure.</p>
<p>This function generates, given a start structure in pair table format, a random walk / path, updates
the pair table to the final structure of the path. Finally, if not requested otherwise by using the
RNA.PATH_NO_TRANSITION_OUTPUT flag in the <cite>options</cite> field, this function returns a list of
individual transitions that lead from the start to the final structure if requested.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as an overloaded method <cite>path_gradient()</cite> to objects of type
<cite>fold_compound</cite>. The optional parameter <cite>options</cite> defaults to RNA.PATH_DEFAULT if it is
omitted. See, e.g.  <a class="reference internal" href="#RNA.fold_compound.path_random" title="RNA.fold_compound.path_random"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.path_random()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – The pair table containing the start structure. Used to update to the final structure after
execution of this function</p></li>
<li><p><strong>steps</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The length of the path, i.e. the total number of transitions / moves</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options to modify the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of transition moves (default), or NULL (if options &amp; RNA.PATH_NO_TRANSITION_OUTPUT)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.move()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.path_gradient" title="RNA.fold_compound.path_gradient"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path_gradient</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.path" title="RNA.fold_compound.path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path</span></code></a>, <a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable_copy</span></code>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.MOVESET_SHIFT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PATH_NO_TRANSITION_OUTPUT</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Since the result is written to the input structure you may want to use RNA.ptable_copy() before
calling this function to keep the initial structure</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.pbacktrack">
<span class="sig-name descname"><span class="pre">pbacktrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span></span></span><a class="headerlink" href="#RNA.fold_compound.pbacktrack" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.StringVector" title="RNA.StringVector"><span class="pre">StringVector</span></a></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pbacktrack_mem</span> <span class="pre">nr_memory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.StringVector" title="RNA.StringVector"><span class="pre">StringVector</span></a></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">Py_None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">unsigned</span> <span class="pre">int</span></span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_samples</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – </p></li>
<li><p><strong>PyFunc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">PyObject</span> <span class="pre">*</span></code>) – </p></li>
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">PyObject</span> <span class="pre">*</span></code>) – </p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – </p></li>
<li><p><strong>pbacktrack</strong><strong>(</strong><strong>self</strong> – </p></li>
<li><p><strong>num_samples</strong> – </p></li>
<li><p><strong>PyFunc</strong> – </p></li>
<li><p><strong>data</strong> – </p></li>
<li><p><strong>nr_memory</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_pbacktrack_mem_t</span> <span class="pre">*</span></code>) – </p></li>
<li><p><strong>int</strong> (<em>options=0</em><em>) </em><em>-&gt; unsigned</em>) – </p></li>
<li><p><strong>num_samples</strong> – </p></li>
<li><p><strong>PyFunc</strong> – </p></li>
<li><p><strong>data</strong> – </p></li>
<li><p><strong>nr_memory</strong> – </p></li>
<li><p><strong>options</strong> – </p></li>
</ul>
</dd>
</dl>
<p>Sample a secondary structure from the Boltzmann ensemble according its probability.</p>
<p>Perform a probabilistic (stochastic) backtracing in the partition function DP arrays to obtain a
secondary structure.</p>
<p>The structure <span class="math notranslate nohighlight">\(s\)</span> with free energy <span class="math notranslate nohighlight">\(E(s)\)</span> is picked from the Boltzmann distributed
ensemble according to its probability</p>
<div class="math notranslate nohighlight">
\[p(s) = \frac{exp(-E(s) / kT)}{Z}\]</div>
<p>with partition function <span class="math notranslate nohighlight">\(Z = \sum_{s} exp(-E(s) / kT)\)</span>, Boltzmann constant <span class="math notranslate nohighlight">\(k\)</span> and
thermodynamic temperature <span class="math notranslate nohighlight">\(T\)</span>.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>Unique multiloop decomposition has to be active upon creation of <cite>fc</cite> with RNA.fold_compound()
or similar. This can be done easily by passing RNA.fold_compound() a model details parameter
with RNA.md().uniq_ML = 1.  RNA.fold_compound.pf() has to be called first to fill the partition</p>
</dd>
<dt>function</dt><dd><p>matrices</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>pbacktrack()</cite> to objects of type
<cite>fold_compound</cite>. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.pbacktrack" title="RNA.fold_compound.pbacktrack"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.pbacktrack()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>
and the  <a class="reference internal" href="examples/python.html#boltzmann-sampling"><span class="std std-ref">Boltzmann Sampling</span></a> Python examples .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A sampled secondary structure in dot-bracket notation (or NULL on error)</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.pbacktrack5" title="RNA.fold_compound.pbacktrack5"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.pbacktrack5</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pbacktrack_num</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pbacktrack_cb</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic. It accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE, and
RNA.FC_TYPE_COMPARATIVE.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.pbacktrack5">
<span class="sig-name descname"><span class="pre">pbacktrack5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">length</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span></span></span><a class="headerlink" href="#RNA.fold_compound.pbacktrack5" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.StringVector" title="RNA.StringVector"><span class="pre">StringVector</span></a></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pbacktrack_mem</span> <span class="pre">nr_memory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.StringVector" title="RNA.StringVector"><span class="pre">StringVector</span></a></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">unsigned</span> <span class="pre">int</span></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pbacktrack_mem</span> <span class="pre">nr_memory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">unsigned</span> <span class="pre">int</span></span></span></dt>
<dd><p>Sample a secondary structure of a subsequence from the Boltzmann ensemble according its probability.</p>
<p>Perform a probabilistic (stochastic) backtracing in the partition function DP arrays to obtain a
secondary structure. The parameter <cite>length</cite> specifies the length of the substructure starting from
the 5’ end.</p>
<p>The structure <span class="math notranslate nohighlight">\(s\)</span> with free energy <span class="math notranslate nohighlight">\(E(s)\)</span> is picked from the Boltzmann distributed
ensemble according to its probability</p>
<div class="math notranslate nohighlight">
\[p(s) = \frac{exp(-E(s) / kT)}{Z}\]</div>
<p>with partition function <span class="math notranslate nohighlight">\(Z = \sum_{s} exp(-E(s) / kT)\)</span>, Boltzmann constant <span class="math notranslate nohighlight">\(k\)</span> and
thermodynamic temperature <span class="math notranslate nohighlight">\(T\)</span>.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>Unique multiloop decomposition has to be active upon creation of <cite>fc</cite> with RNA.fold_compound()
or similar. This can be done easily by passing RNA.fold_compound() a model details parameter
with RNA.md().uniq_ML = 1.  RNA.fold_compound.pf() has to be called first to fill the partition</p>
</dd>
<dt>function</dt><dd><p>matrices</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>pbacktrack5()</cite> to objects of type
<cite>fold_compound</cite>. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.pbacktrack5" title="RNA.fold_compound.pbacktrack5"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.pbacktrack5()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>  and the  <a class="reference internal" href="examples/python.html#boltzmann-sampling"><span class="std std-ref">Boltzmann Sampling</span></a> Python examples .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>length</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The length of the subsequence to consider (starting with 5’ end)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A sampled secondary structure in dot-bracket notation (or NULL on error)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pbacktrack5_num</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pbacktrack5_cb</span></code>, <a class="reference internal" href="#RNA.fold_compound.pbacktrack" title="RNA.fold_compound.pbacktrack"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.pbacktrack</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic. It accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE, and
RNA.FC_TYPE_COMPARATIVE.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.pbacktrack_sub">
<span class="sig-name descname"><span class="pre">pbacktrack_sub</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">end</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span></span></span><a class="headerlink" href="#RNA.fold_compound.pbacktrack_sub" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack_sub</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.StringVector" title="RNA.StringVector"><span class="pre">StringVector</span></a></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack_sub</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pbacktrack_mem</span> <span class="pre">nr_memory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.StringVector" title="RNA.StringVector"><span class="pre">StringVector</span></a></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack_sub</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">unsigned</span> <span class="pre">int</span></span></span></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pbacktrack_sub</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">num_samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pbacktrack_mem</span> <span class="pre">nr_memory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">unsigned</span> <span class="pre">int</span></span></span></dt>
<dd><p>Sample a secondary structure of a subsequence from the Boltzmann ensemble according its probability.</p>
<p>Perform a probabilistic (stochastic) backtracing in the partition function DP arrays to obtain a
secondary structure. The parameters <cite>start</cite> and <cite>end</cite> specify the interval <span class="math notranslate nohighlight">\([start:end]\)</span> of
the subsequence with <span class="math notranslate nohighlight">\(1 \leq start &lt; end \leq n\)</span> for sequence length <span class="math notranslate nohighlight">\(n\)</span>, the
structure <span class="math notranslate nohighlight">\(s_{start,end}\)</span> should be drawn from.</p>
<p>The resulting substructure <span class="math notranslate nohighlight">\(s_{start,end}\)</span> with free energy <span class="math notranslate nohighlight">\(E(s_{start, end})\)</span> is
picked from the Boltzmann distributed sub ensemble of all structures within the interval
<span class="math notranslate nohighlight">\([start:end]\)</span> according to its probability</p>
<div class="math notranslate nohighlight">
\[p(s_{start,end}) = \frac{exp(-E(s_{start,end}) / kT)}{Z_{start,end}}\]</div>
<p>with partition function <span class="math notranslate nohighlight">\(Z_{start,end} = \sum_{s_{start,end}} exp(-E(s_{start,end}) / kT)\)</span>,
Boltzmann constant <span class="math notranslate nohighlight">\(k\)</span> and thermodynamic temperature <span class="math notranslate nohighlight">\(T\)</span>.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>Unique multiloop decomposition has to be active upon creation of <cite>fc</cite> with RNA.fold_compound()
or similar. This can be done easily by passing RNA.fold_compound() a model details parameter
with RNA.md().uniq_ML = 1.  RNA.fold_compound.pf() has to be called first to fill the partition</p>
</dd>
<dt>function</dt><dd><p>matrices</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>pbacktrack_sub()</cite> to objects of type
<em>fold_compound</em>. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.pbacktrack_sub" title="RNA.fold_compound.pbacktrack_sub"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.pbacktrack_sub()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>  and the  <a class="reference internal" href="examples/python.html#boltzmann-sampling"><span class="std std-ref">Boltzmann Sampling</span></a> Python examples .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The start of the subsequence to consider, i.e. 5’-end position(1-based)</p></li>
<li><p><strong>end</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The end of the subsequence to consider, i.e. 3’-end position (1-based)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A sampled secondary structure in dot-bracket notation (or NULL on error)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pbacktrack_sub_num</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pbacktrack_sub_cb</span></code>, <a class="reference internal" href="#RNA.fold_compound.pbacktrack" title="RNA.fold_compound.pbacktrack"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.pbacktrack</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic. It accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE, and
RNA.FC_TYPE_COMPARATIVE.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.pf">
<span class="sig-name descname"><span class="pre">pf</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.pf" title="Permalink to this definition"></a></dt>
<dd><p>Compute the partition function <span class="math notranslate nohighlight">\(Q\)</span> for a given RNA sequence, or sequence alignment.</p>
<p>If <em>structure</em> is not a NULL pointer on input, it contains on return a string consisting of the
letters ” . , | { } ( ) ” denoting bases that are essentially unpaired, weakly paired, strongly
paired without preference, weakly upstream (downstream) paired, or strongly up- (down-)stream paired
bases, respectively. If the model’s compute_bpp is set to 0 base pairing probabilities will not be
computed (saving CPU time), otherwise after calculations took place pr will contain the probability
that bases <em>i</em> and <em>j</em> pair.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>pf()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.pf" title="RNA.fold_compound.pf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.pf()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A pointer to the character array where position-wise pairing propensity will be stored. (Maybe
NULL)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ensemble free energy <span class="math notranslate nohighlight">\(G = -RT \cdot \log(Q)\)</span> in kcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a>, <a class="reference internal" href="#RNA.pf_fold" title="RNA.pf_fold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pf_fold</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pf_circfold</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound_comparative</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pf_alifold</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pf_circalifold</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_from_probs</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.exp_params</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.aln_pinfo</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is polymorphic. It accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE, and
RNA.FC_TYPE_COMPARATIVE. Also, this function may return INF / 100. in case of contradicting
constraints or numerical over-/underflow. In the latter case, a corresponding warning will be issued
to <cite>stdout</cite>.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.pf_dimer">
<span class="sig-name descname"><span class="pre">pf_dimer</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.pf_dimer" title="Permalink to this definition"></a></dt>
<dd><p>Calculate partition function and base pair probabilities of nucleic acid/nucleic acid dimers.</p>
<p>This is the cofold partition function folding.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>pf_dimer()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.pf_dimer" title="RNA.fold_compound.pf_dimer"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.pf_dimer()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Will hold the structure or constraints</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>RNA.dimer_pf() structure containing a set of energies needed for concentration computations.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.dimer_pf()</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound" title="RNA.fold_compound"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function may return INF / 100. for the <cite>FA</cite>, <cite>FB</cite>, <cite>FAB</cite>, <cite>F0AB</cite> members of the output data
structure in case of contradicting constraints or numerical over-/underflow. In the latter case, a
corresponding warning will be issued to <cite>stdout</cite>.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.plist_from_probs">
<span class="sig-name descname"><span class="pre">plist_from_probs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">cutoff</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.ElemProbVector" title="RNA.ElemProbVector"><span class="pre">ElemProbVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.plist_from_probs" title="Permalink to this definition"></a></dt>
<dd><p>Create a RNA.ep() from base pair probability matrix.</p>
<p>The probability matrix provided via the RNA.fold_compound() is parsed and all pair probabilities
above the given threshold are used to create an entry in the plist</p>
<p>The end of the plist is marked by sequence positions i as well as j equal to 0. This condition
should be used to stop looping over its entries</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cut_off</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The cutoff value</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A pointer to the plist that is to be created</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.positional_entropy">
<span class="sig-name descname"><span class="pre">positional_entropy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.positional_entropy" title="Permalink to this definition"></a></dt>
<dd><p>Compute a vector of positional entropies.</p>
<p>This function computes the positional entropies from base pair probabilities as</p>
<div class="math notranslate nohighlight">
\[S(i) = - \sum_{j} p_{ij} \log(p_{ij}) - q_{i} \log(q_{i})\]</div>
<p>with unpaired probabilities <span class="math notranslate nohighlight">\(q_{i} = 1 - \sum_{j} p_{ij}\)</span>.</p>
<p>Low entropy regions have little structural flexibility and the reliability of the predicted
structure is high. High entropy implies many structural alternatives. While these alternatives may
be functionally important, they make structure prediction more difficult and thus less reliable.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>This function requires pre-computed base pair probabilities! Thus, RNA.fold_compound.pf() must</p>
</dd>
<dt>be called</dt><dd><p>beforehand.</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>positional_entropy()</cite> to objects of type <cite>fold_compound</cite>.
See, e.g.  <a class="reference internal" href="#RNA.fold_compound.positional_entropy" title="RNA.fold_compound.positional_entropy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.positional_entropy()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A 1-based vector of positional entropies <span class="math notranslate nohighlight">\(S(i)\)</span>. (position 0 contains the sequence length)</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.pr_energy">
<span class="sig-name descname"><span class="pre">pr_energy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">e</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.pr_energy" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>pr_energy()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.pr_energy" title="RNA.fold_compound.pr_energy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.pr_energy()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.pr_structure">
<span class="sig-name descname"><span class="pre">pr_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.pr_structure" title="Permalink to this definition"></a></dt>
<dd><p>Compute the equilibrium probability of a particular secondary structure.</p>
<p>The probability <span class="math notranslate nohighlight">\(p(s)\)</span> of a particular secondary structure <span class="math notranslate nohighlight">\(s\)</span> can be computed as</p>
<div class="math notranslate nohighlight">
\[p(s) = \frac{exp(-\beta E(s)}{Z}\]</div>
<p>from the structures free energy <span class="math notranslate nohighlight">\(E(s)\)</span> and the partition function</p>
<div class="math notranslate nohighlight">
\[Z = \sum_{s} exp(-\beta E(s)),\quad\mathrm{with}\quad\beta = \frac{1}{RT}\]</div>
<p>where <span class="math notranslate nohighlight">\(R\)</span> is the gas constant and <span class="math notranslate nohighlight">\(T\)</span> the thermodynamic temperature.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>The fold compound <cite>fc</cite> must have went through a call to RNA.fold_compound.pf() to fill the</p>
</dd>
<dt>dynamic</dt><dd><p>programming matrices with the corresponding partition function.</p>
</dd>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>pr_structure()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.  <a class="reference internal" href="#RNA.fold_compound.pr_structure" title="RNA.fold_compound.pr_structure"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.pr_structure()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure to compute the probability for in dot-bracket notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The probability of the input structure (range <span class="math notranslate nohighlight">\([0:1]\)</span>)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.probs_window">
<span class="sig-name descname"><span class="pre">probs_window</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">ulength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.probs_window" title="Permalink to this definition"></a></dt>
<dd><p>Compute various equilibrium probabilities under a sliding window approach.</p>
<p>This function applies a sliding window scan for the sequence provided with the argument <cite>fc</cite> and
reports back equilibrium probabilities through the callback function <cite>cb</cite>. The data reported to the
callback depends on the <cite>options</cite> flag.</p>
<p>#### Options:
.. note:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>The parameter `ulength` only affects computation and resulting data if unpaired probability
computations are requested through the `options` flag.

*   RNA.PROBS_WINDOW_BPP - Trigger base pairing probabilities.
*   RNA.PROBS_WINDOW_UP - Trigger unpaired probabilities.
*   RNA.PROBS_WINDOW_UP_SPLIT - Trigger detailed unpaired probabilities split up into different
    loop type contexts.

Options may be OR-ed together
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ulength</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The maximal length of an unpaired segment (only for unpaired probability computations)</p></li>
<li><p><strong>cb</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.probs_window</span></code>) – The callback function which collects the pair probability data for further processing</p></li>
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code>) – Some arbitrary data structure that is passed to the callback <cite>cb</cite></p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Option flags to control the behavior of this function</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>0 on failure, non-zero on success</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.pfl_fold_cb" title="RNA.pfl_fold_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pfl_fold_cb</span></code></a>, <a class="reference internal" href="#RNA.pfl_fold_up_cb" title="RNA.pfl_fold_up_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pfl_fold_up_cb</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.rotational_symmetry_db">
<span class="sig-name descname"><span class="pre">rotational_symmetry_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.rotational_symmetry_db" title="Permalink to this definition"></a></dt>
<dd><p>Determine the order of rotational symmetry for a dot-bracket structure.</p>
<p>Given a (permutation of multiple) RNA strand(s) and a particular secondary structure in dot-bracket
notation, compute the degree of rotational symmetry. In case there is only a single linear RNA
strand, the structure always has degree 1, as there are no rotational symmetries due to the
direction of the nucleic acid sequence and the fixed positions of 5’ and 3’ ends. However, for
circular RNAs, rotational symmetries might arise if the sequence consists of a concatenation of
<span class="math notranslate nohighlight">\(k\)</span> identical subsequences.</p>
<p>If the argument <cite>positions</cite> is not <cite>NULL</cite>, the function stores an array of string start positions
for rotational shifts that map the string back onto itself. This array has length of order of
rotational symmetry, i.e. the number returned by this function. The first element <cite>positions`[0]
always contains a shift value of `0</cite> representing the trivial rotation.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>rotational_symmetry_db()</cite> to objects of type
<cite>fold_compound</cite> (i.e. RNA.fold_compound()). Thus, the first argument must be omitted. In
contrast to our C-implementation, this function doesn’t simply return the order of rotational
symmetry of the secondary structure, but returns the list <cite>position</cite> of cyclic permutation
shifts that result in a rotationally symmetric structure. The length of the list then determines
the order of rotational symmetry. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.rotational_symmetry_db" title="RNA.fold_compound.rotational_symmetry_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.rotational_symmetry_db()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The dot-bracket structure the degree of rotational symmetry is checked for</p></li>
<li><p><strong>positions</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(list-like(unsigned</span> <span class="pre">int))</span></code>) – A pointer to an (undefined) list of alternative string start positions that lead to an identity
mapping (may be NULL)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The degree of rotational symmetry of the <cite>structure</cite> (0 in case of any errors)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.rotational_symmetry_db</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.rotational_symmetry_pos</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.rotational_symmetry_pos_num</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Do not forget to release the memory occupied by <cite>positions</cite> after a successful execution of this
function.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_SHAPE_deigan">
<span class="sig-name descname"><span class="pre">sc_add_SHAPE_deigan</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DoubleVector</span> <span class="pre">reactivities</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_SHAPE_deigan" title="Permalink to this definition"></a></dt>
<dd><p>Add SHAPE reactivity data as soft constraints (Deigan et al. method)</p>
<p>This approach of SHAPE directed RNA folding uses the simple linear ansatz</p>
<div class="math notranslate nohighlight">
\[\Delta G_{\text{SHAPE}}(i) = m \ln(\text{SHAPE reactivity}(i)+1)+ b\]</div>
<p>to convert SHAPE reactivity values to pseudo energies whenever a nucleotide <span class="math notranslate nohighlight">\(i\)</span> contributes to
a stacked pair. A positive slope <span class="math notranslate nohighlight">\(m\)</span> penalizes high reactivities in paired regions, while a
negative intercept <span class="math notranslate nohighlight">\(b\)</span> results in a confirmatory <a href="#id76"><span class="problematic" id="id77">`</span></a>bonus’ free energy for correctly predicted
base pairs. Since the energy evaluation of a base pair stack involves two pairs, the pseudo energies
are added for all four contributing nucleotides. Consequently, the energy term is applied twice for
pairs inside a helix and only once for pairs adjacent to other structures. For all other loop types
the energy model remains unchanged even when the experimental data highly disagrees with a certain
motif.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add_SHAPE_deigan()</cite> to objects of type <cite>fold_compound</cite>.
See, e.g.   <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_deigan" title="RNA.fold_compound.sc_add_SHAPE_deigan"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_deigan()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reactivities</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">double</span> <span class="pre">*</span></code>) – A vector of normalized SHAPE reactivities</p></li>
<li><p><strong>m</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The slope of the conversion function</p></li>
<li><p><strong>b</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The intercept of the conversion function</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The options flag indicating how/where to store the soft constraints</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 on successful extraction of the method, 0 on errors</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_remove" title="RNA.fold_compound.sc_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_remove</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_zarringhalam" title="RNA.fold_compound.sc_add_SHAPE_zarringhalam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_zarringhalam</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_minimize_pertubation</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For further details, we refer to  <span id="id78">Deigan <em>et al.</em> [<a class="reference internal" href="bibliography.html#id5" title="Katherine E. Deigan, Tian W. Li, David H. Mathews, and Kevin M. Weeks. Accurate SHAPE-directed RNA structure determination. PNAS, 106:97-102, 2009. doi:10.1073/pnas.080692910.">2009</a>]</span> .</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_SHAPE_deigan_ali">
<span class="sig-name descname"><span class="pre">sc_add_SHAPE_deigan_ali</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">shape_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">shape_file_association</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_SHAPE_deigan_ali" title="Permalink to this definition"></a></dt>
<dd><p>Add SHAPE reactivity data from files as soft constraints for consensus structure prediction (Deigan
et al. method)</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add_SHAPE_deigan_ali()</cite> to objects of type
<cite>fold_compound</cite>. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_deigan_ali" title="RNA.fold_compound.sc_add_SHAPE_deigan_ali"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_deigan_ali()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shape_files</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span> <span class="pre">**</span></code>) – A set of filenames that contain normalized SHAPE reactivity data</p></li>
<li><p><strong>shape_file_association</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">int</span> <span class="pre">*</span></code>) – An array of integers that associate the files with sequences in the alignment</p></li>
<li><p><strong>m</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The slope of the conversion function</p></li>
<li><p><strong>b</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The intercept of the conversion function</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The options flag indicating how/where to store the soft constraints</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 on successful extraction of the method, 0 on errors</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_SHAPE_zarringhalam">
<span class="sig-name descname"><span class="pre">sc_add_SHAPE_zarringhalam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DoubleVector</span> <span class="pre">reactivities</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">default_value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">shape_conversion</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_SHAPE_zarringhalam" title="Permalink to this definition"></a></dt>
<dd><p>Add SHAPE reactivity data as soft constraints (Zarringhalam et al. method)</p>
<p>This method first converts the observed SHAPE reactivity of nucleotide <span class="math notranslate nohighlight">\(i\)</span> into a probability
<span class="math notranslate nohighlight">\(q_{i}\)</span> that position <span class="math notranslate nohighlight">\(i\)</span> is unpaired by means of a non-linear map. Then pseudo-energies
of the form</p>
<div class="math notranslate nohighlight">
\[\Delta G_{\text{SHAPE}}(x,i) = \beta\ |x_{i} - q_{i}|\]</div>
<p>are computed, where <span class="math notranslate nohighlight">\(x_{i}=0\)</span> if position <span class="math notranslate nohighlight">\(i\)</span> is unpaired and <span class="math notranslate nohighlight">\(x_{i}=1\)</span> if
<span class="math notranslate nohighlight">\(i\)</span> is paired in a given secondary structure. The parameter <span class="math notranslate nohighlight">\(\beta\)</span> serves as scaling
factor. The magnitude of discrepancy between prediction and experimental observation is represented
by <span class="math notranslate nohighlight">\(|x_{i} - q_{i}|\)</span>.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add_SHAPE_zarringhalam()</cite> to objects of type
<cite>fold_compound</cite>. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_zarringhalam" title="RNA.fold_compound.sc_add_SHAPE_zarringhalam"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_zarringhalam()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reactivities</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">double</span> <span class="pre">*</span></code>) – A vector of normalized SHAPE reactivities</p></li>
<li><p><strong>b</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The scaling factor <span class="math notranslate nohighlight">\(\beta\)</span> of the conversion function</p></li>
<li><p><strong>default_value</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The default value for a nucleotide where reactivity data is missing for</p></li>
<li><p><strong>shape_conversion</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A flag that specifies how to convert reactivities to probabilities</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The options flag indicating how/where to store the soft constraints</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 on successful extraction of the method, 0 on errors</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_remove" title="RNA.fold_compound.sc_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_remove</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_deigan" title="RNA.fold_compound.sc_add_SHAPE_deigan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_deigan</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_minimize_pertubation</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For further details, we refer to  <span id="id79">Zarringhalam <em>et al.</em> [<a class="reference internal" href="bibliography.html#id6" title="Kourosh Zarringhalam, Michelle M. Meyer, Ivan Dotu, Jeffrey H. Chuang, and Peter Clote. Integrating chemical footprinting data into RNA secondary structure prediction. PLOS ONE, 2012. doi:10.1371/journal.pone.0045160.">2012</a>]</span></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_bp">
<span class="sig-name descname"><span class="pre">sc_add_bp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_add_bp" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraints for paired nucleotides.</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><blockquote>
<div><p>This function is attached as an overloaded method <cite>sc_add_bp()</cite> to objects of type
<cite>fold_compound</cite>. The method either takes arguments for a single base pair (i,j) with the
corresponding energy value:</p>
</div></blockquote>
<p>or an entire 2-dimensional matrix with dimensions n x n that stores free energy contributions for</p>
</dd>
</dl>
<p>any base pair (i,j) with <span class="math notranslate nohighlight">\(1 \leq i &lt; j \leq n\)</span>:  In both variants, the optional argument
<cite>options</cite> defaults to RNA.OPTION_DEFAULT. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.sc_add_bp" title="RNA.fold_compound.sc_add_bp"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bp()</span></code></a> in
the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 5’ position of the base pair the soft constraint is added for</p></li>
<li><p><strong>j</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The 3’ position of the base pair the soft constraint is added for</p></li>
<li><p><strong>energy</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The free energy (soft-constraint) in <span class="math notranslate nohighlight">\(kcal / mol\)</span></p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The options flag indicating how/where to store the soft constraints</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful application of the constraint, 0 otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_set_bp" title="RNA.fold_compound.sc_set_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_set_up" title="RNA.fold_compound.sc_set_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_up" title="RNA.fold_compound.sc_add_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_up</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_bt">
<span class="sig-name descname"><span class="pre">sc_add_bt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_bt" title="Permalink to this definition"></a></dt>
<dd><p>Bind a backtracking function pointer for generic soft constraint feature.</p>
<p>This function allows one to easily bind a function pointer to the soft constraint part RNA.sc() of
the RNA.fold_compound(). The provided function should be used for backtracking purposes in loop
regions that were altered via the generic soft constraint feature. It has to return an array of
RNA.basepair() data structures, were the last element in the list is indicated by a value of -1 in
it’s i position.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add_bt()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_add_bt" title="RNA.fold_compound.sc_add_bt"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bt()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>f</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.sc_bt</span></code>) – A pointer to the function that returns additional base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful binding the callback function, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_add_data" title="RNA.fold_compound.sc_add_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_data</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_exp</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_data">
<span class="sig-name descname"><span class="pre">sc_add_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">callback=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_data" title="Permalink to this definition"></a></dt>
<dd><p>Add an auxiliary data structure for the generic soft constraints callback function.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add_data()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <a class="reference internal" href="#RNA.fold_compound.sc_add_data" title="RNA.fold_compound.sc_add_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_data()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code>) – A pointer to the data structure that holds required data for function ‘f’</p></li>
<li><p><strong>free_data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.auxdata_free</span></code>) – A pointer to a function that free’s the memory occupied by <cite>data</cite> (Maybe NULL)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful binding the data (and free-function), 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_exp</span></code>, <a class="reference internal" href="#RNA.fold_compound.sc_add_bt" title="RNA.fold_compound.sc_add_bt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bt</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_exp_f">
<span class="sig-name descname"><span class="pre">sc_add_exp_f</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_exp_f" title="Permalink to this definition"></a></dt>
<dd><p>Bind a function pointer for generic soft constraint feature (PF version)</p>
<p>This function allows one to easily bind a function pointer and corresponding data structure to the
soft constraint part RNA.sc() of the RNA.fold_compound(). The function for evaluating the generic
soft constraint feature has to return a pseudo free energy <span class="math notranslate nohighlight">\(\hat{E}\)</span> as Boltzmann factor,
i.e. <span class="math notranslate nohighlight">\(exp(- \hat{E} / kT)\)</span>. The required unit for <span class="math notranslate nohighlight">\(E\)</span> is <span class="math notranslate nohighlight">\(cal/mol\)</span>.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add_exp()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.   <code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_exp()</span></code> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>exp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.sc_exp</span></code>) – A pointer to the function that evaluates the generic soft constraint feature</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful binding the callback function, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_add_bt" title="RNA.fold_compound.sc_add_bt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bt</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add</span></code>, <a class="reference internal" href="#RNA.fold_compound.sc_add_data" title="RNA.fold_compound.sc_add_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_data</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_f">
<span class="sig-name descname"><span class="pre">sc_add_f</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">callback</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_f" title="Permalink to this definition"></a></dt>
<dd><p>Bind a function pointer for generic soft constraint feature (MFE version)</p>
<p>This function allows one to easily bind a function pointer and corresponding data structure to the
soft constraint part RNA.sc() of the RNA.fold_compound(). The function for evaluating the generic
soft constraint feature has to return a pseudo free energy <span class="math notranslate nohighlight">\(\hat{E}\)</span> in <span class="math notranslate nohighlight">\(dacal/mol\)</span>,
where <span class="math notranslate nohighlight">\(1 dacal/mol = 10 cal/mol\)</span>.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add()</span></code> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>f</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.sc</span></code>) – A pointer to the function that evaluates the generic soft constraint feature</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful binding the callback function, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_add_data" title="RNA.fold_compound.sc_add_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_data</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_bt" title="RNA.fold_compound.sc_add_bt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bt</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_exp</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_hi_motif">
<span class="sig-name descname"><span class="pre">sc_add_hi_motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">FLT_OR_DBL</span> <span class="pre">energy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_hi_motif" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraints for hairpin or interior loop binding motif.</p>
<p>Here is an example that adds a theophylline binding motif. Free energy contribution is derived from
<span class="math notranslate nohighlight">\(k_{d} = 0.1 \mu M\)</span>, taken from Jenison et al. 1994. At <span class="math notranslate nohighlight">\(1M\)</span> concentration the
corresponding binding free energy amounts to <span class="math notranslate nohighlight">\(-9.93~kcal/mol\)</span>.</p>
<p>theo_aptamer.svg</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_add_hi_motif()</cite> to objects of type <cite>fold_compound</cite>. The
last parameter is optional an defaults to <cite>options</cite> = RNA.OPTION_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_add_hi_motif" title="RNA.fold_compound.sc_add_hi_motif"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_hi_motif()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The sequence motif (may be interspaced by ‘&amp;’ character</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The structure motif (may be interspaced by ‘&amp;’ character</p></li>
<li><p><strong>energy</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The free energy of the motif (e.g. binding free energy)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>non-zero value if application of the motif using soft constraints was successful</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_stack">
<span class="sig-name descname"><span class="pre">sc_add_stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">energy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_add_stack" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">sc_add_stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DoubleVector</span> <span class="pre">energies</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_add_up">
<span class="sig-name descname"><span class="pre">sc_add_up</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_add_up" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraints for unpaired nucleotides.</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><blockquote>
<div><p>This function is attached as an overloaded method <cite>sc_add_up()</cite> to objects of type
<cite>fold_compound</cite>. The method either takes arguments for a single nucleotide <span class="math notranslate nohighlight">\(i\)</span> with the
corresponding energy value:</p>
</div></blockquote>
<p>or an entire vector that stores free energy contributions for each nucleotide <span class="math notranslate nohighlight">\(i\)</span> with</p>
</dd>
</dl>
<p><span class="math notranslate nohighlight">\(1 \leq i \leq n\)</span>:  In both variants, the optional argument <cite>options</cite> defaults to
RNA.OPTION_DEFAULT. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.sc_add_up" title="RNA.fold_compound.sc_add_up"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_up()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>
.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The nucleotide position the soft constraint is added for</p></li>
<li><p><strong>energy</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The free energy (soft-constraint) in <span class="math notranslate nohighlight">\(kcal / mol\)</span></p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The options flag indicating how/where to store the soft constraints</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful application of the constraint, 0 otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_set_up" title="RNA.fold_compound.sc_set_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_bp" title="RNA.fold_compound.sc_add_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_set_bp" title="RNA.fold_compound.sc_set_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_bp</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_init">
<span class="sig-name descname"><span class="pre">sc_init</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_init" title="Permalink to this definition"></a></dt>
<dd><p>Initialize an empty soft constraints data structure within a RNA.fold_compound().</p>
<p>This function adds a proper soft constraints data structure to the RNA.fold_compound() data
structure. If soft constraints already exist within the fold compound, they are removed.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_init()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_init" title="RNA.fold_compound.sc_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_init()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_set_bp" title="RNA.fold_compound.sc_set_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_set_up" title="RNA.fold_compound.sc_set_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_deigan" title="RNA.fold_compound.sc_add_SHAPE_deigan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_deigan</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_SHAPE_zarringhalam" title="RNA.fold_compound.sc_add_SHAPE_zarringhalam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_SHAPE_zarringhalam</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_remove" title="RNA.fold_compound.sc_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_remove</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_exp</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_add_pre</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_add_post</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE and RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod">
<span class="sig-name descname"><span class="pre">sc_mod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod" title="Permalink to this definition"></a></dt>
<dd><p>Prepare soft constraint callbacks for modified base as specified in JSON string.</p>
<p>This function takes a RNA.sc_mod_param() data structure as obtained from
RNA.sc_mod_read_from_json() or RNA.sc_mod_read_from_jsonfile() and prepares all requirements to
acknowledge modified bases as specified in the provided <cite>params</cite> data structure. All subsequent
predictions will treat each modification site special and adjust energy contributions if necessary.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod()</cite> to objects of type <cite>fold_compound</cite>
with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.   <a class="reference internal" href="#RNA.fold_compound.sc_mod" title="RNA.fold_compound.sc_mod"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod()</span></code></a>
in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>json</strong> – The JSON formatted string with the modified base parameters</p></li>
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.sc_mod_read_from_json" title="RNA.sc_mod_read_from_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_mod_read_from_json</span></code></a>, <a class="reference internal" href="#RNA.sc_mod_read_from_jsonfile" title="RNA.sc_mod_read_from_jsonfile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_mod_read_from_jsonfile</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_json" title="RNA.fold_compound.sc_mod_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_json</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_jsonfile" title="RNA.fold_compound.sc_mod_jsonfile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_jsonfile</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_m6A" title="RNA.fold_compound.sc_mod_m6A"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_m6A</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_pseudouridine" title="RNA.fold_compound.sc_mod_pseudouridine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_pseudouridine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_inosine" title="RNA.fold_compound.sc_mod_inosine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_inosine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_7DA" title="RNA.fold_compound.sc_mod_7DA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_7DA</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_purine" title="RNA.fold_compound.sc_mod_purine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_purine</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_mod_dihydrouridine</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_7DA">
<span class="sig-name descname"><span class="pre">sc_mod_7DA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_7DA" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraint callbacks for 7-deaza-adenosine (7DA)</p>
<p>This is a convenience wrapper to add support for 7-deaza-adenosine using the soft constraint
callback mechanism. Modification sites are provided as a list of sequence positions (1-based).
Energy parameter corrections are derived from  <span id="id80">Richardson and Znosko [<a class="reference internal" href="bibliography.html#id47" title="Katherine E Richardson and Brent M Znosko. Nearest-neighbor parameters for 7-deaza-adenosine· uridine base pairs in RNA duplexes. RNA, 22(6):934–942, 2016. doi:10.1261/rna.055277.115.">2016</a>]</span> .</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_7DA()</cite> to objects of type <cite>fold_compound</cite>
with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_7DA" title="RNA.fold_compound.sc_mod_7DA"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_7DA()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_dihydrouridine">
<span class="sig-name descname"><span class="pre">sc_mod_dihydrouridine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_dihydrouridine" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraint callbacks for dihydrouridine.</p>
<p>This is a convenience wrapper to add support for dihydrouridine using the soft constraint callback
mechanism. Modification sites are provided as a list of sequence positions (1-based). Energy
parameter corrections are derived from Rosetta/RECESS predictions.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_dihydrouridine()</cite> to objects of type
<cite>fold_compound</cite> with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_dihydrouridine" title="RNA.fold_compound.sc_mod_dihydrouridine"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_dihydrouridine()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_inosine">
<span class="sig-name descname"><span class="pre">sc_mod_inosine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_inosine" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraint callbacks for Inosine.</p>
<p>This is a convenience wrapper to add support for inosine using the soft constraint callback
mechanism. Modification sites are provided as a list of sequence positions (1-based). Energy
parameter corrections are derived from  <span id="id81">Wright <em>et al.</em> [<a class="reference internal" href="bibliography.html#id45" title="Daniel J Wright, Jamie L Rice, Dawn M Yanker, and Brent M Znosko. Nearest Neighbor Parameters for Inosine· Uridine Pairs in RNA Duplexes. Biochemistry, 46(15):4625–4634, 2007. doi:10.1021/bi0616910.">2007</a>]</span>  and  <span id="id82">Wright <em>et al.</em> [<a class="reference internal" href="bibliography.html#id46" title="Daniel J Wright, Christopher R Force, and Brent M Znosko. Stability of RNA duplexes containing inosine· cytosine pairs. Nucleic Acids Research, 46(22):12099–12108, 2018. doi:10.1093/nar/gky907.">2018</a>]</span> .</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_inosine()</cite> to objects of type
<cite>fold_compound</cite> with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_inosine" title="RNA.fold_compound.sc_mod_inosine"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_inosine()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_json">
<span class="sig-name descname"><span class="pre">sc_mod_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_json" title="Permalink to this definition"></a></dt>
<dd><p>Prepare soft constraint callbacks for modified base as specified in JSON string.</p>
<p>This function prepares all requirements to acknowledge modified bases as specified in the provided
<cite>json</cite> string. All subsequent predictions will treat each modification site special and adjust
energy contributions if necessary.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_json()</cite> to objects of type
<cite>fold_compound</cite> with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_json" title="RNA.fold_compound.sc_mod_json"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_json()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>json</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The JSON formatted string with the modified base parameters</p></li>
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_mod_jsonfile" title="RNA.fold_compound.sc_mod_jsonfile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_jsonfile</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod" title="RNA.fold_compound.sc_mod"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_m6A" title="RNA.fold_compound.sc_mod_m6A"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_m6A</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_pseudouridine" title="RNA.fold_compound.sc_mod_pseudouridine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_pseudouridine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_inosine" title="RNA.fold_compound.sc_mod_inosine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_inosine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_7DA" title="RNA.fold_compound.sc_mod_7DA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_7DA</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_purine" title="RNA.fold_compound.sc_mod_purine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_purine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_dihydrouridine" title="RNA.fold_compound.sc_mod_dihydrouridine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_dihydrouridine</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">modified-bases-params</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_jsonfile">
<span class="sig-name descname"><span class="pre">sc_mod_jsonfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_jsonfile" title="Permalink to this definition"></a></dt>
<dd><p>Prepare soft constraint callbacks for modified base as specified in JSON string.</p>
<p>Similar to RNA.fold_compound.sc_mod_json(), this function prepares all requirements to acknowledge modified bases
as specified in the provided <cite>json</cite> file. All subsequent predictions will treat each modification
site special and adjust energy contributions if necessary.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_jsonfile()</cite> to objects of type
<cite>fold_compound</cite> with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_jsonfile" title="RNA.fold_compound.sc_mod_jsonfile"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_jsonfile()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>json</strong> – The JSON formatted string with the modified base parameters</p></li>
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_mod_json" title="RNA.fold_compound.sc_mod_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_json</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod" title="RNA.fold_compound.sc_mod"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_m6A" title="RNA.fold_compound.sc_mod_m6A"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_m6A</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_pseudouridine" title="RNA.fold_compound.sc_mod_pseudouridine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_pseudouridine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_inosine" title="RNA.fold_compound.sc_mod_inosine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_inosine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_7DA" title="RNA.fold_compound.sc_mod_7DA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_7DA</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_purine" title="RNA.fold_compound.sc_mod_purine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_purine</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod_dihydrouridine" title="RNA.fold_compound.sc_mod_dihydrouridine"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_dihydrouridine</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">modified-bases-params</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_m6A">
<span class="sig-name descname"><span class="pre">sc_mod_m6A</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_m6A" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraint callbacks for N6-methyl-adenosine (m6A)</p>
<p>This is a convenience wrapper to add support for m6A using the soft constraint callback mechanism.
Modification sites are provided as a list of sequence positions (1-based). Energy parameter
corrections are derived from  <span id="id83">Kierzek <em>et al.</em> [<a class="reference internal" href="bibliography.html#id44" title="Elzbieta Kierzek, Xiaoju Zhang, Richard M Watson, Scott D Kennedy, Marta Szabat, Ryszard Kierzek, and David H Mathews. Secondary structure prediction for RNA sequences including N6-methyladenosine. Nature communications, 13(1):1–10, 2022. doi:10.1038/s41467-022-28817-4.">2022</a>]</span> .</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_m6A()</cite> to objects of type <cite>fold_compound</cite>
with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_m6A" title="RNA.fold_compound.sc_mod_m6A"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_m6A()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_pseudouridine">
<span class="sig-name descname"><span class="pre">sc_mod_pseudouridine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_pseudouridine" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraint callbacks for Pseudouridine.</p>
<p>This is a convenience wrapper to add support for pseudouridine using the soft constraint callback
mechanism. Modification sites are provided as a list of sequence positions (1-based). Energy
parameter corrections are derived from  <span id="id84">Hudson <em>et al.</em> [<a class="reference internal" href="bibliography.html#id43" title="Graham A Hudson, Richard J Bloomingdale, and Brent M Znosko. Thermodynamic contribution and nearest-neighbor parameters of pseudouridine-adenosine base pairs in oligoribonucleotides. RNA, 19(11):1474–1482, 2013. doi:10.1261/rna.039610.113.">2013</a>]</span> .</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_pseudouridine()</cite> to objects of type
<cite>fold_compound</cite> with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_pseudouridine" title="RNA.fold_compound.sc_mod_pseudouridine"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_pseudouridine()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_mod_purine">
<span class="sig-name descname"><span class="pre">sc_mod_purine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_mod_purine" title="Permalink to this definition"></a></dt>
<dd><p>Add soft constraint callbacks for Purine (a.k.a. nebularine)</p>
<p>This is a convenience wrapper to add support for Purine using the soft constraint callback
mechanism. Modification sites are provided as a list of sequence positions (1-based). Energy
parameter corrections are derived from  <span id="id85">Jolley and Znosko [<a class="reference internal" href="bibliography.html#id48" title="Elizabeth A Jolley and Brent M Znosko. The loss of a hydrogen bond: Thermodynamic contributions of a non-standard nucleotide. Nucleic acids research, 45(3):1479–1487, 2017. doi:10.1093/nar/gkw830.">2017</a>]</span> .</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as overloaded method <cite>sc_mod_purine()</cite> to objects of type
<cite>fold_compound</cite> with default <cite>options</cite> = RNA.SC_MOD_DEFAULT. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_mod_purine" title="RNA.fold_compound.sc_mod_purine"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod_purine()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modification_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">*</span></code>) – A list of modification site, i.e. positions that contain the modified base (1-based, last
element in the list indicated by 0)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitvector of options how to handle the input, e.g. RNA.SC_MOD_DEFAULT</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of sequence positions modified base parameters will be used for</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_FALLBACK</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_CHECK_UNMOD</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_SILENT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.SC_MOD_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_remove">
<span class="sig-name descname"><span class="pre">sc_remove</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sc_remove" title="Permalink to this definition"></a></dt>
<dd><p>Remove soft constraints from RNA.fold_compound().</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_remove()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_remove" title="RNA.fold_compound.sc_remove"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_remove()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Accepts RNA.fold_compound() of type RNA.FC_TYPE_SINGLE and RNA.FC_TYPE_COMPARATIVE</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_set_bp">
<span class="sig-name descname"><span class="pre">sc_set_bp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DoubleDoubleVector</span> <span class="pre">constraints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_set_bp" title="Permalink to this definition"></a></dt>
<dd><p>Set soft constraints for paired nucleotides.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_set_bp()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_set_bp" title="RNA.fold_compound.sc_set_bp"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_bp()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>constraints</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">FLT_OR_DBL</span> <span class="pre">**</span></code>) – A two-dimensional array of pseudo free energies in <span class="math notranslate nohighlight">\(kcal / mol\)</span></p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The options flag indicating how/where to store the soft constraints</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful application of the constraint, 0 otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_add_bp" title="RNA.fold_compound.sc_add_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_set_up" title="RNA.fold_compound.sc_set_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_up" title="RNA.fold_compound.sc_add_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_up</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function replaces any pre-exisitng soft constraints with the ones supplied in <cite>constraints</cite>.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_set_stack">
<span class="sig-name descname"><span class="pre">sc_set_stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DoubleVector</span> <span class="pre">constraints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_set_stack" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">sc_set_stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DoubleDoubleVector</span> <span class="pre">constraints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sc_set_up">
<span class="sig-name descname"><span class="pre">sc_set_up</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DoubleVector</span> <span class="pre">constraints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.fold_compound.sc_set_up" title="Permalink to this definition"></a></dt>
<dd><p>Set soft constraints for unpaired nucleotides.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>sc_set_up()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.sc_set_up" title="RNA.fold_compound.sc_set_up"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_up()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>constraints</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">FLT_OR_DBL</span> <span class="pre">*</span></code>) – A vector of pseudo free energies in <span class="math notranslate nohighlight">\(kcal / mol\)</span></p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The options flag indicating how/where to store the soft constraints</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on successful application of the constraint, 0 otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.sc_add_up" title="RNA.fold_compound.sc_add_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_up</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_set_bp" title="RNA.fold_compound.sc_set_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_set_bp</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_add_bp" title="RNA.fold_compound.sc_add_bp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_add_bp</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function replaces any pre-exisitng soft constraints with the ones supplied in <cite>constraints</cite>.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id86">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#id86" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sequence_add">
<span class="sig-name descname"><span class="pre">sequence_add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sequence_add" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id87">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_encoding</span></span><a class="headerlink" href="#id87" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id88">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_encoding2</span></span><a class="headerlink" href="#id88" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sequence_prepare">
<span class="sig-name descname"><span class="pre">sequence_prepare</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sequence_prepare" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sequence_remove">
<span class="sig-name descname"><span class="pre">sequence_remove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sequence_remove" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.sequence_remove_all">
<span class="sig-name descname"><span class="pre">sequence_remove_all</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.sequence_remove_all" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id89">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand_end</span></span><a class="headerlink" href="#id89" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id90">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand_number</span></span><a class="headerlink" href="#id90" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id91">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand_order</span></span><a class="headerlink" href="#id91" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id92">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand_start</span></span><a class="headerlink" href="#id92" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id93">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strands</span></span><a class="headerlink" href="#id93" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.subopt">
<span class="sig-name descname"><span class="pre">subopt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">delta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">sorted=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">FILE</span> <span class="pre">*</span> <span class="pre">nullfile=None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.SuboptVector" title="RNA.SuboptVector"><span class="pre">SuboptVector</span></a></span></span><a class="headerlink" href="#RNA.fold_compound.subopt" title="Permalink to this definition"></a></dt>
<dd><p>Returns list of subopt structures or writes to fp.</p>
<p>This function produces <strong>all</strong> suboptimal secondary structures within ‘delta’ * 0.01 kcal/mol of the
optimum, see  <span id="id94">Wuchty <em>et al.</em> [<a class="reference internal" href="bibliography.html#id33" title="S. Wuchty, W. Fontana, I. L. Hofacker, and P. Schuster. Complete suboptimal folding of RNA and the stability of secondary structures. Biopolymers, 49(2):145–165, February 1999. doi:10.1002/(SICI)1097-0282(199902)49:2&lt;145::AID-BIP4&gt;3.0.CO;2-G.">1999</a>]</span> . The results are either directly written to a ‘fp’ (if ‘fp’ is
not NULL), or (fp==NULL) returned in a RNA.subopt_solution() * list terminated by an entry were the
‘structure’ member is NULL.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <strong>subopt()</strong> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.subopt" title="RNA.fold_compound.subopt"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>delta</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>sorted</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Sort results by energy in ascending order</p></li>
<li><p><strong>fp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">FILE</span> <span class="pre">*</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.subopt_solution()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.subopt_cb" title="RNA.fold_compound.subopt_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt_cb</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.subopt_zuker" title="RNA.fold_compound.subopt_zuker"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt_zuker</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function requires all multibranch loop DP matrices for unique multibranch loop backtracing.
Therefore, the supplied RNA.fold_compound()`fc` (argument 1) must be initialized with
RNA.md().uniq_ML = 1, for instance like this:</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.subopt_cb">
<span class="sig-name descname"><span class="pre">subopt_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">delta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.subopt_cb" title="Permalink to this definition"></a></dt>
<dd><p>Generate suboptimal structures within an energy band arround the MFE.</p>
<p>This is the most generic implementation of the suboptimal structure generator according to
<span id="id95">Wuchty <em>et al.</em> [<a class="reference internal" href="bibliography.html#id33" title="S. Wuchty, W. Fontana, I. L. Hofacker, and P. Schuster. Complete suboptimal folding of RNA and the stability of secondary structures. Biopolymers, 49(2):145–165, February 1999. doi:10.1002/(SICI)1097-0282(199902)49:2&lt;145::AID-BIP4&gt;3.0.CO;2-G.">1999</a>]</span> . Identical to RNA.fold_compound.subopt(), it computes all secondary structures within an
energy band <cite>delta</cite> arround the MFE. However, this function does not print the resulting structures
and their corresponding free energies to a file pointer, or returns them as a list. Instead, it
calls a user-provided callback function which it passes the structure in dot-bracket format, the
corresponding free energy in kcal/mol, and a user-provided data structure each time a structure was
backtracked successfully. This function indicates the final output, i.e. the end of the backtracking
procedure by passing NULL instead of an actual dot-bracket string to the callback.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <strong>subopt_cb()</strong> to objects of type <cite>fold_compound</cite>. See,
e.g.  <a class="reference internal" href="#RNA.fold_compound.subopt_cb" title="RNA.fold_compound.subopt_cb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt_cb()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>delta</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Energy band arround the MFE in 10cal/mol, i.e. deka-calories</p></li>
<li><p><strong>cb</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.subopt_result</span></code>) – Pointer to a callback function that handles the backtracked structure and its free energy in
kcal/mol</p></li>
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code>) – Pointer to some data structure that is passed along to the callback</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.subopt_result</span></code>, <a class="reference internal" href="#RNA.fold_compound.subopt" title="RNA.fold_compound.subopt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.subopt_zuker" title="RNA.fold_compound.subopt_zuker"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt_zuker</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function requires all multibranch loop DP matrices for unique multibranch loop backtracing.
Therefore, the supplied RNA.fold_compound()`fc` (argument 1) must be initialized with
RNA.md().uniq_ML = 1, for instance like this:</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.subopt_zuker">
<span class="sig-name descname"><span class="pre">subopt_zuker</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.subopt_zuker" title="Permalink to this definition"></a></dt>
<dd><p>Compute Zuker type suboptimal structures.</p>
<p>Compute Suboptimal structures according to  <span id="id96">Zuker [<a class="reference internal" href="bibliography.html#id32" title="M. Zuker. On finding all suboptimal foldings of an RNA molecule. Science, 244(4900):48–52, April 1989. doi:10.1126/science.2468181.">1989</a>]</span>  , i.e. for every possible base
pair the minimum energy structure containing the resp. base pair. Returns a list of these structures
and their energies.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <strong>subopt_zuker()</strong> to objects of type <cite>fold_compound</cite>. See,
e.g.  <a class="reference internal" href="#RNA.fold_compound.subopt_zuker" title="RNA.fold_compound.subopt_zuker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt_zuker()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>List of zuker suboptimal structures</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.subopt_solution()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.subopt" title="RNA.fold_compound.subopt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.subopt</span></code></a>, <a class="reference internal" href="#RNA.zukersubopt" title="RNA.zukersubopt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">zukersubopt</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">zukersubopt_par</span></code></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.fold_compound.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.fold_compound.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id97">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#id97" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.ud_add_motif">
<span class="sig-name descname"><span class="pre">ud_add_motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">motif</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">motif_en</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">name=&quot;&quot;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.ud_add_motif" title="Permalink to this definition"></a></dt>
<dd><p>Add an unstructured domain motif, e.g. for ligand binding.</p>
<p>This function adds a ligand binding motif and the associated binding free energy to the RNA.ud()
attribute of a RNA.fold_compound(). The motif data will then be used in subsequent secondary
structure predictions. Multiple calls to this function with different motifs append all additional
data to a list of ligands, which all will be evaluated. Ligand motif data can be removed from the
RNA.fold_compound() again using the RNA.fold_compound.ud_remove() function. The loop type parameter allows one
to limit the ligand binding to particular loop type, such as the exterior loop, hairpin loops,
interior loops, or multibranch loops.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>motif</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The sequence motif the ligand binds to</p></li>
<li><p><strong>motif_en</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – The binding free energy of the ligand in kcal/mol</p></li>
<li><p><strong>motif_name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The name/id of the motif (may be <cite>NULL</cite>)</p></li>
<li><p><strong>loop_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The loop type the ligand binds to</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.UNSTRUCTURED_DOMAIN_EXT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.UNSTRUCTURED_DOMAIN_HP_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.UNSTRUCTURED_DOMAIN_INT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.UNSTRUCTURED_DOMAIN_MB_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.UNSTRUCTURED_DOMAIN_ALL_LOOPS</span></code>, <a class="reference internal" href="#RNA.fold_compound.ud_remove" title="RNA.fold_compound.ud_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_remove</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.ud_remove">
<span class="sig-name descname"><span class="pre">ud_remove</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.ud_remove" title="Permalink to this definition"></a></dt>
<dd><p>Remove ligand binding to unpaired stretches.</p>
<p>This function removes all ligand motifs that were bound to a RNA.fold_compound() using the
RNA.fold_compound.ud_add_motif() function.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>ud_remove()</cite> to objects of type <cite>fold_compound</cite>. See, e.g.
<a class="reference internal" href="#RNA.fold_compound.ud_remove" title="RNA.fold_compound.ud_remove"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_remove()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.ud_set_data">
<span class="sig-name descname"><span class="pre">ud_set_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">free_cb=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.ud_set_data" title="Permalink to this definition"></a></dt>
<dd><p>Attach an auxiliary data structure.</p>
<p>This function binds an arbitrary, auxiliary data structure for user-implemented ligand binding. The
optional callback <cite>free_cb</cite> will be passed the bound data structure whenever the
RNA.fold_compound() is removed from memory to avoid memory leaks.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>ud_set_data()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.  <a class="reference internal" href="#RNA.fold_compound.ud_set_data" title="RNA.fold_compound.ud_set_data"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_set_data()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code>) – A pointer to the auxiliary data structure</p></li>
<li><p><strong>free_cb</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.auxdata_free</span></code>) – A pointer to a callback function that free’s memory occupied by <cite>data</cite></p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.ud_set_prod_rule_cb" title="RNA.fold_compound.ud_set_prod_rule_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_set_prod_rule_cb</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.ud_set_exp_prod_rule_cb" title="RNA.fold_compound.ud_set_exp_prod_rule_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_set_exp_prod_rule_cb</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.ud_remove" title="RNA.fold_compound.ud_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_remove</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.ud_set_exp_prod_rule_cb">
<span class="sig-name descname"><span class="pre">ud_set_exp_prod_rule_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">prod_cb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">eval_cb</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.ud_set_exp_prod_rule_cb" title="Permalink to this definition"></a></dt>
<dd><p>Attach production rule for partition function.</p>
<p>This function is the partition function companion of RNA.fold_compound.ud_set_prod_rule_cb().</p>
<p>Use it to bind callbacks to (i) fill the <cite>U</cite> production rule dynamic programming matrices and/or
prepare the RNA.unstructured_domain().data, and (ii) provide a callback to retrieve partition
functions for subsegments <span class="math notranslate nohighlight">\([i,j]\)</span>.</p>
<p>B_prod_rule.svg</p>
<p>ligands_up_callback.svg</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>ud_set_exp_prod_rule_cb()</cite> to objects of type
<cite>fold_compound</cite>. See, e.g.  <a class="reference internal" href="#RNA.fold_compound.ud_set_exp_prod_rule_cb" title="RNA.fold_compound.ud_set_exp_prod_rule_cb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_set_exp_prod_rule_cb()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pre_cb</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ud_exp_production</span></code>) – A pointer to a callback function for the <cite>B</cite> production rule</p></li>
<li><p><strong>exp_e_cb</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ud_exp</span></code>) – A pointer to a callback function that retrieves the partition function for a segment
<span class="math notranslate nohighlight">\([i,j]\)</span> that may be bound by one or more ligands.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.ud_set_prod_rule_cb" title="RNA.fold_compound.ud_set_prod_rule_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_set_prod_rule_cb</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.ud_set_prob_cb">
<span class="sig-name descname"><span class="pre">ud_set_prob_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">setter_cb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">getter_cb</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.ud_set_prob_cb" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>ud_set_prob_cb()</cite> to objects of type <cite>fold_compound</cite>. See,
e.g.  <a class="reference internal" href="#RNA.fold_compound.ud_set_prob_cb" title="RNA.fold_compound.ud_set_prob_cb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_set_prob_cb()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.ud_set_prod_rule_cb">
<span class="sig-name descname"><span class="pre">ud_set_prod_rule_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fold_compound</span> <span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">prod_cb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">eval_cb</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PyObject</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.fold_compound.ud_set_prod_rule_cb" title="Permalink to this definition"></a></dt>
<dd><p>Attach production rule callbacks for free energies computations.</p>
<p>Use this function to bind a user-implemented grammar extension for unstructured domains.</p>
<p>The callback <cite>e_cb</cite> needs to evaluate the free energy contribution <span class="math notranslate nohighlight">\(f(i,j)\)</span> of the unpaired
segment <span class="math notranslate nohighlight">\([i,j]\)</span>. It will be executed in each of the regular secondary structure production
rules. Whenever the callback is passed the RNA.UNSTRUCTURED_DOMAIN_MOTIF flag via its <cite>loop_type</cite>
parameter the contribution of any ligand that consecutively binds from position <span class="math notranslate nohighlight">\(i\)</span> to
<span class="math notranslate nohighlight">\(j\)</span> (the white box) is requested. Otherwise, the callback usually performs a lookup in the
precomputed <cite>B</cite> matrices. Which <cite>B</cite> matrix is addressed will be indicated by the flags
RNA.UNSTRUCTURED_DOMAIN_EXT_LOOP,
RNA.UNSTRUCTURED_DOMAIN_HP_LOOPRNA.UNSTRUCTURED_DOMAIN_INT_LOOP, and
RNA.UNSTRUCTURED_DOMAIN_MB_LOOP. As their names already imply, they specify exterior loops (<cite>F</cite>
production rule), hairpin loops and interior loops (<cite>C</cite> production rule), and multibranch loops (<cite>M</cite>
and <cite>M1</cite> production rule).</p>
<p>ligands_up_callback.svg</p>
<p>The <cite>pre_cb</cite> callback will be executed as a pre-processing step right before the regular secondary
structure rules. Usually one would use this callback to fill the dynamic programming matrices <cite>U</cite>
and preparations of the auxiliary data structure RNA.unstructured_domain().data</p>
<p>B_prod_rule.svg</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is attached as method <cite>ud_set_prod_rule_cb()</cite> to objects of type <cite>fold_compound</cite>.
See, e.g.  <a class="reference internal" href="#RNA.fold_compound.ud_set_prod_rule_cb" title="RNA.fold_compound.ud_set_prod_rule_cb"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RNA.fold_compound.ud_set_prod_rule_cb()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pre_cb</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ud_production</span></code>) – A pointer to a callback function for the <cite>B</cite> production rule</p></li>
<li><p><strong>e_cb</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ud</span></code>) – A pointer to a callback function for free energy evaluation</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.zsc_compute">
<span class="sig-name descname"><span class="pre">zsc_compute</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">e</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.zsc_compute" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.zsc_compute_raw">
<span class="sig-name descname"><span class="pre">zsc_compute_raw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">e</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.zsc_compute_raw" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.zsc_filter_free">
<span class="sig-name descname"><span class="pre">zsc_filter_free</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.zsc_filter_free" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.zsc_filter_init">
<span class="sig-name descname"><span class="pre">zsc_filter_init</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.zsc_filter_init" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.zsc_filter_on">
<span class="sig-name descname"><span class="pre">zsc_filter_on</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.zsc_filter_on" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.zsc_filter_threshold">
<span class="sig-name descname"><span class="pre">zsc_filter_threshold</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.zsc_filter_threshold" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.fold_compound.zsc_filter_update">
<span class="sig-name descname"><span class="pre">zsc_filter_update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.fold_compound.zsc_filter_update" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_alifold_arrays">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_alifold_arrays</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_alifold_arrays" title="Permalink to this definition"></a></dt>
<dd><p>Free the memory occupied by MFE alifold functions.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Usage of this function is discouraged! It only affects memory being free’d that was allocated by
an old API function before. Release of memory occupied by the newly introduced
RNA.fold_compound() is handled by RNA.fold_compound_free()</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound_free</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_arrays">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_arrays</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_arrays" title="Permalink to this definition"></a></dt>
<dd><p>Free arrays for mfe folding.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>See RNA.fold(), RNA.circfold(), or RNA.fold_compound.mfe() and RNA.fold_compound() for the usage</p>
</div>
<dl class="simple">
<dt>of the</dt><dd><p>new API!</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_co_arrays">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_co_arrays</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_co_arrays" title="Permalink to this definition"></a></dt>
<dd><p>Free memory occupied by cofold()</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function will only free memory allocated by a prior call of cofold() or cofold_par(). See
RNA.fold_compound.mfe_dimer() for how to use the new API</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fc_destroy</span></code>, <a class="reference internal" href="#RNA.fold_compound.mfe_dimer" title="RNA.fold_compound.mfe_dimer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mfe_dimer</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>folding matrices now reside in the fold compound, and should be free’d there</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_co_pf_arrays">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_co_pf_arrays</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_co_pf_arrays" title="Permalink to this definition"></a></dt>
<dd><p>Free the memory occupied by co_pf_fold()</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function will be removed for the new API soon! See RNA.fold_compound.pf_dimer(),</p>
</div>
<dl class="simple">
<dt>RNA.fold_compound(),</dt><dd><p>and RNA.fold_compound_free() for an alternative</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_path">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_path" title="Permalink to this definition"></a></dt>
<dd><p>Free memory allocated by get_path() function.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.path_free() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.path()</span> <span class="pre">*</span></code>) – pointer to memory to be freed</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_pf_arrays">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_pf_arrays</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_pf_arrays" title="Permalink to this definition"></a></dt>
<dd><p>Free arrays for the partition function recursions.</p>
<p>Call this function if you want to free all allocated memory associated with the partition function
forward recursion.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>See RNA.fold_compound() and its related functions for how to free memory occupied by the
dynamic programming matrices</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Successive calls of pf_fold(), pf_circ_fold() already check if they should free any memory from a
previous run.  <strong>OpenMP notice:</strong></p>
<blockquote>
<div><p>This function should be called before leaving a thread in order to avoid leaking memory</p>
</div></blockquote>
<dl class="simple">
<dt><strong>Postcondition</strong></dt><dd><p>All memory allocated by pf_fold_par(), pf_fold() or pf_circ_fold() will be free’d</p>
</dd>
</dl>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pf_fold_par</span></code>, <a class="reference internal" href="#RNA.pf_fold" title="RNA.pf_fold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pf_fold</span></code></a>, <a class="reference internal" href="#RNA.pf_circ_fold" title="RNA.pf_circ_fold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pf_circ_fold</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_profile">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_profile" title="Permalink to this definition"></a></dt>
<dd><p>free space allocated in Make_bp_profile</p>
<p>Backward compatibility only. You can just use plain free()</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.free_tree">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">free_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.free_tree" title="Permalink to this definition"></a></dt>
<dd><p>Free the memory allocated for Tree t.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>t</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span> <span class="pre">*</span></code>) – </p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_aligned_line">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_aligned_line</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arg1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.get_aligned_line" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_centroid_struct_pl">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_centroid_struct_pl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pl</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.get_centroid_struct_pl" title="Permalink to this definition"></a></dt>
<dd><p>Get the centroid structure of the ensemble.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function was renamed to RNA.centroid_from_plist()</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_centroid_struct_pr">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_centroid_struct_pr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.get_centroid_struct_pr" title="Permalink to this definition"></a></dt>
<dd><p>Get the centroid structure of the ensemble.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function was renamed to RNA.centroid_from_probs()</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_concentrations">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_concentrations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">FcAB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">FcAA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">FcBB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">FEA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">FEB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">BO</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.get_concentrations" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_multi_input_line">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_multi_input_line</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.get_multi_input_line" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_path">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">maxkeep</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.PathVector" title="RNA.PathVector"><span class="pre">PathVector</span></a></span></span><a class="headerlink" href="#RNA.get_path" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_pr">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_pr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.get_pr" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.get_xy_coordinates">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">get_xy_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">structure</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.COORDINATE" title="RNA.COORDINATE"><span class="pre">COORDINATE</span></a></span></span><a class="headerlink" href="#RNA.get_xy_coordinates" title="Permalink to this definition"></a></dt>
<dd><p>Compute nucleotide coordinates for secondary structure plot.</p>
<p>This function takes a secondary structure and computes X-Y coordinates for each nucleotide that then
can be used to create a structure plot. The parameter <cite>plot_type</cite> is used to select the underlying
layout algorithm. Currently, the following selections are provided:</p>
<ul class="simple">
<li><p>RNA.PLOT_TYPE_SIMPLE</p></li>
<li><p>RNA.PLOT_TYPE_NAVIEW</p></li>
<li><p>RNA.PLOT_TYPE_CIRCULAR</p></li>
<li><p>RNA.PLOT_TYPE_TURTLE</p></li>
<li><p>RNA.PLOT_TYPE_PUZZLER</p></li>
</ul>
<p>Passing an unsupported selection leads to the default algorithm RNA.PLOT_TYPE_NAVIEW</p>
<p>Here is a simple example how to use this function, assuming variable <cite>structure</cite> contains a valid
dot-bracket string:</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p></li>
<li><p><strong>x</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span> <span class="pre">**</span></code>) – The address of a pointer of X coordinates (pointer will point to memory, or NULL on failure)</p></li>
<li><p><strong>y</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span> <span class="pre">**</span></code>) – The address of a pointer of Y coordinates (pointer will point to memory, or NULL on failure)</p></li>
<li><p><strong>plot_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The layout algorithm to be used</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The length of the structure on success, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plot_coords_pt</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plot_coords_simple</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plot_coords_naview</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plot_coords_circular</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plot_coords_turtle</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plot_coords_puzzler</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>On success, this function allocates memory for X and Y coordinates and assigns the pointers at
addressess <cite>x</cite> and <cite>y</cite> to the corresponding memory locations. It’s the users responsibility to
cleanup this memory after usage!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.gettype">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">gettype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ident</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.gettype" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.gmlRNA">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">gmlRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ssfile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">option</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.gmlRNA" title="Permalink to this definition"></a></dt>
<dd><p>Produce a secondary structure graph in Graph Meta Language (gml) and write it to a file.</p>
<p>If ‘option’ is an uppercase letter the RNA sequence is used to label nodes, if ‘option’ equals <em>‘X’</em>
or <em>‘x’</em> the resulting file will coordinates for an initial layout of the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p></li>
<li><p><strong>ssfile</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The filename of the gml output</p></li>
<li><p><strong>option</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code>) – The option flag</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 on success, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.hamming">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hamming</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.hamming" title="Permalink to this definition"></a></dt>
<dd><p>Calculate hamming distance between two sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The first sequence</p></li>
<li><p><strong>s2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The second sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The hamming distance between s1 and s2</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.hamming_bound">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hamming_bound</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.hamming_bound" title="Permalink to this definition"></a></dt>
<dd><p>Calculate hamming distance between two sequences up to a specified length.</p>
<p>This function is similar to RNA.hamming_distance() but instead of comparing both sequences up to
their actual length only the first ‘n’ characters are taken into account</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The first sequence</p></li>
<li><p><strong>s2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The second sequence</p></li>
<li><p><strong>n</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The length of the subsequences to consider (starting from the 5’ end)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The hamming distance between s1 and s2</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.hamming_distance">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hamming_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.hamming_distance" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.hamming_distance_bound">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hamming_distance_bound</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.hamming_distance_bound" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.hc">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hc</span></span><a class="headerlink" href="#RNA.hc" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The hard constraints data structure.</p>
<p>The content of this data structure determines the decomposition pattern used in the folding
recursions. Attribute ‘matrix’ is used as source for the branching pattern of the decompositions
during all folding recursions. Any entry in matrix[i,j] consists of the 6 LSB that allows one to
distinguish the following types of base pairs:</p>
<ul class="simple">
<li><p>in the exterior loop (RNA.CONSTRAINT_CONTEXT_EXT_LOOP)</p></li>
<li><p>enclosing a hairpin (RNA.CONSTRAINT_CONTEXT_HP_LOOP)</p></li>
<li><p>enclosing an interior loop (RNA.CONSTRAINT_CONTEXT_INT_LOOP)</p></li>
<li><p>enclosed by an exterior loop (RNA.CONSTRAINT_CONTEXT_INT_LOOP_ENC)</p></li>
<li><p>enclosing a multi branch loop (RNA.CONSTRAINT_CONTEXT_MB_LOOP)</p></li>
<li><p>enclosed by a multi branch loop (RNA.CONSTRAINT_CONTEXT_MB_LOOP_ENC)</p></li>
</ul>
<p>The four linear arrays ‘up_xxx’ provide the number of available unpaired nucleotides (including
position i) 3’ of each position in the sequence.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.hc_init" title="RNA.fold_compound.hc_init"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.hc_init</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.hc_free</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_EXT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_HP_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_INT_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_MB_LOOP</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.CONSTRAINT_CONTEXT_MB_LOOP_ENC</span></code></p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.type">
<span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#RNA.hc.type" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_hc_type_e</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.n">
<span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#RNA.hc.n" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.state">
<span class="sig-name descname"><span class="pre">state</span></span><a class="headerlink" href="#RNA.hc.state" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.mx">
<span class="sig-name descname"><span class="pre">mx</span></span><a class="headerlink" href="#RNA.hc.mx" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">char</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.matrix_local">
<span class="sig-name descname"><span class="pre">matrix_local</span></span><a class="headerlink" href="#RNA.hc.matrix_local" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">char</span> <span class="pre">**</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&#64;23</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">union</span> <span class="pre">vrna_hc_s::&#64;22</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.up_ext">
<span class="sig-name descname"><span class="pre">up_ext</span></span><a class="headerlink" href="#RNA.hc.up_ext" title="Permalink to this definition"></a></dt>
<dd><p>A linear array that holds the number of allowed unpaired nucleotides in an exterior loop.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.up_hp">
<span class="sig-name descname"><span class="pre">up_hp</span></span><a class="headerlink" href="#RNA.hc.up_hp" title="Permalink to this definition"></a></dt>
<dd><p>A linear array that holds the number of allowed unpaired nucleotides in a hairpin loop.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.up_int">
<span class="sig-name descname"><span class="pre">up_int</span></span><a class="headerlink" href="#RNA.hc.up_int" title="Permalink to this definition"></a></dt>
<dd><p>A linear array that holds the number of allowed unpaired nucleotides in an interior loop.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.up_ml">
<span class="sig-name descname"><span class="pre">up_ml</span></span><a class="headerlink" href="#RNA.hc.up_ml" title="Permalink to this definition"></a></dt>
<dd><p>A linear array that holds the number of allowed unpaired nucleotides in a multi branched loop.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.f">
<span class="sig-name descname"><span class="pre">f</span></span><a class="headerlink" href="#RNA.hc.f" title="Permalink to this definition"></a></dt>
<dd><p>A function pointer that returns whether or not a certain decomposition may be evaluated.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_hc_eval_f</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.data">
<span class="sig-name descname"><span class="pre">data</span></span><a class="headerlink" href="#RNA.hc.data" title="Permalink to this definition"></a></dt>
<dd><p>A pointer to some structure where the user may store necessary data to evaluate its generic hard
constraint function.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">void</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.free_data">
<span class="sig-name descname"><span class="pre">free_data</span></span><a class="headerlink" href="#RNA.hc.free_data" title="Permalink to this definition"></a></dt>
<dd><p>A pointer to a function to free memory occupied by auxiliary data.</p>
<p>The function this pointer is pointing to will be called upon destruction of the RNA.hc(), and
provided with the RNA.hc().data pointer that may hold auxiliary data. Hence, to avoid leaking
memory, the user may use this pointer to free memory occupied by auxiliary data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_auxdata_free_f</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.hc.depot">
<span class="sig-name descname"><span class="pre">depot</span></span><a class="headerlink" href="#RNA.hc.depot" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_hc_depot_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/constraints/hard.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id98">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mx</span></span><a class="headerlink" href="#id98" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id99">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#id99" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.hc.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.hc.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id100">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#id100" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id101">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">up_ext</span></span><a class="headerlink" href="#id101" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id102">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">up_hp</span></span><a class="headerlink" href="#id102" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id103">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">up_int</span></span><a class="headerlink" href="#id103" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id104">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">up_ml</span></span><a class="headerlink" href="#id104" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.heat_capacity">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">heat_capacity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T_increment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mpoints</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.heat_capacity" title="Permalink to this definition"></a></dt>
<dd><p>Compute the specific heat for an RNA (simplified variant)</p>
<p>Similar to RNA.fold_compound.heat_capacity(), this function computes an RNAs specific heat in a given temperature
range from the partition function by numeric differentiation. This simplified version, however, only
requires the RNA sequence as input instead of a RNA.fold_compound() data structure. The result is
returned as a list of pairs of temperature in C and specific heat in Kcal/(Mol*K).</p>
<p>Users can specify the temperature range for the computation from <cite>T_min</cite> to <cite>T_max</cite>, as well as the
increment step size <cite>T_increment</cite>. The latter also determines how many times the partition function
is computed. Finally, the parameter <cite>mpoints</cite> determines how smooth the curve should be. The
algorithm itself fits a parabola to <span class="math notranslate nohighlight">\(2 \cdot mpoints + 1\)</span> data points to calculate 2nd
derivatives. Increasing this parameter produces a smoother curve.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>heat_capacity()</cite>. If the optional function
arguments <cite>T_min</cite>, <cite>T_max</cite>, <cite>T_increment</cite>, and <cite>mpoints</cite> are omitted, they default to 0.0,
100.0, 1.0 and 2, respectively. See, e.g.  <code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.head_capacity()</span></code> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence input (must be uppercase)</p></li>
<li><p><strong>T_min</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Lowest temperature in C</p></li>
<li><p><strong>T_max</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Highest temperature in C</p></li>
<li><p><strong>T_increment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Stepsize for temperature incrementation in C (a reasonable choice might be 1C)</p></li>
<li><p><strong>mpoints</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The number of interpolation points to calculate 2nd derivative (a reasonable choice might be 2,
min: 1, max: 100)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of pairs of temperatures and corresponding heat capacity or <em>NULL</em> upon any failure. The
last entry of the list is indicated by a <strong>temperature</strong> field set to a value smaller than
<cite>T_min</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.heat_capacity()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.heat_capacity" title="RNA.fold_compound.heat_capacity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.heat_capacity</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.heat_capacity_cb" title="RNA.fold_compound.heat_capacity_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.heat_capacity_cb</span></code></a>, <a class="reference internal" href="#RNA.heat_capacity" title="RNA.heat_capacity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.heat_capacity</span></code></a>, <a class="reference internal" href="#RNA.heat_capacity" title="RNA.heat_capacity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.heat_capacity</span></code></a></p>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.heat_capacity_result">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">heat_capacity_result</span></span><a class="headerlink" href="#RNA.heat_capacity_result" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.heat_capacity_result.heat_capacity">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">heat_capacity</span></span><a class="headerlink" href="#RNA.heat_capacity_result.heat_capacity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.heat_capacity_result.temperature">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">temperature</span></span><a class="headerlink" href="#RNA.heat_capacity_result.temperature" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.heat_capacity_result.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.heat_capacity_result.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.hx">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">up5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">up3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.hx" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.hx.end">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end</span></span><a class="headerlink" href="#RNA.hx.end" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.hx.length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#RNA.hx.length" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.hx.start">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">start</span></span><a class="headerlink" href="#RNA.hx.start" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.hx.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.hx.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.hx.up3">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">up3</span></span><a class="headerlink" href="#RNA.hx.up3" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.hx.up5">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">up5</span></span><a class="headerlink" href="#RNA.hx.up5" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.hx_from_ptable">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hx_from_ptable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.HelixVector" title="RNA.HelixVector"><span class="pre">HelixVector</span></a></span></span><a class="headerlink" href="#RNA.hx_from_ptable" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">hx_from_ptable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.HelixVector" title="RNA.HelixVector"><span class="pre">HelixVector</span></a></span></span></dt>
<dd><p>Convert a pair table representation of a secondary structure into a helix list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – The secondary structure in pair table representation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The secondary structure represented as a helix list</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.hx()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.init_pf_fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">init_pf_fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.init_pf_fold" title="Permalink to this definition"></a></dt>
<dd><p>Allocate space for pf_fold()</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is obsolete and will be removed soon!</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.init_rand">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">init_rand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.init_rand" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the random number generator with a pre-defined seed.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as an overloaded function <strong>init_rand()</strong> where the argument <cite>seed</cite>
is optional. See, e.g.  <a class="reference internal" href="#RNA.init_rand" title="RNA.init_rand"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.init_rand()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seed</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The seed for the random number generator</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.init_rand" title="RNA.init_rand"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.init_rand</span></code></a>, <a class="reference internal" href="#RNA.urn" title="RNA.urn"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.urn</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.initialize_cofold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">initialize_cofold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.initialize_cofold" title="Permalink to this definition"></a></dt>
<dd><p>allocate arrays for folding</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>{This function is obsolete and will be removed soon!}</p>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.intArray">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">intArray</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.intArray" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.intArray.cast">
<span class="sig-name descname"><span class="pre">cast</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.intArray.cast" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.intArray.frompointer">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">frompointer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.intArray.frompointer" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.intArray.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.intArray.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.intArray_frompointer">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">intArray_frompointer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.intArray_frompointer" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.intP_getitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">intP_getitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.intP_getitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.intP_setitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">intP_setitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.intP_setitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.int_urn">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">int_urn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.int_urn" title="Permalink to this definition"></a></dt>
<dd><p>Generates a pseudo random integer in a specified range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The first number in range</p></li>
<li><p><strong>to</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The last number in range</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A pseudo random number in range [from, to]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.urn" title="RNA.urn"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.urn</span></code></a>, <a class="reference internal" href="#RNA.init_rand" title="RNA.init_rand"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.init_rand</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.inverse_fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">inverse_fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">*</span> <span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">target</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.inverse_fold" title="Permalink to this definition"></a></dt>
<dd><p>Find sequences with predefined structure.</p>
<p>This function searches for a sequence with minimum free energy structure provided in the parameter
‘target’, starting with sequence ‘start’. It returns 0 if the search was successful, otherwise a
structure distance in terms of the energy difference between the search result and the actual target
‘target’ is returned. The found sequence is returned in ‘start’. If give_up is set to 1, the
function will return as soon as it is clear that the search will be unsuccessful, this speeds up the
algorithm if you are only interested in exact solutions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The start sequence</p></li>
<li><p><strong>target</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The target secondary structure in dot-bracket notation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The distance to the target in case a search was unsuccessful, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.inverse_pf_fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">inverse_pf_fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">*</span> <span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">target</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.inverse_pf_fold" title="Permalink to this definition"></a></dt>
<dd><p>Find sequence that maximizes probability of a predefined structure.</p>
<p>This function searches for a sequence with maximum probability to fold into the provided structure
‘target’ using the partition function algorithm. It returns <span class="math notranslate nohighlight">\(-kT \cdot \log(p)\)</span> where
<span class="math notranslate nohighlight">\(p\)</span> is the frequency of ‘target’ in the ensemble of possible structures. This is usually much
slower than inverse_fold().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The start sequence</p></li>
<li><p><strong>target</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The target secondary structure in dot-bracket notation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The distance to the target in case a search was unsuccessful, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.last_parameter_file">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">last_parameter_file</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.last_parameter_file" title="Permalink to this definition"></a></dt>
<dd><p>Get the file name of the parameter file that was most recently loaded.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The file name of the last parameter file, or NULL if parameters are still at defaults</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.loop_energy">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">loop_energy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.loop_energy" title="Permalink to this definition"></a></dt>
<dd><p>Calculate energy of a loop.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.eval_loop_pt() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ptable</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – the pair table of the secondary structure</p></li>
<li><p><strong>s</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
<li><p><strong>s1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – encoded RNA sequence</p></li>
<li><p><strong>i</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – position of covering base pair</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>free energy of the loop in 10cal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.eval_loop_pt" title="RNA.fold_compound.eval_loop_pt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.eval_loop_pt</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.loopidx_from_ptable">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">loopidx_from_ptable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.IntVector" title="RNA.IntVector"><span class="pre">IntVector</span></a></span></span><a class="headerlink" href="#RNA.loopidx_from_ptable" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">loopidx_from_ptable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.varArrayInt" title="RNA.varArrayInt"><span class="pre">varArrayInt</span></a></span></span></dt>
<dd><p>Get a loop index representation of a structure.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.make_loop_index">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">make_loop_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.make_loop_index" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.make_tree">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">make_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">struc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.make_tree" title="Permalink to this definition"></a></dt>
<dd><p>Constructs a Tree ( essentially the postorder list ) of the structure ‘struc’, for use in
tree_edit_distance().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>struc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – may be any rooted structure representation.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.maximum_matching">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">maximum_matching</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.maximum_matching" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as global function <cite>maximum_matching()</cite>. See e.g.
<a class="reference internal" href="#RNA.maximum_matching" title="RNA.maximum_matching"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.maximum_matching()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.md">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">md</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.md" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The data structure that contains the complete model details used throughout the calculations.</p>
<p>For convenience reasons, we provide the type name RNA.md() to address this data structure without
the use of the struct keyword</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>RNA.md.reset(), set_model_details(), RNA.md_update(), RNA.md()</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This data structure is wrapped as an object <cite>md</cite> with multiple related functions attached as
methods.</p>
<p>A new set of default parameters can be obtained by calling the constructure of <cite>md:</cite></p>
<ul class="simple">
<li><p><cite>md()</cite>– Initialize with default settings</p></li>
</ul>
<p>The resulting object has a list of attached methods which directly correspond to functions that
mainly operate on the corresponding <em>C</em> data structure:</p>
<ul class="simple">
<li><p><cite>reset()</cite> - RNA.md.reset()</p></li>
<li><p><cite>set_from_globals()</cite> - set_model_details()</p></li>
<li><p><cite>option_string()</cite> - RNA.md.option_string()</p></li>
</ul>
<p>Note, that default parameters can be modified by directly setting any of the following global
variables. Internally, getting/setting default parameters using their global variable
representative translates into calls of the following functions, therefore these wrappers for
these functions do not exist in the scripting language interface(s):</p>
</dd>
</dl>
</div>
<p>global variable</p>
<p><cite>C getter</cite></p>
<p><cite>C setter</cite></p>
<p>temperature</p>
<p>RNA.md_defaults_temperature_get()</p>
<p>RNA.md_defaults_temperature()</p>
<p>dangles</p>
<p>RNA.md_defaults_dangles_get()</p>
<p>RNA.md_defaults_dangles()</p>
<p>betaScale</p>
<p>RNA.md_defaults_betaScale_get()</p>
<p>RNA.md_defaults_betaScale()</p>
<p>tetra_loop</p>
<p>this is an alias of <em>special_hp</em></p>
<p>special_hp</p>
<p>RNA.md_defaults_special_hp_get()</p>
<p>RNA.md_defaults_special_hp()</p>
<p>noLonelyPairs</p>
<p>this is an alias of <em>noLP</em></p>
<p>noLP</p>
<p>RNA.md_defaults_noLP_get()</p>
<p>RNA.md_defaults_noLP()</p>
<p>noGU</p>
<p>RNA.md_defaults_noGU_get()</p>
<p>RNA.md_defaults_noGU()</p>
<p>no_closingGU</p>
<p>this is an alias of <em>noGUclosure</em></p>
<p>noGUclosure</p>
<p>RNA.md_defaults_noGUclosure_get()</p>
<p>RNA.md_defaults_noGUclosure()</p>
<p>logML</p>
<p>RNA.md_defaults_logML_get()</p>
<p>RNA.md_defaults_logML()</p>
<p>circ</p>
<p>RNA.md_defaults_circ_get()</p>
<p>RNA.md_defaults_circ()</p>
<p>gquad</p>
<p>RNA.md_defaults_gquad_get()</p>
<p>RNA.md_defaults_gquad()</p>
<p>uniq_ML</p>
<p>RNA.md_defaults_uniq_ML_get()</p>
<p>RNA.md_defaults_uniq_ML()</p>
<p>energy_set</p>
<p>RNA.md_defaults_energy_set_get()</p>
<p>RNA.md_defaults_energy_set()</p>
<p>backtrack</p>
<p>RNA.md_defaults_backtrack_get()</p>
<p>RNA.md_defaults_backtrack()</p>
<p>backtrack_type</p>
<p>RNA.md_defaults_backtrack_type_get()</p>
<p>RNA.md_defaults_backtrack_type()</p>
<p>do_backtrack</p>
<p>this is an alias of <em>compute_bpp</em></p>
<p>compute_bpp</p>
<p>RNA.md_defaults_compute_bpp_get()</p>
<p>RNA.md_defaults_compute_bpp()</p>
<p>max_bp_span</p>
<p>RNA.md_defaults_max_bp_span_get()</p>
<p>RNA.md_defaults_max_bp_span()</p>
<p>min_loop_size</p>
<p>RNA.md_defaults_min_loop_size_get()</p>
<p>RNA.md_defaults_min_loop_size()</p>
<p>window_size</p>
<p>RNA.md_defaults_window_size_get()</p>
<p>RNA.md_defaults_window_size()</p>
<p>oldAliEn</p>
<p>RNA.md_defaults_oldAliEn_get()</p>
<p>RNA.md_defaults_oldAliEn()</p>
<p>ribo</p>
<p>RNA.md_defaults_ribo_get()</p>
<p>RNA.md_defaults_ribo()</p>
<p>cv_fact</p>
<p>RNA.md_defaults_cv_fact_get()</p>
<p>RNA.md_defaults_cv_fact()</p>
<p>nc_fact</p>
<p>RNA.md_defaults_nc_fact_get()</p>
<p>RNA.md_defaults_nc_fact()</p>
<p>sfact</p>
<p>RNA.md_defaults_sfact_get()</p>
<p>RNA.md_defaults_sfact()</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.temperature">
<span class="sig-name descname"><span class="pre">temperature</span></span><a class="headerlink" href="#RNA.md.temperature" title="Permalink to this definition"></a></dt>
<dd><p>The temperature used to scale the thermodynamic parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.betaScale">
<span class="sig-name descname"><span class="pre">betaScale</span></span><a class="headerlink" href="#RNA.md.betaScale" title="Permalink to this definition"></a></dt>
<dd><p>A scaling factor for the thermodynamic temperature of the Boltzmann factors.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.pf_smooth">
<span class="sig-name descname"><span class="pre">pf_smooth</span></span><a class="headerlink" href="#RNA.md.pf_smooth" title="Permalink to this definition"></a></dt>
<dd><p>A flat specifying whether energies in Boltzmann factors need to be smoothed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.dangles">
<span class="sig-name descname"><span class="pre">dangles</span></span><a class="headerlink" href="#RNA.md.dangles" title="Permalink to this definition"></a></dt>
<dd><p>Specifies the dangle model used in any energy evaluation (0,1,2 or 3)</p>
<p>If set to 0 no stabilizing energies are assigned to bases adjacent to helices in free ends and
multiloops (so called dangling ends). Normally (dangles = 1) dangling end energies are assigned
only to unpaired bases and a base cannot participate simultaneously in two dangling ends. In the
partition function algorithm RNA.fold_compound.pf() these checks are neglected. To provide comparability
between free energy minimization and partition function algorithms, the default setting is 2.
This treatment of dangling ends gives more favorable energies to helices directly adjacent to
one another, which can be beneficial since such helices often do engage in stabilizing
interactions through co-axial stacking.
If set to 3 co-axial stacking is explicitly included for adjacent helices in multiloops. The
option affects only mfe folding and energy evaluation (RNA.mfe() and RNA.eval_structure()), as
well as suboptimal folding (RNA.subopt()) via re-evaluation of energies. Co-axial stacking with
one intervening mismatch is not considered so far. Note, that some function do not implement all
dangle model but only a subset of (0,1,2,3). In particular, partition function algorithms can
only handle 0 and 2. Read the documentation of the particular recurrences or energy evaluation
function for information about the provided dangle model.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.special_hp">
<span class="sig-name descname"><span class="pre">special_hp</span></span><a class="headerlink" href="#RNA.md.special_hp" title="Permalink to this definition"></a></dt>
<dd><p>Include special hairpin contributions for tri, tetra and hexaloops.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.noLP">
<span class="sig-name descname"><span class="pre">noLP</span></span><a class="headerlink" href="#RNA.md.noLP" title="Permalink to this definition"></a></dt>
<dd><p>Only consider canonical structures, i.e. no ‘lonely’ base pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.noGU">
<span class="sig-name descname"><span class="pre">noGU</span></span><a class="headerlink" href="#RNA.md.noGU" title="Permalink to this definition"></a></dt>
<dd><p>Do not allow GU pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.noGUclosure">
<span class="sig-name descname"><span class="pre">noGUclosure</span></span><a class="headerlink" href="#RNA.md.noGUclosure" title="Permalink to this definition"></a></dt>
<dd><p>Do not allow loops to be closed by GU pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.logML">
<span class="sig-name descname"><span class="pre">logML</span></span><a class="headerlink" href="#RNA.md.logML" title="Permalink to this definition"></a></dt>
<dd><p>Use logarithmic scaling for multiloops.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.circ">
<span class="sig-name descname"><span class="pre">circ</span></span><a class="headerlink" href="#RNA.md.circ" title="Permalink to this definition"></a></dt>
<dd><p>Assume RNA to be circular instead of linear.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.gquad">
<span class="sig-name descname"><span class="pre">gquad</span></span><a class="headerlink" href="#RNA.md.gquad" title="Permalink to this definition"></a></dt>
<dd><p>Include G-quadruplexes in structure prediction.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.uniq_ML">
<span class="sig-name descname"><span class="pre">uniq_ML</span></span><a class="headerlink" href="#RNA.md.uniq_ML" title="Permalink to this definition"></a></dt>
<dd><p>Flag to ensure unique multi-branch loop decomposition during folding.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.energy_set">
<span class="sig-name descname"><span class="pre">energy_set</span></span><a class="headerlink" href="#RNA.md.energy_set" title="Permalink to this definition"></a></dt>
<dd><p>Specifies the energy set that defines set of compatible base pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.backtrack">
<span class="sig-name descname"><span class="pre">backtrack</span></span><a class="headerlink" href="#RNA.md.backtrack" title="Permalink to this definition"></a></dt>
<dd><p>Specifies whether or not secondary structures should be backtraced.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.backtrack_type">
<span class="sig-name descname"><span class="pre">backtrack_type</span></span><a class="headerlink" href="#RNA.md.backtrack_type" title="Permalink to this definition"></a></dt>
<dd><p>Specifies in which matrix to backtrack.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.compute_bpp">
<span class="sig-name descname"><span class="pre">compute_bpp</span></span><a class="headerlink" href="#RNA.md.compute_bpp" title="Permalink to this definition"></a></dt>
<dd><p>Specifies whether or not backward recursions for base pair probability (bpp) computation will be
performed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.nonstandards">
<span class="sig-name descname"><span class="pre">nonstandards</span></span><a class="headerlink" href="#RNA.md.nonstandards" title="Permalink to this definition"></a></dt>
<dd><p>contains allowed non standard bases</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.max_bp_span">
<span class="sig-name descname"><span class="pre">max_bp_span</span></span><a class="headerlink" href="#RNA.md.max_bp_span" title="Permalink to this definition"></a></dt>
<dd><p>maximum allowed base pair span</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.min_loop_size">
<span class="sig-name descname"><span class="pre">min_loop_size</span></span><a class="headerlink" href="#RNA.md.min_loop_size" title="Permalink to this definition"></a></dt>
<dd><p>Minimum size of hairpin loops.</p>
<p>The default value for this field is TURN, however, it may be 0 in cofolding context.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.window_size">
<span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#RNA.md.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Size of the sliding window for locally optimal structure prediction.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.oldAliEn">
<span class="sig-name descname"><span class="pre">oldAliEn</span></span><a class="headerlink" href="#RNA.md.oldAliEn" title="Permalink to this definition"></a></dt>
<dd><p>Use old alifold energy model.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.ribo">
<span class="sig-name descname"><span class="pre">ribo</span></span><a class="headerlink" href="#RNA.md.ribo" title="Permalink to this definition"></a></dt>
<dd><p>Use ribosum scoring table in alifold energy model.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.cv_fact">
<span class="sig-name descname"><span class="pre">cv_fact</span></span><a class="headerlink" href="#RNA.md.cv_fact" title="Permalink to this definition"></a></dt>
<dd><p>Co-variance scaling factor for consensus structure prediction.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.nc_fact">
<span class="sig-name descname"><span class="pre">nc_fact</span></span><a class="headerlink" href="#RNA.md.nc_fact" title="Permalink to this definition"></a></dt>
<dd><p>Scaling factor to weight co-variance contributions of non-canonical pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.sfact">
<span class="sig-name descname"><span class="pre">sfact</span></span><a class="headerlink" href="#RNA.md.sfact" title="Permalink to this definition"></a></dt>
<dd><p>Scaling factor for partition function scaling.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.rtype">
<span class="sig-name descname"><span class="pre">rtype</span></span><a class="headerlink" href="#RNA.md.rtype" title="Permalink to this definition"></a></dt>
<dd><p>Reverse base pair type array.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.alias">
<span class="sig-name descname"><span class="pre">alias</span></span><a class="headerlink" href="#RNA.md.alias" title="Permalink to this definition"></a></dt>
<dd><p>alias of an integer nucleotide representation</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">short</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.pair">
<span class="sig-name descname"><span class="pre">pair</span></span><a class="headerlink" href="#RNA.md.pair" title="Permalink to this definition"></a></dt>
<dd><p>Integer representation of a base pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.pair_dist">
<span class="sig-name descname"><span class="pre">pair_dist</span></span><a class="headerlink" href="#RNA.md.pair_dist" title="Permalink to this definition"></a></dt>
<dd><p>Base pair dissimilarity, a.k.a. distance matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.salt">
<span class="sig-name descname"><span class="pre">salt</span></span><a class="headerlink" href="#RNA.md.salt" title="Permalink to this definition"></a></dt>
<dd><p>Salt (monovalent) concentration (M) in buffer.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.saltMLLower">
<span class="sig-name descname"><span class="pre">saltMLLower</span></span><a class="headerlink" href="#RNA.md.saltMLLower" title="Permalink to this definition"></a></dt>
<dd><p>Lower bound of multiloop size to use in loop salt correction linear fitting.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.saltMLUpper">
<span class="sig-name descname"><span class="pre">saltMLUpper</span></span><a class="headerlink" href="#RNA.md.saltMLUpper" title="Permalink to this definition"></a></dt>
<dd><p>Upper bound of multiloop size to use in loop salt correction linear fitting.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.saltDPXInit">
<span class="sig-name descname"><span class="pre">saltDPXInit</span></span><a class="headerlink" href="#RNA.md.saltDPXInit" title="Permalink to this definition"></a></dt>
<dd><p>User-provided salt correction for duplex initialization (in dcal/mol). If set to 99999 the
default salt correction is used. If set to 0 there is no salt correction for duplex
initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.md.saltDPXInitFact">
<span class="sig-name descname"><span class="pre">saltDPXInitFact</span></span><a class="headerlink" href="#RNA.md.saltDPXInitFact" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<p>helical_rise : float</p>
<p>backbone_length : float</p>
<p>C++ includes: ViennaRNA/model.h</p>
<dl class="py property">
<dt class="sig sig-object py" id="id105">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alias</span></span><a class="headerlink" href="#id105" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.md.backbone_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">backbone_length</span></span><a class="headerlink" href="#RNA.md.backbone_length" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id106">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">backtrack</span></span><a class="headerlink" href="#id106" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id107">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">backtrack_type</span></span><a class="headerlink" href="#id107" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id108">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">betaScale</span></span><a class="headerlink" href="#id108" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id109">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">circ</span></span><a class="headerlink" href="#id109" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id110">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">compute_bpp</span></span><a class="headerlink" href="#id110" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id111">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cv_fact</span></span><a class="headerlink" href="#id111" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id112">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dangles</span></span><a class="headerlink" href="#id112" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id113">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">energy_set</span></span><a class="headerlink" href="#id113" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id114">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gquad</span></span><a class="headerlink" href="#id114" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.md.helical_rise">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">helical_rise</span></span><a class="headerlink" href="#RNA.md.helical_rise" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id115">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">logML</span></span><a class="headerlink" href="#id115" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id116">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_bp_span</span></span><a class="headerlink" href="#id116" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id117">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">min_loop_size</span></span><a class="headerlink" href="#id117" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id118">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nc_fact</span></span><a class="headerlink" href="#id118" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id119">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">noGU</span></span><a class="headerlink" href="#id119" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id120">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">noGUclosure</span></span><a class="headerlink" href="#id120" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id121">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">noLP</span></span><a class="headerlink" href="#id121" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id122">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nonstandards</span></span><a class="headerlink" href="#id122" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id123">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">oldAliEn</span></span><a class="headerlink" href="#id123" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.md.option_string">
<span class="sig-name descname"><span class="pre">option_string</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.md.option_string" title="Permalink to this definition"></a></dt>
<dd><p>Get a corresponding commandline parameter string of the options in a RNA.md().</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is not threadsafe!</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id124">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pair</span></span><a class="headerlink" href="#id124" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id125">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pf_smooth</span></span><a class="headerlink" href="#id125" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.md.reset">
<span class="sig-name descname"><span class="pre">reset</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.md.reset" title="Permalink to this definition"></a></dt>
<dd><p>Apply default model details to a provided RNA.md() data structure.</p>
<p>Use this function to initialize a RNA.md() data structure with its default values</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id126">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ribo</span></span><a class="headerlink" href="#id126" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id127">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rtype</span></span><a class="headerlink" href="#id127" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id128">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">salt</span></span><a class="headerlink" href="#id128" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id129">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">saltDPXInit</span></span><a class="headerlink" href="#id129" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id130">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">saltDPXInitFact</span></span><a class="headerlink" href="#id130" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id131">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">saltMLLower</span></span><a class="headerlink" href="#id131" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id132">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">saltMLUpper</span></span><a class="headerlink" href="#id132" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.md.set_from_globals">
<span class="sig-name descname"><span class="pre">set_from_globals</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.md.set_from_globals" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id133">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sfact</span></span><a class="headerlink" href="#id133" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id134">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">special_hp</span></span><a class="headerlink" href="#id134" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id135">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">temperature</span></span><a class="headerlink" href="#id135" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.md.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.md.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id136">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">uniq_ML</span></span><a class="headerlink" href="#id136" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id137">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#id137" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.mean_bp_distance">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">mean_bp_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.mean_bp_distance" title="Permalink to this definition"></a></dt>
<dd><p>Get the mean base pair distance of the last partition function computation.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.mean_bp_distance() or RNA.mean_bp_distance_pr() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>length</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>mean base pair distance in thermodynamic ensemble</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.mean_bp_distance" title="RNA.fold_compound.mean_bp_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.mean_bp_distance</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.mean_bp_distance_pr</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.memmove">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">memmove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indata</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.memmove" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.move">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">move</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos_5</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos_3</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.move" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An atomic representation of the transition / move from one structure to its neighbor.</p>
<p>An atomic transition / move may be one of the following:</p>
<ul class="simple">
<li><p>a <strong>base pair insertion</strong>,</p></li>
<li><p>a <strong>base pair removal</strong>, or</p></li>
<li><p>a <strong>base pair shift</strong> where an existing base pair changes one of its pairing partner.</p></li>
</ul>
<p>These moves are encoded by two integer values that represent the affected 5’ and 3’ nucleotide
positions. Furthermore, we use the following convention on the signedness of these encodings:</p>
<ul class="simple">
<li><p>both values are positive for <em>insertion moves</em></p></li>
<li><p>both values are negative for <em>base pair removals</em></p></li>
<li><p>both values have different signedness for <em>shift moves</em>, where the positive value indicates the
nucleotide that stays constant, and the others absolute value is the new pairing partner</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A value of 0 in either field is used as list-end indicator and doesn’t represent any valid move.</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.move.pos_5">
<span class="sig-name descname"><span class="pre">pos_5</span></span><a class="headerlink" href="#RNA.move.pos_5" title="Permalink to this definition"></a></dt>
<dd><p>The (absolute value of the) 5’ position of a base pair, or any position of a shifted pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.move.pos_3">
<span class="sig-name descname"><span class="pre">pos_3</span></span><a class="headerlink" href="#RNA.move.pos_3" title="Permalink to this definition"></a></dt>
<dd><p>The (absolute value of the) 3’ position of a base pair, or any position of a shifted pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.move.next">
<span class="sig-name descname"><span class="pre">next</span></span><a class="headerlink" href="#RNA.move.next" title="Permalink to this definition"></a></dt>
<dd><p>The next base pair (if an elementary move changes more than one base pair), or <cite>NULL</cite> Has to be
terminated with move 0,0.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_move_t</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/landscape/move.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.move.compare">
<span class="sig-name descname"><span class="pre">compare</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.move.compare" title="Permalink to this definition"></a></dt>
<dd><p>Compare two moves.</p>
<p>The function compares two moves <cite>m</cite> and <cite>b</cite> and returns whether move <cite>m</cite> is lexicographically
smaller (-1), larger (1) or equal to move <cite>b</cite>.</p>
<p>If any of the moves <cite>m</cite> or <cite>b</cite> is a shift move, this comparison only makes sense in a structure
context. Thus, the third argument with the current structure must be provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>b</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">RNA.move()</span> <span class="pre">*</span></code>) – The second move of the comparison</p></li>
<li><p><strong>pt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – The pair table of the current structure that is compatible with both moves (maybe NULL if moves
are guaranteed to be no shifts)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>-1 if <cite>m</cite> &lt; <cite>b</cite>, 1 if <cite>m</cite> &gt; <cite>b</cite>, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Currently, shift moves are not supported!</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function returns 0 (equality) upon any error, e.g. missing input</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.move.is_insertion">
<span class="sig-name descname"><span class="pre">is_insertion</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.move.is_insertion" title="Permalink to this definition"></a></dt>
<dd><p>Test whether a move is a base pair insertion.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero if the move is a base pair insertion, 0 otherwise</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.move.is_removal">
<span class="sig-name descname"><span class="pre">is_removal</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.move.is_removal" title="Permalink to this definition"></a></dt>
<dd><p>Test whether a move is a base pair removal.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero if the move is a base pair removal, 0 otherwise</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.move.is_shift">
<span class="sig-name descname"><span class="pre">is_shift</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.move.is_shift" title="Permalink to this definition"></a></dt>
<dd><p>Test whether a move is a base pair shift.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero if the move is a base pair shift, 0 otherwise</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id138">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pos_3</span></span><a class="headerlink" href="#id138" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id139">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pos_5</span></span><a class="headerlink" href="#id139" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.move.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.move.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.move_standard">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">move_standard</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">*</span> <span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">*</span> <span class="pre">struc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enum</span> <span class="pre">MOVE_TYPE</span> <span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">verbosity_level</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">shifts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">noLP</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.move_standard" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.mx_mfe">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">mx_mfe</span></span><a class="headerlink" href="#RNA.mx_mfe" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.Fc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Fc</span></span><a class="headerlink" href="#RNA.mx_mfe.Fc" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.FcH">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">FcH</span></span><a class="headerlink" href="#RNA.mx_mfe.FcH" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.FcI">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">FcI</span></span><a class="headerlink" href="#RNA.mx_mfe.FcI" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.FcM">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">FcM</span></span><a class="headerlink" href="#RNA.mx_mfe.FcM" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.c">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">c</span></span><a class="headerlink" href="#RNA.mx_mfe.c" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.f3">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">f3</span></span><a class="headerlink" href="#RNA.mx_mfe.f3" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.f5">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">f5</span></span><a class="headerlink" href="#RNA.mx_mfe.f5" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.fM1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fM1</span></span><a class="headerlink" href="#RNA.mx_mfe.fM1" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.fM2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fM2</span></span><a class="headerlink" href="#RNA.mx_mfe.fM2" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.fML">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fML</span></span><a class="headerlink" href="#RNA.mx_mfe.fML" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.ggg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ggg</span></span><a class="headerlink" href="#RNA.mx_mfe.ggg" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#RNA.mx_mfe.length" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.strands">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strands</span></span><a class="headerlink" href="#RNA.mx_mfe.strands" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.mx_mfe.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_mfe.type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#RNA.mx_mfe.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.mx_pf">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">mx_pf</span></span><a class="headerlink" href="#RNA.mx_pf" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.G">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">G</span></span><a class="headerlink" href="#RNA.mx_pf.G" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.expMLbase">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">expMLbase</span></span><a class="headerlink" href="#RNA.mx_pf.expMLbase" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#RNA.mx_pf.length" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.probs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">probs</span></span><a class="headerlink" href="#RNA.mx_pf.probs" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">q</span></span><a class="headerlink" href="#RNA.mx_pf.q" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.q1k">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">q1k</span></span><a class="headerlink" href="#RNA.mx_pf.q1k" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qb">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qb</span></span><a class="headerlink" href="#RNA.mx_pf.qb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qho">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qho</span></span><a class="headerlink" href="#RNA.mx_pf.qho" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qio">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qio</span></span><a class="headerlink" href="#RNA.mx_pf.qio" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qln">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qln</span></span><a class="headerlink" href="#RNA.mx_pf.qln" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qm</span></span><a class="headerlink" href="#RNA.mx_pf.qm" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qm1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qm1</span></span><a class="headerlink" href="#RNA.mx_pf.qm1" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qm2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qm2</span></span><a class="headerlink" href="#RNA.mx_pf.qm2" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qmo">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qmo</span></span><a class="headerlink" href="#RNA.mx_pf.qmo" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.qo">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qo</span></span><a class="headerlink" href="#RNA.mx_pf.qo" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.scale">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">scale</span></span><a class="headerlink" href="#RNA.mx_pf.scale" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.mx_pf.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.mx_pf.type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#RNA.mx_pf.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.my_PS_rna_plot_snoop_a">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">my_PS_rna_plot_snoop_a</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">relative_access</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">StringVector</span> <span class="pre">seqs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.my_PS_rna_plot_snoop_a" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.my_aln_consensus_sequence2">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">my_aln_consensus_sequence2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md_p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.my_aln_consensus_sequence2" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.naview_xy_coordinates">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">naview_xy_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">arg1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.CoordinateVector" title="RNA.CoordinateVector"><span class="pre">CoordinateVector</span></a></span></span><a class="headerlink" href="#RNA.naview_xy_coordinates" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.new_doubleP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">new_doubleP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.new_doubleP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.new_floatP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">new_floatP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.new_floatP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.new_intP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">new_intP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.new_intP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.new_shortP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">new_shortP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.new_shortP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.new_ushortP">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">new_ushortP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nelements</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.new_ushortP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pack_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pack_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">s</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.pack_structure" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.param">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">param</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_details</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.param" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The datastructure that contains temperature scaled energy parameters.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#RNA.param.id" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.stack">
<span class="sig-name descname"><span class="pre">stack</span></span><a class="headerlink" href="#RNA.param.stack" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.hairpin">
<span class="sig-name descname"><span class="pre">hairpin</span></span><a class="headerlink" href="#RNA.param.hairpin" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.bulge">
<span class="sig-name descname"><span class="pre">bulge</span></span><a class="headerlink" href="#RNA.param.bulge" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.internal_loop">
<span class="sig-name descname"><span class="pre">internal_loop</span></span><a class="headerlink" href="#RNA.param.internal_loop" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.mismatchExt">
<span class="sig-name descname"><span class="pre">mismatchExt</span></span><a class="headerlink" href="#RNA.param.mismatchExt" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.mismatchI">
<span class="sig-name descname"><span class="pre">mismatchI</span></span><a class="headerlink" href="#RNA.param.mismatchI" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.mismatch1nI">
<span class="sig-name descname"><span class="pre">mismatch1nI</span></span><a class="headerlink" href="#RNA.param.mismatch1nI" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.mismatch23I">
<span class="sig-name descname"><span class="pre">mismatch23I</span></span><a class="headerlink" href="#RNA.param.mismatch23I" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.mismatchH">
<span class="sig-name descname"><span class="pre">mismatchH</span></span><a class="headerlink" href="#RNA.param.mismatchH" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.mismatchM">
<span class="sig-name descname"><span class="pre">mismatchM</span></span><a class="headerlink" href="#RNA.param.mismatchM" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.dangle5">
<span class="sig-name descname"><span class="pre">dangle5</span></span><a class="headerlink" href="#RNA.param.dangle5" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.dangle3">
<span class="sig-name descname"><span class="pre">dangle3</span></span><a class="headerlink" href="#RNA.param.dangle3" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.int11">
<span class="sig-name descname"><span class="pre">int11</span></span><a class="headerlink" href="#RNA.param.int11" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.int21">
<span class="sig-name descname"><span class="pre">int21</span></span><a class="headerlink" href="#RNA.param.int21" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.int22">
<span class="sig-name descname"><span class="pre">int22</span></span><a class="headerlink" href="#RNA.param.int22" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.ninio">
<span class="sig-name descname"><span class="pre">ninio</span></span><a class="headerlink" href="#RNA.param.ninio" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.lxc">
<span class="sig-name descname"><span class="pre">lxc</span></span><a class="headerlink" href="#RNA.param.lxc" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.MLbase">
<span class="sig-name descname"><span class="pre">MLbase</span></span><a class="headerlink" href="#RNA.param.MLbase" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.MLintern">
<span class="sig-name descname"><span class="pre">MLintern</span></span><a class="headerlink" href="#RNA.param.MLintern" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.MLclosing">
<span class="sig-name descname"><span class="pre">MLclosing</span></span><a class="headerlink" href="#RNA.param.MLclosing" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.TerminalAU">
<span class="sig-name descname"><span class="pre">TerminalAU</span></span><a class="headerlink" href="#RNA.param.TerminalAU" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.DuplexInit">
<span class="sig-name descname"><span class="pre">DuplexInit</span></span><a class="headerlink" href="#RNA.param.DuplexInit" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.Tetraloop_E">
<span class="sig-name descname"><span class="pre">Tetraloop_E</span></span><a class="headerlink" href="#RNA.param.Tetraloop_E" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.Tetraloops">
<span class="sig-name descname"><span class="pre">Tetraloops</span></span><a class="headerlink" href="#RNA.param.Tetraloops" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.Triloop_E">
<span class="sig-name descname"><span class="pre">Triloop_E</span></span><a class="headerlink" href="#RNA.param.Triloop_E" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.Triloops">
<span class="sig-name descname"><span class="pre">Triloops</span></span><a class="headerlink" href="#RNA.param.Triloops" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.Hexaloop_E">
<span class="sig-name descname"><span class="pre">Hexaloop_E</span></span><a class="headerlink" href="#RNA.param.Hexaloop_E" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.Hexaloops">
<span class="sig-name descname"><span class="pre">Hexaloops</span></span><a class="headerlink" href="#RNA.param.Hexaloops" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.TripleC">
<span class="sig-name descname"><span class="pre">TripleC</span></span><a class="headerlink" href="#RNA.param.TripleC" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.MultipleCA">
<span class="sig-name descname"><span class="pre">MultipleCA</span></span><a class="headerlink" href="#RNA.param.MultipleCA" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.MultipleCB">
<span class="sig-name descname"><span class="pre">MultipleCB</span></span><a class="headerlink" href="#RNA.param.MultipleCB" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.gquad">
<span class="sig-name descname"><span class="pre">gquad</span></span><a class="headerlink" href="#RNA.param.gquad" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.gquadLayerMismatch">
<span class="sig-name descname"><span class="pre">gquadLayerMismatch</span></span><a class="headerlink" href="#RNA.param.gquadLayerMismatch" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.gquadLayerMismatchMax">
<span class="sig-name descname"><span class="pre">gquadLayerMismatchMax</span></span><a class="headerlink" href="#RNA.param.gquadLayerMismatchMax" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.temperature">
<span class="sig-name descname"><span class="pre">temperature</span></span><a class="headerlink" href="#RNA.param.temperature" title="Permalink to this definition"></a></dt>
<dd><p>Temperature used for loop contribution scaling.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.model_details">
<span class="sig-name descname"><span class="pre">model_details</span></span><a class="headerlink" href="#RNA.param.model_details" title="Permalink to this definition"></a></dt>
<dd><p>Model details to be used in the recursions.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">vrna_md_t</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.param_file">
<span class="sig-name descname"><span class="pre">param_file</span></span><a class="headerlink" href="#RNA.param.param_file" title="Permalink to this definition"></a></dt>
<dd><p>The filename the parameters were derived from, or empty string if they represent the default.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.SaltStack">
<span class="sig-name descname"><span class="pre">SaltStack</span></span><a class="headerlink" href="#RNA.param.SaltStack" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.SaltLoop">
<span class="sig-name descname"><span class="pre">SaltLoop</span></span><a class="headerlink" href="#RNA.param.SaltLoop" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.SaltLoopDbl">
<span class="sig-name descname"><span class="pre">SaltLoopDbl</span></span><a class="headerlink" href="#RNA.param.SaltLoopDbl" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.SaltMLbase">
<span class="sig-name descname"><span class="pre">SaltMLbase</span></span><a class="headerlink" href="#RNA.param.SaltMLbase" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.SaltMLintern">
<span class="sig-name descname"><span class="pre">SaltMLintern</span></span><a class="headerlink" href="#RNA.param.SaltMLintern" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.SaltMLclosing">
<span class="sig-name descname"><span class="pre">SaltMLclosing</span></span><a class="headerlink" href="#RNA.param.SaltMLclosing" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.param.SaltDPXInit">
<span class="sig-name descname"><span class="pre">SaltDPXInit</span></span><a class="headerlink" href="#RNA.param.SaltDPXInit" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/params/basic.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id140">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DuplexInit</span></span><a class="headerlink" href="#id140" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id141">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Hexaloop_E</span></span><a class="headerlink" href="#id141" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id142">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Hexaloops</span></span><a class="headerlink" href="#id142" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id143">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">MLbase</span></span><a class="headerlink" href="#id143" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id144">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">MLclosing</span></span><a class="headerlink" href="#id144" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id145">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">MLintern</span></span><a class="headerlink" href="#id145" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id146">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">MultipleCA</span></span><a class="headerlink" href="#id146" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id147">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">MultipleCB</span></span><a class="headerlink" href="#id147" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id148">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltDPXInit</span></span><a class="headerlink" href="#id148" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id149">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltLoop</span></span><a class="headerlink" href="#id149" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id150">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltLoopDbl</span></span><a class="headerlink" href="#id150" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id151">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltMLbase</span></span><a class="headerlink" href="#id151" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id152">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltMLclosing</span></span><a class="headerlink" href="#id152" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id153">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltMLintern</span></span><a class="headerlink" href="#id153" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id154">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SaltStack</span></span><a class="headerlink" href="#id154" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id155">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TerminalAU</span></span><a class="headerlink" href="#id155" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id156">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Tetraloop_E</span></span><a class="headerlink" href="#id156" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id157">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Tetraloops</span></span><a class="headerlink" href="#id157" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id158">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Triloop_E</span></span><a class="headerlink" href="#id158" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id159">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Triloops</span></span><a class="headerlink" href="#id159" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id160">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TripleC</span></span><a class="headerlink" href="#id160" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id161">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bulge</span></span><a class="headerlink" href="#id161" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id162">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dangle3</span></span><a class="headerlink" href="#id162" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id163">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dangle5</span></span><a class="headerlink" href="#id163" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id164">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gquad</span></span><a class="headerlink" href="#id164" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id165">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gquadLayerMismatch</span></span><a class="headerlink" href="#id165" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id166">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gquadLayerMismatchMax</span></span><a class="headerlink" href="#id166" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id167">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">hairpin</span></span><a class="headerlink" href="#id167" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id168">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#id168" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id169">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">int11</span></span><a class="headerlink" href="#id169" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id170">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">int21</span></span><a class="headerlink" href="#id170" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id171">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">int22</span></span><a class="headerlink" href="#id171" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id172">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">internal_loop</span></span><a class="headerlink" href="#id172" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id173">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lxc</span></span><a class="headerlink" href="#id173" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id174">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatch1nI</span></span><a class="headerlink" href="#id174" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id175">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatch23I</span></span><a class="headerlink" href="#id175" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id176">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatchExt</span></span><a class="headerlink" href="#id176" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id177">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatchH</span></span><a class="headerlink" href="#id177" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id178">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatchI</span></span><a class="headerlink" href="#id178" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id179">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatchM</span></span><a class="headerlink" href="#id179" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id180">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">model_details</span></span><a class="headerlink" href="#id180" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id181">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ninio</span></span><a class="headerlink" href="#id181" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id182">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">param_file</span></span><a class="headerlink" href="#id182" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id183">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stack</span></span><a class="headerlink" href="#id183" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id184">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">temperature</span></span><a class="headerlink" href="#id184" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.param.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.param.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename=&quot;&quot;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.params_load" title="Permalink to this definition"></a></dt>
<dd><p>Load energy parameters from a file.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>params_load`(fname=””,
options=RNA.PARAMETER_FORMAT_DEFAULT). Here, the empty filename string indicates to load
default RNA parameters, i.e. this is equivalent to calling RNA.params_load_defaults(). See,
e.g.  :py:func:`RNA.fold_compound.params_load()</cite> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span></code>) – The path to the file containing the energy parameters</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – File format bit-mask (usually RNA.PARAMETER_FORMAT_DEFAULT)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_DNA_Mathews1999">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_DNA_Mathews1999</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.params_load_DNA_Mathews1999" title="Permalink to this definition"></a></dt>
<dd><p>Load Mathews 1999 DNA energy parameter set.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>params_load_DNA_Mathews1999()</cite>. See, e.g.
<a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function also resets the default geometric parameters as stored in RNA.md() to those of DNA.
Only subsequently initialized RNA.md() structures will be affected by this change.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_DNA_Mathews2004">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_DNA_Mathews2004</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.params_load_DNA_Mathews2004" title="Permalink to this definition"></a></dt>
<dd><p>Load Mathews 2004 DNA energy parameter set.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>params_load_DNA_Mathews2004()</cite>. See, e.g.
<a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function also resets the default geometric parameters as stored in RNA.md() to those of DNA.
Only subsequently initialized RNA.md() structures will be affected by this change.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_RNA_Andronescu2007">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_RNA_Andronescu2007</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.params_load_RNA_Andronescu2007" title="Permalink to this definition"></a></dt>
<dd><p>Load Andronsecu 2007 RNA energy parameter set.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>params_load_RNA_Andronescu2007()</cite>. See, e.g.
<a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function also resets the default geometric parameters as stored in RNA.md() to those of RNA.
Only subsequently initialized RNA.md() structures will be affected by this change.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_RNA_Langdon2018">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_RNA_Langdon2018</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.params_load_RNA_Langdon2018" title="Permalink to this definition"></a></dt>
<dd><p>Load Langdon 2018 RNA energy parameter set.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>params_load_RNA_Langdon2018()</cite>. See, e.g.
<a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function also resets the default geometric parameters as stored in RNA.md() to those of RNA.
Only subsequently initialized RNA.md() structures will be affected by this change.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_RNA_Turner1999">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_RNA_Turner1999</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.params_load_RNA_Turner1999" title="Permalink to this definition"></a></dt>
<dd><p>Load Turner 1999 RNA energy parameter set.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>params_load_RNA_Turner1999()</cite>. See, e.g.
<a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function also resets the default geometric parameters as stored in RNA.md() to those of RNA.
Only subsequently initialized RNA.md() structures will be affected by this change.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_RNA_Turner2004">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_RNA_Turner2004</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.params_load_RNA_Turner2004" title="Permalink to this definition"></a></dt>
<dd><p>Load Turner 2004 RNA energy parameter set.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>params_load_RNA_Turner2004()</cite>. See, e.g.
<a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function also resets the default geometric parameters as stored in RNA.md() to those of RNA.
Only subsequently initialized RNA.md() structures will be affected by this change.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_RNA_misc_special_hairpins">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_RNA_misc_special_hairpins</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.params_load_RNA_misc_special_hairpins" title="Permalink to this definition"></a></dt>
<dd><p>Load Misc Special Hairpin RNA energy parameter set.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as function <cite>params_load_RNA_misc_special_hairpins()</cite>. See, e.g.
<a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function also resets the default geometric parameters as stored in RNA.md() to those of RNA.
Only subsequently initialized RNA.md() structures will be affected by this change.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_load_from_string" title="RNA.params_load_from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_from_string</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_load_from_string">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_load_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">parameters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">name=&quot;&quot;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.params_load_from_string" title="Permalink to this definition"></a></dt>
<dd><p>Load energy paramters from string.</p>
<p>The string must follow the default energy parameter file convention! The optional <cite>name</cite> argument
allows one to specify a name for the parameter set which is stored internally.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>params_load_from_string`(string, name=””,
options=RNA.PARAMETER_FORMAT_DEFAULT). See, e.g.  :py:func:`RNA.params_load_from_string()</cite> in
the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A 0-terminated string containing energy parameters</p></li>
<li><p><strong>name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A name for the parameter set in <cite>string</cite> (Maybe <cite>NULL</cite>)</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – File format bit-mask (usually RNA.PARAMETER_FORMAT_DEFAULT)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a>, <a class="reference internal" href="#RNA.params_save" title="RNA.params_save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_save</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_defaults</span></code>, <a class="reference internal" href="#RNA.params_load_RNA_Turner2004" title="RNA.params_load_RNA_Turner2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Turner1999" title="RNA.params_load_RNA_Turner1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Turner1999</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Andronescu2007" title="RNA.params_load_RNA_Andronescu2007"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Andronescu2007</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_Langdon2018" title="RNA.params_load_RNA_Langdon2018"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_Langdon2018</span></code></a>, <a class="reference internal" href="#RNA.params_load_RNA_misc_special_hairpins" title="RNA.params_load_RNA_misc_special_hairpins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_RNA_misc_special_hairpins</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews2004" title="RNA.params_load_DNA_Mathews2004"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews2004</span></code></a>, <a class="reference internal" href="#RNA.params_load_DNA_Mathews1999" title="RNA.params_load_DNA_Mathews1999"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load_DNA_Mathews1999</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.params_save">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">params_save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.params_save" title="Permalink to this definition"></a></dt>
<dd><p>Save energy parameters to a file.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>params_save`(fname,
options=RNA.PARAMETER_FORMAT_DEFAULT). See, e.g.  :py:func:`RNA.params_save()</cite> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span></code>) – A filename (path) for the file where the current energy parameters will be written to</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – File format bit-mask (usually RNA.PARAMETER_FORMAT_DEFAULT)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Non-zero on success, 0 on failure</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.params_load" title="RNA.params_load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.params_load</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.parse_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">parse_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.parse_structure" title="Permalink to this definition"></a></dt>
<dd><p>Collects a statistic of structure elements of the full structure in bracket notation.</p>
<p>The function writes to the following global variables: loop_size, loop_degree, helix_size, loops,
pairs, unpaired</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.path">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.path" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.path.en">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">en</span></span><a class="headerlink" href="#RNA.path.en" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.path.move">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">move</span></span><a class="headerlink" href="#RNA.path.move" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.path.s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">s</span></span><a class="headerlink" href="#RNA.path.s" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.path.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.path.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.path.type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#RNA.path.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.path_options">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">path_options</span></span><a class="headerlink" href="#RNA.path_options" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.path_options.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.path_options.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.path_options_findpath">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">path_options_findpath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.path_options_findpath" title="Permalink to this definition"></a></dt>
<dd><p>Create options data structure for findpath direct (re-)folding path heuristic.</p>
<p>This function returns an options data structure that switches the RNA.path_direct() and
RNA.fold_compound.path_direct() API functions to use the <em>findpath</em> <span id="id185">[<a class="reference internal" href="bibliography.html#id31" title="Christoph Flamm, Ivo L Hofacker, Sebastian Maurer-Stroh, Peter F Stadler, and Martin Zehl. Design of multistable RNA molecules. RNA, 7(02):254–265, 2001. doi:10.1017/s1355838201000863.">Flamm <em>et al.</em>, 2001</a>]</span>  heuristic. The
parameter <cite>width</cite> specifies the width of the breadth-first search while the second parameter <cite>type</cite>
allows one to set the type of the returned (re-)folding path.</p>
<p>Currently, the following return types are available:</p>
<ul class="simple">
<li><p>A list of dot-bracket structures and corresponding free energy (flag:
RNA.PATH_TYPE_DOT_BRACKET)</p></li>
<li><p>A list of transition moves and corresponding free energy changes (flag: RNA.PATH_TYPE_MOVES)</p></li>
</ul>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>path_options_findpath()</cite>. The optional
parameter <cite>width</cite> defaults to 10 if omitted, while the optional parameter <cite>type</cite> defaults to
RNA.PATH_TYPE_DOT_BRACKET. See, e.g.  <a class="reference internal" href="#RNA.path_options_findpath" title="RNA.path_options_findpath"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.path_options_findpath()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>width</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Width of the breath-first search strategy</p></li>
<li><p><strong>type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Setting that specifies how the return (re-)folding path should be encoded</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An options data structure with settings for the findpath direct path heuristic</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#RNA.path_options" title="RNA.path_options"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.path_options()</span></code></a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PATH_TYPE_DOT_BRACKET</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PATH_TYPE_MOVES</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.path_options_free</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.path_direct</span></code>, <a class="reference internal" href="#RNA.fold_compound.path_direct" title="RNA.fold_compound.path_direct"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.path_direct</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pbacktrack">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pbacktrack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.pbacktrack" title="Permalink to this definition"></a></dt>
<dd><p>Sample a secondary structure from the Boltzmann ensemble according its probability.</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>st_back has to be set to 1 before calling pf_fold() or pf_fold_par()  pf_fold_par() or pf_fold()
have to be called first to fill the partition function matrices</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A sampled secondary structure in dot-bracket notation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pbacktrack5">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pbacktrack5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.pbacktrack5" title="Permalink to this definition"></a></dt>
<dd><p>Sample a sub-structure from the Boltzmann ensemble according its probability.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pbacktrack_circ">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pbacktrack_circ</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.pbacktrack_circ" title="Permalink to this definition"></a></dt>
<dd><p>Sample a secondary structure of a circular RNA from the Boltzmann ensemble according its
probability.</p>
<p>This function does the same as pbacktrack() but assumes the RNA molecule to be circular</p>
<dl class="simple">
<dt><strong>Precondition</strong></dt><dd><p>st_back has to be set to 1 before calling pf_fold() or pf_fold_par()  pf_fold_par() or
pf_circ_fold() have to be called first to fill the partition function matrices</p>
</dd>
</dl>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.pbacktrack() instead.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A sampled secondary structure in dot-bracket notation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.pbacktrack_mem">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pbacktrack_mem</span></span><a class="headerlink" href="#RNA.pbacktrack_mem" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.pbacktrack_mem.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.pbacktrack_mem.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pf_add">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pf_add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dG1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dG2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kT</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.pf_add" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pf_circ_fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pf_circ_fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.pf_circ_fold" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pf_float_precision">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pf_float_precision</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.pf_float_precision" title="Permalink to this definition"></a></dt>
<dd><p>Find out whether partition function computations are using single precision floating points.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>1 if single precision is used, 0 otherwise</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">double</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pf_fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pf_fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.pf_fold" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pfl_fold">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pfl_fold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">w</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">L</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double</span> <span class="pre">cutoff</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.ElemProbVector" title="RNA.ElemProbVector"><span class="pre">ElemProbVector</span></a></span></span><a class="headerlink" href="#RNA.pfl_fold" title="Permalink to this definition"></a></dt>
<dd><p>Compute base pair probabilities using a sliding-window approach.</p>
<p>This is a simplified wrapper to RNA.fold_compound.probs_window() that given a nucleid acid sequence, a window
size, a maximum base pair span, and a cutoff value computes the pair probabilities for any base pair
in any window. The pair probabilities are returned as a list and the user has to take care to free()
the memory occupied by the list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The nucleic acid input sequence</p></li>
<li><p><strong>window_size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the sliding window</p></li>
<li><p><strong>max_bp_span</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The maximum distance along the backbone between two nucleotides that form a base pairs</p></li>
<li><p><strong>cutoff</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – A cutoff value that omits all pairs with lower probability</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of base pair probabilities, terminated by an entry with RNA.ep().i and RNA.ep().j set
to 0</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.fold_compound.probs_window" title="RNA.fold_compound.probs_window"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.probs_window</span></code></a>, <a class="reference internal" href="#RNA.pfl_fold_cb" title="RNA.pfl_fold_cb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pfl_fold_cb</span></code></a>, <a class="reference internal" href="#RNA.pfl_fold_up" title="RNA.pfl_fold_up"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pfl_fold_up</span></code></a></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function uses default model settings! For custom model settings, we refer to the function
RNA.fold_compound.probs_window().</p>
<blockquote>
<div><p>In case of any computation errors, this function returns <cite>NULL</cite></p>
</div></blockquote>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pfl_fold_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pfl_fold_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">max_bp_span</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.pfl_fold_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pfl_fold_up">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pfl_fold_up</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">ulength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">max_bp_span</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.DoubleDoubleVector" title="RNA.DoubleDoubleVector"><span class="pre">DoubleDoubleVector</span></a></span></span><a class="headerlink" href="#RNA.pfl_fold_up" title="Permalink to this definition"></a></dt>
<dd><p>Compute probability of contiguous unpaired segments.</p>
<p>This is a simplified wrapper to RNA.fold_compound.probs_window() that given a nucleic acid sequence, a maximum
length of unpaired segments (<cite>ulength</cite>), a window size, and a maximum base pair span computes the
equilibrium probability of any segment not exceeding <cite>ulength</cite>. The probabilities to be unpaired are
returned as a 1-based, 2-dimensional matrix with dimensions <span class="math notranslate nohighlight">\(N \times M\)</span>, where <span class="math notranslate nohighlight">\(N\)</span> is
the length of the sequence and <span class="math notranslate nohighlight">\(M\)</span> is the maximum segment length. As an example, the
probability of a segment of size 5 starting at position 100 is stored in the matrix entry
<span class="math notranslate nohighlight">\(X[100][5]\)</span>.</p>
<p>It is the users responsibility to free the memory occupied by this matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The nucleic acid input sequence</p></li>
<li><p><strong>ulength</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The maximal length of an unpaired segment</p></li>
<li><p><strong>window_size</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the sliding window</p></li>
<li><p><strong>max_bp_span</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The maximum distance along the backbone between two nucleotides that form a base pairs</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The probabilities to be unpaired for any segment not exceeding <cite>ulength</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(list-like(double))</span></code></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function uses default model settings! For custom model settings, we refer to the function
RNA.fold_compound.probs_window().</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pfl_fold_up_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pfl_fold_up_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">ulength</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int</span> <span class="pre">max_bp_span</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">PyFunc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyObject</span> <span class="pre">*</span> <span class="pre">data=Py_None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#RNA.pfl_fold_up_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.plist">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">plist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">float</span> <span class="pre">pr</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.ElemProbVector" title="RNA.ElemProbVector"><span class="pre">ElemProbVector</span></a></span></span><a class="headerlink" href="#RNA.plist" title="Permalink to this definition"></a></dt>
<dd><p>Create a RNA.ep() from a dot-bracket string.</p>
<p>The dot-bracket string is parsed and for each base pair an entry in the plist is created. The
probability of each pair in the list is set by a function parameter.</p>
<p>The end of the plist is marked by sequence positions i as well as j equal to 0. This condition
should be used to stop looping over its entries</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>struc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p></li>
<li><p><strong>pr</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – The probability for each base pair used in the plist</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The plist array</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.plot_dp_EPS">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">plot_dp_EPS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.plot_dp_EPS" title="Permalink to this definition"></a></dt>
<dd><p>Produce an encapsulate PostScript (EPS) dot-plot from one or two lists of base pair probabilities.</p>
<p>This function reads two RNA.ep() lists <cite>upper</cite> and <cite>lower</cite> (e.g. base pair probabilities and a
secondary structure) and produces an EPS “dot plot” with filename <cite>‘filename’</cite> where data from
<cite>upper</cite> is placed in the upper-triangular and data from <cite>lower</cite> is placed in the lower triangular
part of the matrix.</p>
<blockquote>
<div><p>For default output, provide the flag RNA.PLOT_PROBABILITIES_DEFAULT as <cite>options</cite> parameter.</p>
</div></blockquote>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded function <cite>plot_dp_EPS()</cite> where the last three
parameters may be omitted. The default values for these parameters are <cite>lower</cite> = NULL, <cite>auxdata</cite>
= NULL, <cite>options</cite> = RNA.PLOT_PROBABILITIES_DEFAULT. See, e.g.  <a class="reference internal" href="#RNA.plot_dp_EPS" title="RNA.plot_dp_EPS"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.plot_dp_EPS()</span></code></a> in
the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A filename for the EPS output</p></li>
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p></li>
<li><p><strong>upper</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code>) – The base pair probabilities for the upper triangular part</p></li>
<li><p><strong>lower</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.ep()</span> <span class="pre">*</span></code>) – The base pair probabilities for the lower triangular part</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – Options indicating which of the input data should be included in the dot-plot</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 if EPS file was successfully written, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.plist" title="RNA.plist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.plist</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.plist_from_probs" title="RNA.fold_compound.plist_from_probs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.plist_from_probs</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.PLOT_PROBABILITIES_DEFAULT</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.print_bppm">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">print_bppm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.print_bppm" title="Permalink to this definition"></a></dt>
<dd><p>print string representation of probability profile</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.print_tree">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">print_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.print_tree" title="Permalink to this definition"></a></dt>
<dd><p>Print a tree (mainly for debugging)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.profile_edit_distance">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">profile_edit_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.profile_edit_distance" title="Permalink to this definition"></a></dt>
<dd><p>Align the 2 probability profiles T1, T2
.</p>
<p>This is like a Needleman-Wunsch alignment, we should really use affine gap-costs ala Gotoh</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.pt_pk_remove">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pt_pk_remove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">IntVector</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.IntVector" title="RNA.IntVector"><span class="pre">IntVector</span></a></span></span><a class="headerlink" href="#RNA.pt_pk_remove" title="Permalink to this definition"></a></dt>
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">pt_pk_remove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">varArrayShort</span> <span class="pre">pt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.varArrayShort" title="RNA.varArrayShort"><span class="pre">varArrayShort</span></a></span></span></dt>
<dd><p>Remove pseudo-knots from a pair table.</p>
<p>This function removes pseudo-knots from an input structure by determining the minimum number of base
pairs that need to be removed to make the structure pseudo-knot free.</p>
<p>To accomplish that, we use a dynamic programming algorithm similar to the Nussinov maxmimum matching
approach.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ptable</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">short</span> <span class="pre">*</span></code>) – Input structure that may include pseudo-knots</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The input structure devoid of pseudo-knots</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.db_pk_remove" title="RNA.db_pk_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_pk_remove</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ptable">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ptable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.varArrayShort" title="RNA.varArrayShort"><span class="pre">varArrayShort</span></a></span></span><a class="headerlink" href="#RNA.ptable" title="Permalink to this definition"></a></dt>
<dd><p>Create a pair table for a secondary structure string.</p>
<p>This function takes an input string of a secondary structure annotation in dot-bracket-notation or
dot-bracket-ext-notation, and converts it into a pair table representation.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This functions is wrapped as overloaded function <cite>ptable()</cite> that takes an optional argument
<cite>options</cite> to specify which type of matching brackets should be considered during conversion. The
default set is round brackets, i.e. RNA.BRACKETS_RND. See, e.g.  <a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.ptable()</span></code></a> in the
<a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – Secondary structure in dot-bracket-ext-notation</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – A bitmask to specify which brackets are recognized during conversion to pair table</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A pointer to a new pair table of the provided secondary structure</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.ptable" title="RNA.ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable</span></code></a>, <a class="reference internal" href="#RNA.db_from_ptable" title="RNA.db_from_ptable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_from_ptable</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_flatten_to</span></code>, <a class="reference internal" href="#RNA.pt_pk_remove" title="RNA.pt_pk_remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.pt_pk_remove</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_ANG</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_CLY</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_SQR</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_ALPHA</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.BRACKETS_ANY</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function also extracts crossing base pairs, i.e. pseudo-knots if more than a single matching
bracket type is allowed through the bitmask <cite>options</cite>.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ptable_pk">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ptable_pk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.IntVector" title="RNA.IntVector"><span class="pre">IntVector</span></a></span></span><a class="headerlink" href="#RNA.ptable_pk" title="Permalink to this definition"></a></dt>
<dd><p>Create a pair table of a secondary structure (pseudo-knot version)</p>
<p>Returns a newly allocated table, such that table[i]=j if (i.j) pair or 0 if i is unpaired, table[0]
contains the length of the structure.</p>
<p>In contrast to RNA.ptable() this function also recognizes the base pairs denoted by ‘[’ and ‘]’
brackets. Thus, this function behaves like</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in (extended) dot-bracket notation</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A pointer to the created pair_table</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.ptable_from_string</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.random_string">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">random_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symbols</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.random_string" title="Permalink to this definition"></a></dt>
<dd><p>Create a random string using characters from a specified symbol set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>l</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The length of the sequence</p></li>
<li><p><strong>symbols</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span></code>) – The symbol set</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A random string of length ‘l’ containing characters from the symbolset</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.read_parameter_file">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">read_parameter_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.read_parameter_file" title="Permalink to this definition"></a></dt>
<dd><p>Read energy parameters from a file.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.params_load() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span></code>) – The path to the file containing the energy parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.read_record">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">read_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">header</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.read_record" title="Permalink to this definition"></a></dt>
<dd><p>Get a data record from stdin.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>This function is deprecated! Use RNA.file_fasta_read_record() as a replacment.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.rotational_symmetry">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">rotational_symmetry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.rotational_symmetry" title="Permalink to this definition"></a></dt>
<dd><p>Determine the order of rotational symmetry for a NULL-terminated string of ASCII characters.</p>
<p>The algorithm applies a fast search of the provided string within itself, assuming the end of the
string wraps around to connect with it’s start. For example, a string of the form <cite>AABAAB</cite> has
rotational symmetry of order <cite>2</cite></p>
<p>If the argument <cite>positions</cite> is not <cite>NULL</cite>, the function stores an array of string start positions
for rotational shifts that map the string back onto itself. This array has length of order of
rotational symmetry, i.e. the number returned by this function. The first element <cite>positions`[0]
always contains a shift value of `0</cite> representing the trivial rotation.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as overloaded global function <cite>rotational_symmetry()</cite>. It merges the
functionalities of RNA.rotational_symmetry(), RNA.rotational_symmetry_pos(),
RNA.rotational_symmetry_num(), and RNA.rotational_symmetry_pos_num(). In contrast to our
C-implementation, this function doesn’t return the order of rotational symmetry as a single
value, but returns a list of cyclic permutation shifts that result in a rotationally symmetric
string. The length of the list then determines the order of rotational symmetry. See, e.g.
<a class="reference internal" href="#RNA.rotational_symmetry" title="RNA.rotational_symmetry"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.rotational_symmetry()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A NULL-terminated string of characters</p></li>
<li><p><strong>positions</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(list-like(unsigned</span> <span class="pre">int))</span></code>) – A pointer to an (undefined) list of alternative string start positions that lead to an identity
mapping (may be NULL)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The order of rotational symmetry</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.rotational_symmetry" title="RNA.rotational_symmetry"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.rotational_symmetry</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.rotational_symmetry_num</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.rotational_symmetry_num_pos</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Do not forget to release the memory occupied by <cite>positions</cite> after a successful execution of this
function.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.salt_duplex_init">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">salt_duplex_init</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">md</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.salt_duplex_init" title="Permalink to this definition"></a></dt>
<dd><p>Get salt correction for duplex initialization at a given salt concentration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – Model details data structure that specfifies salt concentration in buffer (M)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Rounded correction for duplex initialization in dcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.salt_loop">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">salt_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">L</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">salt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backbonelen</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.salt_loop" title="Permalink to this definition"></a></dt>
<dd><p>Get salt correction for a loop at a given salt concentration and temperature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>L</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – backbone number in loop</p></li>
<li><p><strong>salt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – salt concentration (M)</p></li>
<li><p><strong>T</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – absolute temperature (K)</p></li>
<li><p><strong>backbonelen</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – Backbone Length, phosphate-to-phosphate distance (typically 6 for RNA, 6.76 for DNA)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Salt correction for loop in dcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.salt_loop_int">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">salt_loop_int</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">L</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">salt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backbonelen</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.salt_loop_int" title="Permalink to this definition"></a></dt>
<dd><p>Get salt correction for a loop at a given salt concentration and temperature.</p>
<p>This functions is same as RNA.salt_loop but returns rounded salt correction in integer</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>L</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – backbone number in loop</p></li>
<li><p><strong>salt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – salt concentration (M)</p></li>
<li><p><strong>T</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – absolute temperature (K)</p></li>
<li><p><strong>backbonelen</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – Backbone Length, phosphate-to-phosphate distance (typically 6 for RNA, 6.76 for DNA)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Rounded salt correction for loop in dcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.salt_loop" title="RNA.salt_loop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.salt_loop</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.salt_ml">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">salt_ml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">saltLoop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.salt_ml" title="Permalink to this definition"></a></dt>
<dd><p>Fit linear function to loop salt correction.</p>
<p>For a given range of loop size (backbone number), we perform a linear fitting on loop salt
correction</p>
<div class="math notranslate nohighlight">
\[\text{Loop correction} \approx m \cdot L + b.\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>saltLoop</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – List of loop salt correction of size from 1</p></li>
<li><p><strong>lower</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Define the size lower bound for fitting</p></li>
<li><p><strong>upper</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Define the size upper bound for fitting</p></li>
<li><p><strong>m</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code>) – pointer to store the parameter m in fitting result</p></li>
<li><p><strong>b</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span> <span class="pre">*</span></code>) – pointer to store the parameter b in fitting result</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.salt_loop" title="RNA.salt_loop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.salt_loop</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.salt_stack">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">salt_stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">salt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hrise</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.salt_stack" title="Permalink to this definition"></a></dt>
<dd><p>Get salt correction for a stack at a given salt concentration and temperature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>salt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – salt concentration (M)</p></li>
<li><p><strong>T</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – absolute temperature (K)</p></li>
<li><p><strong>hrise</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code>) – Helical Rise (typically 2.8 for RNA, 3.4 for DNA)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Rounded salt correction for stack in dcal/mol</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.sc_add_bt_pycallback">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_add_bt_pycallback</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyFunc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_add_bt_pycallback" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.sc_add_exp_f_pycallback">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_add_exp_f_pycallback</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyFunc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_add_exp_f_pycallback" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.sc_add_f_pycallback">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_add_f_pycallback</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callback</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_add_f_pycallback" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.sc_add_pydata">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_add_pydata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callback</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_add_pydata" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.sc_mod_param">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_mod_param</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">json</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_mod_param" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.available">
<span class="sig-name descname"><span class="pre">available</span></span><a class="headerlink" href="#RNA.sc_mod_param.available" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#RNA.sc_mod_param.name" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.one_letter_code">
<span class="sig-name descname"><span class="pre">one_letter_code</span></span><a class="headerlink" href="#RNA.sc_mod_param.one_letter_code" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.unmodified">
<span class="sig-name descname"><span class="pre">unmodified</span></span><a class="headerlink" href="#RNA.sc_mod_param.unmodified" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.fallback">
<span class="sig-name descname"><span class="pre">fallback</span></span><a class="headerlink" href="#RNA.sc_mod_param.fallback" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.pairing_partners">
<span class="sig-name descname"><span class="pre">pairing_partners</span></span><a class="headerlink" href="#RNA.sc_mod_param.pairing_partners" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">char</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.pairing_partners_encoding">
<span class="sig-name descname"><span class="pre">pairing_partners_encoding</span></span><a class="headerlink" href="#RNA.sc_mod_param.pairing_partners_encoding" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.unmodified_encoding">
<span class="sig-name descname"><span class="pre">unmodified_encoding</span></span><a class="headerlink" href="#RNA.sc_mod_param.unmodified_encoding" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.fallback_encoding">
<span class="sig-name descname"><span class="pre">fallback_encoding</span></span><a class="headerlink" href="#RNA.sc_mod_param.fallback_encoding" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.num_ptypes">
<span class="sig-name descname"><span class="pre">num_ptypes</span></span><a class="headerlink" href="#RNA.sc_mod_param.num_ptypes" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">size_t</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.ptypes">
<span class="sig-name descname"><span class="pre">ptypes</span></span><a class="headerlink" href="#RNA.sc_mod_param.ptypes" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">size_t</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.stack_dG">
<span class="sig-name descname"><span class="pre">stack_dG</span></span><a class="headerlink" href="#RNA.sc_mod_param.stack_dG" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.stack_dH">
<span class="sig-name descname"><span class="pre">stack_dH</span></span><a class="headerlink" href="#RNA.sc_mod_param.stack_dH" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.dangle5_dG">
<span class="sig-name descname"><span class="pre">dangle5_dG</span></span><a class="headerlink" href="#RNA.sc_mod_param.dangle5_dG" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.dangle5_dH">
<span class="sig-name descname"><span class="pre">dangle5_dH</span></span><a class="headerlink" href="#RNA.sc_mod_param.dangle5_dH" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.dangle3_dG">
<span class="sig-name descname"><span class="pre">dangle3_dG</span></span><a class="headerlink" href="#RNA.sc_mod_param.dangle3_dG" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.dangle3_dH">
<span class="sig-name descname"><span class="pre">dangle3_dH</span></span><a class="headerlink" href="#RNA.sc_mod_param.dangle3_dH" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.mismatch_dG">
<span class="sig-name descname"><span class="pre">mismatch_dG</span></span><a class="headerlink" href="#RNA.sc_mod_param.mismatch_dG" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.mismatch_dH">
<span class="sig-name descname"><span class="pre">mismatch_dH</span></span><a class="headerlink" href="#RNA.sc_mod_param.mismatch_dH" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.terminal_dG">
<span class="sig-name descname"><span class="pre">terminal_dG</span></span><a class="headerlink" href="#RNA.sc_mod_param.terminal_dG" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.sc_mod_param.terminal_dH">
<span class="sig-name descname"><span class="pre">terminal_dH</span></span><a class="headerlink" href="#RNA.sc_mod_param.terminal_dH" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.sc_mod_param.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.sc_mod_param.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.sc_mod_parameters_free">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_mod_parameters_free</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_mod_parameters_free" title="Permalink to this definition"></a></dt>
<dd><p>Release memory occupied by a modified base parameter data structure.</p>
<p>Properly free a RNA.sc_mod_param() data structure</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>params</strong> (<a class="reference internal" href="#RNA.sc_mod_param" title="RNA.sc_mod_param"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.sc_mod_param()</span></code></a>) – The data structure to free</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.sc_mod_read_from_json">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_mod_read_from_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">json</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_mod_read_from_json" title="Permalink to this definition"></a></dt>
<dd><p>Parse and extract energy parameters for a modified base from a JSON string.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as an overloaded function <cite>sc_mod_read_from_json()</cite> where the <cite>md</cite>
parameter may be omitted and defaults to <cite>NULL</cite>. See, e.g.
<a class="reference internal" href="#RNA.sc_mod_read_from_json" title="RNA.sc_mod_read_from_json"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.sc_mod_read_from_json()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> – The JSON file containing the specifications of the modified base</p></li>
<li><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – A model-details data structure (for look-up of canonical base pairs)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Parameters of the modified base</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#RNA.sc_mod_param" title="RNA.sc_mod_param"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.sc_mod_param()</span></code></a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.sc_mod_read_from_jsonfile" title="RNA.sc_mod_read_from_jsonfile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_mod_read_from_jsonfile</span></code></a>, <a class="reference internal" href="#RNA.sc_mod_parameters_free" title="RNA.sc_mod_parameters_free"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_mod_parameters_free</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod" title="RNA.fold_compound.sc_mod"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">modified-</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">bases-</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">params</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.sc_mod_read_from_jsonfile">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">sc_mod_read_from_jsonfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.sc_mod_read_from_jsonfile" title="Permalink to this definition"></a></dt>
<dd><p>Parse and extract energy parameters for a modified base from a JSON file.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>This function is available as an overloaded function <cite>sc_mod_read_from_jsonfile()</cite> where the
<cite>md</cite> parameter may be omitted and defaults to <cite>NULL</cite>. See, e.g.
<a class="reference internal" href="#RNA.sc_mod_read_from_jsonfile" title="RNA.sc_mod_read_from_jsonfile"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.sc_mod_read_from_jsonfile()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a> .</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The JSON file containing the specifications of the modified base</p></li>
<li><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – A model-details data structure (for look-up of canonical base pairs)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Parameters of the modified base</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#RNA.sc_mod_param" title="RNA.sc_mod_param"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.sc_mod_param()</span></code></a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.sc_mod_read_from_json" title="RNA.sc_mod_read_from_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_mod_read_from_json</span></code></a>, <a class="reference internal" href="#RNA.sc_mod_parameters_free" title="RNA.sc_mod_parameters_free"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.sc_mod_parameters_free</span></code></a>, <a class="reference internal" href="#RNA.fold_compound.sc_mod" title="RNA.fold_compound.sc_mod"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.fold_compound.sc_mod</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">modified-</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">bases-params</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.seq_encode">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">seq_encode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md</span> <span class="pre">md_p=None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.IntVector" title="RNA.IntVector"><span class="pre">IntVector</span></a></span></span><a class="headerlink" href="#RNA.seq_encode" title="Permalink to this definition"></a></dt>
<dd><p>Get a numerical representation of the nucleotide sequence.</p>
<dl class="simple">
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>In the target scripting language, this function is wrapped as overloaded function <cite>seq_encode()</cite>
where the last parameter, the <cite>model_details</cite> data structure, is optional. If it is omitted,
default model settings are applied, i.e. default nucleotide letter conversion. The wrapped
function returns a list/tuple of integer representations of the input sequence. See, e.g.
<a class="reference internal" href="#RNA.seq_encode" title="RNA.seq_encode"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.seq_encode()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The input sequence in upper-case letters</p></li>
<li><p><strong>md</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">RNA.md()</span> <span class="pre">*</span></code>) – A pointer to a RNA.md() data structure that specifies the conversion type</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of integer encodings for each sequence letter (1-based). Position 0 denotes the length of
the list</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.settype">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">settype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.settype" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.shortP_getitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">shortP_getitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.shortP_getitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.shortP_setitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">shortP_setitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.shortP_setitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.simple_circplot_coordinates">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">simple_circplot_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">std::string</span> <span class="pre">arg1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#RNA.CoordinateVector" title="RNA.CoordinateVector"><span class="pre">CoordinateVector</span></a></span></span><a class="headerlink" href="#RNA.simple_circplot_coordinates" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.simple_xy_coordinates">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">simple_xy_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.simple_xy_coordinates" title="Permalink to this definition"></a></dt>
<dd><p>Calculate nucleotide coordinates for secondary structure plot the <em>Simple way</em></p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Consider switching to RNA.plot_coords_simple_pt() instead!</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_pair_table</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">rna_plot_type</span></code>, <a class="reference internal" href="#RNA.simple_circplot_coordinates" title="RNA.simple_circplot_coordinates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">simple_circplot_coordinates</span></code></a>, <a class="reference internal" href="#RNA.naview_xy_coordinates" title="RNA.naview_xy_coordinates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">naview_xy_coordinates</span></code></a>, <a class="reference internal" href="#RNA.file_PS_rnaplot_a" title="RNA.file_PS_rnaplot_a"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_PS_rnaplot_a</span></code></a>, <a class="reference internal" href="#RNA.file_PS_rnaplot" title="RNA.file_PS_rnaplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.file_PS_rnaplot</span></code></a>, <a class="reference internal" href="#RNA.svg_rna_plot" title="RNA.svg_rna_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">svg_rna_plot</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pair_table</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code>) – The pair table of the secondary structure</p></li>
<li><p><strong>X</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – a pointer to an array with enough allocated space to hold the x coordinates</p></li>
<li><p><strong>Y</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(double)</span></code>) – a pointer to an array with enough allocated space to hold the y coordinates</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>length of sequence on success, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ssv_rna_plot">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ssv_rna_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ssfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ssv_rna_plot" title="Permalink to this definition"></a></dt>
<dd><p>Produce a secondary structure graph in SStructView format.</p>
<p>Write coord file for SStructView</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p></li>
<li><p><strong>ssfile</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The filename of the ssv output</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 on success, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.string_edit_distance">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">string_edit_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.string_edit_distance" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the string edit distance of T1 and T2.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>T1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">swString</span> <span class="pre">*</span></code>) – </p></li>
<li><p><strong>T2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">swString</span> <span class="pre">*</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.strtrim">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">strtrim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">*</span> <span class="pre">seq_mutable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">delimiters=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">keep=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsigned</span> <span class="pre">int</span> <span class="pre">options=</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">unsigned</span> <span class="pre">int</span></span></span><a class="headerlink" href="#RNA.strtrim" title="Permalink to this definition"></a></dt>
<dd><p>Trim a string by removing (multiple) occurences of a particular character.</p>
<p>This function removes (multiple) consecutive occurences of a set of characters (<cite>delimiters</cite>) within
an input string. It may be used to remove leading and/or trailing whitespaces or to restrict the
maximum number of consecutive occurences of the delimiting characters <cite>delimiters</cite>. Setting <cite>keep=0</cite>
removes all occurences, while other values reduce multiple consecutive occurences to at most <cite>keep</cite>
delimiters. This might be useful if one would like to reduce multiple whitespaces to a single one,
or to remove empty fields within a comma-separated value string.</p>
<p>The parameter <cite>delimiters</cite> may be a pointer to a 0-terminated char string containing a set of any
ASCII character. If <em>NULL</em> is passed as delimiter set or an empty char string, all whitespace
characters are trimmed. The <cite>options</cite> parameter is a bit vector that specifies which part of the
string should undergo trimming. The implementation distinguishes the leading (RNA.TRIM_LEADING),
trailing (RNA.TRIM_TRAILING), and in-between (RNA.TRIM_IN_BETWEEN) part with respect to the
delimiter set. Combinations of these parts can be specified by using logical-or operator.</p>
<p>The following example code removes all leading and trailing whitespace characters from the input
string:</p>
<dl>
<dt><strong>SWIG Wrapper Notes</strong></dt><dd><p>Since many scripting languages treat strings as immutable objects, this function does not modify
the input string directly. Instead, it returns the modified string as second return value,
together with the number of removed delimiters.</p>
<p>The scripting language interface provides an overloaded version of this function, with default
parameters <cite>delimiters=NULL</cite>, <cite>keep=0</cite>, and <cite>options=RNA.TRIM_DEFAULT</cite>. See, e.g.
<a class="reference internal" href="#RNA.strtrim" title="RNA.strtrim"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.strtrim()</span></code></a> in the <a class="reference internal" href="#"><span class="doc">Python API</span></a>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The ‘0’-terminated input string to trim</p></li>
<li><p><strong>delimiters</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The delimiter characters as 0-terminated char array (or <em>NULL</em>)</p></li>
<li><p><strong>keep</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The maximum number of consecutive occurences of the delimiter in the output string</p></li>
<li><p><strong>options</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code>) – The option bit vector specifying the mode of operation</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of delimiters removed from the string</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.TRIM_LEADING</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.TRIM_TRAILING</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.TRIM_IN_BETWEEN</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.TRIM_SUBST_BY_FIRST</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.TRIM_DEFAULT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.TRIM_ALL</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The delimiter always consists of a single character from the set of characters provided. In case of
alternative delimiters and non-null <cite>keep</cite> parameter, the first <cite>keep</cite> delimiters are preserved
within the string. Use RNA.TRIM_SUBST_BY_FIRST to substitute all remaining delimiting characters
with the first from the <cite>delimiters</cite> list.</p>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.struct_en">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">struct_en</span></span><a class="headerlink" href="#RNA.struct_en" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Data structure for energy_of_move()</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.struct_en.energy">
<span class="sig-name descname"><span class="pre">energy</span></span><a class="headerlink" href="#RNA.struct_en.energy" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="RNA.struct_en.structure">
<span class="sig-name descname"><span class="pre">structure</span></span><a class="headerlink" href="#RNA.struct_en.structure" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">list-like(int)</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">C++</span> <span class="pre">includes</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">ViennaRNA/move_set.h</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id186">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">energy</span></span><a class="headerlink" href="#id186" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id187">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">structure</span></span><a class="headerlink" href="#id187" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.struct_en.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.struct_en.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.subopt">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">subopt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.subopt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.subopt_solution">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">subopt_solution</span></span><a class="headerlink" href="#RNA.subopt_solution" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="RNA.subopt_solution.energy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">energy</span></span><a class="headerlink" href="#RNA.subopt_solution.energy" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.subopt_solution.structure">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">structure</span></span><a class="headerlink" href="#RNA.subopt_solution.structure" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.subopt_solution.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.subopt_solution.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.svg_rna_plot">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">svg_rna_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ssfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.svg_rna_plot" title="Permalink to this definition"></a></dt>
<dd><p>Produce a secondary structure plot in SVG format and write it to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p></li>
<li><p><strong>ssfile</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The filename of the svg output</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 on success, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.tree_edit_distance">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">tree_edit_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.tree_edit_distance" title="Permalink to this definition"></a></dt>
<dd><p>Calculates the edit distance of the two trees.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>T1</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span> <span class="pre">*</span></code>) – </p></li>
<li><p><strong>T2</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span> <span class="pre">*</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.tree_string_to_db">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">tree_string_to_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.tree_string_to_db" title="Permalink to this definition"></a></dt>
<dd><p>Convert a linear tree string representation of a secondary structure back to Dot-Bracket notation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tree</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A linear tree string representation of a secondary structure</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dot-bracket notation of the secondary structure provided in <cite>tree</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function only accepts <em>Expanded</em> and <em>HIT</em> tree representations!</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.db_to_tree_string" title="RNA.db_to_tree_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_to_tree_string</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.STRUCTURE_TREE_EXPANDED</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.STRUCTURE_TREE_HIT</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">sec_structure_representations_tree</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.tree_string_unweight">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">tree_string_unweight</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.tree_string_unweight" title="Permalink to this definition"></a></dt>
<dd><p>Remove weights from a linear string tree representation of a secondary structure.</p>
<p>This function strips the weights of a linear string tree representation such as <cite>HIT</cite>, or Coarse
Grained Tree sensu  <span id="id188">Shapiro [<a class="reference internal" href="bibliography.html#id27" title="B.A. Shapiro. An algorithm for comparing multiple RNA secondary structures. Computer applications in the biosciences: CABIOS, 4(3):387–393, 1988. doi:10.1093/bioinformatics/4.3.387.">1988</a>]</span></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – A linear string tree representation of a secondary structure with weights</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A linear string tree representation of a secondary structure without weights</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.db_to_tree_string" title="RNA.db_to_tree_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.db_to_tree_string</span></code></a></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ubf_eval_ext_int_loop">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ubf_eval_ext_int_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ubf_eval_ext_int_loop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ubf_eval_int_loop">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ubf_eval_int_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rtype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ij</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ubf_eval_int_loop" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ubf_eval_int_loop2">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ubf_eval_int_loop2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">j1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">si</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rtype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ij</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ss</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">P</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ubf_eval_int_loop2" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ud_set_exp_prod_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ud_set_exp_prod_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prod_cb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eval_cb</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ud_set_exp_prod_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ud_set_prob_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ud_set_prob_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">setter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">getter</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ud_set_prob_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ud_set_prod_cb">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ud_set_prod_cb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prod_cb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eval_cb</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ud_set_prod_cb" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ud_set_pydata">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ud_set_pydata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PyFuncOrNone</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ud_set_pydata" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.unexpand_Full">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">unexpand_Full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ffull</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.unexpand_Full" title="Permalink to this definition"></a></dt>
<dd><p>Restores the bracket notation from an expanded full or HIT tree, that is any tree using only
identifiers ‘U’ ‘P’ and ‘R’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ffull</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.unexpand_aligned_F">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">unexpand_aligned_F</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">align</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.unexpand_aligned_F" title="Permalink to this definition"></a></dt>
<dd><p>Converts two aligned structures in expanded notation.</p>
<p>Takes two aligned structures as produced by tree_edit_distance() function back to bracket notation
with ‘_’ as the gap character. The result overwrites the input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>align</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.unpack_structure">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">unpack_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span> <span class="pre">const</span> <span class="pre">*</span> <span class="pre">packed</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">char</span> <span class="pre">*</span></span></span><a class="headerlink" href="#RNA.unpack_structure" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.unweight">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">unweight</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wcoarse</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.unweight" title="Permalink to this definition"></a></dt>
<dd><p>Strip weights from any weighted tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>wcoarse</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.update_co_pf_params">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">update_co_pf_params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.update_co_pf_params" title="Permalink to this definition"></a></dt>
<dd><p>Recalculate energy parameters.</p>
<p>This function recalculates all energy parameters given the current model settings.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.exp_params_subst() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>length</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Length of the current RNA sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.update_cofold_params">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">update_cofold_params</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.update_cofold_params" title="Permalink to this definition"></a></dt>
<dd><p>Recalculate parameters.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>See RNA.fold_compound.params_subst() for an alternative using the new API</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.update_fold_params">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">update_fold_params</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.update_fold_params" title="Permalink to this definition"></a></dt>
<dd><p>Recalculate energy parameters.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>For non-default model settings use the new API with RNA.fold_compound.params_subst() and</p>
</div>
<p>RNA.fold_compound.mfe() instead!</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.update_pf_params">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">update_pf_params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.update_pf_params" title="Permalink to this definition"></a></dt>
<dd><p>Recalculate energy parameters.</p>
<p>Call this function to recalculate the pair matrix and energy parameters after a change in folding
parameters like temperature</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.fold_compound.exp_params_subst() instead</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.urn">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">urn</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.urn" title="Permalink to this definition"></a></dt>
<dd><p>get a random number from [0..1]</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A random number in range [0..1]</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">double</span></code></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#RNA.int_urn" title="RNA.int_urn"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.int_urn</span></code></a>, <a class="reference internal" href="#RNA.init_rand" title="RNA.init_rand"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.init_rand</span></code></a>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA.init_rand_seed</span></code></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Usually implemented by calling <em>erand48()</em>.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ushortP_getitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ushortP_getitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ushortP_getitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.ushortP_setitem">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">ushortP_setitem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.ushortP_setitem" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.varArrayChar">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">varArrayChar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayChar" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayChar.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayChar.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayChar.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayChar.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.varArrayChar.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.varArrayChar.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayChar.type">
<span class="sig-name descname"><span class="pre">type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayChar.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.varArrayFLTorDBL">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">varArrayFLTorDBL</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayFLTorDBL" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayFLTorDBL.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayFLTorDBL.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayFLTorDBL.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayFLTorDBL.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.varArrayFLTorDBL.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.varArrayFLTorDBL.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayFLTorDBL.type">
<span class="sig-name descname"><span class="pre">type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayFLTorDBL.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.varArrayInt">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">varArrayInt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayInt" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayInt.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayInt.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayInt.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayInt.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.varArrayInt.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.varArrayInt.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayInt.type">
<span class="sig-name descname"><span class="pre">type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayInt.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.varArrayMove">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">varArrayMove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayMove" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayMove.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayMove.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayMove.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayMove.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.varArrayMove.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.varArrayMove.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayMove.type">
<span class="sig-name descname"><span class="pre">type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayMove.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.varArrayShort">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">varArrayShort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayShort" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayShort.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayShort.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayShort.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayShort.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.varArrayShort.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.varArrayShort.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayShort.type">
<span class="sig-name descname"><span class="pre">type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayShort.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.varArrayUChar">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">varArrayUChar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUChar" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayUChar.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUChar.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayUChar.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUChar.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.varArrayUChar.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.varArrayUChar.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayUChar.type">
<span class="sig-name descname"><span class="pre">type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUChar.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.varArrayUInt">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">varArrayUInt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUInt" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayUInt.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUInt.get" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayUInt.size">
<span class="sig-name descname"><span class="pre">size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUInt.size" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="RNA.varArrayUInt.thisown">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thisown</span></span><a class="headerlink" href="#RNA.varArrayUInt.thisown" title="Permalink to this definition"></a></dt>
<dd><p>The membership flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="RNA.varArrayUInt.type">
<span class="sig-name descname"><span class="pre">type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.varArrayUInt.type" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="RNA.var_array_Iterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">var_array_Iterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">var_arr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.var_array_Iterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="RNA.var_array_Iterator.next">
<span class="sig-name descname"><span class="pre">next</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#RNA.var_array_Iterator.next" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.write_parameter_file">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">write_parameter_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.write_parameter_file" title="Permalink to this definition"></a></dt>
<dd><p>Write energy parameters to a file.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>Use RNA.params_save() instead!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fname</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">const</span> <span class="pre">char</span></code>) – A filename (path) for the file where the current energy parameters will be written to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.xrna_plot">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">xrna_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ssfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.xrna_plot" title="Permalink to this definition"></a></dt>
<dd><p>Produce a secondary structure plot for further editing in XRNA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The RNA sequence</p></li>
<li><p><strong>structure</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The secondary structure in dot-bracket notation</p></li>
<li><p><strong>ssfile</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – The filename of the xrna output</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 on success, 0 otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="RNA.zukersubopt">
<span class="sig-prename descclassname"><span class="pre">RNA.</span></span><span class="sig-name descname"><span class="pre">zukersubopt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.zukersubopt" title="Permalink to this definition"></a></dt>
<dd><p>Compute Zuker type suboptimal structures.</p>
<p>Compute Suboptimal structures according to M. Zuker, i.e. for every possible base pair the minimum
energy structure containing the resp. base pair. Returns a list of these structures and their
energies.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 2.6.4: </span>use RNA.zukersubopt() instead</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>string</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">string</span></code>) – RNA sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of zuker suboptimal structures</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">SOLUTION</span> <span class="pre">*</span></code></p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="wrappers.html" class="btn btn-neutral float-left" title="SWIG Wrappers" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="news.html" class="btn btn-neutral float-right" title="News" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>