<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bibliography &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="_static/vrna_32.png"/>
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="How to cite the ViennaRNA Package" href="citing.html" />
    <link rel="prev" title="Changelog" href="changelog.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="manpages.html">Manpages</a></li>
<li class="toctree-l1"><a class="reference internal" href="using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1"><a class="reference internal" href="io.html">I/O Formats</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Bibliography</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bibliography">
<h1>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline"></a></h1>
<p id="id1"><dl class="citation">
<dt class="label" id="id29"><span class="brackets">1</span></dt>
<dd><p>Ronny Lorenz, Stephan H. Bernhart, Christian Höner zu Siederdissen, Hakim Tafer, Christoph Flamm, Peter F. Stadler, and Ivo L. Hofacker. ViennaRNA package 2.0. <em>Algorithms for Molecular Biology</em>, 6(1):26, 2011. <a class="reference external" href="https://doi.org/10.1186/1748-7188-6-26">doi:10.1186/1748-7188-6-26</a>.</p>
</dd>
<dt class="label" id="id19"><span class="brackets">2</span></dt>
<dd><p>I.L. Hofacker, W. Fontana, P.F. Stadler, L.S. Bonhoeffer, M. Tacker, and P. Schuster. Fast folding and comparison of RNA secondary structures. <em>Monatshefte für Chemie/Chemical Monthly</em>, 125(2):167–188, 1994. URL: <a class="reference external" href="https://www.academia.edu/download/48689421/Fast_Folding_and_Comparison_of_RNA_Secon20160908-13624-1yg70az.pdf">https://www.academia.edu/download/48689421/Fast_Folding_and_Comparison_of_RNA_Secon20160908-13624-1yg70az.pdf</a>.</p>
</dd>
<dt class="label" id="id20"><span class="brackets">3</span></dt>
<dd><p>I.L. Hofacker, M. Fekete, and P.F. Stadler. Secondary structure prediction for aligned RNA sequences. <em>Journal of molecular biology</em>, 319(5):1059–1066, 2002. <a class="reference external" href="https://doi.org/10.1016/S0022-2836(02)00308-X">doi:10.1016/S0022-2836(02)00308-X</a>.</p>
</dd>
<dt class="label" id="id22"><span class="brackets">4</span></dt>
<dd><p>S.H. Bernhart, I.L. Hofacker, S. Will, A.R. Gruber, and P.F. Stadler. RNAalifold: improved consensus structure prediction for RNA alignments. <em>BMC bioinformatics</em>, 9(1):474, 2008. <a class="reference external" href="https://doi.org/10.1186/1471-2105-9-474">doi:10.1186/1471-2105-9-474</a>.</p>
</dd>
<dt class="label" id="id52"><span class="brackets">5</span></dt>
<dd><p>Christine E Hajdin, Stanislav Bellaousov, Wayne Huggins, Christopher W Leonard, David H Mathews, and Kevin M Weeks. Accurate SHAPE-directed RNA secondary structure modeling, including pseudoknots. <em>Proceedings of the National Academy of Sciences</em>, 110(14):5498–5503, 2013. <a class="reference external" href="https://doi.org/10.1073/pnas.1219988110">doi:10.1073/pnas.1219988110</a>.</p>
</dd>
<dt class="label" id="id53"><span class="brackets">6</span></dt>
<dd><p>Robert D Jenison, Stanley C Gill, Arthur Pardi, and Barry Polisky. High-resolution molecular discrimination by RNA. <em>Science</em>, 263(5152):1425–1429, 1994. <a class="reference external" href="https://doi.org/10.1126/science.7510417">doi:10.1126/science.7510417</a>.</p>
</dd>
<dt class="label" id="id54"><span class="brackets">7</span></dt>
<dd><p>Amy YQ Zhang, Anthony Bugaut, and Shankar Balasubramanian. A sequence-independent analysis of the loop length dependence of intramolecular RNA G-quadruplex stability and topology. <em>Biochemistry</em>, 50(33):7251–7258, 2011. <a class="reference external" href="https://doi.org/10.1021/bi200805j">doi:10.1021/bi200805j</a>.</p>
</dd>
<dt class="label" id="id55"><span class="brackets">8</span></dt>
<dd><p>Robert A Forties and Ralf Bundschuh. Modeling the interplay of single-stranded binding proteins and nucleic acid secondary structure. <em>Bioinformatics</em>, 26(1):61–67, 2010. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp627">doi:10.1093/bioinformatics/btp627</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets">9</span></dt>
<dd><p>Stefan Washietl, Ivo L. Hofacker, Peter F. Stadler, and Manolis Kellis. RNA folding with soft constraints: reconciliation of probing data and thermodynamics secondary structure prediction. <em>Nucleic Acids Research</em>, 40(10):4261–4272, 2012. <a class="reference external" href="https://doi.org/10.1093/nar/gks009">doi:10.1093/nar/gks009</a>.</p>
</dd>
<dt class="label" id="id33"><span class="brackets">10</span></dt>
<dd><p>S. Wuchty, W. Fontana, I. L. Hofacker, and P. Schuster. Complete suboptimal folding of RNA and the stability of secondary structures. <em>Biopolymers</em>, 49(2):145–165, February 1999. <a class="reference external" href="https://doi.org/10.1002/(SICI)1097-0282(199902)49:2&lt;145::AID-BIP4&gt;3.0.CO;2-G">doi:10.1002/(SICI)1097-0282(199902)49:2&lt;145::AID-BIP4&gt;3.0.CO;2-G</a>.</p>
</dd>
<dt class="label" id="id56"><span class="brackets">11</span></dt>
<dd><p>Ye Ding and Charles E. Lawrence. A statistical sampling algorithm for RNA secondary structure prediction. <em>Nucleic Acids Research</em>, 31(24):7280–7301, 12 2003. <a class="reference external" href="https://doi.org/10.1093/nar/gkg938">doi:10.1093/nar/gkg938</a>.</p>
</dd>
<dt class="label" id="id31"><span class="brackets">12</span></dt>
<dd><p>Christoph Flamm, Ivo L Hofacker, Sebastian Maurer-Stroh, Peter F Stadler, and Martin Zehl. Design of multistable RNA molecules. <em>RNA</em>, 7(02):254–265, 2001. <a class="reference external" href="https://doi.org/10.1017/s1355838201000863">doi:10.1017/s1355838201000863</a>.</p>
</dd>
<dt class="label" id="id30"><span class="brackets">13</span></dt>
<dd><p>Ronny Lorenz, Christoph Flamm, and Ivo L. Hofacker. 2d projections of RNA folding landscapes. In Ivo Grosse, Steffen Neumann, Stefan Posch, Falk Schreiber, and Peter F. Stadler, editors, <em>German Conference on Bioinformatics 2009</em>, volume 157 of Lecture Notes in Informatics, 11–20. Bonn, September 2009. Gesellschaft f. Informatik. URL: <a class="reference external" href="https://dl.gi.de/items/8f88acfe-c389-4dfe-b975-84a638900683">https://dl.gi.de/items/8f88acfe-c389-4dfe-b975-84a638900683</a>.</p>
</dd>
<dt class="label" id="id42"><span class="brackets">14</span></dt>
<dd><p>Robert Giegerich, Björn Voß, and Marc Rehmsmeier. Abstract shapes of RNA. <em>Nucleic Acids Research</em>, 32(16):4843–4851, 2004. <a class="reference external" href="https://doi.org/10.1093/nar/gkh779">doi:10.1093/nar/gkh779</a>.</p>
</dd>
<dt class="label" id="id25"><span class="brackets">15</span></dt>
<dd><p>W. Fontana, P.F. Stadler, E.G. Bornberg-Bauer, T. Griesmacher, I.L. Hofacker, M. Tacker, P. Tarazona, E.D. Weinberger, and P. Schuster. RNA folding and combinatory landscapes. <em>Physical review E</em>, 47(3):2083, 1993. <a class="reference external" href="https://doi.org/10.1103/PhysRevE.47.2083">doi:10.1103/PhysRevE.47.2083</a>.</p>
</dd>
<dt class="label" id="id27"><span class="brackets">16</span></dt>
<dd><p>B.A. Shapiro. An algorithm for comparing multiple RNA secondary structures. <em>Computer applications in the biosciences: CABIOS</em>, 4(3):387–393, 1988. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/4.3.387">doi:10.1093/bioinformatics/4.3.387</a>.</p>
</dd>
<dt class="label" id="id51"><span class="brackets">17</span></dt>
<dd><p>Thomas R Einert and Roland R Netz. Theory for RNA folding, stretching, and melting including loops and salt. <em>Biophysical journal</em>, 100(11):2745–2753, 2011. <a class="reference external" href="https://doi.org/10.1016/j.bpj.2011.04.038">doi:10.1016/j.bpj.2011.04.038</a>.</p>
</dd>
<dt class="label" id="id34"><span class="brackets">18</span></dt>
<dd><p>Ronny Lorenz, Ivo L. Hofacker, and Peter F. Stadler. RNA folding with hard and soft constraints. <em>Algorithms for Molecular Biology</em>, 11(1):1–13, 2016. <a class="reference external" href="https://doi.org/10.1186/s13015-016-0070-z">doi:10.1186/s13015-016-0070-z</a>.</p>
</dd>
<dt class="label" id="id2"><span class="brackets">19</span></dt>
<dd><p>M. Zuker and P. Stiegler. Optimal computer folding of large RNA sequences using thermodynamics and auxiliary information. <em>Nucleic acids research</em>, 9(1):133–148, 1981. <a class="reference external" href="https://doi.org/10.1093/nar/9.1.133">doi:10.1093/nar/9.1.133</a>.</p>
</dd>
<dt class="label" id="id21"><span class="brackets">20</span></dt>
<dd><p>I.L. Hofacker and P.F. Stadler. Memory efficient folding algorithms for circular RNA secondary structures. <em>Bioinformatics</em>, 22(10):1172–1176, 2006. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btl023">doi:10.1093/bioinformatics/btl023</a>.</p>
</dd>
<dt class="label" id="id3"><span class="brackets">21</span></dt>
<dd><p>J.S. McCaskill. The equilibrium partition function and base pair binding probabilities for RNA secondary structure. <em>Biopolymers</em>, 29(6-7):1105–1119, 1990. <a class="reference external" href="https://doi.org/10.1002/bip.360290621">doi:10.1002/bip.360290621</a>.</p>
</dd>
<dt class="label" id="id38"><span class="brackets">22</span></dt>
<dd><p>Stephan H Bernhart, Ivo L Hofacker, and Peter F Stadler. Local RNA base pairing probabilities in large sequences. <em>Bioinformatics</em>, 22(5):614–615, 2005. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btk014">doi:10.1093/bioinformatics/btk014</a>.</p>
</dd>
<dt class="label" id="id37"><span class="brackets">23</span></dt>
<dd><p>Stephan H Bernhart, Ullrike Mückstein, and Ivo L Hofacker. RNA accessibility in cubic time. <em>Algorithms for Molecular Biology</em>, 6(1):3, 2011. <a class="reference external" href="https://doi.org/10.1186/1748-7188-6-3">doi:10.1186/1748-7188-6-3</a>.</p>
</dd>
<dt class="label" id="id32"><span class="brackets">24</span></dt>
<dd><p>M. Zuker. On finding all suboptimal foldings of an RNA molecule. <em>Science</em>, 244(4900):48–52, April 1989. <a class="reference external" href="https://doi.org/10.1126/science.2468181">doi:10.1126/science.2468181</a>.</p>
</dd>
<dt class="label" id="id57"><span class="brackets">25</span></dt>
<dd><p>G Steger, H Hofmann, J Förtsch, HJ Gross, JW Randies, HL Sänger, and D Riesner. Conformational transitions in viroids and virusoids: comparison of results from energy minimization algorithm and from experimental data. <em>Journal of Biomolecular Structure and Dynamics</em>, 2(3):543–571, 1984. <a class="reference external" href="https://doi.org/10.1080/07391102.1984.10507591">doi:10.1080/07391102.1984.10507591</a>.</p>
</dd>
<dt class="label" id="id40"><span class="brackets">26</span></dt>
<dd><p>Juraj Michálik, Hélène Touzet, and Yann Ponty. Efficient approximations of RNA kinetics landscape using non-redundant sampling. <em>Bioinformatics</em>, 33(14):i283–i292, 2017. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btx269">doi:10.1093/bioinformatics/btx269</a>.</p>
</dd>
<dt class="label" id="id23"><span class="brackets">27</span></dt>
<dd><p>S.H. Bernhart, H. Tafer, U. Mückstein, C. Flamm, P.F. Stadler, and I.L. Hofacker. Partition function and base pairing probabilities of RNA heterodimers. <em>Algorithms for Molecular Biology</em>, 1(1):3, 2006. <a class="reference external" href="https://doi.org/10.1186/1748-7188-1-3">doi:10.1186/1748-7188-1-3</a>.</p>
</dd>
<dt class="label" id="id35"><span class="brackets">28</span></dt>
<dd><p>Ronny Lorenz, Dominik Luntzer, Ivo L. Hofacker, Peter F. Stadler, and Michael T. Wolfinger. Shape directed rna folding. <em>Bioinformatics</em>, 32(1):145–147, 2016. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv523">doi:10.1093/bioinformatics/btv523</a>.</p>
</dd>
<dt class="label" id="id5"><span class="brackets">29</span></dt>
<dd><p>Katherine E. Deigan, Tian W. Li, David H. Mathews, and Kevin M. Weeks. Accurate SHAPE-directed RNA structure determination. <em>PNAS</em>, 106:97–102, 2009. <a class="reference external" href="https://doi.org/10.1073/pnas.080692910">doi:10.1073/pnas.080692910</a>.</p>
</dd>
<dt class="label" id="id6"><span class="brackets">30</span></dt>
<dd><p>Kourosh Zarringhalam, Michelle M. Meyer, Ivan Dotu, Jeffrey H. Chuang, and Peter Clote. Integrating chemical footprinting data into RNA secondary structure prediction. <em>PLOS ONE</em>, 2012. <a class="reference external" href="https://doi.org/10.1371/journal.pone.0045160">doi:10.1371/journal.pone.0045160</a>.</p>
</dd>
<dt class="label" id="id50"><span class="brackets">31</span></dt>
<dd><p>Pietro Boccaletto, Filip Stefaniak, Angana Ray, Andrea Cappannini, Sunandan Mukherjee, Elżbieta Purta, Małgorzata Kurkowska, Niloofar Shirvanizadeh, Eliana Destefanis, Paula Groza, and others. MODOMICS: a database of RNA modification pathways. 2021 update. <em>Nucleic Acids Research</em>, 50(D1):D231–D235, 2022. <a class="reference external" href="https://doi.org/10.1093/nar/gkab1083">doi:10.1093/nar/gkab1083</a>.</p>
</dd>
<dt class="label" id="id44"><span class="brackets">32</span></dt>
<dd><p>Elzbieta Kierzek, Xiaoju Zhang, Richard M Watson, Scott D Kennedy, Marta Szabat, Ryszard Kierzek, and David H Mathews. Secondary structure prediction for RNA sequences including N6-methyladenosine. <em>Nature communications</em>, 13(1):1–10, 2022. <a class="reference external" href="https://doi.org/10.1038/s41467-022-28817-4">doi:10.1038/s41467-022-28817-4</a>.</p>
</dd>
<dt class="label" id="id43"><span class="brackets">33</span></dt>
<dd><p>Graham A Hudson, Richard J Bloomingdale, and Brent M Znosko. Thermodynamic contribution and nearest-neighbor parameters of pseudouridine-adenosine base pairs in oligoribonucleotides. <em>RNA</em>, 19(11):1474–1482, 2013. <a class="reference external" href="https://doi.org/10.1261/rna.039610.113">doi:10.1261/rna.039610.113</a>.</p>
</dd>
<dt class="label" id="id45"><span class="brackets">34</span></dt>
<dd><p>Daniel J Wright, Jamie L Rice, Dawn M Yanker, and Brent M Znosko. Nearest Neighbor Parameters for Inosine· Uridine Pairs in RNA Duplexes. <em>Biochemistry</em>, 46(15):4625–4634, 2007. <a class="reference external" href="https://doi.org/10.1021/bi0616910">doi:10.1021/bi0616910</a>.</p>
</dd>
<dt class="label" id="id46"><span class="brackets">35</span></dt>
<dd><p>Daniel J Wright, Christopher R Force, and Brent M Znosko. Stability of RNA duplexes containing inosine· cytosine pairs. <em>Nucleic Acids Research</em>, 46(22):12099–12108, 2018. <a class="reference external" href="https://doi.org/10.1093/nar/gky907">doi:10.1093/nar/gky907</a>.</p>
</dd>
<dt class="label" id="id47"><span class="brackets">36</span></dt>
<dd><p>Katherine E Richardson and Brent M Znosko. Nearest-neighbor parameters for 7-deaza-adenosine· uridine base pairs in RNA duplexes. <em>RNA</em>, 22(6):934–942, 2016. <a class="reference external" href="https://doi.org/10.1261/rna.055277.115">doi:10.1261/rna.055277.115</a>.</p>
</dd>
<dt class="label" id="id48"><span class="brackets">37</span></dt>
<dd><p>Elizabeth A Jolley and Brent M Znosko. The loss of a hydrogen bond: Thermodynamic contributions of a non-standard nucleotide. <em>Nucleic acids research</em>, 45(3):1479–1487, 2017. <a class="reference external" href="https://doi.org/10.1093/nar/gkw830">doi:10.1093/nar/gkw830</a>.</p>
</dd>
<dt class="label" id="id39"><span class="brackets">38</span></dt>
<dd><p>Eva Freyhult, Vincent Moulton, and Paul Gardner. Predicting RNA structure using mutual information. <em>Applied bioinformatics</em>, 4(1):53–59, 2005. <a class="reference external" href="https://doi.org/10.2165/00822942-200504010-00006">doi:10.2165/00822942-200504010-00006</a>.</p>
</dd>
<dt class="label" id="id41"><span class="brackets">39</span></dt>
<dd><p>Daniel Wiegreffe, Daniel Alexander, Peter F Stadler, and Dirk Zeckzer. RNApuzzler: efficient outerplanar drawing of RNA-secondary structures. <em>Bioinformatics</em>, 35(8):1342–1349, 2019. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty817">doi:10.1093/bioinformatics/bty817</a>.</p>
</dd>
<dt class="label" id="id15"><span class="brackets">40</span></dt>
<dd><p>R.E. Bruccoleri and G. Heinrich. An improved algorithm for nucleic acid secondary structure display. <em>Computer applications in the biosciences: CABIOS</em>, 4(1):167–173, 1988. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/4.1.167">doi:10.1093/bioinformatics/4.1.167</a>.</p>
</dd>
<dt class="label" id="id36"><span class="brackets">41</span></dt>
<dd><p>Joe Sawada. A fast algorithm to generate necklaces with fixed content. <em>Theoretical Computer Science</em>, 301(1):477–489, 2003. <a class="reference external" href="https://doi.org/10.1016/S0304-3975(03)00049-5">doi:10.1016/S0304-3975(03)00049-5</a>.</p>
</dd>
</dl>
</p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="changelog.html" class="btn btn-neutral float-left" title="Changelog" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="citing.html" class="btn btn-neutral float-right" title="How to cite the ViennaRNA Package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>