<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Changelog &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="_static/vrna_32.png"/>
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
        <script defer="defer" src="_static/js/mathjax/tex-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]]}, "options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bibliography" href="bibliography.html" />
    <link rel="prev" title="News" href="news.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="manpages.html">Manpages</a></li>
<li class="toctree-l1"><a class="reference internal" href="using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1"><a class="reference internal" href="io.html">I/O Formats</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="news.html">What's new</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Changelog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#version-2-6-x">Version 2.6.x</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#unreleased">Unreleased</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-6-4">Version 2.6.4</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#programs">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#library">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#package">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-6-3">Version 2.6.3</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id2">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-6-2-release-date-2023-06-21">Version 2.6.2 (Release date: 2023-06-21)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id3">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id4">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id5">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-6-1-release-date-2023-06-12">Version 2.6.1 (Release date: 2023-06-12)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id6">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id7">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-6-0-release-date-2023-06-09">Version 2.6.0 (Release date: 2023-06-09)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id8">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id9">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id10">Package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#version-2-5-x">Version 2.5.x</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#version-2-5-1-release-date-2022-06-02">Version 2.5.1 (Release date: 2022-06-02)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id11">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id12">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id13">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-5-0-release-date-2021-11-08">Version 2.5.0 (Release date: 2021-11-08)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id14">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id15">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id16">Package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#version-2-4-x">Version 2.4.x</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-18-release-date-2021-04-22">Version 2.4.18 (Release date: 2021-04-22)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id17">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id18">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id19">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-17-release-date-2020-11-25">Version 2.4.17 (Release date: 2020-11-25)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id20">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id21">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id22">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-16-release-date-2020-10-09">Version 2.4.16 (Release date: 2020-10-09)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id23">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id24">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id25">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-15-release-date-2020-08-18">Version 2.4.15 (Release date: 2020-08-18)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id26">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id27">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id28">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-14-release-date-2019-08-13">Version 2.4.14 (Release date: 2019-08-13)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id29">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id30">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id31">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-13-release-date-2019-05-30">Version 2.4.13 (Release date: 2019-05-30)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id32">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id33">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id34">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-12-release-date-2019-04-16">Version 2.4.12 (Release date: 2019-04-16)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id35">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id36">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id37">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-11-release-date-2018-12-17">Version 2.4.11 (Release date: 2018-12-17)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id38">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id39">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id40">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-10-release-date-2018-09-26">Version 2.4.10 (Release date: 2018-09-26)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id41">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id42">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id43">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-9-release-date-2018-07-11">Version 2.4.9 (Release date: 2018-07-11)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id44">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id45">Library</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-8-release-date-2018-06-23">Version 2.4.8 (Release date: 2018-06-23)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id46">Programs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id47">Library</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id48">Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-7-release-date-2018-06-13">Version 2.4.7 (Release date: 2018-06-13)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-6-release-date-2018-04-19">Version 2.4.6 (Release date: 2018-04-19)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-5-release-date-2018-04-17">Version 2.4.5 (Release date: 2018-04-17)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-4-release-date-2018-03-06">Version 2.4.4 (Release date: 2018-03-06)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-3-release-date-2017-11-14">Version 2.4.3 (Release date: 2017-11-14)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-2-release-date-2017-10-13">Version 2.4.2 (Release date: 2017-10-13)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-1-release-date-2017-08-23">Version 2.4.1 (Release date: 2017-08-23)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-4-0-release-date-2017-08-01">Version 2.4.0 (Release date: 2017-08-01)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#version-2-3-x">Version 2.3.x</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#version-2-3-5-release-date-2017-04-14">Version 2.3.5 (Release date: 2017-04-14)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-3-4-release-date-2017-03-10">Version 2.3.4 (Release date: 2017-03-10)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-3-3-release-date-2017-01-24">Version 2.3.3 (Release date: 2017-01-24)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-3-2-release-date-2017-01-18">Version 2.3.2 (Release date: 2017-01-18)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-3-1-release-date-2016-11-15">Version 2.3.1 (Release date: 2016-11-15)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-3-0-release-date-2016-11-01">Version 2.3.0 (Release date: 2016-11-01)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#version-2-2-x">Version 2.2.x</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-10-release-date-2016-09-06">Version 2.2.10 (Release date: 2016-09-06)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-9-release-date-2016-09-01">Version 2.2.9 (Release date: 2016-09-01)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-8-release-date-2016-08-01">Version 2.2.8 (Release date: 2016-08-01)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-7-release-date-2016-06-30">Version 2.2.7 (Release date: 2016-06-30)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-6-release-date-2016-06-19">Version 2.2.6 (Release date: 2016-06-19)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-5-release-date-2016-04-09">Version 2.2.5 (Release date: 2016-04-09)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-4-release-date-2016-02-19">Version 2.2.4 (Release date: 2016-02-19)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-3-release-date-2016-02-13">Version 2.2.3 (Release date: 2016-02-13)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-2-release-date-2016-02-08">Version 2.2.2 (Release date: 2016-02-08)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-1-release-date-2016-02-06">Version 2.2.1 (Release date: 2016-02-06)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-2-0-release-date-2016-01-25">Version 2.2.0 (Release date: 2016-01-25)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#version-2-1-x">Version 2.1.x</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-9">Version 2.1.9</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-8">Version 2.1.8</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-7">Version 2.1.7</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-6">Version 2.1.6</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-5">Version 2.1.5</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-4">Version 2.1.4</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-3">Version 2.1.3</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-2">Version 2.1.2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-1">Version 2.1.1</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-1-0">Version 2.1.0</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#version-2-0-x">Version 2.0.x</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#version-2-0-7">Version 2.0.7</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-0-6">Version 2.0.6</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-0-5">Version 2.0.5</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-0-4">Version 2.0.4</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-0-3">Version 2.0.3</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-0-2">Version 2.0.2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#version-2-0-1">Version 2.0.1</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#history">History</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Changelog</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section class="tex2jax_ignore mathjax_ignore" id="changelog">
<h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this headline"></a></h1>
<p>Below, you’ll find a list of notable changes for each version of the ViennaRNA Package.</p>
<section id="version-2-6-x">
<h2>Version 2.6.x<a class="headerlink" href="#version-2-6-x" title="Permalink to this headline"></a></h2>
<section id="unreleased">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.6.4...HEAD">Unreleased</a><a class="headerlink" href="#unreleased" title="Permalink to this headline"></a></h3>
</section>
<section id="version-2-6-4">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.6.3...v2.6.4">Version 2.6.4</a><a class="headerlink" href="#version-2-6-4" title="Permalink to this headline"></a></h3>
<section id="programs">
<h4>Programs<a class="headerlink" href="#programs" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix C++17 compilation issue with <code class="docutils literal notranslate"><span class="pre">kinwalker</span></code></p></li>
<li><p>Fix potential compilation issues with C++20 in <code class="docutils literal notranslate"><span class="pre">RNAforester</span></code> frontend</p></li>
<li><p>Refactor and correct spelling issues in man pages for several executable programs</p></li>
</ul>
</section>
<section id="library">
<h4>Library<a class="headerlink" href="#library" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Add shift move support to <code class="docutils literal notranslate"><span class="pre">vrna_move_neighbor_diff*()</span></code> functions</p></li>
<li><p>API: Fix char array initialization in <code class="docutils literal notranslate"><span class="pre">snoop.c</span></code></p></li>
<li><p>API: Fix potentially leaking file pointer in <code class="docutils literal notranslate"><span class="pre">vrna_file_msa_read()</span></code></p></li>
<li><p>API: Fix potentially leaking memory in <code class="docutils literal notranslate"><span class="pre">rnaplot_EPS()</span></code></p></li>
<li><p>API: Fix potential use of uninitialized variable in <code class="docutils literal notranslate"><span class="pre">vrna_rotational_symmetry_db_pos()</span></code></p></li>
<li><p>API: Fix soft constraints issue in external loop of <code class="docutils literal notranslate"><span class="pre">vrna_subopt*()</span></code></p></li>
<li><p>SWIG: Add swig class output parameter typemap for Python</p></li>
<li><p>SWIG: Add <code class="docutils literal notranslate"><span class="pre">__hash__()</span></code> and <code class="docutils literal notranslate"><span class="pre">__eq__()</span></code> methods for wrapped <code class="docutils literal notranslate"><span class="pre">_vrna_move_t</span></code> in Python</p></li>
<li><p>SWIG: Return <code class="docutils literal notranslate"><span class="pre">var_array&lt;vrna_move_t&gt;</span></code> objects in Python wrapped <code class="docutils literal notranslate"><span class="pre">vrna_neighbors()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_move_neighbor_diff()</span></code></p></li>
<li><p>SWIG: Refactor file handle wrapping between Python 3 and C</p></li>
<li><p>SWIG: Fix <code class="docutils literal notranslate"><span class="pre">var_array</span></code> Python slices and associated memory leak</p></li>
<li><p>SWIG: Fix bogus <code class="docutils literal notranslate"><span class="pre">delete/free()</span></code> calls in swig interface</p></li>
<li><p>Add requirements to build <code class="docutils literal notranslate"><span class="pre">RNAlib</span></code> with MSVC for Windows</p></li>
<li><p>Remove unused code in <code class="docutils literal notranslate"><span class="pre">RNApuzzler</span></code></p></li>
</ul>
</section>
<section id="package">
<h4>Package<a class="headerlink" href="#package" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>DOC: Transition reference manual from <code class="docutils literal notranslate"><span class="pre">doxygen</span></code> to <code class="docutils literal notranslate"><span class="pre">sphinx</span></code> via <code class="docutils literal notranslate"><span class="pre">breathe</span></code> bridge</p></li>
<li><p>DOC: Merge documentation of C-API and Python API</p></li>
<li><p>DOC: Merge parts of tutorial into reference manual</p></li>
<li><p>AUTOCONF: Refactor autoconf checks for capability to build reference manual</p></li>
<li><p>AUTOCONF: Deactivate build of <code class="docutils literal notranslate"><span class="pre">RNAxplorer</span></code> if <code class="docutils literal notranslate"><span class="pre">lapack</span></code> requirements are not met</p></li>
</ul>
</section>
</section>
<section id="version-2-6-3">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.6.2...v2.6.3">Version 2.6.3</a><a class="headerlink" href="#version-2-6-3" title="Permalink to this headline"></a></h3>
<section id="id1">
<h4>Library<a class="headerlink" href="#id1" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Make JSON parser integral part of ViennaRNA library</p></li>
<li><p>API: Move modified energy parameters into ‘modified_base’ object in JSON file(s)</p></li>
<li><p>SWIG: Enable stand-alone build of Python interface (for PyPI)</p></li>
</ul>
</section>
<section id="id2">
<h4>Package<a class="headerlink" href="#id2" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Add enthalpy and terminal end values for predicted stacks with dihydrouridine</p></li>
<li><p>TESTS: Allow for using <code class="docutils literal notranslate"><span class="pre">pytest</span></code> to test the Python 3 interface</p></li>
</ul>
</section>
</section>
<section id="version-2-6-2-release-date-2023-06-21">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.6.1...v2.6.2">Version 2.6.2</a> (Release date: 2023-06-21)<a class="headerlink" href="#version-2-6-2-release-date-2023-06-21" title="Permalink to this headline"></a></h3>
<section id="id3">
<h4>Programs<a class="headerlink" href="#id3" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix preparation of input sequences for modified base support in <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code></p></li>
</ul>
</section>
<section id="id4">
<h4>Library<a class="headerlink" href="#id4" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix energy corrections for modified base support when unmodified base is not the same as fallback base, e.g. in the case of inosine</p></li>
<li><p>Add soft constraints to multifold external loop decomposition</p></li>
<li><p>Add soft constraints preparation stage callback</p></li>
<li><p>SWIG: Fix <code class="docutils literal notranslate"><span class="pre">fc.sc_add_bp()</span></code> propagation of constraint values</p></li>
<li><p>SWIG: Wrap energy parameter file strings</p></li>
</ul>
</section>
<section id="id5">
<h4>Package<a class="headerlink" href="#id5" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>TESTS: Add modified base tests on duplex data with I-C and A-Psi pairs from publications</p></li>
</ul>
</section>
</section>
<section id="version-2-6-1-release-date-2023-06-12">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.6.0...v2.6.1">Version 2.6.1</a> (Release date: 2023-06-12)<a class="headerlink" href="#version-2-6-1-release-date-2023-06-12" title="Permalink to this headline"></a></h3>
<section id="id6">
<h4>Programs<a class="headerlink" href="#id6" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix double free corruption in <code class="docutils literal notranslate"><span class="pre">RNAdos</span></code></p></li>
<li><p>Fix compilation issues due to use of <code class="docutils literal notranslate"><span class="pre">uint</span></code> instead of <code class="docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code> for <code class="docutils literal notranslate"><span class="pre">RNAxplorer</span></code></p></li>
<li><p>Fix compilation issues for <code class="docutils literal notranslate"><span class="pre">RNAxplorer</span></code> when OpenMP is unavailable</p></li>
</ul>
</section>
<section id="id7">
<h4>Package<a class="headerlink" href="#id7" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>AUTOCONF: Update autoconf macros</p></li>
<li><p>Update Debian-based packaging rules</p></li>
</ul>
</section>
</section>
<section id="version-2-6-0-release-date-2023-06-09">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.5.1...v2.6.0">Version 2.6.0</a> (Release date: 2023-06-09)<a class="headerlink" href="#version-2-6-0-release-date-2023-06-09" title="Permalink to this headline"></a></h3>
<section id="id8">
<h4>Programs<a class="headerlink" href="#id8" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Add modified base input support to <code class="docutils literal notranslate"><span class="pre">RNAfold</span></code></p></li>
<li><p>Add modified base input support to <code class="docutils literal notranslate"><span class="pre">RNAplfold</span></code></p></li>
<li><p>Add modified base input support to <code class="docutils literal notranslate"><span class="pre">RNALfold</span></code></p></li>
<li><p>Add modified base input support to <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code></p></li>
<li><p>Add modified base input support to <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code></p></li>
<li><p>Fix missing strand separators in <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code> when applied to multiple interacting sequences</p></li>
<li><p>Fix sorted output in <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code> with <code class="docutils literal notranslate"><span class="pre">--gquad</span></code> option</p></li>
<li><p>Allow for only <code class="docutils literal notranslate"><span class="pre">-Fp</span></code> in <code class="docutils literal notranslate"><span class="pre">RNAinverse</span></code> instead of always activating <code class="docutils literal notranslate"><span class="pre">-Fm</span></code></p></li>
<li><p>Fix default value of <code class="docutils literal notranslate"><span class="pre">RNAinverse</span> <span class="pre">-R</span></code> option in manpage</p></li>
<li><p>Restructure <code class="docutils literal notranslate"><span class="pre">--*help</span></code> output and man pages for most executable programs</p></li>
<li><p>Allow for cation concentration (Na+) changes in most executable programs (default 1.021M)</p></li>
<li><p>Allow for at least as many threads as CPUs are configured if maximum thread number detection fails</p></li>
<li><p>Fix alignment input parsing in <code class="docutils literal notranslate"><span class="pre">refold.pl</span></code></p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">RNAxplorer</span></code> program to the distribution</p></li>
</ul>
</section>
<section id="id9">
<h4>Library<a class="headerlink" href="#id9" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Extend <code class="docutils literal notranslate"><span class="pre">model_details</span></code> to allow for salt concentration changes</p></li>
<li><p>API: Add functions for salt concentration change derived energy corrections in <code class="docutils literal notranslate"><span class="pre">ViennaRNA/params/salt.h</span></code></p></li>
<li><p>API: Add arbitrary modified base support (<code class="docutils literal notranslate"><span class="pre">vrna_sc_mod()</span></code>) via soft constraints mechanism and JSON input data</p></li>
<li><p>API: Add Pseuoduridine-A parameters via soft constraints callback</p></li>
<li><p>API: Add Dihydrouridine parameters via soft constraints callback</p></li>
<li><p>API: Add inosine-U and inosine-C parameters via soft constraints callback</p></li>
<li><p>API: Add m6A parameters via soft constraints callback mechanism</p></li>
<li><p>API: Add 7DA modification support via soft constraints</p></li>
<li><p>API: Add Purine (nebularine) modification support</p></li>
<li><p>API: Add new soft constraints multi-callback dispatcher</p></li>
<li><p>API: Add dynamic array data structure utilities</p></li>
<li><p>API: Add string data structure utilities</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_strchr()</span></code> function</p></li>
<li><p>API: Fix potential problems in free_dp_matrices() of LPfold.c</p></li>
<li><p>API: Fix z-score initialization in <code class="docutils literal notranslate"><span class="pre">vrna_Lfoldz()</span></code> amd <code class="docutils literal notranslate"><span class="pre">vrna_mfe_window_zscore_cb()</span></code></p></li>
<li><p>API: Fix file close issue in <code class="docutils literal notranslate"><span class="pre">vrna_file_commands_read()</span></code></p></li>
<li><p>API: Fix backtracking issue in Zuker subopt</p></li>
<li><p>API: Fix missing soft constraints callback execution in Zuker subopt</p></li>
<li><p>API: Fix enumeration of G-quadruplexes in <code class="docutils literal notranslate"><span class="pre">vrna_subopt()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_subopt_cb()</span></code></p></li>
<li><p>API: Fix constraints bug for exterior loop in boltzmann sampling</p></li>
<li><p>API: Allow for enforcing ‘must pair’ constraint (<code class="docutils literal notranslate"><span class="pre">|</span></code>) in dot-bracket constraints strings</p></li>
<li><p>API: Fix discrepancy between global and local folding in how hard constraints for unpaired bases and non-specific pairing are applied</p></li>
<li><p>API: Refactor function typdefs to make them actual function pointer typedefs</p></li>
<li><p>SWIG: Fix Python 3 wrapper suffix issue</p></li>
<li><p>SWIG: Fix Perl 5 wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_ud_prob_get()</span></code></p></li>
<li><p>SWIG: Only accept upper triangular part of matrix input in fc.sc_bp_add()</p></li>
<li><p>SWIG: Use var_array instead of tuples for Python RNA.ptable()</p></li>
<li><p>SWIG: Add Python wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_move_neighbor_diff()</span></code></p></li>
<li><p>SWIG: Add Python docstrings generated from doxygen documentation of C-library</p></li>
</ul>
</section>
<section id="id10">
<h4>Package<a class="headerlink" href="#id10" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Update <code class="docutils literal notranslate"><span class="pre">libsvm</span></code> to version 3.31</p></li>
<li><p>Update <code class="docutils literal notranslate"><span class="pre">dlib</span></code> to version 19.24</p></li>
<li><p>Adapt Debian dependencies</p></li>
<li><p>Fix compilation issues with RNAforester</p></li>
<li><p>AUTOCONF: Fix requirement checks when SVM support is deactivated and swig is missing</p></li>
<li><p>AUTOMAKE: Add <code class="docutils literal notranslate"><span class="pre">auto</span></code> parameters for <code class="docutils literal notranslate"><span class="pre">-flto</span></code> compile/link flags</p></li>
<li><p>AUTOCONF: Require C++17 due to dependencies to compile <code class="docutils literal notranslate"><span class="pre">DLIB</span></code></p></li>
<li><p>AUTOCONF: Deactivate Python 2 bindings by default</p></li>
</ul>
</section>
</section>
</section>
<section id="version-2-5-x">
<h2>Version 2.5.x<a class="headerlink" href="#version-2-5-x" title="Permalink to this headline"></a></h2>
<section id="version-2-5-1-release-date-2022-06-02">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.5.0...v2.5.1">Version 2.5.1</a> (Release date: 2022-06-02)<a class="headerlink" href="#version-2-5-1-release-date-2022-06-02" title="Permalink to this headline"></a></h3>
<section id="id11">
<h4>Programs<a class="headerlink" href="#id11" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Refactor <code class="docutils literal notranslate"><span class="pre">ct2db</span></code> program to allow for pseudoknots in output structure</p></li>
</ul>
</section>
<section id="id12">
<h4>Library<a class="headerlink" href="#id12" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix MEA computation for G-quadruplex predictions</p></li>
<li><p>API: Fix memory leak in hard constraints container</p></li>
<li><p>API: Fix RNApuzzler edge-case that resulted in segmentation faults</p></li>
<li><p>API: Fix invalid memory access in <code class="docutils literal notranslate"><span class="pre">vrna_strjoin()</span></code></p></li>
<li><p>API: Revisit generic soft constraints for sliding-window base pair probability computations</p></li>
<li><p>API: Enable to overwrite automatic unpaired probability determination in MEA computation</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">#VRNA_PLIST_TYPE_UNPAIRED</span></code> and <code class="docutils literal notranslate"><span class="pre">#VRNA_PLIST_TYPE_TRIPLE</span></code> identifiers for <code class="docutils literal notranslate"><span class="pre">vrna_ep_t</span></code></p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_init_rand_seed()</span></code> to initialize RNG with seed</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_zsc_compute_raw()</span></code> to obtain mean and sd for Z-score computation</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_file_connect_read_record()</span></code> function to parse connectivity table (<code class="docutils literal notranslate"><span class="pre">*.ct</span></code>) files</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_strtrim()</span></code> function</p></li>
<li><p>API: Update sanity checks for input in <code class="docutils literal notranslate"><span class="pre">vrna_pbacktrack_sub*()</span></code></p></li>
<li><p>API: Allow for pseudo-knots in <code class="docutils literal notranslate"><span class="pre">vrna_db_from_ptable()</span></code></p></li>
<li><p>API: Do not use <code class="docutils literal notranslate"><span class="pre">min_loop_size</span> <span class="pre">=</span> <span class="pre">0</span></code> for multi strand interaction prediction</p></li>
<li><p>API: Remove unnecessary uses of <code class="docutils literal notranslate"><span class="pre">min_loop_size</span></code> at multiple locations</p></li>
<li><p>API: Deprecate cutpoint member of vrna_fold_compound_t and prepare for 5’/3’ encoding</p></li>
<li><p>API: Refactor sequence addition/preparation for <code class="docutils literal notranslate"><span class="pre">vrna_fold_compound_t</span></code></p></li>
<li><p>DOC: Update documentation</p></li>
<li><p>SWIG: Add simple dot-plot file wrapper <code class="docutils literal notranslate"><span class="pre">plot_dp_EPS()</span></code></p></li>
<li><p>SWIG: Add <code class="docutils literal notranslate"><span class="pre">sequence</span></code>, <code class="docutils literal notranslate"><span class="pre">sequence_encoding</span></code> and <code class="docutils literal notranslate"><span class="pre">sequence_encoding2</span></code> attributes to <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code> objects</p></li>
<li><p>SWIG: Fix RNG wrapping and initialize RNG upon module load and update associated functions</p></li>
<li><p>SWIG: Add more access to member variable arrays for various objects used throughout the library</p></li>
<li><p>SWIG: Add memory efficient wrapper for dynamically allocated arrays and matrices</p></li>
<li><p>SWIG: Shadow pair table data structure for efficient interactions between C and target languages</p></li>
<li><p>SWIG: Expose hard constraints members in <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code> objects</p></li>
<li><p>SWIG: Add <code class="docutils literal notranslate"><span class="pre">exp_E_ext_stem()</span></code> method (<code class="docutils literal notranslate"><span class="pre">vrna_exp_E_ext_stem()</span></code>) to <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code> objects</p></li>
<li><p>SWIG: Expose DP matrices within <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code> objects</p></li>
<li><p>SWIG: Fix memory leak in wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_db_from_ptable()</span></code></p></li>
</ul>
</section>
<section id="id13">
<h4>Package<a class="headerlink" href="#id13" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Update dlib to version 19.23</p></li>
<li><p>DOC: Update doxygen.conf for version 1.9.2</p></li>
<li><p>AUTOCONF: Factor-out Naview layout algorithm to allow for deactivating the Naview layout algorithm at configure-time</p></li>
<li><p>AUTOCONF: Make LaTeX checks more portable and update LaTeX package checks</p></li>
<li><p>AUTOCONF: Check whether we can build the swig interface when SVM support is deactivated</p></li>
<li><p>AUTOCONF: Fix condition check for CLA build</p></li>
</ul>
</section>
</section>
<section id="version-2-5-0-release-date-2021-11-08">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.18...v2.5.0">Version 2.5.0</a> (Release date: 2021-11-08)<a class="headerlink" href="#version-2-5-0-release-date-2021-11-08" title="Permalink to this headline"></a></h3>
<section id="id14">
<h4>Programs<a class="headerlink" href="#id14" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Add <code class="docutils literal notranslate"><span class="pre">RNAmultifold</span></code> program to compute secondary structures for multiple interacting RNAs</p></li>
<li><p>Add multistrand capabilities to <code class="docutils literal notranslate"><span class="pre">RNAeval</span></code></p></li>
<li><p>Add multistrand capabilities to <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code></p></li>
<li><p>Replace <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code> with a wrapper to <code class="docutils literal notranslate"><span class="pre">RNAmultifold</span></code></p></li>
<li><p>Fix computation of BB homodimer base pair probabilities in <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code></p></li>
</ul>
</section>
<section id="id15">
<h4>Library<a class="headerlink" href="#id15" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix use of undefined values in deprecated function <code class="docutils literal notranslate"><span class="pre">PS_dot_plot()</span></code></p></li>
<li><p>API: Fix probability computations for unstructured domains within multibranch loops</p></li>
<li><p>API: Fix index error in ensemble defect computations</p></li>
<li><p>API: Fix hard constraints behavior on non-specific base pairing</p></li>
<li><p>API: Fix segmentation fault for short input sequences in <code class="docutils literal notranslate"><span class="pre">vrna_hx_from_ptable()</span></code></p></li>
<li><p>API: Fix memory leak in static <code class="docutils literal notranslate"><span class="pre">rna_layout()</span></code> function</p></li>
<li><p>API: Fix corner-case in covariance score computation on sequence alignments that determines which alignment columns may pair and which don’t</p></li>
<li><p>API: Add MFE computations for multiple interacting strands</p></li>
<li><p>API: Add partition function computations for multiple interacting strands</p></li>
<li><p>API: Add base pair probability computations for multiple interacting strands</p></li>
<li><p>API: Add suboptimal structure prediction for multiple interacting strands</p></li>
<li><p>API: Add multistrand capabilities to <code class="docutils literal notranslate"><span class="pre">vrna_eval*()</span></code> functions</p></li>
<li><p>API: Add new function <code class="docutils literal notranslate"><span class="pre">vrna_equilibrium_conc()</span></code> fir concentration dependency computations of multiple interacting strands with <code class="docutils literal notranslate"><span class="pre">dlib</span></code> backend</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_equilibrium_constants()</span></code> function to obtain equilibrium constants for different complexes of multiple interacting strands</p></li>
<li><p>API: Add function <code class="docutils literal notranslate"><span class="pre">vrna_pf_add()</span></code> to add ensemble free energies of two ensembles</p></li>
<li><p>API: Add function <code class="docutils literal notranslate"><span class="pre">vrna_pf_substrands()</span></code> to get ensemble free energies for complexes up to a specific number of interacting strands</p></li>
<li><p>API: Add function <code class="docutils literal notranslate"><span class="pre">vrna_n_multichoose_k()</span></code> to obtain a list of k-combinations with repetition</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_cstr_discard()</span></code> function to allow for discarding char streams prior to flushing</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_bp_distance_pt()</span></code> function to allow for base pair distance computation with pseudo-knots</p></li>
<li><p>API: Add functions <code class="docutils literal notranslate"><span class="pre">vrna_pbacktrack_sub*()</span></code> to allow for stochastic backtracing within arbitrary sequence intervals</p></li>
<li><p>API: Add functions <code class="docutils literal notranslate"><span class="pre">vrna_boustrophedon()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_boustrophedon_pos()</span></code> to generate lists of or obtain values from sequences of Boustrophedon distributed integer numbers</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_pscore()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_pscore_freq()</span></code> functions to obtain covariance score for particular alignment columns</p></li>
<li><p>API: Rewrite Zuker suboptimals implementation</p></li>
<li><p>API: Remove old cofold implementations</p></li>
<li><p>API: Make <code class="docutils literal notranslate"><span class="pre">type</span></code> attribute of <code class="docutils literal notranslate"><span class="pre">vrna_mx_mfe_t</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_mx_pf_t</span></code> a constant</p></li>
<li><p>API: Guard more functions in <code class="docutils literal notranslate"><span class="pre">utils/structure_utils.c</span></code> against <code class="docutils literal notranslate"><span class="pre">NULL</span></code> input</p></li>
<li><p>API: Rename <code class="docutils literal notranslate"><span class="pre">vrna_E_ext_loop()</span></code> to <code class="docutils literal notranslate"><span class="pre">vrna_eval_ext_stem()</span></code></p></li>
<li><p>API: Use v3 typedefs in dot-plot function declarations</p></li>
<li><p>SWIG: Fix Python 3 file handle as optional argument in <code class="docutils literal notranslate"><span class="pre">eval*</span></code> functions and methods</p></li>
<li><p>SWIG: Add wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_pf_add()</span></code></p></li>
<li><p>SWIG: Add wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_hx_from_ptable()</span></code></p></li>
<li><p>SWIG: Add wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_db_from_probs()</span></code></p></li>
</ul>
</section>
<section id="id16">
<h4>Package<a class="headerlink" href="#id16" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Update <code class="docutils literal notranslate"><span class="pre">libsvm</span></code> to version 3.25</p></li>
<li><p>Make Python 3.x the default Python for the scripting languange interfaces</p></li>
<li><p>Add Python3 capability for Mac OS X installer builds</p></li>
<li><p>TESTS: Create TAP driver output for all unit tests (library, executables, SWIG interfaces)</p></li>
<li><p>Remove compile-time switch to deactivate Boustrophedon backtracing scheme (this is the status-quo now)</p></li>
<li><p>Add Contributors License Agreement (CLA) to the Package in <code class="docutils literal notranslate"><span class="pre">doc/CLA/</span></code></p></li>
</ul>
</section>
</section>
</section>
<section id="version-2-4-x">
<h2>Version 2.4.x<a class="headerlink" href="#version-2-4-x" title="Permalink to this headline"></a></h2>
<section id="version-2-4-18-release-date-2021-04-22">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.17...v2.4.18">Version 2.4.18</a> (Release date: 2021-04-22)<a class="headerlink" href="#version-2-4-18-release-date-2021-04-22" title="Permalink to this headline"></a></h3>
<section id="id17">
<h4>Programs<a class="headerlink" href="#id17" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix and refactor <code class="docutils literal notranslate"><span class="pre">RNApkplex</span></code> program</p></li>
<li><p>Fix occasional backtracing errors in <code class="docutils literal notranslate"><span class="pre">RNALalifold</span></code></p></li>
<li><p>Restrict available dangling end models in <code class="docutils literal notranslate"><span class="pre">RNALalifold</span></code> to 0 and 2</p></li>
<li><p>Prevent segmentation faults upon bogus input data in <code class="docutils literal notranslate"><span class="pre">RNAfold</span></code>, <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code>, <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code>, <code class="docutils literal notranslate"><span class="pre">RNAheat</span></code>, and <code class="docutils literal notranslate"><span class="pre">RNAeval</span></code></p></li>
<li><p>Free MFE DP matrices in <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code> Boltzmann sampling when not required anymore</p></li>
</ul>
</section>
<section id="id18">
<h4>Library<a class="headerlink" href="#id18" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_abstract_shapes()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_abstract_shapes_pt()</span></code> functions to convert secondary structures into their respective abstract shape notation ala Giegerich et al. 2004</p></li>
<li><p>API: Add functions <code class="docutils literal notranslate"><span class="pre">vrna_seq_reverse()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_DNA_complement()</span></code> to create reverse complements of a sequence</p></li>
<li><p>API: Add more soft constraint handling to comparative structure prediction</p></li>
<li><p>API: Add generic soft constraints for sliding window comparative MFE backtracing</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_ensemble_defect_pt()</span></code> that accepts pair table input instead of dot-bracket string to allow for non-nested reference structures</p></li>
<li><p>API: Add failure/success return values to generic soft constraints application functions</p></li>
<li><p>API: Refactor <code class="docutils literal notranslate"><span class="pre">RNAPKplex</span></code> implementation by better using constraints framework and moving out many parts from <code class="docutils literal notranslate"><span class="pre">RNAPKplex.c</span></code> into <code class="docutils literal notranslate"><span class="pre">RNAlib</span></code> as separate re-usable functions</p></li>
<li><p>API: Fix energy contributions used in <code class="docutils literal notranslate"><span class="pre">RNAPKplex</span></code> implementations</p></li>
<li><p>API: Fix energy evaluation for cofolding with dangle model 1</p></li>
<li><p>API: Fix wrong arithmetic usage for PF variant of combined generic and simple soft constraints applied to external loops</p></li>
<li><p>API: Fix memory size in #vrna_fold_compound_t initialization</p></li>
<li><p>API: Fix bogus memory access for comparative prediction when preparing hard constraints</p></li>
<li><p>API: Fix wrong index usage in hard constraints for comparative base pair probability computations of internal loops</p></li>
<li><p>API: Fix G-Quadruplex contributions as part of multibranch loops in single sequence base pair probability computations</p></li>
<li><p>API: Fix multibranch loop MFE decomposition step for multiple strand cases</p></li>
<li><p>API: Fix external loop generic hard constraint index updating for partition function computations</p></li>
<li><p>API: Fix memory allocation for auxiliary grammar data structure</p></li>
<li><p>API: Fix incorporation of auxiliary grammar contrib for closing pairs in sliding-window MFE computation</p></li>
<li><p>API: Fix DP matrix intitialization in sliding window MFE computations (fixes occasional backtracing issues in comparative sliding-window MFE computations)</p></li>
<li><p>API: Make <code class="docutils literal notranslate"><span class="pre">vrna_sc_t.type</span></code> attribute a constant</p></li>
<li><p>API: Remove upper-triangular hard constraint matrix in favor of full matrix</p></li>
<li><p>API: Always ensure sane base pair span settings after <code class="docutils literal notranslate"><span class="pre">vrna_fold_compound_prepare()</span></code></p></li>
<li><p>API: Return INF on predictions of <code class="docutils literal notranslate"><span class="pre">vrna_mfe_dimer()</span></code> that fail due to unsatisfiable constraints</p></li>
<li><p>API: Rename internally used hard and soft constraints API symbols</p></li>
<li><p>API: Fix header file inclusions to prevent #include cycles</p></li>
<li><p>SWIG: Add wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_file_fasta_read_record()</span></code></p></li>
<li><p>SWIG: Fix memory leak in wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_probs_window()</span></code></p></li>
<li><p>SWIG: Refactor and therefore fix soft constraint binding functions for use in comparative structure predictions</p></li>
<li><p>SWIG: Fix typo that prevented properly wrapping <code class="docutils literal notranslate"><span class="pre">vrna_params_load_RNA_Andronescu2007()</span></code></p></li>
<li><p>SWIG: Unify wrappers for <code class="docutils literal notranslate"><span class="pre">vrna_ptable()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_ptable_from_string()</span></code></p></li>
</ul>
</section>
<section id="id19">
<h4>Package<a class="headerlink" href="#id19" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>REFMAN: Refactored structure annotation documentation</p></li>
<li><p>REFMAN: Update Mac OS X install section</p></li>
<li><p>Replace <code class="docutils literal notranslate"><span class="pre">DEF</span></code> placeholders in energy parameter files with their value of -50</p></li>
<li><p>Update <code class="docutils literal notranslate"><span class="pre">RNAlocmin</span></code> subpackage to properly compile with more stringent C++ compilers</p></li>
<li><p>Update <code class="docutils literal notranslate"><span class="pre">RNAforester</span></code> subpackage to properly compile with more stringent C++ compilers</p></li>
<li><p>Update autotools framework, e.g. checks for pthreads</p></li>
<li><p>Update universal binary build instructions for Mac OS X builds to enable ARM compilation for M1 CPUs</p></li>
</ul>
</section>
</section>
<section id="version-2-4-17-release-date-2020-11-25">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.16...v2.4.17">Version 2.4.17</a> (Release date: 2020-11-25)<a class="headerlink" href="#version-2-4-17-release-date-2020-11-25" title="Permalink to this headline"></a></h3>
<section id="id20">
<h4>Programs<a class="headerlink" href="#id20" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix <code class="docutils literal notranslate"><span class="pre">RNAup</span> <span class="pre">-b</span></code> mode with shorter sequence first</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--backtrack-global</span></code> option to <code class="docutils literal notranslate"><span class="pre">RNALfold</span></code> (currently only available for dangles == 2 | 0)</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--zscore-pre-filter</span></code> and <code class="docutils literal notranslate"><span class="pre">--zscore-report-subsumed</span></code> options to <code class="docutils literal notranslate"><span class="pre">RNALfold</span></code></p></li>
</ul>
</section>
<section id="id21">
<h4>Library<a class="headerlink" href="#id21" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix multiloop backtracing with soft constraints for unpaired positions in <code class="docutils literal notranslate"><span class="pre">vrna_subopt()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_subopt_cb()</span></code></p></li>
<li><p>API: Fix parameter parse in <code class="docutils literal notranslate"><span class="pre">vrna_params_load_from_string()</span></code></p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_heat_capacity()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_head_capacity_cb()</span></code> functions to RNAlib</p></li>
<li><p>API: Add backtracing function <code class="docutils literal notranslate"><span class="pre">vrna_backtrack_window()</span></code> for global MFE structure to sliding-window predictions</p></li>
<li><p>API: Add SVG support for <code class="docutils literal notranslate"><span class="pre">RNApuzzler</span></code> structure layouts</p></li>
<li><p>API: Make <code class="docutils literal notranslate"><span class="pre">vrna_md_t</span></code> argument to vrna_fold_compound() a constant pointer</p></li>
<li><p>API: Remove missing symbols from header file <code class="docutils literal notranslate"><span class="pre">ViennaRNA/params/default.h</span></code></p></li>
<li><p>API: Refactor z-score threshold filter handling for sliding-window MFE prediction</p></li>
<li><p>SWIG: Fix typo in interface functions to load DNA parameters</p></li>
<li><p>SWIG: Add python-3.9 autoconf checks</p></li>
<li><p>SWIG: Add <code class="docutils literal notranslate"><span class="pre">vrna_head_capacity*()</span></code> wrappers</p></li>
<li><p>SWIG: Add access to raw energy parameters</p></li>
<li><p>SWIG: Add <code class="docutils literal notranslate"><span class="pre">alias</span></code> and <code class="docutils literal notranslate"><span class="pre">pair</span></code> attribute to objects of type <code class="docutils literal notranslate"><span class="pre">md</span></code></p></li>
<li><p>SWIG: Add out/varout typemaps for 2-dimensional int-like arrays</p></li>
<li><p>SWIG: Add all data fields to objects of type ‘param’ and ‘exp_param’</p></li>
</ul>
</section>
<section id="id22">
<h4>Package<a class="headerlink" href="#id22" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix Debian and Windows installer files</p></li>
</ul>
</section>
</section>
<section id="version-2-4-16-release-date-2020-10-09">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.15...v2.4.16">Version 2.4.16</a> (Release date: 2020-10-09)<a class="headerlink" href="#version-2-4-16-release-date-2020-10-09" title="Permalink to this headline"></a></h3>
<section id="id23">
<h4>Programs<a class="headerlink" href="#id23" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix backtracing errors in <code class="docutils literal notranslate"><span class="pre">RNALalifold</span></code> for alignments with more than <code class="docutils literal notranslate"><span class="pre">32768</span></code> columns</p></li>
<li><p>Fix backtracing errors in <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code> and <code class="docutils literal notranslate"><span class="pre">RNALalifold</span></code> for rare cases when two alignment columns may pair due to covariance score threshold but still yield infinite energies due to energy model</p></li>
<li><p>Refactored manpages/help options for <code class="docutils literal notranslate"><span class="pre">RNAplfold</span></code>, <code class="docutils literal notranslate"><span class="pre">RNAplot</span></code>, <code class="docutils literal notranslate"><span class="pre">RNApvmin</span></code>, <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code>, and <code class="docutils literal notranslate"><span class="pre">RNAup</span></code></p></li>
</ul>
</section>
<section id="id24">
<h4>Library<a class="headerlink" href="#id24" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix undefined behavior due to short int overflows when accessing alignment lengths with alignments larger than 32768 columns. This fixes occasional backtracing errors in <code class="docutils literal notranslate"><span class="pre">RNALalifold</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_mfe_window()</span></code></p></li>
<li><p>API: Fix adding pscore to base pairs that yield INF energy in comparative global and local MFE prediction</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_convert_kcal_to_dcal()</span></code> and vice-versa function for safely converting integer to float representations of energy values</p></li>
<li><p>SWIG: Add a reasonable <code class="docutils literal notranslate"><span class="pre">Python</span></code> interface for objects of type <code class="docutils literal notranslate"><span class="pre">vrna_path_t</span></code></p></li>
<li><p>SWIG: Add a wrapper for <code class="docutils literal notranslate"><span class="pre">vrna_seq_encode()</span></code></p></li>
</ul>
</section>
<section id="id25">
<h4>Package<a class="headerlink" href="#id25" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Move <code class="docutils literal notranslate"><span class="pre">units.h</span></code> include file to <code class="docutils literal notranslate"><span class="pre">ViennaRNA/utils/units.h</span></code></p></li>
</ul>
</section>
</section>
<section id="version-2-4-15-release-date-2020-08-18">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.14...v2.4.15">Version 2.4.15</a> (Release date: 2020-08-18)<a class="headerlink" href="#version-2-4-15-release-date-2020-08-18" title="Permalink to this headline"></a></h3>
<section id="id26">
<h4>Programs<a class="headerlink" href="#id26" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix compilation of <code class="docutils literal notranslate"><span class="pre">Kinfold</span></code> with GCC 10</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--en-only</span></code> flag to <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code> to allow for sorting by energy only</p></li>
<li><p>Prevent <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code> to process input with more than two strands</p></li>
<li><p>Add cutpoint marker to dot-plots created with <code class="docutils literal notranslate"><span class="pre">RNAcofold</span> <span class="pre">-a</span></code></p></li>
<li><p>Update <code class="docutils literal notranslate"><span class="pre">Kinfold</span></code> to version <code class="docutils literal notranslate"><span class="pre">1.4</span></code></p></li>
</ul>
</section>
<section id="id27">
<h4>Library<a class="headerlink" href="#id27" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix removal of strand delimiter in <code class="docutils literal notranslate"><span class="pre">vrna_plot_dp_PS_list()</span></code></p></li>
<li><p>API: Fix <code class="docutils literal notranslate"><span class="pre">vrna_enumerate_necklaces()</span></code></p></li>
<li><p>API: Fix bogus backtracing for co-folded structures in <code class="docutils literal notranslate"><span class="pre">vrna_subopt()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_subopt_cb()</span></code></p></li>
<li><p>API: Fix storing co-folded structures for sorted output in <code class="docutils literal notranslate"><span class="pre">vrna_subopt()</span></code></p></li>
<li><p>API: Fix multibranch loop component hard constraints for multi-strand cases</p></li>
<li><p>API: Prevent adding internal loop energy contributions to enclosed parts with <code class="docutils literal notranslate"><span class="pre">energy=INF</span></code></p></li>
<li><p>API: Adapt <code class="docutils literal notranslate"><span class="pre">vrna_db_pack()/vrna_db_unpack()</span></code> functions to produce comparable strings</p></li>
<li><p>API: Add sorting modes <code class="docutils literal notranslate"><span class="pre">VRNA_UNSORTED</span></code>, <code class="docutils literal notranslate"><span class="pre">VRNA_SORT_BY_ENERGY_LEXICOGRAPHIC_ASC</span></code>, and <code class="docutils literal notranslate"><span class="pre">VRNA_SORT_BY_ENERGY_ASC</span></code> to <code class="docutils literal notranslate"><span class="pre">vrna_subopt()</span></code></p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_strjoin()</span></code> function</p></li>
<li><p>API: Add missing case to external loop hard constraints</p></li>
<li><p>API: Make hard constrains strand-aware</p></li>
<li><p>SWIG: Fix invalid memory access when using <code class="docutils literal notranslate"><span class="pre">MEA_from_plist()</span></code> in <code class="docutils literal notranslate"><span class="pre">Perl</span> <span class="pre">5</span></code> or <code class="docutils literal notranslate"><span class="pre">Python</span></code></p></li>
<li><p>SWIG: Enable keyword argument features in <code class="docutils literal notranslate"><span class="pre">Python</span></code> interface of constructors for <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code>, <code class="docutils literal notranslate"><span class="pre">md</span></code>, <code class="docutils literal notranslate"><span class="pre">move</span></code>, <code class="docutils literal notranslate"><span class="pre">param</span></code>, and <code class="docutils literal notranslate"><span class="pre">exp_param</span></code> objects</p></li>
<li><p>SWIG: Enable autodoc feature for <code class="docutils literal notranslate"><span class="pre">Python</span></code> interface of constructors for <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code>, <code class="docutils literal notranslate"><span class="pre">md</span></code>, and <code class="docutils literal notranslate"><span class="pre">move</span></code> objects</p></li>
<li><p>SWIG: Enable <code class="docutils literal notranslate"><span class="pre">toString</span></code> conversion for <code class="docutils literal notranslate"><span class="pre">Python</span></code> interface for objects of type <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code>, <code class="docutils literal notranslate"><span class="pre">md</span></code>, <code class="docutils literal notranslate"><span class="pre">move</span></code>, <code class="docutils literal notranslate"><span class="pre">params</span></code>, <code class="docutils literal notranslate"><span class="pre">exp_params</span></code>, and <code class="docutils literal notranslate"><span class="pre">subopt_solution</span></code></p></li>
<li><p>SWIG: Add (read-only) attributes <code class="docutils literal notranslate"><span class="pre">type</span></code>, <code class="docutils literal notranslate"><span class="pre">length</span></code>, <code class="docutils literal notranslate"><span class="pre">strands</span></code>, <code class="docutils literal notranslate"><span class="pre">params</span></code>, and <code class="docutils literal notranslate"><span class="pre">exp_params</span></code> to objects of type <code class="docutils literal notranslate"><span class="pre">fold_compound</span></code></p></li>
<li><p>SWIG: Make attributes of objects of type <code class="docutils literal notranslate"><span class="pre">param</span></code> and <code class="docutils literal notranslate"><span class="pre">exp_param</span></code> read-only</p></li>
<li><p>Add array of strand nicks to EPS dot plot files instead of single cutpoint</p></li>
<li><p>Draw separator line for each strand nick in EPS dot-plots</p></li>
<li><p>Update <code class="docutils literal notranslate"><span class="pre">libsvm</span></code> to version 3.24</p></li>
</ul>
</section>
<section id="id28">
<h4>Package<a class="headerlink" href="#id28" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Disable Link-Time-Optimization (<code class="docutils literal notranslate"><span class="pre">LTO</span></code>) for third-party programs linking against <code class="docutils literal notranslate"><span class="pre">RNAlib</span></code> using <code class="docutils literal notranslate"><span class="pre">pkg-config</span></code></p></li>
<li><p>TESTS: Fix results dir path for out-of-tree builds</p></li>
<li><p>TESTS: Set default timeout for library tests to 20s</p></li>
</ul>
</section>
</section>
<section id="version-2-4-14-release-date-2019-08-13">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.13...v2.4.14">Version 2.4.14</a> (Release date: 2019-08-13)<a class="headerlink" href="#version-2-4-14-release-date-2019-08-13" title="Permalink to this headline"></a></h3>
<section id="id29">
<h4>Programs<a class="headerlink" href="#id29" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix <code class="docutils literal notranslate"><span class="pre">RNApvmin</span></code> pertubation vector computation</p></li>
<li><p>Add non-redundant sampling option to <code class="docutils literal notranslate"><span class="pre">RNApvmin</span></code></p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">RNAdos</span></code> program to compute density of states</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">-P</span> <span class="pre">DNA</span></code> convenience command line parameter to most programs to quickly load DNA parameters without any input file</p></li>
<li><p>MAN: Add example section to man-page of <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code></p></li>
</ul>
</section>
<section id="id30">
<h4>Library<a class="headerlink" href="#id30" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix memory leak in <code class="docutils literal notranslate"><span class="pre">vrna_path_gradient()</span></code></p></li>
<li><p>API: Fix release of memory fir <code class="docutils literal notranslate"><span class="pre">vrna_sequence_remove_all()</span></code></p></li>
<li><p>API: Fix soft-constraints application in <code class="docutils literal notranslate"><span class="pre">vrna_sc_minimize_pertubation()</span></code> that prevented proper computation of the pertubation vector</p></li>
<li><p>API: Add 5’ and 3’ neighbor nucleotide encoding arrays and name string to <code class="docutils literal notranslate"><span class="pre">vrna_seq_t</span></code></p></li>
<li><p>API: Add new data structure for multiple sequence alignments</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_sequence_order_update()</span></code> function</p></li>
<li><p>API: Add non-redundant sampling mode to <code class="docutils literal notranslate"><span class="pre">vrna_sc_minimize_pertubation()</span></code> through passing negative sample-sizes</p></li>
<li><p>API: Add v3.0 API functions for maximum expected accuracy (MEA) computation</p></li>
<li><p>API: Include energy parameter sets into <code class="docutils literal notranslate"><span class="pre">RNAlib</span></code> and provide functions to load them at runtime</p></li>
<li><p>API: Prepare sequence data in <code class="docutils literal notranslate"><span class="pre">vrna_fold_compound_t</span></code> with <code class="docutils literal notranslate"><span class="pre">vrna_sequence_add()</span></code></p></li>
<li><p>API: Use <code class="docutils literal notranslate"><span class="pre">vrna_pbacktrack_num()</span></code> instead of <code class="docutils literal notranslate"><span class="pre">vrna_pbacktrack()</span></code> in <code class="docutils literal notranslate"><span class="pre">vrna_sc_minimize_pertubation()</span></code> to speed-up sample generation</p></li>
<li><p>Reduce use of global variable <code class="docutils literal notranslate"><span class="pre">cut_point</span></code> in <code class="docutils literal notranslate"><span class="pre">RNAlib</span></code></p></li>
<li><p>SWIG: Use <code class="docutils literal notranslate"><span class="pre">importlib</span></code> in favor of <code class="docutils literal notranslate"><span class="pre">imp</span></code> to determine Python 3 tag extension</p></li>
<li><p>SWIG: Update various wrapper functions</p></li>
<li><p>SWIG: Add wrappers for MEA computation with <code class="docutils literal notranslate"><span class="pre">vrna_MEA()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_MEA_from_plist</span></code></p></li>
<li><p>SWIG: Add wrappers for <code class="docutils literal notranslate"><span class="pre">vrna_pr_structure()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_pr_energy()</span></code></p></li>
</ul>
</section>
<section id="id31">
<h4>Package<a class="headerlink" href="#id31" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>REFMAN: Fix LaTeX code in <code class="docutils literal notranslate"><span class="pre">units.h</span></code> that prevented proper compilation with <code class="docutils literal notranslate"><span class="pre">pdflatex</span></code></p></li>
<li><p>Add an R script to create 2D landscape plots from <code class="docutils literal notranslate"><span class="pre">RNA2Dfold</span></code> output</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">gengetopt</span></code> to configure-time requirements to build man-pages</p></li>
<li><p>Add new energy parameter file <code class="docutils literal notranslate"><span class="pre">rna_misc_special_hairpins.par</span></code> with additional UV-melting derived parameters for Tri- and Tetra-loops</p></li>
<li><p>Update RNA Tutorial</p></li>
<li><p>Colorize final configure script message</p></li>
<li><p>REFMAN: Always use <code class="docutils literal notranslate"><span class="pre">pdflatex</span></code> to compile reference manual and tutorial</p></li>
<li><p>EXAMPLES: Add Python script that performs computations equivalent to <code class="docutils literal notranslate"><span class="pre">RNAfold</span> <span class="pre">-p</span> <span class="pre">--MEA</span></code></p></li>
</ul>
</section>
</section>
<section id="version-2-4-13-release-date-2019-05-30">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.12...v2.4.13">Version 2.4.13</a> (Release date: 2019-05-30)<a class="headerlink" href="#version-2-4-13-release-date-2019-05-30" title="Permalink to this headline"></a></h3>
<section id="id32">
<h4>Programs<a class="headerlink" href="#id32" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix centroid structure prediction for <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code></p></li>
<li><p>Fix <code class="docutils literal notranslate"><span class="pre">--noLP</span></code> option for <code class="docutils literal notranslate"><span class="pre">RNALalifold</span></code></p></li>
</ul>
</section>
<section id="id33">
<h4>Library<a class="headerlink" href="#id33" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Refactor and fix collision handling in <code class="docutils literal notranslate"><span class="pre">vrna_hash_table_t</span></code></p></li>
<li><p>API: Fix one access using wrong index for odd dangles in <code class="docutils literal notranslate"><span class="pre">loops/external.c</span></code></p></li>
<li><p>API: Add two missing <code class="docutils literal notranslate"><span class="pre">MLbase</span></code> contributions for MFE prediction in <code class="docutils literal notranslate"><span class="pre">loops/multibranch.c</span></code></p></li>
<li><p>API: Refactor multiloop MFE backtracking for odd dangles</p></li>
<li><p>API: Add function <code class="docutils literal notranslate"><span class="pre">vrna_backtrack5()</span></code> to allow for MFE backtracking of sub-sequences starting at the 5’-end</p></li>
<li><p>API: Reduce usage of global macro <code class="docutils literal notranslate"><span class="pre">TURN</span></code> by replacing it with <code class="docutils literal notranslate"><span class="pre">min_loop_size</span></code> field of <code class="docutils literal notranslate"><span class="pre">vrna_md_t</span></code></p></li>
<li><p>API: Add functions <code class="docutils literal notranslate"><span class="pre">vrna_path_direct()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_path_direct_ub()</span></code> that may also return move lists instead of dot-bracket lists</p></li>
<li><p>API: Add functions <code class="docutils literal notranslate"><span class="pre">vrna_pt_pk_remove()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_db_pk_remove()</span></code> that remove pseudoknots from an input structure</p></li>
<li><p>API: Fix invalid memory access for lonely pair mode (<code class="docutils literal notranslate"><span class="pre">--noLP</span></code>) in comparative sliding-window MFE prediction</p></li>
<li><p>SWIG: Fix access to global variable <code class="docutils literal notranslate"><span class="pre">pf_smooth</span></code> and <code class="docutils literal notranslate"><span class="pre">pf_smooth</span></code> attribute in <code class="docutils literal notranslate"><span class="pre">model_details</span></code> object</p></li>
<li><p>SWIG: Fix Python reference counting for <code class="docutils literal notranslate"><span class="pre">Py_None</span></code> in <code class="docutils literal notranslate"><span class="pre">interfaces/findpath.i</span></code> wrapper</p></li>
<li><p>SWIG: Refactor reference counting for all Python2 and Python3 wrappers</p></li>
<li><p>REFMAN: Larger updates and restructuring of reference manual</p></li>
</ul>
</section>
<section id="id34">
<h4>Package<a class="headerlink" href="#id34" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Install example scripts and source code files, e.g. to $prefix/share/ViennaRNA/examples</p></li>
<li><p>Properly pass GSL, PTHREADS, and MPFR flags to sub-projects</p></li>
<li><p>Fix <code class="docutils literal notranslate"><span class="pre">RNApuzzler</span></code> header file installation</p></li>
<li><p>SWIG: Include Python 3.7 and 3.8 in list of autoconf-probed python interpreters</p></li>
<li><p>SWIG: Fix wrapper building for <code class="docutils literal notranslate"><span class="pre">swig</span> <span class="pre">&gt;=</span> <span class="pre">4.0.0</span></code></p></li>
</ul>
</section>
</section>
<section id="version-2-4-12-release-date-2019-04-16">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.11...v2.4.12">Version 2.4.12</a> (Release date: 2019-04-16)<a class="headerlink" href="#version-2-4-12-release-date-2019-04-16" title="Permalink to this headline"></a></h3>
<section id="id35">
<h4>Programs<a class="headerlink" href="#id35" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Add non-redundant stochastic backtracing option for <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code></p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--noDP</span></code> option to suppress dot-plot output in <code class="docutils literal notranslate"><span class="pre">RNAfold</span></code> and <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code></p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">RNApuzzler</span></code> (4) and <code class="docutils literal notranslate"><span class="pre">RNAturtle</span></code> (3) secondary structure layout algorithm options to <code class="docutils literal notranslate"><span class="pre">RNAfold</span></code> and <code class="docutils literal notranslate"><span class="pre">RNAplot</span></code></p></li>
<li><p>Update help/man page of <code class="docutils literal notranslate"><span class="pre">RNALfold</span></code></p></li>
<li><p>Allow for multiple input files and parallel input processing in <code class="docutils literal notranslate"><span class="pre">RNAheat</span></code></p></li>
</ul>
</section>
<section id="id36">
<h4>Library<a class="headerlink" href="#id36" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix declaration of <code class="docutils literal notranslate"><span class="pre">vrna_move_apply_db()</span></code></p></li>
<li><p>API: Fix <code class="docutils literal notranslate"><span class="pre">vrna_path()</span></code> lexicographical ordering in gradient walks</p></li>
<li><p>API: Enable non-redundant stochastic backtracing for comparative structure prediction</p></li>
<li><p>API: Enable stochastic backtracing for circular comparative structure prediction</p></li>
<li><p>API: Enable stochastic backtracing of subsequences (5’ prefixes) for comparative structure prediction</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">pf_smooth</span></code> attribute to <code class="docutils literal notranslate"><span class="pre">vrna_md_t</span></code> data stucture to allow for disabling Boltzmann factor energy smoothing</p></li>
<li><p>API: Add functions to allow for resuming non-redundant stochastic backtracing</p></li>
<li><p>API: Add functions to retrieve multiple stochastically backtraced structures (list and callback variants)</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">vrna_positional_entropy</span></code> to compute vector of positional entropies</p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">RNApuzzler</span></code> and <code class="docutils literal notranslate"><span class="pre">RNAturtle</span></code> secondary structure layout algorithm (Wiegreffe et al. 2018)</p></li>
<li><p>API: Add v3.0 API for secondary structure layout/coordinate algorithms</p></li>
<li><p>API: Add more helper/utility functions for <code class="docutils literal notranslate"><span class="pre">vrna_move_t</span></code> data structures</p></li>
<li><p>API: Add callback-based neighborhood update function for (subsequent) <code class="docutils literal notranslate"><span class="pre">vrna_move_t</span></code> application</p></li>
<li><p>API: Add abstract heap data structure available as <code class="docutils literal notranslate"><span class="pre">&lt;ViennaRNA/datastructures/heap.h&gt;</span></code></p></li>
<li><p>API: Refactor and speed-up gradient walk implementation available as <code class="docutils literal notranslate"><span class="pre">vrna_path_gradient()</span></code></p></li>
<li><p>API: Substitute <code class="docutils literal notranslate"><span class="pre">vrna_file_PS_aln_sub()</span></code> alignment plot function by <code class="docutils literal notranslate"><span class="pre">vrna_file_PS_aln_slice()</span></code> that actually slices out a sub-alignment</p></li>
<li><p>API: Rename <code class="docutils literal notranslate"><span class="pre">vrna_annotate_covar_struct()</span></code> to <code class="docutils literal notranslate"><span class="pre">vrna_annotate_covar_db()</span></code> and add new function <code class="docutils literal notranslate"><span class="pre">vrna_annotate_covar_db_extended()</span></code> to support more bracket types</p></li>
<li><p>API: Calling <code class="docutils literal notranslate"><span class="pre">vrna_params_reset()</span></code> now implies a call to <code class="docutils literal notranslate"><span class="pre">vrna_exp_params_reset()</span></code> as well</p></li>
<li><p>API: Move landscape implementations into separate directory, thus headers should be included as <code class="docutils literal notranslate"><span class="pre">&lt;ViennaRNA/landscape/move.h&gt;</span></code>, <code class="docutils literal notranslate"><span class="pre">&lt;ViennaRNA/landscape/neighbor.h&gt;</span></code>, etc.</p></li>
<li><p>Ensure proper rescaling of energy parameters upon temperature changes</p></li>
<li><p>Refactor soft constraints implementation in stochastic backtracing</p></li>
<li><p>SWIG: Wrap all non-redundant stochastic backtracing functions to scripting language interface(s)</p></li>
<li><p>SWIG: Refactor stochastic backtracing interface(s)</p></li>
<li><p>SWIG: Add proper constructor for objects of type <code class="docutils literal notranslate"><span class="pre">vrna_ep_t</span></code></p></li>
<li><p>SWIG: Sanitize alignment plot function interface(s)</p></li>
</ul>
</section>
<section id="id37">
<h4>Package<a class="headerlink" href="#id37" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Update Ubuntu/Debian and OpenSUSE build instructions</p></li>
<li><p>Reduce intra-package dependency on non-v3.0 API</p></li>
</ul>
</section>
</section>
<section id="version-2-4-11-release-date-2018-12-17">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.10...v2.4.11">Version 2.4.11</a> (Release date: 2018-12-17)<a class="headerlink" href="#version-2-4-11-release-date-2018-12-17" title="Permalink to this headline"></a></h3>
<section id="id38">
<h4>Programs<a class="headerlink" href="#id38" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--commands</span></code> option to <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code></p></li>
<li><p>Add non-redundant Boltzmann sampling mode for <code class="docutils literal notranslate"><span class="pre">RNAsubopt</span></code></p></li>
</ul>
</section>
<section id="id39">
<h4>Library<a class="headerlink" href="#id39" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix wrong access to base pair soft constraints in equilibrium probability computations</p></li>
<li><p>API: Fix behavior of <code class="docutils literal notranslate"><span class="pre">vrna_nucleotide_encode()</span></code> with lowercase characters in sequence</p></li>
<li><p>API: Fix behavior of <code class="docutils literal notranslate"><span class="pre">encode_char()</span></code> with lowercase characters in sequence</p></li>
<li><p>API: Fix forbidden <code class="docutils literal notranslate"><span class="pre">GU</span></code> pairs behavior in pscore computation for comparative folding</p></li>
<li><p>API: Fix potential errors due to uninitialized <code class="docutils literal notranslate"><span class="pre">next</span></code> pointers in <code class="docutils literal notranslate"><span class="pre">vrna_move_t</span></code> of <code class="docutils literal notranslate"><span class="pre">vrna_eval_move_shift_pt</span></code></p></li>
<li><p>API: Add <code class="docutils literal notranslate"><span class="pre">AVX</span> <span class="pre">512</span></code> optimized version of MFE multibranch loop decomposition</p></li>
<li><p>API: Add functions for CPU SIMD feature detection</p></li>
<li><p>API: Add dispatcher to automatically delegate exterior-/multibranch loop MFE decomposition to supported SIMD optimized implementation</p></li>
<li><p>API: Add function <code class="docutils literal notranslate"><span class="pre">vrna_dist_mountain()</span></code> to compute mountain distance between two structures</p></li>
<li><p>API: Add function <code class="docutils literal notranslate"><span class="pre">vrna_ensemble_defect()</span></code> to compute ensemble defect given a target structure</p></li>
<li><p>API: Add non-redundant Boltzmann sampling</p></li>
<li><p>API: Change behavior of <code class="docutils literal notranslate"><span class="pre">vrna_cstr_free()</span></code> and <code class="docutils literal notranslate"><span class="pre">vrna_cstr_close()</span></code> to always flush output before unregistering the stream</p></li>
<li><p>SWIG: Add interface for <code class="docutils literal notranslate"><span class="pre">vrna_loopidx_from_ptable()</span></code></p></li>
</ul>
</section>
<section id="id40">
<h4>Package<a class="headerlink" href="#id40" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Activate compilation for compile-time supported SIMD optimized implementations by default</p></li>
<li><p>Replace <code class="docutils literal notranslate"><span class="pre">--enable-sse</span></code> configure script option with <code class="docutils literal notranslate"><span class="pre">--disable-simd</span></code></p></li>
</ul>
</section>
</section>
<section id="version-2-4-10-release-date-2018-09-26">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.9...v2.4.10">Version 2.4.10</a> (Release date: 2018-09-26)<a class="headerlink" href="#version-2-4-10-release-date-2018-09-26" title="Permalink to this headline"></a></h3>
<section id="id41">
<h4>Programs<a class="headerlink" href="#id41" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix wrong output filename for binary opening energies in <code class="docutils literal notranslate"><span class="pre">RNAplfold</span></code></p></li>
<li><p>Enable G-Quadruplex support for partition function computation in <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code></p></li>
</ul>
</section>
<section id="id42">
<h4>Library<a class="headerlink" href="#id42" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix broken SSE4.1 support for multibranch loop MFE computation that resulted in increased run times</p></li>
<li><p>Fix redundant output issue in subopt backtracking with unusually high delta energies (&gt;=INF)</p></li>
<li><p>Restore default behavior of ‘|’ symbol in dot-bracket hard constraint strings that got lost with version 2.2.0</p></li>
<li><p>Add faster (cache-optimized) version of Nussinov Maximum Matching algorithm</p></li>
<li><p>Change default linker- and loop length computations for G-Quadruplex predictions in comparative prediction modes</p></li>
<li><p>Add hard constraints warning for base pairs that violate the <code class="docutils literal notranslate"><span class="pre">min_loop_size</span></code> of the model</p></li>
<li><p>Update <code class="docutils literal notranslate"><span class="pre">libsvm</span></code> to version 3.23</p></li>
<li><p>API: Add functions to set auxiliary grammar extension rules</p></li>
<li><p>API: Replace upper-triangular hard constraints matrix with full matrix for cache-optimized access</p></li>
<li><p>API: Add G-Quadruplex prediction support for comparative partition function</p></li>
<li><p>API: Remove <code class="docutils literal notranslate"><span class="pre">VRNA_GQUAD_MISMATCH_PENALTY</span></code> and <code class="docutils literal notranslate"><span class="pre">VRNA_GQUAD_MISMATCH_NUM_ALI</span></code> macros</p></li>
<li><p>SWIG: Fix invalid memory access in <code class="docutils literal notranslate"><span class="pre">subopt()</span></code> method of fold_compound object when writing to file</p></li>
<li><p>SWIG: Add wrapper for Nussinov Maximum Matching algorithm</p></li>
</ul>
</section>
<section id="id43">
<h4>Package<a class="headerlink" href="#id43" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Add <code class="docutils literal notranslate"><span class="pre">-ftree-vectorize</span></code> compile flag by default if supported</p></li>
</ul>
</section>
</section>
<section id="version-2-4-9-release-date-2018-07-11">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.8...v2.4.9">Version 2.4.9</a> (Release date: 2018-07-11)<a class="headerlink" href="#version-2-4-9-release-date-2018-07-11" title="Permalink to this headline"></a></h3>
<section id="id44">
<h4>Programs<a class="headerlink" href="#id44" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix interactive mode behavior for multiple sequence alignment input in <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code>, <code class="docutils literal notranslate"><span class="pre">RNALalifold</span></code></p></li>
<li><p>Allow for Stockholm formatted multiple sequence alignment input in <code class="docutils literal notranslate"><span class="pre">RNAeval</span></code> and <code class="docutils literal notranslate"><span class="pre">RNAplot</span></code></p></li>
<li><p>Allow for multiple input files in <code class="docutils literal notranslate"><span class="pre">RNAeval</span></code> and <code class="docutils literal notranslate"><span class="pre">RNAplot</span></code></p></li>
<li><p>Allow for parallel processing of input batch jobs in <code class="docutils literal notranslate"><span class="pre">RNAeval</span></code> and <code class="docutils literal notranslate"><span class="pre">RNAplot</span></code></p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">-g</span></code> option to activate G-Quadruplex support in <code class="docutils literal notranslate"><span class="pre">RNAheat</span></code></p></li>
<li><p>Warn on unsatisfiable hard constraints from dot-bracket string input in <code class="docutils literal notranslate"><span class="pre">RNAfold</span></code>, <code class="docutils literal notranslate"><span class="pre">RNAcofold</span></code>, and <code class="docutils literal notranslate"><span class="pre">RNAalifold</span></code></p></li>
</ul>
</section>
<section id="id45">
<h4>Library<a class="headerlink" href="#id45" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix parameter order bug in <code class="docutils literal notranslate"><span class="pre">vrna_path_findpath*</span></code> functions that resulted in too large search widths</p></li>
<li><p>Fix wrong application of base pair soft constraints in partition function computations</p></li>
<li><p>Fix position ruler string in EPS alignment output files</p></li>
<li><p>Fix MFE backtracking errors that might appear under specific hard constrained base pair patterns</p></li>
<li><p>Refrain from reading anything other than <code class="docutils literal notranslate"><span class="pre">#=GC</span> <span class="pre">SS_cons</span></code> to retrieve structures when parsing Stockholm 1.0 format</p></li>
<li><p>Complete soft constraints additions to Boltzmann sampling implementation for single sequences</p></li>
<li><p>Allow for disabling alignment wrapping in <code class="docutils literal notranslate"><span class="pre">vrna_file_PS_aln*</span></code> functions</p></li>
<li><p>Do not remove G-Quadruplex annotation from WUSS formatted structure strings upon calls to <code class="docutils literal notranslate"><span class="pre">vrna_db_from_WUSS</span></code></p></li>
<li><p>Enable G-Quadruplex related average loop energy correction terms in verbose output of <code class="docutils literal notranslate"><span class="pre">vrna_eval_*</span></code> functions</p></li>
<li><p>Speed-up backward compatibility layer for energy evaluation functions that unnecessarily slowed down third-party tools using the old API</p></li>
<li><p>Allow for passing dot-bracket strings with <code class="docutils literal notranslate"><span class="pre">'&amp;'</span></code> strand-end identifier to simple <code class="docutils literal notranslate"><span class="pre">vrna_eval_*</span></code> functions</p></li>
<li><p>Remove implicit <code class="docutils literal notranslate"><span class="pre">exit()</span></code> calls from global MFE backtracking implementation.</p></li>
</ul>
</section>
</section>
<section id="version-2-4-8-release-date-2018-06-23">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.7...v2.4.8">Version 2.4.8</a> (Release date: 2018-06-23)<a class="headerlink" href="#version-2-4-8-release-date-2018-06-23" title="Permalink to this headline"></a></h3>
<section id="id46">
<h4>Programs<a class="headerlink" href="#id46" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>Fix compilation of RNAforester with C++17 standard</p></li>
<li><p>Fix tty input detection in RNAcofold</p></li>
<li><p>Fix bad memory access with RNAcofold -p</p></li>
</ul>
</section>
<section id="id47">
<h4>Library<a class="headerlink" href="#id47" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>API: Fix incorrect unpaired probability computations in vrna_probs_window()</p></li>
<li><p>API: Fix potential out-of-bounds access situations (for circular RNA folding) in eval.c</p></li>
<li><p>API: Fix comparative exterior internal loop partition function computation for circfold</p></li>
<li><p>SWIG: Fix false-positive use of uninitialized value in Python3/file_py3.i</p></li>
</ul>
</section>
<section id="id48">
<h4>Package<a class="headerlink" href="#id48" title="Permalink to this headline"></a></h4>
<ul class="simple">
<li><p>TESTS: Add tests for special features in RNAalifold</p></li>
<li><p>TESTS: Add test case for RNAcofold -p</p></li>
</ul>
</section>
</section>
<section id="version-2-4-7-release-date-2018-06-13">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.6...v2.4.7">Version 2.4.7</a> (Release date: 2018-06-13)<a class="headerlink" href="#version-2-4-7-release-date-2018-06-13" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Allow for parallel processing across multiple input files in RNAfold</p></li>
<li><p>Allow for arbitrary number of input files in RNAalifold</p></li>
<li><p>Allow for parallel processing of input data in RNAalifold</p></li>
<li><p>Allow for arbitrary number of input files in RNAcofold</p></li>
<li><p>Allow for parallel processing of input data in RNAcofold</p></li>
<li><p>Enable parallel processing in RNAfold, RNAcofold, RNAalifold for MS Windows build</p></li>
<li><p>Add centroid and MEA structure computation to RNAcofold</p></li>
<li><p>Add configure time check for LTO capabilities of the linker</p></li>
<li><p>Include ligand binding energies in centroid and MEA structure output of RNAfold</p></li>
<li><p>Refactor ct2db program to process multiple structures from single .ct file</p></li>
<li><p>API: Enable processing of comparative fold_compound with vrna_pr_*() functions</p></li>
<li><p>API: Refactor vrna_ostream_t to enable NULL input in vrna_ostream_provide()</p></li>
<li><p>API: Major refactoring in loop energy evaluations (MFE and PF)</p></li>
<li><p>API: Make vrna_mx_pf_aux_el_t and vrna_mx_pf_aux_ml_s opaque pointers</p></li>
<li><p>API: Make fold_compound field <code class="docutils literal notranslate"><span class="pre">type</span></code> a const attribute</p></li>
<li><p>API: Refactor MFE post-processing for circular RNAs</p></li>
<li><p>API: Add motif name/id support for unstructured domains</p></li>
<li><p>API: Remove major part of implicit exit() calls in RNAlib</p></li>
<li><p>API: Add implementations of Boyer-Moore-Horspool search algorithm</p></li>
<li><p>API: Add implementations to determine number of rotational symmetry for strings (of objects)</p></li>
<li><p>API: Make vrna_cmd_t an opaque pointer</p></li>
<li><p>API: Move headers for constraints, datastructures, io, loop energy evaluation, energy parameters, plotting, search, and utilities into separate subdirectories (backward compatibility is maintained)</p></li>
<li><p>API: Add hash table data structure</p></li>
<li><p>API: Fix discrepancy between comparative and single sequence –noLP predictions</p></li>
<li><p>API: Add functions to replace ‘old API’ interface of RNAstruct.h</p></li>
<li><p>API: Add functions to replace ‘old API’ interface of aln_util.h</p></li>
<li><p>API: Add generic soft constraints support to suboptimal structure prediction sensu Wuchty et al.</p></li>
<li><p>SWIG: Refactor callback execution for Python 2 / 3 interface to reduce overhead</p></li>
<li><p>SWIG: Fix configure-time check for Python 3 interface build</p></li>
<li><p>SWIG: Fix Python 3 IO file stream to C FILE * conversion</p></li>
<li><p>Cosmetic changes in final configure notice</p></li>
<li><p>Major changes in source tree structure of the library</p></li>
<li><p>Add autoconf checks for maintainer tools</p></li>
<li><p>Generate C strings from static PostScript files at configure time (for structure- and dot plots)</p></li>
<li><p>REFMAN: Large updates in API documentation and structure of reference manual</p></li>
</ul>
</section>
<section id="version-2-4-6-release-date-2018-04-19">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.5...v2.4.6">Version 2.4.6</a> (Release date: 2018-04-19)<a class="headerlink" href="#version-2-4-6-release-date-2018-04-19" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Stabilize rounding of free energy output in RNAalifold</p></li>
<li><p>API: Fix potential rounding errors for comparative free energies in eval.c and mfe.c</p></li>
<li><p>API: Fix regression in exterior loop dangling end contributions for comparative base pair probabilities and Boltzmann sampling (introduced with v2.4.4)</p></li>
<li><p>API: Fix regression with hard constrained base pairs for comparative structure prediction (introduced with v2.4.4)</p></li>
<li><p>TESTS: Add basic tests for RNAalifold executable</p></li>
<li><p>TESTS: Ignore ‘frequency of MFE structure’ in RNAcofold partition function checks</p></li>
</ul>
</section>
<section id="version-2-4-5-release-date-2018-04-17">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.4...v2.4.5">Version 2.4.5</a> (Release date: 2018-04-17)<a class="headerlink" href="#version-2-4-5-release-date-2018-04-17" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Allow for arbitrary number of input files in RNAfold</p></li>
<li><p>Allow for parallel processing of input data in RNAfold (UNIX only, no Windows support yet)</p></li>
<li><p>Add SHAPE reactivity support through commandline options for RNAplfold</p></li>
<li><p>Fix unstructured domain motif detection in MFE, centroid, and MEA structures computed by RNAfold</p></li>
<li><p>Limit allowed set of commands in command file for RNAcofold to hard and soft constraints</p></li>
<li><p>API: Add functions to compute equilibrium probability of particular secondary structures</p></li>
<li><p>API: Add dynamic string stream data type and associated functions</p></li>
<li><p>API: Add priority-queue like data structure with unordered fill capability and ordered output callback execution</p></li>
<li><p>API: Add functions to detect unstructured domain motifs in MFE, centroid, and MEA structures</p></li>
<li><p>API: Fix bug in sliding-window partition function computation with SHAPE reactivity and Deigan et al. conversion method</p></li>
<li><p>API: Fix application of ‘&lt;’ and ‘&gt;’ constraint symbols in dot-bracket provided constraints (was broken since v2.4.2)</p></li>
<li><p>API: Fix MEA structure computation in the presence of unstructured domains</p></li>
<li><p>API: Stabilize order of probability entries in EPS dot-plot files</p></li>
<li><p>Fix compiler warnings on wrong type of printf() in naview.c</p></li>
<li><p>Define VRNA_VERSION macro as string literal and add macros for major, minor, and patch numbers</p></li>
<li><p>Stabilize parallel make of Mac OS X installer</p></li>
<li><p>Add energy parameter set from Langdon et al. 2018</p></li>
<li><p>Add autoconf checks for POSIX threads compiler/linker support</p></li>
<li><p>SWIG: Fix ‘next’ is a perl keyword warnings for Perl5 wrapper</p></li>
<li><p>SWIG: Catch errors and throw execptions whenever scripting language provided callback functions are not applicable or fail</p></li>
<li><p>SWIG: Add keyword arguments and autodoc feature for Python/Python3 wrappers</p></li>
</ul>
</section>
<section id="version-2-4-4-release-date-2018-03-06">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.3...v2.4.4">Version 2.4.4</a> (Release date: 2018-03-06)<a class="headerlink" href="#version-2-4-4-release-date-2018-03-06" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Change verbose output for soft-constraints derived ligand binding motifs in RNAfold</p></li>
<li><p>Allow for lowercase letters in ct2db input</p></li>
<li><p>Fix bug in interior-like G-Quadruplex MFE computation for single sequences</p></li>
<li><p>Fix autoconf switch to enable deprecation warnings</p></li>
<li><p>Fix bug in eval_int_loop() that prevented propagation of energy evaluation for loops with nick in strands</p></li>
<li><p>Fix several bugs for SHAPE reactivity related comparative partition function computations</p></li>
<li><p>Fix annotation of PostScript output for soft-constraint derived ligand binding motifs in RNAfold</p></li>
<li><p>Fix constraint indices for multibranch loops in unpaired probability computations of LPfold.c</p></li>
<li><p>Fix dangling end contributions in comparative partition function for exterior loops</p></li>
<li><p>API: Add simplified interface for vrna_pf_dimer()</p></li>
<li><p>API: Move concentraton dependent implementation for co-folding to separate compile unit</p></li>
<li><p>API: Add new API functions for exterior loop evaluations</p></li>
<li><p>API: Add simplified interfaces for energy evaluation with G-Quadruplexes and circular RNAs</p></li>
<li><p>API: Add findpath functions that allow for specification of an upper bound for the saddle point</p></li>
<li><p>Add configure-time linker check for Python3 interface</p></li>
<li><p>Add automatic CPP suggestions for deprecated function substitutes</p></li>
<li><p>Major restucturing and constraints feature additions in loop type dependent energy evaluation functions</p></li>
<li><p>Major restructuring in MFE implementations</p></li>
<li><p>Major restructuring in PF implementations</p></li>
<li><p>Minor fixes in Boltzmann sampling implementation</p></li>
<li><p>SWIG: Fix wrappers for findpath() implementation</p></li>
<li><p>SWIG: Add tons of energy evaluation wrappers</p></li>
<li><p>SWIG: Fix configure-time check of Perl5 interface build capabilities</p></li>
<li><p>SWIG: Wrap functions from walk.c and neighbor.c</p></li>
<li><p>DOC: Add some missing references to manpages of executable programs</p></li>
<li><p>REFMAN: Heavy re-ordering of the RNAlib reference manual</p></li>
</ul>
</section>
<section id="version-2-4-3-release-date-2017-11-14">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.2...v2.4.3">Version 2.4.3</a> (Release date: 2017-11-14)<a class="headerlink" href="#version-2-4-3-release-date-2017-11-14" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix handling of dangling end contribution at sequence boundaries for sliding window base pair probability computations</p></li>
<li><p>Fix handling of base pair hard constraints in sliding-window implementations</p></li>
<li><p>Fix sliding-window pair probability computations with multibranch-loop unpaired constraints</p></li>
<li><p>Fix sliding-window non-specific base pair hard constraint implementation</p></li>
<li><p>Fix probability computation for stochastic backtracking in RNAsubopt –stochBT_en output</p></li>
<li><p>Fix regression in comparative structure prediction for circular RNAs</p></li>
<li><p>Fix LDFLAGS for scripting language interfaces in corresponding Makefiles</p></li>
<li><p>Stabilize partition function scaling by always using sfact scaling factor from model details</p></li>
<li><p>Add –pf_scale commandling parameter to RNAplfold</p></li>
<li><p>Add constraint framework for single sequence circular RNA structure prediction</p></li>
<li><p>Add RNAfold test suite to check for working implementation of constraints for circular RNAs</p></li>
<li><p>Add a brief contribution guideline CONTRIBUTING.md</p></li>
<li><p>Prevent RNAplfold from creating inf/-inf output when solution set is empty with particular hard constraints</p></li>
<li><p>Include RNAforester v2.0.1</p></li>
</ul>
</section>
<section id="version-2-4-2-release-date-2017-10-13">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.1...v2.4.2">Version 2.4.2</a> (Release date: 2017-10-13)<a class="headerlink" href="#version-2-4-2-release-date-2017-10-13" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix G-Quadruplex energy corrections in comparative structure energy evaluations</p></li>
<li><p>Fix discrepancy in comparative exterior loop dangling end contribution of eval vs. MFE predictions</p></li>
<li><p>Fix regression in RNAup unstructuredness and interaction energy computations</p></li>
<li><p>Fix sequence length confusions when FASTA input contains carriage returns</p></li>
<li><p>Fix build problems of RNAlocmin with older compilers</p></li>
<li><p>Fix sliding-window hard constraints where single nucleotides are prohibited from pairing</p></li>
<li><p>Fix dot-bracket output string length in sliding-window MFE with G-Quadruplexes</p></li>
<li><p>Fix unpaired probability computations for separate individual loop types in LPfold.c</p></li>
<li><p>Fix bad memory access in RNAsubopt with dot-bracket constraint</p></li>
<li><p>Add full WUSS support for –SS_cons constraint option in RNAalifold</p></li>
<li><p>Add commandline option to RNALalifold that enables splitting of energy contributions into separate parts</p></li>
<li><p>Add missing hard constraint cases to sliding-window partition function implementation</p></li>
<li><p>Add CSV output option to RNAcofold</p></li>
<li><p>Use the same model details for SCI computations in RNAalifold</p></li>
<li><p>Abort computations in vrna_eval_structure_v() if structure has unexpected length</p></li>
<li><p>Use original MSA in all output generated by RNAalifold and RNALalifold</p></li>
<li><p>API: Add new functions to convert dot-bracket like structure annotations</p></li>
<li><p>API: Add various new utility functions for alignment handling and comparative structure predictions</p></li>
<li><p>API: Add function vrna_strsplit() to split string into tokens</p></li>
<li><p>API: Do not convert sequences of input MSA to uppercase letters in vrna_file_msa_read_record()</p></li>
<li><p>API: Rename vrna_annotate_bp_covar() and vrna_annotate_pr_covar()</p></li>
<li><p>API: Add new noLP neighbor generation</p></li>
<li><p>SWIG: Add wrapper for functions in file_utils_msa.h</p></li>
<li><p>SWIG: Add wrappers for vrna_pbacktrack() and vrna_pbacktrack5()</p></li>
<li><p>SWIG: Add vrna_db_to_element_string() to scripting language interface</p></li>
<li><p>REFMAN: Fix formula to image conversion in HTML output</p></li>
</ul>
</section>
<section id="version-2-4-1-release-date-2017-08-23">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.4.0...v2.4.1">Version 2.4.1</a> (Release date: 2017-08-23)<a class="headerlink" href="#version-2-4-1-release-date-2017-08-23" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix memory leak in fold_compound methods of SWIG interface</p></li>
<li><p>Fix memory leaks in double ** returning functions of SWIG Perl5 interface</p></li>
<li><p>Fix memory leak in vrna_ep_t to-string() function of SWIG interface</p></li>
<li><p>Regression: Fix reverting pf_scale to defaults after vrna_exp_params_rescale()</p></li>
<li><p>Regression: Fix homo-dimer partition function computation in RNAcofold</p></li>
<li><p>Add unit tests for RNAcofold executable</p></li>
<li><p>Add SHAPE reactivity support to RNAcofold</p></li>
<li><p>Add SHAPE reactivity support to RNALalifold</p></li>
</ul>
</section>
<section id="version-2-4-0-release-date-2017-08-01">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.3.5...v2.4.0">Version 2.4.0</a> (Release date: 2017-08-01)<a class="headerlink" href="#version-2-4-0-release-date-2017-08-01" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Bump libsvm to version 3.22</p></li>
<li><p>Print G-Quadruplex corrections in verbose mode of RNAeval</p></li>
<li><p>Change behavior of RNAfold –outfile option to something more predictable</p></li>
<li><p>Unify max_bp_span usage among sliding window prediction algorithms: RNAplfold, RNALfold, and RNALalifold now consider any base pair (i,j) with (j - i + 1) &lt;= max_bp_span</p></li>
<li><p>Add SHAPE reactivity data support to RNALfold</p></li>
<li><p>Add commands-file support for RNALfold, RNAplfold (hard/soft constraints)</p></li>
<li><p>Add RNAlocmin - Calculate local minima from structures via gradient walks</p></li>
<li><p>Add RNA Bioinformatics tutorial (PDF version)</p></li>
<li><p>Add hard constraints to sliding-window MFE implementations (RNALfold, RNALalifold)</p></li>
<li><p>Add hard constraints to sliding-window PF implementations (RNAplfold)</p></li>
<li><p>Add soft constraints to sliding-window MFE implementation for single sequences (RNALfold)</p></li>
<li><p>Add soft constraints to sliding-window PF implementations (RNAplfold)</p></li>
<li><p>Add SWIG interfaces for sliding-window MFE/PF implementations</p></li>
<li><p>Add proper SWIG interface for alignment and structure plotting functions</p></li>
<li><p>Add proper SWIG interface for duplexfold, duplex_subopt, and its comparative variants</p></li>
<li><p>Add SWIG wrapper for vrna_exp_params_rescale()</p></li>
<li><p>Add explicit destructor for SWIG generated vrna_md_t objects</p></li>
<li><p>Add SWIG perl5 typemap for simple nested STL vectors</p></li>
<li><p>Add dummy field in vrna_structured_domains_s</p></li>
<li><p>Add note about SSE optimized code in reference manual</p></li>
<li><p>Add SWIG interface for findpath implementation</p></li>
<li><p>Add prepare() functions for ptypes-arrays and vrna_(exp_)param_t</p></li>
<li><p>Add warnings for ignored commands in function vrna_commands_apply()</p></li>
<li><p>Add callback featured functions for sliding window MFE and PF implementations</p></li>
<li><p>Change default behavior of adding soft constraints to a vrna_fold_compound_t (store only)</p></li>
<li><p>Several fixes with respect to G-Quadruplex prediction in sliding-window MFE recursions (single sequence and comparative implementation)</p></li>
<li><p>Replace comparative sliding-window MFE recursions (All hits are reported to callback and can be filtered in a post-processing step)</p></li>
<li><p>API: Remove E_mb_loop_stack() and introduce new function vrna_E_mb_loop_stack() as a replacement</p></li>
<li><p>API: change data type of all constraint bit-flags from <code class="docutils literal notranslate"><span class="pre">char</span></code> to <code class="docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">char</span></code></p></li>
<li><p>API: change data type of a2s array in comparative structure prediction from <code class="docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">short</span></code> to <code class="docutils literal notranslate"><span class="pre">unsigned</span> <span class="pre">int</span></code></p></li>
<li><p>API: Change function parameter order in vrna_probs_window() to follow the style of other callback-aware functions in RNAlib</p></li>
<li><p>Move sliding-window MFE implementations to new file mfe_window.c</p></li>
<li><p>Fix building PDF Reference manual with non-standard executable paths</p></li>
<li><p>Fix redefinition of macro ON_SAME_STRAND() in subopt.c</p></li>
<li><p>Fix dangling end issues in sliding-window MFE implementations</p></li>
<li><p>Fix regression for –canonicalBPonly switch in RNAfold/RNAcofold/RNAsubopt</p></li>
<li><p>Fix building sliding-window MFE implementation without SVM support</p></li>
<li><p>Fix parsing of STOCKHOLM 1.0 MSA files that contain MSA spanning multiple blocks</p></li>
<li><p>Fix Alidot link in RNAalifold manpage</p></li>
<li><p>Fix wrong pre-processor flags when enabling single-precision PF computations</p></li>
<li><p>Fix unit testing perl5 interface by including builddir/tests in PERL5LIB path</p></li>
<li><p>Fix buffer overflow in hairpin loop sequence motif extraction for circular RNAs</p></li>
<li><p>Fix out-of-bounds memory access in neighbor.c</p></li>
<li><p>Restore capability to compile stand-alone findpath utility</p></li>
<li><p>Restore capability to use non-standard alphabets for structure prediction</p></li>
<li><p>Restore old-API random number functions in SWIG interface</p></li>
<li><p>Allow additional control characters in MAF MSA input that do not end a block</p></li>
<li><p>Improve reference manual</p></li>
<li><p>Make functions in pair_mat.h static inline</p></li>
<li><p>Prevent users from adding out-of-range base pair soft constraints</p></li>
<li><p>Inline print functions in color_output.inc</p></li>
<li><p>Start documenting callback features in reference manual</p></li>
<li><p>Re-write large portions of sliding-window PF implementation</p></li>
<li><p>Introduce soft-constraint state flag</p></li>
<li><p>Clean-up SWIG unit test framework</p></li>
<li><p>Remove obsolete scripts ct2b.pl and colorrna.pl from src/Utils directory</p></li>
<li><p>Remove old RNAfold tutorial</p></li>
</ul>
</section>
</section>
<section id="version-2-3-x">
<h2>Version 2.3.x<a class="headerlink" href="#version-2-3-x" title="Permalink to this headline"></a></h2>
<section id="version-2-3-5-release-date-2017-04-14">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.3.4...v2.3.5">Version 2.3.5</a> (Release date: 2017-04-14)<a class="headerlink" href="#version-2-3-5-release-date-2017-04-14" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix duplication of output filename prefix in RNAfold</p></li>
<li><p>Add V3.0 API for sliding window partition function (a.k.a. RNAPLfold)</p></li>
<li><p>Add G-Quadruplex prediction to RNALalifold</p></li>
<li><p>Add SWIG wrappers for callback-based sliding window comparative MFE prediction</p></li>
<li><p>Add SSE4.1 multiloop decomposition for single sequence MFE prediction</p></li>
<li><p>Enable RNAfold unit tests to run in paralllel</p></li>
<li><p>Enable users to turn-off base pair probability computations in RNAcofold with -a option</p></li>
<li><p>Split move set in neighbor.c</p></li>
</ul>
</section>
<section id="version-2-3-4-release-date-2017-03-10">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.3.3...v2.3.4">Version 2.3.4</a> (Release date: 2017-03-10)<a class="headerlink" href="#version-2-3-4-release-date-2017-03-10" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix G-Quadruplex probability computation for single sequences</p></li>
<li><p>Fix double-free when using SHAPE reactivity data in RNAalifold</p></li>
<li><p>Fix out-of-bounds access in strand_number array</p></li>
<li><p>Fix weighting of SHAPE reactivity data in consensus structure prediction when fewer data than sequences are present</p></li>
<li><p>Fix z-score output in RNALfold</p></li>
<li><p>Substitute field name ‘A0’/’B0’ in data structure vrna_dimer_conc_s by ‘Ac_start’/’Bc_start’ to avoid clashes with termios.h (Mac OSX Python wrapper bug)</p></li>
<li><p>Minimize usage of ‘unsafe’ sprintf() calls</p></li>
<li><p>Enhance auto-id feature in executable programs</p></li>
<li><p>Always sanitize output file names to avoid problems due to strange FASTA headers</p></li>
<li><p>Lift restrictions of FASTA header length in RNAfold, RNAcofold, and RNAeval</p></li>
<li><p>Add ViennaRNA/config.h with pre-processor definitions of configure time choices</p></li>
<li><p>Add test-suite for RNAfold</p></li>
<li><p>Add functions to procude colored EPS structure alignments</p></li>
<li><p>Add function to write Stockholm 1.0 formatted alignments</p></li>
<li><p>Add function to sanitize file names</p></li>
<li><p>Add callback based implementation for sliding-window MFE prediction (single sequences, comparative structure prediction)</p></li>
<li><p>Add fast API 3.0 implementations to generate structural neighbors and perform steepest descent / random walks (Thanks to Gregor!)</p></li>
<li><p>Add parameter option to RNALalifold for colored EPS structure alignment and structure plot output</p></li>
<li><p>Add parameter option to RNALalifold to write hits into Stockholm file</p></li>
<li><p>Add parameter option to RNAalifold to write Stockholm 1.0 formatted output</p></li>
<li><p>Add parameter option to RNAalifold to suppress stderr spam</p></li>
<li><p>Add auto-id feature to RNAplot, RNALfold, RNAsubopt, RNAplfold, RNAheat</p></li>
<li><p>Add SHAPE reactivity derived pseudo-energies as separate output in RNAalifold</p></li>
<li><p>Add colored output to RNA2Dfold, RNALalifold, RNALfold, RNAduplex, RNAheat, RNAinverse, RNAplfold, and RNAsubopt</p></li>
<li><p>Add command line parameters to RNAsubopt to allow for specification of input/output files</p></li>
</ul>
</section>
<section id="version-2-3-3-release-date-2017-01-24">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.3.2...v2.3.3">Version 2.3.3</a> (Release date: 2017-01-24)<a class="headerlink" href="#version-2-3-3-release-date-2017-01-24" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix multiloop contributions for comparative partition function</p></li>
<li><p>Fix building python2 extension module for OSX</p></li>
</ul>
</section>
<section id="version-2-3-2-release-date-2017-01-18">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.3.1...v2.3.2">Version 2.3.2</a> (Release date: 2017-01-18)<a class="headerlink" href="#version-2-3-2-release-date-2017-01-18" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix pair probability plist creation with G-Quadruplexes</p></li>
<li><p>Allow for specification of python2/3-config at configure time</p></li>
<li><p>Fix init of vrna_md_t data structure after call to set_model_details()</p></li>
<li><p>Fix bug in consensus partition function with hard constraints that force nucleotides to be paired</p></li>
<li><p>Fix compilation of functions that use ellipsis/va_list</p></li>
<li><p>Enable generic hard constraints by default</p></li>
<li><p>Fix init of partition function DP matrices for unusually short RNAs</p></li>
<li><p>Fix behavior of RNAplfold for unusually short RNAs</p></li>
<li><p>Report SCI of 0 in RNAalifold when sum of single sequence MFEs is 0</p></li>
<li><p>Avoid multiple includes of pair_mat.h</p></li>
<li><p>Add configure flag to build entirely static executables</p></li>
</ul>
</section>
<section id="version-2-3-1-release-date-2016-11-15">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.3.0...v2.3.1">Version 2.3.1</a> (Release date: 2016-11-15)<a class="headerlink" href="#version-2-3-1-release-date-2016-11-15" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Add description for how to use unstructured domains through command files to reference manual and RNAfold manpage</p></li>
<li><p>Fix compilation issue for Windows platforms with MingW</p></li>
<li><p>Add missing newline in non-TTY-color output of vrna_message_info()</p></li>
<li><p>Fix regression in vrna_md_update() that resulted in incomplete init of reverse-basepair type array</p></li>
<li><p>Extend coverage of generic hard constraints for partition function computations</p></li>
<li><p>Fix scaling of secondary structure in EPS plot such that it always fits into bounding box</p></li>
<li><p>Several fixes and improvements for SWIG generated scripting language interface(s)</p></li>
</ul>
</section>
<section id="version-2-3-0-release-date-2016-11-01">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.10...v2.3.0">Version 2.3.0</a> (Release date: 2016-11-01)<a class="headerlink" href="#version-2-3-0-release-date-2016-11-01" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Add grammar extension with structured and unstructured domains</p></li>
<li><p>Add default implementation for unstructured domains to allow for ligand/protein binding to unpaired structure segments (MFE and PF for single sequences)</p></li>
<li><p>Introduced command files that subsume constraint definition files (currently used in RNAfold and RNAcofold)</p></li>
<li><p>Replace explicit calls to asprintf() with portable equivalent functions in the library</p></li>
<li><p>Fix configure script to deal with situations where Perl module can’t be build</p></li>
<li><p>Fix bug in doc/Makefile.am that prevented HTML installation due to long argument list</p></li>
<li><p>Added utility functions that deal with conversion between different units</p></li>
<li><p>Bugfix in SWIG wrapped generic soft constraint feature</p></li>
<li><p>Add subopt() and subopt_zuker() methods to SWIG wrapped fold_compound objects</p></li>
<li><p>Bugfix multiloop decomposition in MFE for circular RNAs</p></li>
<li><p>Add separate function to compute pscore for alignments</p></li>
<li><p>Renamed VRNA_VC_TYPE_* macros to VRNA_FC_TYPE_*</p></li>
<li><p>Bugfix regression that prevented programs to fail on too long input sequences</p></li>
<li><p>Extend EPS dot-plot in RNAfold to include motif/binding probabilities from unstructured domains</p></li>
<li><p>Add variadic functions for error/warning/info message</p></li>
<li><p>Add ID manipulation feature to RNAeval</p></li>
<li><p>Extend API for soft constraint feature for more fine-grained control</p></li>
<li><p>Add section on SWIG wrapped functions in reference manual</p></li>
<li><p>Fix bug in interior loop computations when hard constraints result in non-canonical base pairs</p></li>
</ul>
</section>
</section>
<section id="version-2-2-x">
<h2>Version 2.2.x<a class="headerlink" href="#version-2-2-x" title="Permalink to this headline"></a></h2>
<section id="version-2-2-10-release-date-2016-09-06">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.9...v2.2.10">Version 2.2.10</a> (Release date: 2016-09-06)<a class="headerlink" href="#version-2-2-10-release-date-2016-09-06" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Do not ‘forget’ subopt results when output is not written to file handle and sorting is switched off</p></li>
<li><p>Fix bad memory access in vrna_subopt() with sorted output</p></li>
<li><p>Add SWIG wrappers for vrna_subopt_cb()</p></li>
<li><p>Correctly show if C11 features are activated in configure status</p></li>
<li><p>Fix autoconf checks to allow for cross compilation again</p></li>
</ul>
</section>
<section id="version-2-2-9-release-date-2016-09-01">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.8...v2.2.9">Version 2.2.9</a> (Release date: 2016-09-01)<a class="headerlink" href="#version-2-2-9-release-date-2016-09-01" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix bug in partition function scaling for backward compatibility of ali_pf_fold()</p></li>
<li><p>Stabilize v3.0 API when building RNAlib and third party program linking against it with compilers that use different C/C++ standards</p></li>
<li><p>Add details on how to link against RNAlib to the reference manual</p></li>
<li><p>Fix RNAlib2.pc</p></li>
<li><p>Fix bug for temperature setting in RNAplfold</p></li>
<li><p>Use -fflat-lto-objects for static RNAlib library to allow linking without LTO</p></li>
<li><p>Fix interpretation of ‘P’ hard constraint for single nucleotides in constraint definition files</p></li>
<li><p>Add ‘A’ command for hard constraints</p></li>
<li><p>Fix several hard constraint corner-cases in MFE and partition function computation when nucleotides must not be unpaired</p></li>
<li><p>Fix order of hard constraints when read from input file</p></li>
<li><p>Allow for non-canonical base pairs in MFE and partition function computations if hard constraints demand it</p></li>
<li><p>Fix behavior of –without-swig configure script option</p></li>
<li><p>Fix bug in hard constraints usage of exterior loop MFE prediction with odd dangles</p></li>
<li><p>Add parsers for Clustal, Stockholm, FASTA, and MAF formatted alignment files</p></li>
<li><p>Enable RNAalifold to use Clustal, Stockholm, FASTA, or MAF alignments as input</p></li>
<li><p>Lift restriction of sequence number in alignments for RNAalifold</p></li>
<li><p>Enable ANSI colors for TTY output in RNAfold, RNAcofold, RNAalifold, RNAsubopt, and warnings/errors issued by RNAlib</p></li>
<li><p>Add various new commandline options to manipulate sequence/alignment IDs in RNAfold, RNAcofold and RNAalifold</p></li>
</ul>
</section>
<section id="version-2-2-8-release-date-2016-08-01">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.7...v2.2.8">Version 2.2.8</a> (Release date: 2016-08-01)<a class="headerlink" href="#version-2-2-8-release-date-2016-08-01" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix bad memory access in RNAalifold</p></li>
<li><p>Fix regression in RNAalifold to restore covariance contribution ratio determination for circular RNA alignments</p></li>
<li><p>Changed output of RNAsubopt in energy-band enumeration mode to print MFE and energy range in kcal/mol instead of 10cal/mol</p></li>
<li><p>Include latest Kinfold sources that make use of v3.0 API, therefore speeding up runtime substantially</p></li>
<li><p>Re-activate warnings in RNAeval when non-canonical base pairs are encountered</p></li>
<li><p>Fix syntactic incompatibilities that potentially prevented compilation with compilers other than gcc</p></li>
<li><p>dd function to compare nucleotides encoded in IUPAC format</p></li>
<li><p>Fix regression in energy evaluation for circular RNA sequences</p></li>
<li><p>Fix regression in suboptimal structure enumeration for circular RNAs</p></li>
<li><p>Allow for P i-j k-l commands in constraint definition files</p></li>
<li><p>Make free energy evaluation functions polymorphic</p></li>
<li><p>Add free energy evaluation functions that allow for specifying verbosity level</p></li>
<li><p>Secure functions in alphabet.c against NULL pointer arguments</p></li>
<li><p>Fix incompatibility with swig &gt;= 3.0.9</p></li>
<li><p>Fix memory leak in swig-generated scripting language interface(s) for user-provided target language soft-constraint callbacks</p></li>
<li><p>Expose additional functions to swig-generated scripting language interface(s)</p></li>
<li><p>Build Python3 interface by default</p></li>
<li><p>Start of more comprehensive scripting language interface documentation</p></li>
<li><p>Fix linking of python2/python3 interfaces when libpython is in non-standard directory</p></li>
<li><p>Restructured viennarna.spec for RPM based distributions</p></li>
<li><p>Several syntactic changes in the implementation to minimize compiler warnings</p></li>
<li><p>Fix –with-<em>/–without-</em> and –enable-<em>/–disable-</em> configure script behavior</p></li>
</ul>
</section>
<section id="version-2-2-7-release-date-2016-06-30">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.6...v2.2.7">Version 2.2.7</a> (Release date: 2016-06-30)<a class="headerlink" href="#version-2-2-7-release-date-2016-06-30" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix partition function scaling for long sequences in RNAfold, RNAalifold, and RNAup</p></li>
<li><p>Fix backtracking issue in RNAcofold when –noLP option is activated</p></li>
<li><p>Fix hard constraints issue for circular RNAs in generating suboptimal structures</p></li>
<li><p>Rebuild reference manual only when actually required</p></li>
</ul>
</section>
<section id="version-2-2-6-release-date-2016-06-19">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.5...v2.2.6">Version 2.2.6</a> (Release date: 2016-06-19)<a class="headerlink" href="#version-2-2-6-release-date-2016-06-19" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Plugged memory leak in RNAcofold</p></li>
<li><p>Fixed partition function rescaling bug in RNAup</p></li>
<li><p>Fixed bug in RNALfold with window sizes larger than sequence length</p></li>
<li><p>Re-added SCI parameter for RNAalifold</p></li>
<li><p>Fixed backtracking issue for large G-quadruplexes in RNAalifold</p></li>
<li><p>Fixed missing FASTA id in RNAeval output</p></li>
<li><p>Added option to RNAalifold that allows to specify prefix for output files</p></li>
<li><p>Several fixes and additional functions/methods in scripting language interface(s)</p></li>
<li><p>Added version information for scripting language interface(s)</p></li>
<li><p>Some changes to allow for compilation with newer compilers, such as gcc 6.1</p></li>
</ul>
</section>
<section id="version-2-2-5-release-date-2016-04-09">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.4...v2.2.5">Version 2.2.5</a> (Release date: 2016-04-09)<a class="headerlink" href="#version-2-2-5-release-date-2016-04-09" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fixed regression in RNAcofold that prohibited output of concentration computations</p></li>
<li><p>Fixed behavior of RNAfold and RNAcofold when hard constraints create empty solution set (programs now abort with error message)</p></li>
<li><p>Added optional Python 3 interface</p></li>
<li><p>Added RNA::Params Perl 5 sub-package</p></li>
<li><p>Update RNA::Design Perl 5 sub-package</p></li>
<li><p>Simplified usage of v3.0 API with default options</p></li>
<li><p>Wrap more functions of v3.0 API in SWIG generated scripting language interfaces</p></li>
<li><p>Plugged some memory leaks in SWIG generated scripting language interfaces</p></li>
<li><p>Changed parameters of recursion status callback in vrna_fold_compound_t</p></li>
<li><p>Enable definition and binding of callback functions from within SWIG target language</p></li>
<li><p>Added optional subpackage Kinwalker</p></li>
<li><p>Added several configure options to ease building and packaging under MacOS X</p></li>
<li><p>Added new utility script RNAdesign.pl</p></li>
</ul>
</section>
<section id="version-2-2-4-release-date-2016-02-19">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.3...v2.2.4">Version 2.2.4</a> (Release date: 2016-02-19)<a class="headerlink" href="#version-2-2-4-release-date-2016-02-19" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fixed bug in RNAsubopt that occasionally produced cofolded structures twice</p></li>
<li><p>Removed debugging output in preparations of consensus structure prediction datastructures</p></li>
</ul>
</section>
<section id="version-2-2-3-release-date-2016-02-13">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.2...v2.2.3">Version 2.2.3</a> (Release date: 2016-02-13)<a class="headerlink" href="#version-2-2-3-release-date-2016-02-13" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Added postscipt annotations for found ligand motifs in RNAfold</p></li>
<li><p>Added more documentation for the constraints features in RNAfold and RNAalifold</p></li>
<li><p>Restore backward compatibility of get_alipf_arrays()</p></li>
</ul>
</section>
<section id="version-2-2-2-release-date-2016-02-08">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.1...v2.2.2">Version 2.2.2</a> (Release date: 2016-02-08)<a class="headerlink" href="#version-2-2-2-release-date-2016-02-08" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix regression bug that occasionally prevented backtracking with RNAcofold –noLP</p></li>
</ul>
</section>
<section id="version-2-2-1-release-date-2016-02-06">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.2.0...v2.2.1">Version 2.2.1</a> (Release date: 2016-02-06)<a class="headerlink" href="#version-2-2-1-release-date-2016-02-06" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix regression bug that made RNAcofold -a unusable</p></li>
<li><p>Fix regression bug that prohibited RNAfold to compute the MEA structure when G-Quadruplex support was switched on</p></li>
<li><p>Fix bug in Kinfold to enable loading energy parameters from file</p></li>
<li><p>Fix potential use of uninitialized value in RNApdist</p></li>
<li><p>Add manpage for ct2db</p></li>
<li><p>Fix MEA computation when G-Quadruplex support is activated</p></li>
<li><p>Allow for vendor installation of the perl interface using INSTALLDIRS=vendor at configure time</p></li>
<li><p>Install architecture dependent and independent files of the perl and python interface to their correct file system locations</p></li>
</ul>
</section>
<section id="version-2-2-0-release-date-2016-01-25">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.9...v2.2.0">Version 2.2.0</a> (Release date: 2016-01-25)<a class="headerlink" href="#version-2-2-0-release-date-2016-01-25" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>RNAforester is now of version 2.0</p></li>
<li><p>New program RNApvmin to compute pseudo-energy pertubation vector that minimizes discrepancy between observed and predicted pairing probabilities</p></li>
<li><p>SHAPE reactivity support for RNAfold, RNAsubopt, and RNAalifold</p></li>
<li><p>Ligand binding to hairpin- and interior-loop motif support in RNAfold</p></li>
<li><p>New commandline option to limit maximum base pair span for RNAfold, RNAsubopt, RNAcofold, and RNAalifold</p></li>
<li><p>Bugfix in RNAheat to remove numerical instabilities</p></li>
<li><p>Bugfix in RNAplex to allow for computation of interactions without length limitation</p></li>
<li><p>Bugfix in RNAplot for simple layouts and hairpins of size 0</p></li>
<li><p>(generic) hard- and soft-constraints for MFE, partition function, base pair probabilities, stochastic backtracking, and suboptimal secondary structures of single sequences, sequence alignments, and sequence dimers</p></li>
<li><p>libsvm version as required for z-scoring in RNALfold is now 3.20</p></li>
<li><p>Stochastic backtracking for single sequences is faster due to usage of Boustrophedon scheme</p></li>
<li><p>First polymorphic functions vrna_mfe(), vrna_pf(), and vrna_pbacktrack().</p></li>
<li><p>The FLT_OR_DBL macro is now a typedef</p></li>
<li><p>New functions to convert between different secondary structure representations, such as helix lists, and RNAshapes abstractions</p></li>
<li><p>First object-oriented interface for new API functions in the scripting language interfaces</p></li>
<li><p>new ViennaRNA-perl submodule that augments the Perl interface to RNAlib</p></li>
<li><p>Ligand binding to hairpin- and interior-loop motif support in C-library and scripting language interfaces.</p></li>
<li><p>Libraries are generated using libtool</p></li>
<li><p>Linking of libraries and executables defaults to use Link Time Optimization (LTO)</p></li>
<li><p>Large changes in directory structure of the source code files</p></li>
</ul>
</section>
</section>
<section id="version-2-1-x">
<h2>Version 2.1.x<a class="headerlink" href="#version-2-1-x" title="Permalink to this headline"></a></h2>
<section id="version-2-1-9">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.8...v2.1.9">Version 2.1.9</a><a class="headerlink" href="#version-2-1-9" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fixed integer underflow bug in RNALfold</p></li>
<li><p>Added Sequence Conservation index (SCI) option to RNAalifold</p></li>
<li><p>Fixed bug in energy evaluation of dangling ends / terminal mismatches of exterior loops and multibranch loops</p></li>
<li><p>Fixed bug in alifold partition function for circular RNAs</p></li>
<li><p>Fixed bug in alifold that scrambled backtracing with activated G-Quadruplex support</p></li>
<li><p>Fixed bug in alifold backtracking for larger G-Quadruplexes</p></li>
</ul>
</section>
<section id="version-2-1-8">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.7...v2.1.8">Version 2.1.8</a><a class="headerlink" href="#version-2-1-8" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Repaired incorporation of RNAinverse user provided alphabet</p></li>
<li><p>Fix missing FASTA ID in RNAeval output</p></li>
<li><p>prevent race condition in parallel calls of Lfold()</p></li>
<li><p>Fixed memory bug in Lfold() that occured using long sequences and activated G-Quad support</p></li>
<li><p>Added latest version of switch.pl</p></li>
</ul>
</section>
<section id="version-2-1-7">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.6...v2.1.7">Version 2.1.7</a><a class="headerlink" href="#version-2-1-7" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fixed bug in RNALfold -z</p></li>
<li><p>Python and Perl interface are compiling again under MacOSX</p></li>
<li><p>Fixed handling of C arrays in Python interface</p></li>
<li><p>Added latest version of switch.pl</p></li>
<li><p>Make relplot.pl work with RNAcofold output</p></li>
</ul>
</section>
<section id="version-2-1-6">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.5...v2.1.6">Version 2.1.6</a><a class="headerlink" href="#version-2-1-6" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>New commandline switches allow for elimination of non-canonical base pairs from constraint structures in RNAfold, RNAalifold and RNAsubopt</p></li>
<li><p>updated moveset functions</p></li>
<li><p>final fix for discrepancy of tri-loop evaluation between partition function and mfe</p></li>
<li><p>pkg-config file now includes the OpenMP linker flag if necessary</p></li>
<li><p>New program ct2db allows for conversion of .ct files into dot-bracket notation (incl. pseudo-knot removal)</p></li>
</ul>
</section>
<section id="version-2-1-5">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.4...v2.1.5">Version 2.1.5</a><a class="headerlink" href="#version-2-1-5" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix for discrepancy between special hairpin loop evaluation in partition functions and MFE</p></li>
</ul>
</section>
<section id="version-2-1-4">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.3...v2.1.4">Version 2.1.4</a><a class="headerlink" href="#version-2-1-4" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Fix of G-quadruplex support in subopt()</p></li>
<li><p>Fix for discrepancy between special hairpin loop evaluation in partition functions and MFE</p></li>
</ul>
</section>
<section id="version-2-1-3">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.2...v2.1.3">Version 2.1.3</a><a class="headerlink" href="#version-2-1-3" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>RNAfold: Bugfix for ignoring user specified energy parameter files</p></li>
<li><p>RNAcofold: Bugfix for crashing upon constrained folding without specifying a constraint structure</p></li>
<li><p>RNAsubopt: Added G-quadruplex support</p></li>
<li><p>RNAalifold: Added parameter option to specify base pair probability threshold in dotplot</p></li>
<li><p>Fix of several G-quadruplex related bugs</p></li>
<li><p>Added G-quadruplex support in subopt()</p></li>
</ul>
</section>
<section id="version-2-1-2">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.1...v2.1.2">Version 2.1.2</a><a class="headerlink" href="#version-2-1-2" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>RNAfold: Bugfix for randomly missing probabilities in dot-plot during batch job execution</p></li>
<li><p>RNAeval: Bugfix for misinterpreted G-quadruplex containing sequences where the quadruplex starts at nucleotide 1</p></li>
<li><p>RNAsubopt: Slight changes to the output of stochastic backtracking and zuker subopt</p></li>
<li><p>Fix of some memory leaks</p></li>
<li><p>Bugfixes in zukersubopt(), assign_plist_from_pr()</p></li>
<li><p>New threadsafe variants of putoutpU_prob*() for LPfold()</p></li>
<li><p>Provision of python2 interface support.</p></li>
</ul>
</section>
<section id="version-2-1-1">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.1.0...v2.1.1">Version 2.1.1</a><a class="headerlink" href="#version-2-1-1" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Bugfix to restore backward compatibility with ViennaRNA Package 1.8.x API (this bug also affected proper usage of the the perl interface)</p></li>
</ul>
</section>
<section id="version-2-1-0">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.0.7...v2.1.0">Version 2.1.0</a><a class="headerlink" href="#version-2-1-0" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>G-Quadruplex support in RNAfold, RNAcofold, RNALfold, RNAalifold, RNAeval and RNAplot</p></li>
<li><p>LPfold got a new option to output its computations in split-mode</p></li>
<li><p>several G-Quadruplex related functions were introduced with this release</p></li>
<li><p>several functions for moves in an RNA landscape were introduced</p></li>
<li><p>new function in alipfold.c now enables access to the partition function matrices of alipf_fold()</p></li>
<li><p>different numeric approach was implement for concentration dependend co-folding to avoid instabilities which occured under certain circumstances</p></li>
</ul>
</section>
</section>
<section id="version-2-0-x">
<h2>Version 2.0.x<a class="headerlink" href="#version-2-0-x" title="Permalink to this headline"></a></h2>
<section id="version-2-0-7">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.0.6...v2.0.7">Version 2.0.7</a><a class="headerlink" href="#version-2-0-7" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Bugfix for RNAplfold where segfault happened upon usage of -O option</p></li>
<li><p>Corrected misbehavior of RNAeval and RNAplot in tty mode</p></li>
</ul>
</section>
<section id="version-2-0-6">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.0.5...v2.0.6">Version 2.0.6</a><a class="headerlink" href="#version-2-0-6" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Bugfix for bad type casting with gcc under MacOSX (resulted in accidental “sequence too long” errors)</p></li>
<li><p>Bugfix for disappearing tri-/hexaloop contributions when read in from certain parameter files</p></li>
<li><p>Bugfix for RNALfold that segfaulted on short strange sequences like AT+ repeats</p></li>
<li><p>Change of RNA2Dfold output format for stochastic backtracking</p></li>
</ul>
</section>
<section id="version-2-0-5">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.0.4...v2.0.5">Version 2.0.5</a><a class="headerlink" href="#version-2-0-5" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Restored z-score computation capabilities in RNALfold</p></li>
</ul>
</section>
<section id="version-2-0-4">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.0.3...v2.0.4">Version 2.0.4</a><a class="headerlink" href="#version-2-0-4" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Bugfix for RNAcofold partition function</p></li>
<li><p>Perl wrapper compatibility to changed RNAlib has been restored</p></li>
<li><p>Backward compatibility for partition function calls has been restored</p></li>
</ul>
</section>
<section id="version-2-0-3">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.0.2...v2.0.3">Version 2.0.3</a><a class="headerlink" href="#version-2-0-3" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Bugfix for RNAalifold partition function and base pair probabilities in v2.0.3b</p></li>
<li><p>Added Boltzmann factor scaling in RNAsubopt, RNAalifold, RNAplfold and RNAcofold</p></li>
<li><p>Bugfix for alipfold() in v2.0.3b</p></li>
<li><p>Restored threadsafety of folding matrix access in LPfold.c, alipfold.c, part_func.c, part_func_co.c and part_func_up.c</p></li>
<li><p>Added several new functions regarding threadsafe function calls in terms of concurrently changing the model details</p></li>
<li><p>Added pkg-config file in the distribution to allow easy checks for certain RNAlib2 versions, compiler flags and linker flags.</p></li>
</ul>
</section>
<section id="version-2-0-2">
<h3><a class="reference external" href="https://github.com/ViennaRNA/ViennaRNA/compare/v2.0.1...v2.0.2">Version 2.0.2</a><a class="headerlink" href="#version-2-0-2" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>added support for Boltzmann factor scaling in RNAfold</p></li>
<li><p>fixed fastaheader to filename bug</p></li>
<li><p>plugged some memory leaks</p></li>
</ul>
</section>
<section id="version-2-0-1">
<h3>Version 2.0.1<a class="headerlink" href="#version-2-0-1" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>First official release of version 2.0</p></li>
<li><p>included latest bugfixes</p></li>
</ul>
</section>
</section>
<section id="history">
<h2>History<a class="headerlink" href="#history" title="Permalink to this headline"></a></h2>
<p>2011-03-10  Ronny Lorenz  <a class="reference external" href="mailto:ronny&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ronny<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>new naming scheme for all shipped energy parameter files</p></li>
<li><p>fixed bugs that appear while compiling with gcc under MacOS X</p></li>
<li><p>fixed bug in RNAup –interaction-first where the longer of the first two
sequences was taken as target</p></li>
<li><p>added full FASTA input support to RNAfold, RNAcofold, RNAheat, RNAplfold
RNALfoldz, RNAsubopt and RNALfold</p></li>
</ul>
<p>2010-11-24  Ronny Lorenz  <a class="reference external" href="mailto:ronny&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ronny<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>first full pre-release of version 2.0</p></li>
</ul>
<p>2009-11-03  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Fix memory corruption in PS_color_aln()</p></li>
</ul>
<p>2009-09-09  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Fix bug in RNAplfold when -u and -L parameters are equal</p></li>
<li><p>Fix double call to free_arrays() in RNAfold.c</p></li>
<li><p>Improve drawing of cofolded structures</p></li>
</ul>
<p>2009-05-14  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Fix occasional segfault in RNAalifold’s print_aliout()</p></li>
</ul>
<p>2009-02-24  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Add -MEA options to RNAfold and RNAalifold</p></li>
<li><p>change energy_of_alistruct to return float not void</p></li>
</ul>
<p>2009-02-24  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>RNAfold will draw structures unless -noPS is used (no more
“structure too long” messages)</p></li>
<li><p>Restore the “alifold.out” output from RNAalifold -p</p></li>
<li><p>RNAalifold -circ did not work due to wrong return type</p></li>
<li><p>Accessibility calculation with RNAplfold would give wrong
results for u&lt;=30</p></li>
</ul>
<p>2008-12-03  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Add zuker style suboptimals to RNAsubopt (-z)</p></li>
<li><p>get_line() should be much faster when reading huge sequences
(e.g. whole chromosomes for RNALfold)</p></li>
</ul>
<p>2008-08-12  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Add Ribosum matrices for covariance scoring in RNAalifold</p></li>
</ul>
<p>2008-06-27  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Change RNAalifold to used berni’s new energy evaluation w/o gaps</p></li>
<li><p>Add stochastic backtracking in RNAalifold</p></li>
</ul>
<p>2008-07-04  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>modify output of RNAup (again).
Program reading RNAup output will have to updated!</p></li>
</ul>
<p>2008-07-02  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>RNAplfold now computes accessibilities for all regions up to a
max length simultaneously. Slightly slower when only 1 value is
needed, but much faster if all of them are wanted.
This entails a new output format. Programs reading accessibility
output from RNAplfold need to be updated!</p></li>
</ul>
<p>2008-03-31  Stephan Bernhart  <a class="reference external" href="mailto:berni&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:berni<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add cofolding to RNAsubopt</p></li>
</ul>
<p>2008-01-08  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>ensure circfold works even for open chain</p></li>
</ul>
<p>2007-12-13  Ulli Mueckstein  <a class="reference external" href="mailto:ulli&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ulli<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>upate RNAup related files
RNAup can now include the intramolecular structure of both
molecules and handles constraints.</p></li>
</ul>
<p>2007-12-05  Ronny Lorenz  <a class="reference external" href="mailto:ronny&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ronny<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add circfold variants in part_func.c alipfold.c subopt.c</p></li>
</ul>
<p>2007-09-19  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>compute the controid structure of the ensemble in RNAfold -p</p></li>
<li><p>fix a missing factor 2 in mean_bp_dist().
CAUTION ensemble diversities returned by RNAfold -p are now
twice as large as in earlier versions.</p></li>
</ul>
<p>2007-09-04  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;blini&#46;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>blini<span>&#46;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>fix a bug in Lfold() where base number n-max-4 would never pair</p></li>
</ul>
<p>2007-08-26  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add RNAaliduplex the alignment version of RNAduplex</p></li>
<li><p>introduce a minimal distance between hits produced by  duplex_subopt()</p></li>
</ul>
<p>2007-07-03  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add a loop_energy() function to compute energy of a single loop</p></li>
</ul>
<p>2007-06-23  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add aliLfold() and RNALalifold, alignment variant of Lfold()</p></li>
</ul>
<p>2007-04-30  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add RNAup to distribution</p></li>
</ul>
<p>2007-04-15  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>fix segfault in colorps output (thanks to Andres Varon)</p></li>
</ul>
<p>2007-03-03  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>avoid unnormalized doubles in scale[], big speedup for pf_fold()
on very long sequences</p></li>
</ul>
<p>2007-02-03  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>RNAalifold can now produce colored structure plots and
alignment plots</p></li>
</ul>
<p>2007-02-01 Ivo Hofacker <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Fix segfault in RNAplfold because of missing prototype</p></li>
</ul>
<p>2006-12-01 Ivo Hofacker <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>RNAduplex would segfault when no structure base pairs are possible</p></li>
</ul>
<p>2006-08-22  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add computation stacking probabilities using RNAfold -p2</p></li>
<li><p>add -noPS option for NRAfold to supress drawing structures</p></li>
</ul>
<p>2006-08-09  Stephan Bernhart  <a class="reference external" href="mailto:berni&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:berni<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>RNAplfold can now compute probabilites of unpaired regions
(scanning version of RNAup)</p></li>
</ul>
<p>2006-06-14  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>compile library with -fpic (if available) for use as shared
library in the Perl module.</p></li>
<li><p>fix another bug when calling Lfold() repeatedly</p></li>
<li><p>fix switch cmdline parsing in RNAalifold (-mis implied -4)</p></li>
<li><p>fix bug in cofold() with dangles=0</p></li>
</ul>
<p>2006-05-08  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>fix segfault in Lfold() when calling repeatedly</p></li>
<li><p>fix structure parsing in RNAstruct.c
(thanks to Michael Pheasant for reporting both bugs)</p></li>
<li><p>add duplexfold() and alifold() to Perl module</p></li>
<li><p>distinguish window size and max pair span in LPfold</p></li>
</ul>
<p>2006-04-05  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>fix performance bug in co_pf_fold()</p></li>
<li><p>use relative error for termination of Newton iteration</p></li>
</ul>
<p>2006-03-02  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add circular folding in alifold()</p></li>
</ul>
<p>2006-01-18  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>cleanup berni partition cofold code, including several bug fixes</p></li>
</ul>
<p>2006-01-16  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>update RNAplfold to working version</p></li>
<li><p>add PS_dot_plot_turn() in  PS_dot.c</p></li>
</ul>
<p>2005-11-07  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add new utilities colorna and coloraln</p></li>
</ul>
<p>2005-10-11  Christoph Flamm  <a class="reference external" href="mailto:xtof&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:xtof<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>adapt PS_rna_plot() for drawing co-folded structures</p></li>
</ul>
<p>2005-07-24  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>fix a few memory problems in structure comparison routines</p></li>
</ul>
<p>2005-04-30  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;blini&#46;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>blini<span>&#46;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add folding of circular RNAs</p></li>
</ul>
<p>2005-03-11  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;blini&#46;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>blini<span>&#46;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add -mis option to RNAalifold to give “most informative
sequence” as consensus</p></li>
</ul>
<p>2005-02-10  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>move alifold() into the library</p></li>
</ul>
<p>2004-12-22  Stephan Bernhart  <a class="reference external" href="mailto:berni&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:berni<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add partition function version of RNAcofold</p></li>
</ul>
<p>2004-12-23  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add RNApaln for fast structural alignments (RNApdist improvement)</p></li>
</ul>
<p>2004-08-12  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>fix constrained folding in stochastic backtracking</p></li>
</ul>
<p>2004-07-21  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add RNAduplex, to compute hybrid structures without
intra-molecular pairs</p></li>
</ul>
<p>2004-02-09  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>fix bug in fold that caused segfaults when using Intel compiler</p></li>
<li><p>add computation of ensemble diversity to RNAfold</p></li>
</ul>
<p>2003-09-10  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add annotation options to RNAplot</p></li>
</ul>
<p>2003-08-04  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>stochastic backtracking finally works. Try e.g.
RNAsubopt -p 10</p></li>
</ul>
<p>2003-07-18  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add relplot.pl and rotate_ss.pl utilities for reliability
annotation and rotation of rna structure plots</p></li>
</ul>
<p>2003-01-29  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add RNALfold program to compute locally optimal structures with
maximum pair span.</p></li>
<li><p>add RNAcofold for computing hybrid structure</p></li>
</ul>
<p>2002-11-07  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>change Make_bp_profile() and profile_edit_distance() to use
simple (float *) arrays; makes Perl access much easier.
RNApdist -B now works again</p></li>
</ul>
<p>2002-10-28  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Improved Perl module with pod documentation; allow to write
things like
(<span class="math notranslate nohighlight">\(structure, \)</span>energy) = RNA::fold($seq);
Compatibility warning: the ptrvalue() and related functions are
gone, see the pod documentation for alternatives.</p></li>
</ul>
<p>2002-10-29  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>added svg structure plots in PS_dot.c and RNAplot</p></li>
</ul>
<p>2002-08-15  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>Improve reading of clustal files (alifold)</p></li>
<li><p>add a sample alifold.cgi script</p></li>
</ul>
<p>2001-09-18  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>moved suboptimal folding into the library, thus it’s now
accessible from the Perl module</p></li>
</ul>
<p>2001-08-31  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>added co-folding support in energy_of_struct(), and thus RNAeval</p></li>
</ul>
<p>2001-04-30  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>switch from handcrafted makefiles to automake and autoconf</p></li>
</ul>
<p>2001-04-05  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>added PS_rna_plot_a to produce structure plots with annotation</p></li>
</ul>
<p>2001-03-03  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add alifold; predict consensus structures from alignment</p></li>
</ul>
<p>2000-09-28  Ivo Hofacker  <a class="reference external" href="mailto:ivo&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">mailto:ivo<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a></p>
<ul class="simple">
<li><p>add -d3 option to RNAfold for co-axial stacking</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="news.html" class="btn btn-neutral float-left" title="News" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="bibliography.html" class="btn btn-neutral float-right" title="Bibliography" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>