<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>RNAcofold &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/vrna_32.png"/>
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="RNAdistance" href="RNAdistance.html" />
    <link rel="prev" title="RNAalifold" href="RNAalifold.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../manpages.html">Manpages</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="RNA2Dfold.html">RNA2Dfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAaliduplex.html">RNAaliduplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAalifold.html">RNAalifold</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">RNAcofold</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#synopsis">Synopsis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#description">DESCRIPTION</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#i-o-options">I/O Options:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#algorithms">Algorithms:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#structure-constraints">Structure Constraints:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#energy-parameters">Energy Parameters:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#model-details">Model Details:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#plotting">Plotting:</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#references">REFERENCES</a></li>
<li class="toctree-l3"><a class="reference internal" href="#author">AUTHOR</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reporting-bugs">REPORTING BUGS</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="RNAdistance.html">RNAdistance</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAdos.html">RNAdos</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAduplex.html">RNAduplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAeval.html">RNAeval</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAfold.html">RNAfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAheat.html">RNAheat</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAinverse.html">RNAinverse</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNALalifold.html">RNALalifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNALfold.html">RNALfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAmultifold.html">RNAmultifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApaln.html">RNApaln</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAparconv.html">RNAparconv</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApdist.html">RNApdist</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAPKplex.html">RNAPKplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplex.html">RNAplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplfold.html">RNAplfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplot.html">RNAplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApvmin.html">RNApvmin</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAsnoop.html">RNAsnoop</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAsubopt.html">RNAsubopt</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAup.html">RNAup</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1"><a class="reference internal" href="../io.html">I/O Formats</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../manpages.html">Manpages</a> &raquo;</li>
      <li>RNAcofold</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnacofold">
<h1>RNAcofold<a class="headerlink" href="#rnacofold" title="Permalink to this headline"></a></h1>
<p><strong class="program">RNAcofold</strong> - manual page for RNAcofold 2.6.4</p>
<section id="synopsis">
<h2>Synopsis<a class="headerlink" href="#synopsis" title="Permalink to this headline"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>RNAcofold <span class="o">[</span>OPTION<span class="o">]</span>... <span class="o">[</span>FILE<span class="o">]</span>...
</pre></div>
</div>
</section>
<section id="description">
<h2>DESCRIPTION<a class="headerlink" href="#description" title="Permalink to this headline"></a></h2>
<p>RNAcofold 2.6.4</p>
<p>calculate secondary structures of two RNAs with dimerization</p>
<p>The program works much like RNAfold, but allows one to specify two RNA
sequences which are then allowed to form a dimer structure. RNA sequences are
read from stdin in the usual format, i.e. each line of input corresponds to one
sequence, except for lines starting with <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> which contain the name of the next
sequence.
To compute the hybrid structure of two molecules, the two sequences must be
concatenated using the <code class="docutils literal notranslate"><span class="pre">&amp;</span></code> character as separator.
RNAcofold can compute minimum free energy (mfe) structures, as well as
partition function (pf) and base pairing probability matrix (using the <a class="reference internal" href="RNAsubopt.html#cmdoption-p"><code class="xref std std-option docutils literal notranslate"><span class="pre">-p</span></code></a>
switch)
Since dimer formation is concentration dependent, RNAcofold can be used to
compute equilibrium concentrations for all five monomer and (homo/hetero)-dimer
species, given input concentrations for the monomers.
Output consists of the mfe structure in bracket notation as well as PostScript
structure plots and “dot plot” files containing the pair probabilities, see
the RNAfold man page for details. In the dot plots a cross marks the chain
break between the two concatenated sequences.
The program will continue to read new sequences until a line consisting of the
single character <code class="docutils literal notranslate"><span class="pre">&#64;</span></code> or an end of file condition is encountered.</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-h">
<span id="cmdoption-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-h" title="Permalink to this definition"></a></dt>
<dd><p>Print help and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-detailed-help">
<span class="sig-name descname"><span class="pre">--detailed-help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-detailed-help" title="Permalink to this definition"></a></dt>
<dd><p>Print help, including all details and hidden options, and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-full-help">
<span class="sig-name descname"><span class="pre">--full-help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-full-help" title="Permalink to this definition"></a></dt>
<dd><p>Print help, including hidden options, and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-V">
<span id="cmdoption-v"></span><span id="cmdoption-version"></span><span class="sig-name descname"><span class="pre">-V</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--version</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-V" title="Permalink to this definition"></a></dt>
<dd><p>Print version and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-0">
<span id="cmdoption-verbose"></span><span class="sig-name descname"><span class="pre">-v</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--verbose</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-0" title="Permalink to this definition"></a></dt>
<dd><p>Be verbose.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<section id="i-o-options">
<h3>I/O Options:<a class="headerlink" href="#i-o-options" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Command line options for input and output (pre-)processing</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-output-format">
<span class="sig-name descname"><span class="pre">--output-format</span></span><span class="sig-prename descclassname"><span class="pre">=format-character</span></span><a class="headerlink" href="#cmdoption-output-format" title="Permalink to this definition"></a></dt>
<dd><p>Change the default output format.</p>
<p><em>(default=”V”)</em></p>
<p>The following output formats are currently supported:</p>
<p>ViennaRNA format (<code class="docutils literal notranslate"><span class="pre">V</span></code>), Delimiter-separated format (<code class="docutils literal notranslate"><span class="pre">D</span></code>) also known as <code class="docutils literal notranslate"><span class="pre">CSV</span></code></p>
<p>format.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-csv-delim">
<span class="sig-name descname"><span class="pre">--csv-delim</span></span><span class="sig-prename descclassname"><span class="pre">=delimiter</span></span><a class="headerlink" href="#cmdoption-csv-delim" title="Permalink to this definition"></a></dt>
<dd><p>Change the delimiting character for Delimiter-separated output format, such as <code class="docutils literal notranslate"><span class="pre">CSV</span></code>.</p>
<p><em>(default=”,”)</em></p>
<p>Delimiter-separated output defaults to comma separated values (<code class="docutils literal notranslate"><span class="pre">CSV</span></code>), i.e.
all data in one data set is delimited by a comma character. This option
allows one to change the delimiting character to something else. Note, to
switch to tab-separated data, use <code class="docutils literal notranslate"><span class="pre">$'\t'</span></code> as delimiting character.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-csv-noheader">
<span class="sig-name descname"><span class="pre">--csv-noheader</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-csv-noheader" title="Permalink to this definition"></a></dt>
<dd><p>Do not print header for Delimiter-separated output, such as CSV.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-j">
<span id="cmdoption-jobs"></span><span class="sig-name descname"><span class="pre">-j</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--jobs</span></span><span class="sig-prename descclassname"><span class="pre">[=number]</span></span><a class="headerlink" href="#cmdoption-j" title="Permalink to this definition"></a></dt>
<dd><p>Split batch input into jobs and start processing in parallel using multiple threads. A value of 0 indicates to use as many parallel threads as computation cores are available.</p>
<p><em>(default=”0”)</em></p>
<p>Default processing of input data is performed in a serial fashion, i.e. one
sequence pair at a time. Using this switch, a user can instead start the
computation for many sequence pairs in the input in parallel. RNAcofold will
create as many parallel computation slots as specified and assigns input
sequences of the input file(s) to the available slots. Note, that this
increases memory consumption since input alignments have to be kept in memory
until an empty compute slot is available and each running job requires its
own dynamic programming matrices.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-unordered">
<span class="sig-name descname"><span class="pre">--unordered</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-unordered" title="Permalink to this definition"></a></dt>
<dd><p>Do not try to keep output in order with input while parallel processing is in place.</p>
<p><em>(default=off)</em></p>
<p>When parallel input processing (<a class="reference internal" href="RNAplot.html#cmdoption-j"><code class="xref std std-option docutils literal notranslate"><span class="pre">--jobs</span></code></a> flag) is enabled, the order in which
input is processed depends on the host machines job scheduler. Therefore, any
output to stdout or files generated by this program will most likely not
follow the order of the corresponding input data set. The default of
RNAcofold is to use a specialized data structure to still keep the results
output in order with the input data. However, this comes with a trade-off in
terms of memory consumption, since all output must be kept in memory for as
long as no chunks of consecutive, ordered output are available. By setting
this flag, RNAcofold will not buffer individual results but print them as
soon as they have been computated.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noconv">
<span class="sig-name descname"><span class="pre">--noconv</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noconv" title="Permalink to this definition"></a></dt>
<dd><p>Do not automatically substitute nucleotide “T” with “U”.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-auto-id">
<span class="sig-name descname"><span class="pre">--auto-id</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-auto-id" title="Permalink to this definition"></a></dt>
<dd><p>Automatically generate an ID for each sequence. <em>(default=off)</em></p>
<p>The default mode of RNAcofold is to automatically determine an ID from the
input sequence data if the input file format allows to do that. Sequence IDs
are usually given in the FASTA header of input sequences. If this flag is
active, RNAcofold ignores any IDs retrieved from the input and automatically
generates an ID for each sequence. This ID consists of a prefix and an
increasing number. This flag can also be used to add a FASTA header to the
output even if the input has none.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-prefix">
<span class="sig-name descname"><span class="pre">--id-prefix</span></span><span class="sig-prename descclassname"><span class="pre">=STRING</span></span><a class="headerlink" href="#cmdoption-id-prefix" title="Permalink to this definition"></a></dt>
<dd><p>Prefix for automatically generated IDs (as used in output file names).</p>
<p><em>(default=”sequence”)</em></p>
<p>If this parameter is set, each sequence will be prefixed with the provided
string. Hence, the output files will obey the following naming scheme:
“prefix_xxxx_ss.ps” (secondary structure plot), “prefix_xxxx_dp.ps”
(dot-plot), “prefix_xxxx_dp2.ps” (stack probabilities), etc. where xxxx is
the sequence number. Note: Setting this parameter implies <a class="reference internal" href="RNAsubopt.html#cmdoption-auto-id"><code class="xref std std-option docutils literal notranslate"><span class="pre">--auto-id</span></code></a>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-delim">
<span class="sig-name descname"><span class="pre">--id-delim</span></span><span class="sig-prename descclassname"><span class="pre">=CHAR</span></span><a class="headerlink" href="#cmdoption-id-delim" title="Permalink to this definition"></a></dt>
<dd><p>Change the delimiter between prefix and increasing number for automatically generated IDs (as used in output file names).</p>
<p><em>(default=”_”)</em></p>
<p>This parameter can be used to change the default delimiter “_” between the
prefix string and the increasing number for automatically generated ID.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-digits">
<span class="sig-name descname"><span class="pre">--id-digits</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-id-digits" title="Permalink to this definition"></a></dt>
<dd><p>Specify the number of digits of the counter in automatically generated alignment IDs.</p>
<p><em>(default=”4”)</em></p>
<p>When alignments IDs are automatically generated, they receive an increasing
number, starting with 1. This number will always be left-padded by leading
zeros, such that the number takes up a certain width. Using this parameter,
the width can be specified to the users need. We allow numbers in the range
[1:18]. This option implies <a class="reference internal" href="RNAsubopt.html#cmdoption-auto-id"><code class="xref std std-option docutils literal notranslate"><span class="pre">--auto-id</span></code></a>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-start">
<span class="sig-name descname"><span class="pre">--id-start</span></span><span class="sig-prename descclassname"><span class="pre">=LONG</span></span><a class="headerlink" href="#cmdoption-id-start" title="Permalink to this definition"></a></dt>
<dd><p>Specify the first number in automatically generated IDs.</p>
<p><em>(default=”1”)</em></p>
<p>When sequence IDs are automatically generated, they receive an increasing
number, usually starting with 1. Using this parameter, the first number can
be specified to the users requirements. Note: negative numbers are not
allowed.
Note: Setting this parameter implies to ignore any IDs retrieved from the
input data, i.e. it activates the <a class="reference internal" href="RNAsubopt.html#cmdoption-auto-id"><code class="xref std std-option docutils literal notranslate"><span class="pre">--auto-id</span></code></a> flag.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-filename-delim">
<span class="sig-name descname"><span class="pre">--filename-delim</span></span><span class="sig-prename descclassname"><span class="pre">=CHAR</span></span><a class="headerlink" href="#cmdoption-filename-delim" title="Permalink to this definition"></a></dt>
<dd><p>Change the delimiting character used in sanitized filenames.</p>
<p><em>(default=”ID-delimiter”)</em></p>
<p>This parameter can be used to change the delimiting character used while
sanitizing filenames, i.e. replacing invalid characters. Note, that the
default delimiter ALWAYS is the first character of the “ID delimiter” as
supplied through the <a class="reference internal" href="RNAsubopt.html#cmdoption-id-delim"><code class="xref std std-option docutils literal notranslate"><span class="pre">--id-delim</span></code></a> option. If the delimiter is a whitespace
character or empty, invalid characters will be simply removed rather than
substituted. Currently, we regard the following characters as illegal for use
in filenames: backslash <code class="docutils literal notranslate"><span class="pre">\</span></code>, slash <code class="docutils literal notranslate"><span class="pre">/</span></code>, question mark <code class="docutils literal notranslate"><span class="pre">?</span></code>, percent sign <code class="docutils literal notranslate"><span class="pre">%</span></code>,
asterisk <code class="docutils literal notranslate"><span class="pre">*</span></code>, colon <code class="docutils literal notranslate"><span class="pre">:</span></code>, pipe symbol <code class="docutils literal notranslate"><span class="pre">|</span></code>, double quote <code class="docutils literal notranslate"><span class="pre">&quot;</span></code>, triangular
brackets <code class="docutils literal notranslate"><span class="pre">&lt;</span></code> and <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-filename-full">
<span class="sig-name descname"><span class="pre">--filename-full</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-filename-full" title="Permalink to this definition"></a></dt>
<dd><p>Use full FASTA header to create filenames. <em>(default=off)</em></p>
<p>This parameter can be used to deactivate the default behavior of limiting
output filenames to the first word of the sequence ID. Consider the following
example: An input with FASTA header <code class="docutils literal notranslate"><span class="pre">&gt;NM_0001</span> <span class="pre">Homo</span> <span class="pre">Sapiens</span> <span class="pre">some</span> <span class="pre">gene</span></code> usually
produces output files with the prefix “NM_0001” without the additional data
available in the FASTA header, e.g. “NM_0001_ss.ps” for secondary structure
plots. With this flag set, no truncation of the output filenames is done,
i.e. output filenames receive the full FASTA header data as prefixes. Note,
however, that invalid characters (such as whitespace) will be substituted by
a delimiting character or simply removed, (see also the parameter option
<a class="reference internal" href="RNAsubopt.html#cmdoption-filename-delim"><code class="xref std std-option docutils literal notranslate"><span class="pre">--filename-delim</span></code></a>).</p>
</dd></dl>

</section>
<section id="algorithms">
<h3>Algorithms:<a class="headerlink" href="#algorithms" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Select additional algorithms which should be included in the calculations.
The Minimum free energy (MFE) and a structure representative are calculated
in any case.</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-p">
<span id="cmdoption-partfunc"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--partfunc</span></span><span class="sig-prename descclassname"><span class="pre">[=INT]</span></span><a class="headerlink" href="#cmdoption-p" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the partition function and base pairing probability matrix in addition to the mfe structure. Default is calculation of mfe structure only.</p>
<p><em>(default=”1”)</em></p>
<p>In addition to the MFE structure we print a coarse representation of the pair
probabilities in form of a pseudo bracket notation, followed by the ensemble
free energy, as well as the centroid structure derived from the pair
probabilities together with its free energy and distance to the ensemble.
Finally it prints the frequency of the mfe structure, and the structural
diversity (mean distance between the structures in the ensemble).
See the description of pf_fold() and mean_bp_dist() and centroid() in the
RNAlib documentation for details.
Note that unless you also specify <code class="xref std std-option docutils literal notranslate"><span class="pre">-d2</span></code> or <code class="xref std std-option docutils literal notranslate"><span class="pre">-d0</span></code>, the partition function and mfe
calculations will use a slightly different energy model. See the discussion
of dangling end options below.</p>
<p>An additionally passed value to this option changes the behavior of partition
function calculation:</p>
<p>In order to calculate the partition function but not the pair probabilities</p>
<p>use the <code class="xref std std-option docutils literal notranslate"><span class="pre">-p0</span></code> option and save about</p>
<p>50% in runtime. This prints the ensemble free energy <code class="docutils literal notranslate"><span class="pre">dG=-kT</span> <span class="pre">ln(Z)</span></code>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-a">
<span id="cmdoption-all_pf"></span><span id="cmdoption-all-pf"></span><span class="sig-name descname"><span class="pre">-a</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--all_pf</span></span><span class="sig-prename descclassname"><span class="pre">[=INT]</span></span><a class="headerlink" href="#cmdoption-a" title="Permalink to this definition"></a></dt>
<dd><p>Compute the partition function and free energies not only of the hetero-dimer consisting of the two input sequences (the <code class="docutils literal notranslate"><span class="pre">AB</span> <span class="pre">dimer</span></code>), but also of the homo-dimers AA and BB as well as A and B monomers.</p>
<p><em>(default=”1”)</em></p>
<p>The output will contain the free energies for each of these species, as well
as 5 dot plots containing the conditional pair probabilities, called
“ABname5.ps”, “AAname5.ps” and so on. For later use, these dot plot files
also contain the free energy of the ensemble as a comment. Using <a class="reference internal" href="RNAsnoop.html#cmdoption-2"><code class="xref std std-option docutils literal notranslate"><span class="pre">-a</span></code></a>
automatically switches on the <a class="reference internal" href="RNAsubopt.html#cmdoption-p"><code class="xref std std-option docutils literal notranslate"><span class="pre">-p</span></code></a> option. Base pair probability computations
may be turned off altogether by providing <code class="docutils literal notranslate"><span class="pre">0</span></code> as an argument to this
parameter. In that case, no dot plot files will be generated.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-betaScale">
<span id="cmdoption-betascale"></span><span class="sig-name descname"><span class="pre">--betaScale</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-betaScale" title="Permalink to this definition"></a></dt>
<dd><p>Set the scaling of the Boltzmann factors. <em>(default=”1.”)</em></p>
<p>The argument provided with this option is used to scale the thermodynamic
temperature in the Boltzmann factors independently from the temperature of
the individual loop energy contributions. The Boltzmann factors then become
<code class="docutils literal notranslate"><span class="pre">exp(-</span> <span class="pre">dG/(kT*betaScale))</span></code> where <code class="docutils literal notranslate"><span class="pre">k</span></code> is the Boltzmann constant, <code class="docutils literal notranslate"><span class="pre">dG</span></code> the free
energy contribution of the state and <code class="docutils literal notranslate"><span class="pre">T</span></code> the absolute temperature.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-S">
<span id="cmdoption-s"></span><span id="cmdoption-pfScale"></span><span id="cmdoption-pfscale"></span><span class="sig-name descname"><span class="pre">-S</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--pfScale</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-S" title="Permalink to this definition"></a></dt>
<dd><p>In the calculation of the pf use scale*mfe as an estimate for the ensemble free energy (used to avoid overflows).</p>
<p><em>(default=”1.07”)</em></p>
<p>The default is 1.07, useful values are 1.0 to 1.2. Occasionally needed for
long sequences.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-c">
<span id="cmdoption-concentrations"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--concentrations</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-c" title="Permalink to this definition"></a></dt>
<dd><p>In addition to everything listed under the <a class="reference internal" href="RNAsnoop.html#cmdoption-2"><code class="xref std std-option docutils literal notranslate"><span class="pre">-a</span></code></a> option, read in initial monomer concentrations and compute the expected equilibrium concentrations of the 5 possible species (AB, AA, BB, A, B).</p>
<p><em>(default=off)</em></p>
<p>Start concentrations are read from stdin (unless the <a class="reference internal" href="RNAsnoop.html#cmdoption-f"><code class="xref std std-option docutils literal notranslate"><span class="pre">-f</span></code></a> option is used) in
[mol/l], equilibrium concentrations are given realtive to the sum of the two
inputs. An arbitrary number of initial concentrations can be specified (one
pair of concentrations per line).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-f">
<span id="cmdoption-concfile"></span><span class="sig-name descname"><span class="pre">-f</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--concfile</span></span><span class="sig-prename descclassname"><span class="pre">=filename</span></span><a class="headerlink" href="#cmdoption-f" title="Permalink to this definition"></a></dt>
<dd><p>Specify a file with initial concentrations for the two sequences.</p>
<p>The table consits of arbitrary many lines with just two numbers (the
concentration of sequence A and B). This option will automatically toggle the
<a class="reference internal" href="RNAup.html#cmdoption-c"><code class="xref std std-option docutils literal notranslate"><span class="pre">-c</span></code></a> (and thus <a class="reference internal" href="RNAsnoop.html#cmdoption-2"><code class="xref std std-option docutils literal notranslate"><span class="pre">-a</span></code></a> and <a class="reference internal" href="RNAsubopt.html#cmdoption-p"><code class="xref std std-option docutils literal notranslate"><span class="pre">-p</span></code></a>) options (see above).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-centroid">
<span class="sig-name descname"><span class="pre">--centroid</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-centroid" title="Permalink to this definition"></a></dt>
<dd><p>Compute the centroid structure. <em>(default=off)</em></p>
<p>Additionally to the MFE structure, compute the centroid representative of the
structure ensemble. Here, we apply the base pair distance as distance
measure, and report the structure that minimizes its Boltzmann weighted base
pair distance to the rest of the ensemble. Computing the centroid structure
requires equilibrium base pair probabilities. Therefore, this option implies
the <a class="reference internal" href="RNAsubopt.html#cmdoption-p"><code class="xref std std-option docutils literal notranslate"><span class="pre">-p</span></code></a> switch. For historical reasons, the centroid structure output is
deactivated by default.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-MEA">
<span id="cmdoption-mea"></span><span class="sig-name descname"><span class="pre">--MEA</span></span><span class="sig-prename descclassname"><span class="pre">[=gamma]</span></span><a class="headerlink" href="#cmdoption-MEA" title="Permalink to this definition"></a></dt>
<dd><p>Compute MEA (maximum expected accuracy) structure.</p>
<p><em>(default=”1.”)</em></p>
<p>The expected accuracy is computed from the pair probabilities: each base pair
<code class="docutils literal notranslate"><span class="pre">(i,j)</span></code> receives a score <code class="docutils literal notranslate"><span class="pre">2*gamma*p_ij</span></code> and the score of an unpaired base is
given by the probability of not forming a pair. The parameter gamma tunes the
importance of correctly predicted pairs versus unpaired bases. Thus, for
small values of gamma the MEA structure will contain only pairs with very
high probability. Using <a class="reference internal" href="RNAfold.html#cmdoption-MEA"><code class="xref std std-option docutils literal notranslate"><span class="pre">--MEA</span></code></a> implies <a class="reference internal" href="RNAsubopt.html#cmdoption-p"><code class="xref std std-option docutils literal notranslate"><span class="pre">-p</span></code></a> for computing the pair
probabilities.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-bppmThreshold">
<span id="cmdoption-bppmthreshold"></span><span class="sig-name descname"><span class="pre">--bppmThreshold</span></span><span class="sig-prename descclassname"><span class="pre">=cutoff</span></span><a class="headerlink" href="#cmdoption-bppmThreshold" title="Permalink to this definition"></a></dt>
<dd><p>Set the threshold/cutoff for base pair probabilities included in the postscript output.</p>
<p><em>(default=”1e-5”)</em></p>
<p>By setting the threshold the base pair probabilities that are included in the
output can be varied. By default only those exceeding <code class="docutils literal notranslate"><span class="pre">1e-5</span></code> in probability
will be shown as squares in the dot plot. Changing the threshold to any other
value allows for increase or decrease of data.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-g">
<span id="cmdoption-gquad"></span><span class="sig-name descname"><span class="pre">-g</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--gquad</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-g" title="Permalink to this definition"></a></dt>
<dd><p>Incoorporate G-Quadruplex formation into the structure prediction algorithm.</p>
<p><em>(default=off)</em></p>
</dd></dl>

</section>
<section id="structure-constraints">
<h3>Structure Constraints:<a class="headerlink" href="#structure-constraints" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Command line options to interact with the structure constraints feature of
this program</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-maxBPspan">
<span id="cmdoption-maxbpspan"></span><span class="sig-name descname"><span class="pre">--maxBPspan</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-maxBPspan" title="Permalink to this definition"></a></dt>
<dd><p>Set the maximum base pair span.</p>
<p><em>(default=”-1”)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-C">
<span id="cmdoption-constraint"></span><span class="sig-name descname"><span class="pre">-C</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--constraint</span></span><span class="sig-prename descclassname"><span class="pre">[=filename]</span></span><a class="headerlink" href="#cmdoption-C" title="Permalink to this definition"></a></dt>
<dd><p>Calculate structures subject to constraints. <em>(default=””)</em></p>
<p>The program reads first the sequence, then a string containing constraints on
the structure encoded with the symbols:</p>
<p><code class="docutils literal notranslate"><span class="pre">.</span></code> (no constraint for this base)</p>
<p><code class="docutils literal notranslate"><span class="pre">|</span></code> (the corresponding base has to be paired</p>
<p><code class="docutils literal notranslate"><span class="pre">x</span></code> (the base is unpaired)</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;</span></code> (base i is paired with a base j&gt;i)</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span></code> (base i is paired with a base j&lt;i)</p>
<p>and matching brackets <code class="docutils literal notranslate"><span class="pre">(</span></code> <code class="docutils literal notranslate"><span class="pre">)</span></code> (base i pairs base j)</p>
<p>With the exception of <code class="docutils literal notranslate"><span class="pre">|</span></code>, constraints will disallow all pairs conflicting
with the constraint. This is usually sufficient to enforce the constraint,
but occasionally a base may stay unpaired in spite of constraints. PF folding
ignores constraints of type <code class="docutils literal notranslate"><span class="pre">|</span></code>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-batch">
<span class="sig-name descname"><span class="pre">--batch</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-batch" title="Permalink to this definition"></a></dt>
<dd><p>Use constraints for multiple sequences. <em>(default=off)</em></p>
<p>Usually, constraints provided from input file only apply to a single input
sequence. Therefore, RNAcofold will stop its computation and quit after the
first input sequence was processed. Using this switch, RNAcofold processes
multiple input sequences and applies the same provided constraints to each of
them.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-canonicalBPonly">
<span id="cmdoption-canonicalbponly"></span><span class="sig-name descname"><span class="pre">--canonicalBPonly</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-canonicalBPonly" title="Permalink to this definition"></a></dt>
<dd><p>Remove non-canonical base pairs from the structure constraint.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-enforceConstraint">
<span id="cmdoption-enforceconstraint"></span><span class="sig-name descname"><span class="pre">--enforceConstraint</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-enforceConstraint" title="Permalink to this definition"></a></dt>
<dd><p>Enforce base pairs given by round brackets <code class="docutils literal notranslate"><span class="pre">(</span></code> <code class="docutils literal notranslate"><span class="pre">)</span></code> in structure constraint.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-shape">
<span class="sig-name descname"><span class="pre">--shape</span></span><span class="sig-prename descclassname"><span class="pre">=filename</span></span><a class="headerlink" href="#cmdoption-shape" title="Permalink to this definition"></a></dt>
<dd><p>Use SHAPE reactivity data to guide structure predictions.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-shapeMethod">
<span id="cmdoption-shapemethod"></span><span class="sig-name descname"><span class="pre">--shapeMethod</span></span><span class="sig-prename descclassname"><span class="pre">=method</span></span><a class="headerlink" href="#cmdoption-shapeMethod" title="Permalink to this definition"></a></dt>
<dd><p>Select SHAPE reactivity data incorporation strategy.</p>
<p><em>(default=”D”)</em></p>
<p>The following methods can be used to convert SHAPE reactivities into pseudo
energy contributions.</p>
<p><code class="docutils literal notranslate"><span class="pre">D</span></code>: Convert by using the linear equation according to Deigan et al 2009.</p>
<p>Derived pseudo energy terms will be applied for every nucleotide involved in
a stacked pair. This method is recognized by a capital <code class="docutils literal notranslate"><span class="pre">D</span></code> in the provided
parameter, i.e.: <code class="xref std std-option docutils literal notranslate"><span class="pre">--shapeMethod=</span></code>”D” is the default setting. The slope <code class="docutils literal notranslate"><span class="pre">m</span></code>
and the intercept <code class="docutils literal notranslate"><span class="pre">b</span></code> can be set to a non-default value if necessary,
otherwise m=1.8 and b=-0.6. To alter these parameters, e.g. m=1.9 and b=-0.7,
use a parameter string like this: <code class="xref std std-option docutils literal notranslate"><span class="pre">--shapeMethod=</span></code>”Dm1.9b-0.7”. You may also
provide only one of the two parameters like: <code class="xref std std-option docutils literal notranslate"><span class="pre">--shapeMethod=</span></code>”Dm1.9” or
<code class="xref std std-option docutils literal notranslate"><span class="pre">--shapeMethod=</span></code>”Db-0.7”.</p>
<p><code class="docutils literal notranslate"><span class="pre">Z</span></code>: Convert SHAPE reactivities to pseudo energies according to Zarringhalam</p>
<p>et al 2012. SHAPE reactivities will be converted to pairing probabilities by
using linear mapping. Aberration from the observed pairing probabilities will
be penalized during the folding recursion. The magnitude of the penalties can
affected by adjusting the factor beta (e.g. <code class="xref std std-option docutils literal notranslate"><span class="pre">--shapeMethod=</span></code>”Zb0.8”).</p>
<p><code class="docutils literal notranslate"><span class="pre">W</span></code>: Apply a given vector of perturbation energies to unpaired nucleotides</p>
<p>according to Washietl et al 2012. Perturbation vectors can be calculated by
using RNApvmin.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-shapeConversion">
<span id="cmdoption-shapeconversion"></span><span class="sig-name descname"><span class="pre">--shapeConversion</span></span><span class="sig-prename descclassname"><span class="pre">=method</span></span><a class="headerlink" href="#cmdoption-shapeConversion" title="Permalink to this definition"></a></dt>
<dd><p>Select method for SHAPE reactivity conversion.</p>
<p><em>(default=”O”)</em></p>
<p>This parameter is useful when dealing with the SHAPE incorporation according
to Zarringhalam et al. The following methods can be used to convert SHAPE
reactivities into the probability for a certain nucleotide to be unpaired.</p>
<p><code class="docutils literal notranslate"><span class="pre">M</span></code>: Use linear mapping according to Zarringhalam et al.
<code class="docutils literal notranslate"><span class="pre">C</span></code>: Use a cutoff-approach to divide into paired and unpaired nucleotides
(e.g. “C0.25”)
<code class="docutils literal notranslate"><span class="pre">S</span></code>: Skip the normalizing step since the input data already represents
probabilities for being unpaired rather than raw reactivity values
<code class="docutils literal notranslate"><span class="pre">L</span></code>: Use a linear model to convert the reactivity into a probability for
being unpaired (e.g. “Ls0.68i0.2” to use a slope of 0.68 and an intercept
of 0.2)
<code class="docutils literal notranslate"><span class="pre">O</span></code>: Use a linear model to convert the log of the reactivity into a
probability for being unpaired (e.g. “Os1.6i-2.29” to use a slope of 1.6
and an intercept of <code class="xref std std-option docutils literal notranslate"><span class="pre">-2</span></code>.29)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-commands">
<span class="sig-name descname"><span class="pre">--commands</span></span><span class="sig-prename descclassname"><span class="pre">=filename</span></span><a class="headerlink" href="#cmdoption-commands" title="Permalink to this definition"></a></dt>
<dd><p>Read additional commands from file</p>
<p>Commands include hard and soft constraints, but also structure motifs in
hairpin and interior loops that need to be treeted differently. Furthermore,
commands can be set for unstructured and structured domains.</p>
</dd></dl>

</section>
<section id="energy-parameters">
<h3>Energy Parameters:<a class="headerlink" href="#energy-parameters" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Energy parameter sets can be adapted or loaded from user-provided input files</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-T">
<span id="cmdoption-t"></span><span id="cmdoption-temp"></span><span class="sig-name descname"><span class="pre">-T</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--temp</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-T" title="Permalink to this definition"></a></dt>
<dd><p>Rescale energy parameters to a temperature of temp C. Default is 37C.</p>
<p><em>(default=”37.0”)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-P">
<span id="cmdoption-paramFile"></span><span id="cmdoption-paramfile"></span><span class="sig-name descname"><span class="pre">-P</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--paramFile</span></span><span class="sig-prename descclassname"><span class="pre">=paramfile</span></span><a class="headerlink" href="#cmdoption-P" title="Permalink to this definition"></a></dt>
<dd><p>Read energy parameters from paramfile, instead of using the default parameter set.</p>
<p>Different sets of energy parameters for RNA and DNA should accompany your
distribution.
See the RNAlib documentation for details on the file format. The placeholder
file name <code class="docutils literal notranslate"><span class="pre">DNA</span></code> can be used to load DNA parameters without the need to
actually specify any input file.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-4">
<span id="cmdoption-noTetra"></span><span id="cmdoption-notetra"></span><span class="sig-name descname"><span class="pre">-4</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--noTetra</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-4" title="Permalink to this definition"></a></dt>
<dd><p>Do not include special tabulated stabilizing energies for tri-, tetra- and hexaloop hairpins.</p>
<p><em>(default=off)</em></p>
<p>Mostly for testing.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-salt">
<span class="sig-name descname"><span class="pre">--salt</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-salt" title="Permalink to this definition"></a></dt>
<dd><p>Set salt concentration in molar (M). Default is 1.021M.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-saltInit">
<span id="cmdoption-saltinit"></span><span class="sig-name descname"><span class="pre">--saltInit</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-saltInit" title="Permalink to this definition"></a></dt>
<dd><p>Provide salt correction for duplex initialization (in kcal/mol).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-m">
<span id="cmdoption-modifications"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--modifications</span></span><span class="sig-prename descclassname"><span class="pre">[=STRING]</span></span><a class="headerlink" href="#cmdoption-m" title="Permalink to this definition"></a></dt>
<dd><p>Allow for modified bases within the RNA sequence string.</p>
<p><em>(default=”7I6P9D”)</em></p>
<p>Treat modified bases within the RNA sequence differently, i.e. use
corresponding energy corrections and/or pairing partner rules if available.
For that, the modified bases in the input sequence must be marked by their
corresponding one-letter code. If no additional arguments are supplied, all
available corrections are performed. Otherwise, the user may limit the
modifications to a particular subset of modifications, resp. one-letter
codes, e.g. <code class="xref std std-option docutils literal notranslate"><span class="pre">-mP6</span></code> will only correct for pseudouridine and m6A bases.</p>
<p>Currently supported one-letter codes and energy corrections are:</p>
<p><code class="docutils literal notranslate"><span class="pre">7</span></code>: 7-deaza-adenonsine (7DA)</p>
<p><code class="docutils literal notranslate"><span class="pre">I</span></code>: Inosine</p>
<p><code class="docutils literal notranslate"><span class="pre">6</span></code>: N6-methyladenosine (m6A)</p>
<p><code class="docutils literal notranslate"><span class="pre">P</span></code>: Pseudouridine</p>
<p><code class="docutils literal notranslate"><span class="pre">9</span></code>: Purine (a.k.a. nebularine)</p>
<p><code class="docutils literal notranslate"><span class="pre">D</span></code>: Dihydrouridine</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-mod-file">
<span class="sig-name descname"><span class="pre">--mod-file</span></span><span class="sig-prename descclassname"><span class="pre">=STRING</span></span><a class="headerlink" href="#cmdoption-mod-file" title="Permalink to this definition"></a></dt>
<dd><p>Use additional modified base data from JSON file.</p>
</dd></dl>

</section>
<section id="model-details">
<h3>Model Details:<a class="headerlink" href="#model-details" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Tweak the energy model and pairing rules additionally using the following
parameters</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-d">
<span id="cmdoption-dangles"></span><span class="sig-name descname"><span class="pre">-d</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--dangles</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-d" title="Permalink to this definition"></a></dt>
<dd><p>How to treat “dangling end” energies for bases adjacent to helices in free ends and multi-loops.</p>
<p><em>(default=”2”)</em></p>
<p>With <code class="xref std std-option docutils literal notranslate"><span class="pre">-d1</span></code> only unpaired bases can participate in at most one dangling end.
With <code class="xref std std-option docutils literal notranslate"><span class="pre">-d2</span></code> this check is ignored, dangling energies will be added for the bases
adjacent to a helix on both sides in any case; this is the default for mfe
and partition function folding (<a class="reference internal" href="RNAsubopt.html#cmdoption-p"><code class="xref std std-option docutils literal notranslate"><span class="pre">-p</span></code></a>).
The option <code class="xref std std-option docutils literal notranslate"><span class="pre">-d0</span></code> ignores dangling ends altogether (mostly for debugging).
With <code class="xref std std-option docutils literal notranslate"><span class="pre">-d3</span></code> mfe folding will allow coaxial stacking of adjacent helices in
multi-loops. At the moment the implementation will not allow coaxial stacking
of the two interior pairs in a loop of degree 3 and works only for mfe
folding.</p>
<p>Note that with <code class="xref std std-option docutils literal notranslate"><span class="pre">-d1</span></code> and <code class="xref std std-option docutils literal notranslate"><span class="pre">-d3</span></code> only the MFE computations will be using this
setting while partition function uses <code class="xref std std-option docutils literal notranslate"><span class="pre">-d2</span></code> setting, i.e. dangling ends will be
treated differently.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noLP">
<span id="cmdoption-nolp"></span><span class="sig-name descname"><span class="pre">--noLP</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noLP" title="Permalink to this definition"></a></dt>
<dd><p>Produce structures without lonely pairs (helices of length 1).</p>
<p><em>(default=off)</em></p>
<p>For partition function folding this only disallows pairs that can only occur
isolated. Other pairs may still occasionally occur as helices of length 1.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noGU">
<span id="cmdoption-nogu"></span><span class="sig-name descname"><span class="pre">--noGU</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noGU" title="Permalink to this definition"></a></dt>
<dd><p>Do not allow GU pairs.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noClosingGU">
<span id="cmdoption-noclosinggu"></span><span class="sig-name descname"><span class="pre">--noClosingGU</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noClosingGU" title="Permalink to this definition"></a></dt>
<dd><p>Do not allow GU pairs at the end of helices.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-nsp">
<span class="sig-name descname"><span class="pre">--nsp</span></span><span class="sig-prename descclassname"><span class="pre">=STRING</span></span><a class="headerlink" href="#cmdoption-nsp" title="Permalink to this definition"></a></dt>
<dd><p>Allow other pairs in addition to the usual AU,GC,and GU pairs.</p>
<p>Its argument is a comma separated list of additionally allowed pairs. If the
first character is a “-” then AB will imply that AB and BA are allowed
pairs, e.g. <code class="xref std std-option docutils literal notranslate"><span class="pre">--nsp=</span></code>”-GA”  will allow GA and AG pairs. Nonstandard pairs are
given 0 stacking energy.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-e">
<span id="cmdoption-energyModel"></span><span id="cmdoption-energymodel"></span><span class="sig-name descname"><span class="pre">-e</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--energyModel</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-e" title="Permalink to this definition"></a></dt>
<dd><p>Set energy model.</p>
<p>Rarely used option to fold sequences from the artificial ABCD… alphabet,
where A pairs B, C-D etc.  Use the energy parameters for GC (<a class="reference internal" href="RNAup.html#cmdoption-e"><code class="xref std std-option docutils literal notranslate"><span class="pre">-e</span></code></a> 1) or AU (<a class="reference internal" href="RNAup.html#cmdoption-e"><code class="xref std std-option docutils literal notranslate"><span class="pre">-e</span></code></a>
2) pairs.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-helical-rise">
<span class="sig-name descname"><span class="pre">--helical-rise</span></span><span class="sig-prename descclassname"><span class="pre">=FLOAT</span></span><a class="headerlink" href="#cmdoption-helical-rise" title="Permalink to this definition"></a></dt>
<dd><p>Set the helical rise of the helix in units of Angstrom.</p>
<p><em>(default=”2.8”)</em></p>
<p>Use with caution! This value will be re-set automatically to 3.4 in case DNA
parameters are loaded via <a class="reference internal" href="RNAup.html#cmdoption-P"><code class="xref std std-option docutils literal notranslate"><span class="pre">-P</span></code></a> DNA and no further value is provided.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-backbone-length">
<span class="sig-name descname"><span class="pre">--backbone-length</span></span><span class="sig-prename descclassname"><span class="pre">=FLOAT</span></span><a class="headerlink" href="#cmdoption-backbone-length" title="Permalink to this definition"></a></dt>
<dd><p>Set the average backbone length for looped regions in units of Angstrom.</p>
<p><em>(default=”6.0”)</em></p>
<p>Use with caution! This value will be re-set automatically to 6.76 in case DNA
parameters are loaded via <a class="reference internal" href="RNAup.html#cmdoption-P"><code class="xref std std-option docutils literal notranslate"><span class="pre">-P</span></code></a> DNA and no further value is provided.</p>
</dd></dl>

</section>
<section id="plotting">
<h3>Plotting:<a class="headerlink" href="#plotting" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Command line options for changing the default behavior of structure layout
and pairing probability plots</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noPS">
<span id="cmdoption-nops"></span><span class="sig-name descname"><span class="pre">--noPS</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noPS" title="Permalink to this definition"></a></dt>
<dd><p>Do not produce postscript drawing of the mfe structure.</p>
<p><em>(default=off)</em></p>
</dd></dl>

</section>
</section>
<section id="references">
<h2>REFERENCES<a class="headerlink" href="#references" title="Permalink to this headline"></a></h2>
<p><em>If you use this program in your work you might want to cite:</em></p>
<p>R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011),
“ViennaRNA Package 2.0”,
Algorithms for Molecular Biology: 6:26</p>
<p>I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster (1994),
“Fast Folding and Comparison of RNA Secondary Structures”,
Monatshefte f. Chemie: 125, pp 167-188</p>
<p>R. Lorenz, I.L. Hofacker, P.F. Stadler (2016),
“RNA folding with hard and soft constraints”,
Algorithms for Molecular Biology 11:1 pp 1-13</p>
<p>S.H.Bernhart, Ch. Flamm, P.F. Stadler, I.L. Hofacker, (2006),
“Partition Function and Base Pairing Probabilities of RNA Heterodimers”,
Algorithms Mol. Biol.</p>
<p><em>The energy parameters are taken from:</em></p>
<p>D.H. Mathews, M.D. Disney, D. Matthew, J.L. Childs, S.J. Schroeder, J. Susan, M. Zuker, D.H. Turner (2004),
“Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure”,
Proc. Natl. Acad. Sci. USA: 101, pp 7287-7292</p>
<p>D.H Turner, D.H. Mathews (2009),
“NNDB: The nearest neighbor parameter database for predicting stability of nucleic acid secondary structure”,
Nucleic Acids Research: 38, pp 280-282</p>
</section>
<section id="author">
<h2>AUTHOR<a class="headerlink" href="#author" title="Permalink to this headline"></a></h2>
<p>Ivo L Hofacker, Peter F Stadler, Stephan Bernhart, Ronny Lorenz</p>
</section>
<section id="reporting-bugs">
<h2>REPORTING BUGS<a class="headerlink" href="#reporting-bugs" title="Permalink to this headline"></a></h2>
<p>If in doubt our program is right, nature is at fault.
Comments should be sent to <a class="reference external" href="mailto:rna&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">rna<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="RNAalifold.html" class="btn btn-neutral float-left" title="RNAalifold" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="RNAdistance.html" class="btn btn-neutral float-right" title="RNAdistance" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>