<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>RNAplot &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/vrna_32.png"/>
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="RNApvmin" href="RNApvmin.html" />
    <link rel="prev" title="RNAplfold" href="RNAplfold.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../manpages.html">Manpages</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="RNA2Dfold.html">RNA2Dfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAaliduplex.html">RNAaliduplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAalifold.html">RNAalifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAcofold.html">RNAcofold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAdistance.html">RNAdistance</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAdos.html">RNAdos</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAduplex.html">RNAduplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAeval.html">RNAeval</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAfold.html">RNAfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAheat.html">RNAheat</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAinverse.html">RNAinverse</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNALalifold.html">RNALalifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNALfold.html">RNALfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAmultifold.html">RNAmultifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApaln.html">RNApaln</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAparconv.html">RNAparconv</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApdist.html">RNApdist</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAPKplex.html">RNAPKplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplex.html">RNAplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplfold.html">RNAplfold</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">RNAplot</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#synopsis">Synopsis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#description">DESCRIPTION</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#i-o-options">I/O Options:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#plotting">Plotting:</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#references">REFERENCES</a></li>
<li class="toctree-l3"><a class="reference internal" href="#author">AUTHOR</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reporting-bugs">REPORTING BUGS</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="RNApvmin.html">RNApvmin</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAsnoop.html">RNAsnoop</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAsubopt.html">RNAsubopt</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAup.html">RNAup</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1"><a class="reference internal" href="../io.html">I/O Formats</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../manpages.html">Manpages</a> &raquo;</li>
      <li>RNAplot</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnaplot">
<h1>RNAplot<a class="headerlink" href="#rnaplot" title="Permalink to this headline"></a></h1>
<p><strong class="program">RNAplot</strong> - manual page for RNAplot 2.6.4</p>
<section id="synopsis">
<h2>Synopsis<a class="headerlink" href="#synopsis" title="Permalink to this headline"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>RNAplot <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>&lt;input0&gt;<span class="o">]</span> <span class="o">[</span>&lt;input1&gt;<span class="o">]</span>...
</pre></div>
</div>
</section>
<section id="description">
<h2>DESCRIPTION<a class="headerlink" href="#description" title="Permalink to this headline"></a></h2>
<p>RNAplot 2.6.4</p>
<p>Draw RNA Secondary Structures</p>
<p>The program reads (aligned) RNA sequences and structures in the format as
produced by RNAfold or Stockholm 1.0 and produces drawings of the secondary
structure graph.
Coordinates for the structure graphs are produced using either E. Bruccoleri’s
naview routines, or a simple radial layout method.
For aligned sequences and consensus structures (<a class="reference internal" href="#cmdoption-a"><code class="xref std std-option docutils literal notranslate"><span class="pre">--msa</span></code></a> option) the graph may be
annotated by covariance information. Additionally, a color-annotated EPS
alignment figure can be produced, similar to that obtained by RNAalifold and
RNALalifold.
If the sequence was preceded by a FASTA header, or if the multiple sequence
alignment contains an ID field, these IDs will be taken as names for the output
file(s): “name_ss.ps” and “name_aln.ps”. Otherwise “rna.ps” and
“aln.ps” will be used. This behavior may be over-ruled by explicitly setting
a filename prefix using the <a class="reference internal" href="RNAsubopt.html#cmdoption-auto-id"><code class="xref std std-option docutils literal notranslate"><span class="pre">--auto-id</span></code></a> option.
Existing files of the same name will be overwritten.</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-h">
<span id="cmdoption-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-h" title="Permalink to this definition"></a></dt>
<dd><p>Print help and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-detailed-help">
<span class="sig-name descname"><span class="pre">--detailed-help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-detailed-help" title="Permalink to this definition"></a></dt>
<dd><p>Print help, including all details and hidden options, and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-full-help">
<span class="sig-name descname"><span class="pre">--full-help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-full-help" title="Permalink to this definition"></a></dt>
<dd><p>Print help, including hidden options, and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-V">
<span id="cmdoption-v"></span><span id="cmdoption-version"></span><span class="sig-name descname"><span class="pre">-V</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--version</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-V" title="Permalink to this definition"></a></dt>
<dd><p>Print version and exit</p>
</dd></dl>

<section id="i-o-options">
<h3>I/O Options:<a class="headerlink" href="#i-o-options" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Command line options for input and output (pre-)processing</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-i">
<span id="cmdoption-infile"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--infile</span></span><span class="sig-prename descclassname"><span class="pre">=&lt;filename&gt;</span></span><a class="headerlink" href="#cmdoption-i" title="Permalink to this definition"></a></dt>
<dd><p>Read a file instead of reading from stdin.</p>
<p>The default behavior of RNAplot is to read input from stdin or the file(s)
that follow(s) the RNAplot command. Using this parameter the user can specify
input file names where data is read from. Note, that any additional files
supplied to RNAplot are still processed as well.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-a">
<span id="cmdoption-msa"></span><span class="sig-name descname"><span class="pre">-a</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-a" title="Permalink to this definition"></a></dt>
<dd><p>Input is multiple sequence alignment in Stockholm 1.0 format. <em>(default=off)</em></p>
<p>Using this flag indicates that the input is a multiple sequence alignment
(MSA) instead of (a) single sequence(s). Note, that only STOCKHOLM format
allows one to specify a consensus structure. Therefore, this is the only
supported MSA format for now!</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-mis">
<span class="sig-name descname"><span class="pre">--mis</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-mis" title="Permalink to this definition"></a></dt>
<dd><p>Output “most informative sequence” instead of simple consensus <em>(default=off)</em></p>
<p>For each column of the alignment output this is the set of nucleotides with
frequency greater than average in IUPAC notation.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-j">
<span id="cmdoption-jobs"></span><span class="sig-name descname"><span class="pre">-j</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--jobs</span></span><span class="sig-prename descclassname"><span class="pre">[=number]</span></span><a class="headerlink" href="#cmdoption-j" title="Permalink to this definition"></a></dt>
<dd><p>Split batch input into jobs and start processing in parallel using multiple threads. <em>(default=”0”)</em></p>
<p>Default processing of input data is performed in a serial fashion, i.e. one
sequence at a time. Using this switch, a user can instead start the
computation for many sequences in the input in parallel. RNAplot will create
as many parallel computation slots as specified and assigns input sequences
of the input file(s) to the available slots. Note, that this increases memory
consumption since input alignments have to be kept in memory until an empty
compute slot is available and each running job requires its own dynamic
programming matrices. A value of 0 indicates to use as many parallel threads
as computation cores are available.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-o">
<span id="cmdoption-output-format"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--output-format</span></span><span class="sig-prename descclassname"><span class="pre">=ps|gml|xrna|svg</span></span><a class="headerlink" href="#cmdoption-o" title="Permalink to this definition"></a></dt>
<dd><p>Specify output format. <em>(default=”ps”)</em></p>
<p>Available formats are: PostScript (<code class="docutils literal notranslate"><span class="pre">ps</span></code>), Graph Meta Language (<code class="docutils literal notranslate"><span class="pre">gml</span></code>),
Scalable Vector Graphics (<code class="docutils literal notranslate"><span class="pre">svg</span></code>), and XRNA save file (<code class="docutils literal notranslate"><span class="pre">xrna</span></code>). Output
filenames will end in “.ps” “.gml” “.svg” “.ss”, respectively.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-pre">
<span class="sig-name descname"><span class="pre">--pre</span></span><span class="sig-prename descclassname"><span class="pre">=string</span></span><a class="headerlink" href="#cmdoption-pre" title="Permalink to this definition"></a></dt>
<dd><p>Add annotation macros to postscript file, and add the postscript code in “string” just before the code to draw the structure. This is an easy way to add annotation.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-post">
<span class="sig-name descname"><span class="pre">--post</span></span><span class="sig-prename descclassname"><span class="pre">=string</span></span><a class="headerlink" href="#cmdoption-post" title="Permalink to this definition"></a></dt>
<dd><p>Same as <a class="reference internal" href="#cmdoption-pre"><code class="xref std std-option docutils literal notranslate"><span class="pre">--pre</span></code></a> but in contrast to adding the annotation macros. E.g to mark position 15 with circle use <code class="xref std std-option docutils literal notranslate"><span class="pre">--post=</span></code>”15 cmark”.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-auto-id">
<span class="sig-name descname"><span class="pre">--auto-id</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-auto-id" title="Permalink to this definition"></a></dt>
<dd><p>Automatically generate an ID for each sequence. <em>(default=off)</em></p>
<p>The default mode of RNAfold is to automatically determine an ID from the
input sequence data if the input file format allows to do that. Sequence IDs
are usually given in the FASTA header of input sequences. If this flag is
active, RNAfold ignores any IDs retrieved from the input and automatically
generates an ID for each sequence. This ID consists of a prefix and an
increasing number. This flag can also be used to add a FASTA header to the
output even if the input has none.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-prefix">
<span class="sig-name descname"><span class="pre">--id-prefix</span></span><span class="sig-prename descclassname"><span class="pre">=STRING</span></span><a class="headerlink" href="#cmdoption-id-prefix" title="Permalink to this definition"></a></dt>
<dd><p>Prefix for automatically generated IDs (as used in output file names).</p>
<p><em>(default=”sequence”)</em></p>
<p>If this parameter is set, each sequence will be prefixed with the provided
string. Hence, the output files will obey the following naming scheme:
“prefix_xxxx_ss.ps” (secondary structure plot), “prefix_xxxx_dp.ps”
(dot-plot), “prefix_xxxx_dp2.ps” (stack probabilities), etc. where xxxx is
the sequence number. Note: Setting this parameter implies <a class="reference internal" href="RNAsubopt.html#cmdoption-auto-id"><code class="xref std std-option docutils literal notranslate"><span class="pre">--auto-id</span></code></a>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-delim">
<span class="sig-name descname"><span class="pre">--id-delim</span></span><span class="sig-prename descclassname"><span class="pre">=CHAR</span></span><a class="headerlink" href="#cmdoption-id-delim" title="Permalink to this definition"></a></dt>
<dd><p>Change the delimiter between prefix and increasing number for automatically generated IDs (as used in output file names).</p>
<p><em>(default=”_”)</em></p>
<p>This parameter can be used to change the default delimiter “_” between the
prefix string and the increasing number for automatically generated ID.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-digits">
<span class="sig-name descname"><span class="pre">--id-digits</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-id-digits" title="Permalink to this definition"></a></dt>
<dd><p>Specify the number of digits of the counter in automatically generated alignment IDs.</p>
<p><em>(default=”4”)</em></p>
<p>When alignments IDs are automatically generated, they receive an increasing
number, starting with 1. This number will always be left-padded by leading
zeros, such that the number takes up a certain width. Using this parameter,
the width can be specified to the users need. We allow numbers in the range
[1:18]. This option implies <a class="reference internal" href="RNAsubopt.html#cmdoption-auto-id"><code class="xref std std-option docutils literal notranslate"><span class="pre">--auto-id</span></code></a>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-id-start">
<span class="sig-name descname"><span class="pre">--id-start</span></span><span class="sig-prename descclassname"><span class="pre">=LONG</span></span><a class="headerlink" href="#cmdoption-id-start" title="Permalink to this definition"></a></dt>
<dd><p>Specify the first number in automatically generated IDs.</p>
<p><em>(default=”1”)</em></p>
<p>When sequence IDs are automatically generated, they receive an increasing
number, usually starting with 1. Using this parameter, the first number can
be specified to the users requirements. Note: negative numbers are not
allowed.
Note: Setting this parameter implies to ignore any IDs retrieved from the
input data, i.e. it activates the <a class="reference internal" href="RNAsubopt.html#cmdoption-auto-id"><code class="xref std std-option docutils literal notranslate"><span class="pre">--auto-id</span></code></a> flag.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-filename-delim">
<span class="sig-name descname"><span class="pre">--filename-delim</span></span><span class="sig-prename descclassname"><span class="pre">=CHAR</span></span><a class="headerlink" href="#cmdoption-filename-delim" title="Permalink to this definition"></a></dt>
<dd><p>Change the delimiting character used in sanitized filenames.</p>
<p><em>(default=”ID-delimiter”)</em></p>
<p>This parameter can be used to change the delimiting character used while
sanitizing filenames, i.e. replacing invalid characters. Note, that the
default delimiter ALWAYS is the first character of the “ID delimiter” as
supplied through the <a class="reference internal" href="RNAsubopt.html#cmdoption-id-delim"><code class="xref std std-option docutils literal notranslate"><span class="pre">--id-delim</span></code></a> option. If the delimiter is a whitespace
character or empty, invalid characters will be simply removed rather than
substituted. Currently, we regard the following characters as illegal for use
in filenames: backslash <code class="docutils literal notranslate"><span class="pre">\</span></code>, slash <code class="docutils literal notranslate"><span class="pre">/</span></code>, question mark <code class="docutils literal notranslate"><span class="pre">?</span></code>, percent sign <code class="docutils literal notranslate"><span class="pre">%</span></code>,
asterisk <code class="docutils literal notranslate"><span class="pre">*</span></code>, colon <code class="docutils literal notranslate"><span class="pre">:</span></code>, pipe symbol <code class="docutils literal notranslate"><span class="pre">|</span></code>, double quote <code class="docutils literal notranslate"><span class="pre">&quot;</span></code>, triangular
brackets <code class="docutils literal notranslate"><span class="pre">&lt;</span></code> and <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-filename-full">
<span class="sig-name descname"><span class="pre">--filename-full</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-filename-full" title="Permalink to this definition"></a></dt>
<dd><p>Use full FASTA header to create filenames. <em>(default=off)</em></p>
<p>This parameter can be used to deactivate the default behavior of limiting
output filenames to the first word of the sequence ID. Consider the following
example: An input with FASTA header <code class="docutils literal notranslate"><span class="pre">&gt;NM_0001</span> <span class="pre">Homo</span> <span class="pre">Sapiens</span> <span class="pre">some</span> <span class="pre">gene</span></code> usually
produces output files with the prefix “NM_0001” without the additional data
available in the FASTA header, e.g. “NM_0001_ss.ps” for secondary structure
plots. With this flag set, no truncation of the output filenames is done,
i.e. output filenames receive the full FASTA header data as prefixes. Note,
however, that invalid characters (such as whitespace) will be substituted by
a delimiting character or simply removed, (see also the parameter option
<a class="reference internal" href="RNAsubopt.html#cmdoption-filename-delim"><code class="xref std std-option docutils literal notranslate"><span class="pre">--filename-delim</span></code></a>).</p>
</dd></dl>

</section>
<section id="plotting">
<h3>Plotting:<a class="headerlink" href="#plotting" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Command line options for changing the default behavior of structure layout
and pairing probability plots</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-covar">
<span class="sig-name descname"><span class="pre">--covar</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-covar" title="Permalink to this definition"></a></dt>
<dd><p>Annotate covariance of base pairs in consensus structure.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-aln">
<span class="sig-name descname"><span class="pre">--aln</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-aln" title="Permalink to this definition"></a></dt>
<dd><p>Produce a colored and structure annotated alignment in PostScript format in the file “aln.ps” in the current directory.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-aln-EPS-cols">
<span id="cmdoption-aln-eps-cols"></span><span class="sig-name descname"><span class="pre">--aln-EPS-cols</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-aln-EPS-cols" title="Permalink to this definition"></a></dt>
<dd><p>Number of columns in colored EPS alignment output.</p>
<p><em>(default=”60”)</em></p>
<p>A value less than 1 indicates that the output should not be wrapped at all.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-t">
<span id="cmdoption-layout-type"></span><span class="sig-name descname"><span class="pre">-t</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--layout-type</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-t" title="Permalink to this definition"></a></dt>
<dd><p>Choose the plotting layout algorithm. <em>(default=”1”)</em></p>
<p>Select the layout algorithm that computes the nucleotide coordinates.
Currently, the following algorithms are available:</p>
<p><code class="docutils literal notranslate"><span class="pre">0</span></code>: simple radial layout</p>
<p><code class="docutils literal notranslate"><span class="pre">1</span></code>: Naview layout (Bruccoleri et al. 1988)</p>
<p><code class="docutils literal notranslate"><span class="pre">2</span></code>: circular layout</p>
<p><code class="docutils literal notranslate"><span class="pre">3</span></code>: RNAturtle (Wiegreffe et al. 2018)</p>
<p><code class="docutils literal notranslate"><span class="pre">4</span></code>: RNApuzzler (Wiegreffe et al. 2018)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noOptimization">
<span id="cmdoption-nooptimization"></span><span class="sig-name descname"><span class="pre">--noOptimization</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noOptimization" title="Permalink to this definition"></a></dt>
<dd><p>Disable the drawing space optimization of RNApuzzler.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-ignoreExteriorIntersections">
<span id="cmdoption-ignoreexteriorintersections"></span><span class="sig-name descname"><span class="pre">--ignoreExteriorIntersections</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-ignoreExteriorIntersections" title="Permalink to this definition"></a></dt>
<dd><p>Ignore intersections with the exterior loop</p>
<p>within the RNA-tree.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-ignoreAncestorIntersections">
<span id="cmdoption-ignoreancestorintersections"></span><span class="sig-name descname"><span class="pre">--ignoreAncestorIntersections</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-ignoreAncestorIntersections" title="Permalink to this definition"></a></dt>
<dd><p>Ignore ancestor intersections within the</p>
<p>RNA-tree.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-ignoreSiblingIntersections">
<span id="cmdoption-ignoresiblingintersections"></span><span class="sig-name descname"><span class="pre">--ignoreSiblingIntersections</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-ignoreSiblingIntersections" title="Permalink to this definition"></a></dt>
<dd><p>Ignore sibling intersections within the</p>
<p>RNA-tree.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-allowFlipping">
<span id="cmdoption-allowflipping"></span><span class="sig-name descname"><span class="pre">--allowFlipping</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-allowFlipping" title="Permalink to this definition"></a></dt>
<dd><p>Allow flipping of exterior loop branches to resolve exterior branch intersections.</p>
<p><em>(default=off)</em></p>
</dd></dl>

</section>
</section>
<section id="references">
<h2>REFERENCES<a class="headerlink" href="#references" title="Permalink to this headline"></a></h2>
<p><em>If you use this program in your work you might want to cite:</em></p>
<p>R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011),
“ViennaRNA Package 2.0”,
Algorithms for Molecular Biology: 6:26</p>
<p>I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster (1994),
“Fast Folding and Comparison of RNA Secondary Structures”,
Monatshefte f. Chemie: 125, pp 167-188</p>
<p>R. Lorenz, I.L. Hofacker, P.F. Stadler (2016),
“RNA folding with hard and soft constraints”,
Algorithms for Molecular Biology 11:1 pp 1-13</p>
<p><em>The energy parameters are taken from:</em></p>
<p>D.H. Mathews, M.D. Disney, D. Matthew, J.L. Childs, S.J. Schroeder, J. Susan, M. Zuker, D.H. Turner (2004),
“Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure”,
Proc. Natl. Acad. Sci. USA: 101, pp 7287-7292</p>
<p>D.H Turner, D.H. Mathews (2009),
“NNDB: The nearest neighbor parameter database for predicting stability of nucleic acid secondary structure”,
Nucleic Acids Research: 38, pp 280-282</p>
</section>
<section id="author">
<h2>AUTHOR<a class="headerlink" href="#author" title="Permalink to this headline"></a></h2>
<p>Ivo L Hofacker, Ronny Lorenz</p>
</section>
<section id="reporting-bugs">
<h2>REPORTING BUGS<a class="headerlink" href="#reporting-bugs" title="Permalink to this headline"></a></h2>
<p>If in doubt our program is right, nature is at fault.
Comments should be sent to <a class="reference external" href="mailto:rna&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">rna<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="RNAplfold.html" class="btn btn-neutral float-left" title="RNAplfold" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="RNApvmin.html" class="btn btn-neutral float-right" title="RNApvmin" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>