<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>RNAup &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/vrna_32.png"/>
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Using RNAlib" href="../using_rnalib.html" />
    <link rel="prev" title="RNAsubopt" href="RNAsubopt.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../manpages.html">Manpages</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="RNA2Dfold.html">RNA2Dfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAaliduplex.html">RNAaliduplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAalifold.html">RNAalifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAcofold.html">RNAcofold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAdistance.html">RNAdistance</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAdos.html">RNAdos</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAduplex.html">RNAduplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAeval.html">RNAeval</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAfold.html">RNAfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAheat.html">RNAheat</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAinverse.html">RNAinverse</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNALalifold.html">RNALalifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNALfold.html">RNALfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAmultifold.html">RNAmultifold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApaln.html">RNApaln</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAparconv.html">RNAparconv</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApdist.html">RNApdist</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAPKplex.html">RNAPKplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplex.html">RNAplex</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplfold.html">RNAplfold</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAplot.html">RNAplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNApvmin.html">RNApvmin</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAsnoop.html">RNAsnoop</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNAsubopt.html">RNAsubopt</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">RNAup</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#synopsis">Synopsis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#description">DESCRIPTION</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#i-o-options">I/O Options:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#algorithms">Algorithms:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#calculations-of-rna-rna-interactions">Calculations of RNA-RNA interactions:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#structure-constraints">Structure Constraints:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#energy-parameters">Energy Parameters:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#model-details">Model Details:</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#references">REFERENCES</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">EXAMPLES</a></li>
<li class="toctree-l3"><a class="reference internal" href="#author">AUTHOR</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reporting-bugs">REPORTING BUGS</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1"><a class="reference internal" href="../io.html">I/O Formats</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../manpages.html">Manpages</a> &raquo;</li>
      <li>RNAup</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnaup">
<h1>RNAup<a class="headerlink" href="#rnaup" title="Permalink to this headline"></a></h1>
<p><strong class="program">RNAup</strong> - manual page for RNAup 2.6.4</p>
<section id="synopsis">
<h2>Synopsis<a class="headerlink" href="#synopsis" title="Permalink to this headline"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>RNAup <span class="o">[</span>OPTION<span class="o">]</span>...
</pre></div>
</div>
</section>
<section id="description">
<h2>DESCRIPTION<a class="headerlink" href="#description" title="Permalink to this headline"></a></h2>
<p>RNAup 2.6.4</p>
<p>Calculate the thermodynamics of RNA-RNA interactions</p>
<p>RNAup calculates the thermodynamics of RNA-RNA interactions, by decomposing the
binding into two stages. (1) First the probability that a potential binding
sites remains unpaired (equivalent to the free energy needed to open the site)
is computed. (2) Then this accessibility is combined with the interaction
energy to obtain the total binding energy. All calculations are done by
computing partition functions over all possible conformations.</p>
<p>RNAup provides two different modes: By default RNAup computes accessibilities,
in terms of the free energies needed to open a region (default length 4). It
prints the region of highest accessibility and its opening energy to stdout,
opening energies for all other regions are written to a file.</p>
<p>.br
In interaction mode the interaction between two RNAs is calculated. It is
invoked if the input consists of two sequences concatenated with an <code class="docutils literal notranslate"><span class="pre">&amp;</span></code>,
or if the options -X[pf] or -b are given. Unless the -b option is specified
RNAup assumes that the longer RNA is a structured target sequence
while the shorter one is an unstructured small RNA.
.br
Additionally, for every position along the target sequence we write the best
free energy of binding for an interaction that includes this position to the
the output file.
Output to stdout consists of the location and free energy, dG,
for the optimal region of interaction. The binding energy dG is also split into
its components the interaction energy dGint and the opening energy dGu_l (and
possibly dGu_s for the shorter sequence).
.br
In addition we print the optimal interaction structure as computed by RNAduplex
for this region. Note that it can happen that the RNAduplex computed optimal
interaction does not coincide with the optimal RNAup region. If the two
predictions don’t match the structure string is replaced by a run of “.”
and a message is written to stderr.
.br</p>
<p>Each sequence should be in 5`` to 3`` direction. If the sequence is preceded
by a line of the form
.. code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">name</span>
</pre></div>
</div>
<p>the output file “name_ux_up.out” is produced, where the “x” in “ux” is the
value set by the -u option. Otherwise the file name defaults to
RNA_ux_up.out. The output is concatenated if a file with the same name exists.
.br</p>
<p>RNA sequences are read from stdin as strings of characters. White space and
newline within a sequence cause an error! Newline is used to separate
sequences. The program will continue to read new sequences until a line
consisting of the single character &#64; or an end of file condition is
encountered.</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-h">
<span id="cmdoption-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-h" title="Permalink to this definition"></a></dt>
<dd><p>Print help and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-detailed-help">
<span class="sig-name descname"><span class="pre">--detailed-help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-detailed-help" title="Permalink to this definition"></a></dt>
<dd><p>Print help, including all details and hidden options, and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-full-help">
<span class="sig-name descname"><span class="pre">--full-help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-full-help" title="Permalink to this definition"></a></dt>
<dd><p>Print help, including hidden options, and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-V">
<span id="cmdoption-v"></span><span id="cmdoption-version"></span><span class="sig-name descname"><span class="pre">-V</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--version</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-V" title="Permalink to this definition"></a></dt>
<dd><p>Print version and exit</p>
</dd></dl>

<section id="i-o-options">
<h3>I/O Options:<a class="headerlink" href="#i-o-options" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Command line options for input and output (pre-)processing</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-o">
<span id="cmdoption-no_output_file"></span><span id="cmdoption-no-output-file"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--no_output_file</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-o" title="Permalink to this definition"></a></dt>
<dd><p>Do not produce an output file.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-no_header">
<span id="cmdoption-no-header"></span><span class="sig-name descname"><span class="pre">--no_header</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-no_header" title="Permalink to this definition"></a></dt>
<dd><p>Do not produce a header with the command line parameters used in the outputfile.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noconv">
<span class="sig-name descname"><span class="pre">--noconv</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noconv" title="Permalink to this definition"></a></dt>
<dd><p>Do not automatically substitute nucleotide “T” with “U”.</p>
<p><em>(default=off)</em></p>
</dd></dl>

</section>
<section id="algorithms">
<h3>Algorithms:<a class="headerlink" href="#algorithms" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Select additional algorithms which should be included in the calculations.</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-u">
<span id="cmdoption-ulength"></span><span class="sig-name descname"><span class="pre">-u</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ulength</span></span><span class="sig-prename descclassname"><span class="pre">=length</span></span><a class="headerlink" href="#cmdoption-u" title="Permalink to this definition"></a></dt>
<dd><p>Specify the length of the unstructured region in the output.</p>
<p><em>(default=”4”)</em></p>
<p>The probability of being unpaired is plotted on the right border of the
unpaired region. You can specify up to 20 different length values: use “-”
to specify a range of continuous values (e.g. <a class="reference internal" href="#cmdoption-u"><code class="xref std std-option docutils literal notranslate"><span class="pre">-u</span></code></a> 4-8) or specify a list of
comma separated values (e.g. <a class="reference internal" href="#cmdoption-u"><code class="xref std std-option docutils literal notranslate"><span class="pre">-u</span></code></a> 4,8,15).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-c">
<span id="cmdoption-contributions"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--contributions</span></span><span class="sig-prename descclassname"><span class="pre">=SHIME</span></span><a class="headerlink" href="#cmdoption-c" title="Permalink to this definition"></a></dt>
<dd><p>Specify the contributions listed in the output. <em>(default=”S”)</em></p>
<p>By default only the full probability of being unpaired is plotted. The <a class="reference internal" href="#cmdoption-c"><code class="xref std std-option docutils literal notranslate"><span class="pre">-c</span></code></a>
option allows one to get the different contributions (c) to the probability
of being unpaired: The full probability of being unpaired (“S” is the sum
of the probability of being unpaired in the exterior loop (“E”), within a
hairpin loop (“H”), within an interior loop (“I”) and within a multiloop
(“M”). Any combination of these letters may be given.</p>
</dd></dl>

</section>
<section id="calculations-of-rna-rna-interactions">
<h3>Calculations of RNA-RNA interactions:<a class="headerlink" href="#calculations-of-rna-rna-interactions" title="Permalink to this headline"></a></h3>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-w">
<span id="cmdoption-window"></span><span class="sig-name descname"><span class="pre">-w</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--window</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-w" title="Permalink to this definition"></a></dt>
<dd><p>Set the maximal length of the region of interaction.</p>
<p><em>(default=”25”)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-b">
<span id="cmdoption-include_both"></span><span id="cmdoption-include-both"></span><span class="sig-name descname"><span class="pre">-b</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--include_both</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-b" title="Permalink to this definition"></a></dt>
<dd><p>Include the probability of unpaired regions in both (b) RNAs.</p>
<p><em>(default=off)</em></p>
<p>By default only the probability of being unpaired in the longer RNA (target)
is used.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-5">
<span id="cmdoption-extend5"></span><span class="sig-name descname"><span class="pre">-5</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--extend5</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-5" title="Permalink to this definition"></a></dt>
<dd><p>Extend the region of interaction in the target to some residues on the 5’ side.</p>
<p>The underlying assumption is that it is favorable for an interaction if not
only the direct region of contact is unpaired but also a few residues 5’</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-3">
<span id="cmdoption-extend3"></span><span class="sig-name descname"><span class="pre">-3</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--extend3</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-3" title="Permalink to this definition"></a></dt>
<dd><p>Extend the region of interaction in the target to some residues on the 3’ side.</p>
<p>The underlying assumption is that it is favorable for an interaction if not
only the direct region of contact is unpaired but also a few residues 3’</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-interaction_pairwise">
<span id="cmdoption-interaction-pairwise"></span><span class="sig-name descname"><span class="pre">--interaction_pairwise</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-interaction_pairwise" title="Permalink to this definition"></a></dt>
<dd><p>Activate pairwise interaction mode. <em>(default=off)</em></p>
<p>The first sequence interacts with the 2nd, the third with the 4th etc. If
activated, two interacting sequences may be given in a single line separated
by “&amp;” or each sequence may be given on an extra line.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-interaction_first">
<span id="cmdoption-interaction-first"></span><span class="sig-name descname"><span class="pre">--interaction_first</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-interaction_first" title="Permalink to this definition"></a></dt>
<dd><p>Activate interaction mode using first sequence only.</p>
<p><em>(default=off)</em></p>
<p>The interaction of each sequence with the first one is calculated (e.g.
interaction of one mRNA with many small RNAs). Each sequence has to be given
on an extra line</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-S">
<span id="cmdoption-s"></span><span id="cmdoption-pfScale"></span><span id="cmdoption-pfscale"></span><span class="sig-name descname"><span class="pre">-S</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--pfScale</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-S" title="Permalink to this definition"></a></dt>
<dd><p>In the calculation of the pf use scale*mfe as an estimate for the ensemble free energy (used to avoid overflows).</p>
<p><em>(default=”1.07”)</em></p>
<p>The default is 1.07, useful values are 1.0 to 1.2. Occasionally needed for
long sequences.</p>
</dd></dl>

</section>
<section id="structure-constraints">
<h3>Structure Constraints:<a class="headerlink" href="#structure-constraints" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Command line options to interact with the structure constraints feature of
this program</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-C">
<span id="cmdoption-constraint"></span><span class="sig-name descname"><span class="pre">-C</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--constraint</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-C" title="Permalink to this definition"></a></dt>
<dd><p>Apply structural constraint(s) during prediction.</p>
<p><em>(default=off)</em></p>
<p>The program first reads the sequence(s), then a dot-bracket like string
containing constraints on the structure. The following symbols are
recognized:</p>
<p><code class="docutils literal notranslate"><span class="pre">.</span></code> … no constraint for this base</p>
<p><code class="docutils literal notranslate"><span class="pre">x</span></code> … the base is unpaired</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;</span></code> … the base pairs downstream, i.e. i is paired with j &gt; i</p>
<p><code class="docutils literal notranslate"><span class="pre">&gt;</span></code> … the base pairs upstream, i.e. i is paired with j &lt; i</p>
<p><code class="docutils literal notranslate"><span class="pre">()</span></code> … base i pairs with base j</p>
<p><code class="docutils literal notranslate"><span class="pre">|</span></code> … the corresponding base has to be paired intermolecularily (only for</p>
<p>interaction mode)</p>
</dd></dl>

</section>
<section id="energy-parameters">
<h3>Energy Parameters:<a class="headerlink" href="#energy-parameters" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Energy parameter sets can be adapted or loaded from user-provided input files</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-T">
<span id="cmdoption-t"></span><span id="cmdoption-temp"></span><span class="sig-name descname"><span class="pre">-T</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--temp</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-T" title="Permalink to this definition"></a></dt>
<dd><p>Rescale energy parameters to a temperature of temp C. Default is 37C.</p>
<p><em>(default=”37.0”)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-P">
<span id="cmdoption-p"></span><span id="cmdoption-paramFile"></span><span id="cmdoption-paramfile"></span><span class="sig-name descname"><span class="pre">-P</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--paramFile</span></span><span class="sig-prename descclassname"><span class="pre">=paramfile</span></span><a class="headerlink" href="#cmdoption-P" title="Permalink to this definition"></a></dt>
<dd><p>Read energy parameters from paramfile, instead of using the default parameter set.</p>
<p>Different sets of energy parameters for RNA and DNA should accompany your
distribution.
See the RNAlib documentation for details on the file format. The placeholder
file name <code class="docutils literal notranslate"><span class="pre">DNA</span></code> can be used to load DNA parameters without the need to
actually specify any input file.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-4">
<span id="cmdoption-noTetra"></span><span id="cmdoption-notetra"></span><span class="sig-name descname"><span class="pre">-4</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--noTetra</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-4" title="Permalink to this definition"></a></dt>
<dd><p>Do not include special tabulated stabilizing energies for tri-, tetra- and hexaloop hairpins.</p>
<p><em>(default=off)</em></p>
<p>Mostly for testing.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-salt">
<span class="sig-name descname"><span class="pre">--salt</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-salt" title="Permalink to this definition"></a></dt>
<dd><p>Set salt concentration in molar (M). Default is 1.021M.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-saltInit">
<span id="cmdoption-saltinit"></span><span class="sig-name descname"><span class="pre">--saltInit</span></span><span class="sig-prename descclassname"><span class="pre">=DOUBLE</span></span><a class="headerlink" href="#cmdoption-saltInit" title="Permalink to this definition"></a></dt>
<dd><p>Provide salt correction for duplex initialization (in kcal/mol).</p>
</dd></dl>

</section>
<section id="model-details">
<h3>Model Details:<a class="headerlink" href="#model-details" title="Permalink to this headline"></a></h3>
<blockquote>
<div><p>Tweak the energy model and pairing rules additionally using the following
parameters</p>
</div></blockquote>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-d">
<span id="cmdoption-dangles"></span><span class="sig-name descname"><span class="pre">-d</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--dangles</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-d" title="Permalink to this definition"></a></dt>
<dd><p>Specify “dangling end” model for bases adjacent to helices in free ends and multi-loops.</p>
<p><em>(default=”2”)</em></p>
<p>With <code class="xref std std-option docutils literal notranslate"><span class="pre">-d2</span></code> dangling energies will be added for the bases adjacent to a helix on
both sides in any case.</p>
<p>The option <code class="xref std std-option docutils literal notranslate"><span class="pre">-d0</span></code> ignores dangling ends altogether (mostly for debugging).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noLP">
<span id="cmdoption-nolp"></span><span class="sig-name descname"><span class="pre">--noLP</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noLP" title="Permalink to this definition"></a></dt>
<dd><p>Produce structures without lonely pairs (helices of length 1).</p>
<p><em>(default=off)</em></p>
<p>For partition function folding this only disallows pairs that can only occur
isolated. Other pairs may still occasionally occur as helices of length 1.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noGU">
<span id="cmdoption-nogu"></span><span class="sig-name descname"><span class="pre">--noGU</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noGU" title="Permalink to this definition"></a></dt>
<dd><p>Do not allow GU pairs.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-noClosingGU">
<span id="cmdoption-noclosinggu"></span><span class="sig-name descname"><span class="pre">--noClosingGU</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-noClosingGU" title="Permalink to this definition"></a></dt>
<dd><p>Do not allow GU pairs at the end of helices.</p>
<p><em>(default=off)</em></p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-nsp">
<span class="sig-name descname"><span class="pre">--nsp</span></span><span class="sig-prename descclassname"><span class="pre">=STRING</span></span><a class="headerlink" href="#cmdoption-nsp" title="Permalink to this definition"></a></dt>
<dd><p>Allow other pairs in addition to the usual AU,GC,and GU pairs.</p>
<p>Its argument is a comma separated list of additionally allowed pairs. If the
first character is a “-” then AB will imply that AB and BA are allowed
pairs, e.g. <code class="xref std std-option docutils literal notranslate"><span class="pre">--nsp=</span></code>”-GA”  will allow GA and AG pairs. Nonstandard pairs are
given 0 stacking energy.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-e">
<span id="cmdoption-energyModel"></span><span id="cmdoption-energymodel"></span><span class="sig-name descname"><span class="pre">-e</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--energyModel</span></span><span class="sig-prename descclassname"><span class="pre">=INT</span></span><a class="headerlink" href="#cmdoption-e" title="Permalink to this definition"></a></dt>
<dd><p>Set energy model.</p>
<p>Rarely used option to fold sequences from the artificial ABCD… alphabet,
where A pairs B, C-D etc.  Use the energy parameters for GC (<a class="reference internal" href="#cmdoption-e"><code class="xref std std-option docutils literal notranslate"><span class="pre">-e</span></code></a> 1) or AU (<a class="reference internal" href="#cmdoption-e"><code class="xref std std-option docutils literal notranslate"><span class="pre">-e</span></code></a>
2) pairs.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-helical-rise">
<span class="sig-name descname"><span class="pre">--helical-rise</span></span><span class="sig-prename descclassname"><span class="pre">=FLOAT</span></span><a class="headerlink" href="#cmdoption-helical-rise" title="Permalink to this definition"></a></dt>
<dd><p>Set the helical rise of the helix in units of Angstrom.</p>
<p><em>(default=”2.8”)</em></p>
<p>Use with caution! This value will be re-set automatically to 3.4 in case DNA
parameters are loaded via <a class="reference internal" href="#cmdoption-P"><code class="xref std std-option docutils literal notranslate"><span class="pre">-P</span></code></a> DNA and no further value is provided.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-backbone-length">
<span class="sig-name descname"><span class="pre">--backbone-length</span></span><span class="sig-prename descclassname"><span class="pre">=FLOAT</span></span><a class="headerlink" href="#cmdoption-backbone-length" title="Permalink to this definition"></a></dt>
<dd><p>Set the average backbone length for looped regions in units of Angstrom.</p>
<p><em>(default=”6.0”)</em></p>
<p>Use with caution! This value will be re-set automatically to 6.76 in case DNA
parameters are loaded via <a class="reference internal" href="#cmdoption-P"><code class="xref std std-option docutils literal notranslate"><span class="pre">-P</span></code></a> DNA and no further value is provided.</p>
</dd></dl>

</section>
</section>
<section id="references">
<h2>REFERENCES<a class="headerlink" href="#references" title="Permalink to this headline"></a></h2>
<p><em>If you use this program in your work you might want to cite:</em></p>
<p>R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011),
“ViennaRNA Package 2.0”,
Algorithms for Molecular Biology: 6:26</p>
<p>I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster (1994),
“Fast Folding and Comparison of RNA Secondary Structures”,
Monatshefte f. Chemie: 125, pp 167-188</p>
<p>R. Lorenz, I.L. Hofacker, P.F. Stadler (2016),
“RNA folding with hard and soft constraints”,
Algorithms for Molecular Biology 11:1 pp 1-13</p>
<p>U. Mueckstein, H. Tafer, J. Hackermueller, S.H. Bernhart, P.F. Stadler, and I.L. Hofacker (2006),
“Thermodynamics of RNA-RNA Binding”,
Bioinformatics: 22(10), pp 1177-1182</p>
<p><em>The energy parameters are taken from:</em></p>
<p>D.H. Mathews, M.D. Disney, D. Matthew, J.L. Childs, S.J. Schroeder, J. Susan, M. Zuker, D.H. Turner (2004),
“Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure”,
Proc. Natl. Acad. Sci. USA: 101, pp 7287-7292</p>
<p>D.H Turner, D.H. Mathews (2009),
“NNDB: The nearest neighbor parameter database for predicting stability of nucleic acid secondary structure”,
Nucleic Acids Research: 38, pp 280-282</p>
</section>
<section id="examples">
<h2>EXAMPLES<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h2>
<p>.B Output to stdout:</p>
<p>In Interaction mode RNAup prints the most favorable interaction energy
between the two sequences to stdout. The most favorable interaction energy
(dG) depends on the position in the longer sequence (region [i,j]) and the
position in the shorter sequence (region[k,l]): dG[i,j;k,l].  dG[i,j;k,l] is the
largest contribution to dG[i,j] = sum_kl dG[i,j;k,l] which is given in the
output file: therefore dG[i,j;k,l] &lt;= dG[i,j].</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>``....,....1....,....2....,....3....,....4....,....5....,....6....,....7....,....8``
&gt; franz
GGAGUAGGUUAUCCUCUGUU
&gt; sissi
AGGACAACCU
dG = dGint + dGu_l
(((((.((((&amp;)))).)))))   6,15  :   1,10  (-6.66 = -9.89 + 3.23)
AGGUUAUCCU&amp;AGGACAACCU
RNAup output in file: franz_sissi_w25_u3_4_up.out
</pre></div>
</div>
<p>where the result line contains following information</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">RNAduplex</span> <span class="n">results</span>       <span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">]</span>     <span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="n">l</span><span class="p">]</span>    <span class="n">dG</span> <span class="o">=</span> <span class="n">dGint</span> <span class="o">+</span> <span class="n">dGu_l</span>
<span class="p">(((((</span><span class="o">.</span><span class="p">((((</span><span class="o">&amp;</span><span class="p">))))</span><span class="o">.</span><span class="p">)))))</span>   <span class="mi">6</span><span class="p">,</span><span class="mi">15</span>   <span class="p">:</span>  <span class="mi">1</span><span class="p">,</span><span class="mi">10</span>     <span class="p">(</span><span class="o">-</span><span class="mf">6.66</span><span class="o">=-</span><span class="mf">9.89</span><span class="o">+</span><span class="mf">3.23</span><span class="p">)</span>
</pre></div>
</div>
<p>.RD
.B Output to file:</p>
<p>Output to file contains a header including date, the command line of the
call to RNAup, length and names of the input sequence(s) followed
by the sequence(s). The first sequence is the target sequence.
Printing of the header can be turned off using the -nh option.</p>
<p>The line directly after the header gives the column names for the output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">position</span>     <span class="n">dGu_l</span> <span class="k">for</span> <span class="o">-</span><span class="n">u</span> <span class="mi">3</span>      <span class="n">dGu_l</span> <span class="k">for</span> <span class="o">-</span><span class="n">u</span> <span class="mi">4</span>       <span class="n">dG</span>
<span class="c1">#     pos      u3S       u3H       u4S       u4H        dG</span>
</pre></div>
</div>
<p>where all information refers to the target sequence. The dGu_l column contains
information about the -u value (u=3 or u=4) and the contribution to the free
energy to open all structures “S” or only hairpin loops “H”, see option -c.
NA means that no results is possible (e.g. column u3S row 2: no region of
length 3 ending at position 2 exists).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#  Thu Apr 10 09:15:11 2008</span>
<span class="c1">#  RNAup -u 3,4 -c SH -b</span>
<span class="c1">#  20 franz</span>
<span class="c1">#  GGAGUAGGUUAUCCUCUGUU</span>
<span class="c1">#  10 sissi</span>
<span class="c1">#  AGGACAACCU</span>
<span class="c1">#     pos      u3S       u3H       u4S       u4H        dG</span>
<span class="mi">1</span>        <span class="n">NA</span>        <span class="n">NA</span>        <span class="n">NA</span>        <span class="n">NA</span>    <span class="o">-</span><span class="mf">1.540</span>
<span class="mi">2</span>        <span class="n">NA</span>        <span class="n">NA</span>        <span class="n">NA</span>        <span class="n">NA</span>    <span class="o">-</span><span class="mf">1.540</span>
<span class="mi">3</span>     <span class="mf">1.371</span>        <span class="n">NA</span>        <span class="n">NA</span>        <span class="n">NA</span>    <span class="o">-</span><span class="mf">1.217</span>
<span class="mi">4</span>     <span class="mf">1.754</span>     <span class="mf">5.777</span>     <span class="mf">1.761</span>        <span class="n">NA</span>    <span class="o">-</span><span class="mf">1.393</span>
<span class="mi">5</span>     <span class="mf">1.664</span>     <span class="mf">3.140</span>     <span class="mf">1.811</span>     <span class="mf">5.800</span>    <span class="o">-</span><span class="mf">1.393</span>
</pre></div>
</div>
<p>If the -b option is selected position and dGu_s values for the shorter sequence
are written after the information for the target sequence.</p>
</section>
<section id="author">
<h2>AUTHOR<a class="headerlink" href="#author" title="Permalink to this headline"></a></h2>
<p>Ivo L Hofacker, Peter F Stadler, Ulrike Mueckstein, Ronny Lorenz</p>
</section>
<section id="reporting-bugs">
<h2>REPORTING BUGS<a class="headerlink" href="#reporting-bugs" title="Permalink to this headline"></a></h2>
<p>If in doubt our program is right, nature is at fault.
Comments should be sent to <a class="reference external" href="mailto:rna&#37;&#52;&#48;tbi&#46;univie&#46;ac&#46;at">rna<span>&#64;</span>tbi<span>&#46;</span>univie<span>&#46;</span>ac<span>&#46;</span>at</a>.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="RNAsubopt.html" class="btn btn-neutral float-left" title="RNAsubopt" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../using_rnalib.html" class="btn btn-neutral float-right" title="Using RNAlib" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>