<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Multiple Sequence Alignment Utilities &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/vrna_32.png"/>
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script defer="defer" src="../_static/js/mathjax/tex-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]]}, "options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Deprecated Interface for Multiple Sequence Alignment Utilities" href="msa_deprecated.html" />
    <link rel="prev" title="Deprecated Interface for Secondary Structure Utilities" href="struct/deprecated.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manpages.html">Manpages</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1"><a class="reference internal" href="../io.html">I/O Formats</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Concepts and Algorithms</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../eval.html">Free Energy Evaluation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../grammar.html">The RNA Folding Grammar</a></li>
<li class="toctree-l2"><a class="reference internal" href="../landscape.html">The RNA Secondary Structure Landscape</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mfe.html">Minimum Free Energy (MFE) Algorithms</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pf_fold.html">Partition Function and Equilibrium Properties</a></li>
<li class="toctree-l2"><a class="reference internal" href="../subopt_and_representatives.html">Suboptimals and Representative Structures</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cofold.html">RNA-RNA Interaction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../class_fold.html">Classified Dynamic Programming Variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="../inverse.html">Inverse Folding (Design)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../probing_data.html">Experimental Structure Probing Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ligand_binding.html">Ligands Binding to RNA Structures</a></li>
<li class="toctree-l2"><a class="reference internal" href="../paired_modules.html">Structure Modules and Pseudoknots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../modified_bases.html">Post-transcriptional Base Modifications</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../utils.html">Utilities</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="alphabet.html">Utilities to deal with Nucleotide Alphabets</a></li>
<li class="toctree-l3"><a class="reference internal" href="strings.html">(Nucleic Acid Sequence) String Utilitites</a></li>
<li class="toctree-l3"><a class="reference internal" href="structures.html">Secondary Structure Utilities</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Multiple Sequence Alignment Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="msa_deprecated.html">Deprecated Interface for Multiple Sequence Alignment Utilities</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="io.html">Files and I/O</a></li>
<li class="toctree-l3"><a class="reference internal" href="plots.html">Plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="search.html">Search Algorithms</a></li>
<li class="toctree-l3"><a class="reference internal" href="combinatorics.html">Combinatorics Algorithms</a></li>
<li class="toctree-l3"><a class="reference internal" href="data_structures.html">(Abstract) Data Structures</a></li>
<li class="toctree-l3"><a class="reference internal" href="messages.html">Messages</a></li>
<li class="toctree-l3"><a class="reference internal" href="units.html">Unit Conversion</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../vrna_3.0.html">RNAlib API v3.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="../callbacks.html">Callback Functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="../deprecated.html">Deprecated List</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../api.html">Concepts and Algorithms</a> &raquo;</li>
          <li><a href="../utils.html">Utilities</a> &raquo;</li>
      <li>Multiple Sequence Alignment Utilities</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="multiple-sequence-alignment-utilities">
<h1>Multiple Sequence Alignment Utilities<a class="headerlink" href="#multiple-sequence-alignment-utilities" title="Permalink to this headline"></a></h1>
<p>Functions to extract features from and to manipulate multiple
sequence alignments (MSA).</p>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Specialized Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="msa_deprecated.html">Deprecated Interface for Multiple Sequence Alignment Utilities</a></li>
</ul>
</div>
<dl>
<dt class="sig sig-object cpp">
<span class="target" id="group__aln__utils"></span></dt>
<dd><div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric" id="breathe-section-title-defines">Defines</p>
<dl class="c macro">
<dt class="sig sig-object c" id="c.VRNA_ALN_DEFAULT">
<span class="target" id="group__aln__utils_1ga7437bcbc3142b266f1f3b086eb669092"></span><span class="sig-name descname"><span class="n"><span class="pre">VRNA_ALN_DEFAULT</span></span></span><a class="headerlink" href="#c.VRNA_ALN_DEFAULT" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Use default alignment settings. </p>
</dd></dl>

<dl class="c macro">
<dt class="sig sig-object c" id="c.VRNA_ALN_RNA">
<span class="target" id="group__aln__utils_1ga1a2aa60bcc51dc8e48c07bf146bd28c1"></span><span class="sig-name descname"><span class="n"><span class="pre">VRNA_ALN_RNA</span></span></span><a class="headerlink" href="#c.VRNA_ALN_RNA" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Convert to RNA alphabet. </p>
</dd></dl>

<dl class="c macro">
<dt class="sig sig-object c" id="c.VRNA_ALN_DNA">
<span class="target" id="group__aln__utils_1ga36affc310a5adedf1bdbf9669dd4bf85"></span><span class="sig-name descname"><span class="n"><span class="pre">VRNA_ALN_DNA</span></span></span><a class="headerlink" href="#c.VRNA_ALN_DNA" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Convert to DNA alphabet. </p>
</dd></dl>

<dl class="c macro">
<dt class="sig sig-object c" id="c.VRNA_ALN_UPPERCASE">
<span class="target" id="group__aln__utils_1ga0de72fc917d72acafe862750a3a5e0bc"></span><span class="sig-name descname"><span class="n"><span class="pre">VRNA_ALN_UPPERCASE</span></span></span><a class="headerlink" href="#c.VRNA_ALN_UPPERCASE" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Convert to uppercase nucleotide letters. </p>
</dd></dl>

<dl class="c macro">
<dt class="sig sig-object c" id="c.VRNA_ALN_LOWERCASE">
<span class="target" id="group__aln__utils_1ga27eb303be3c8368c9e8be76a4d6ca35e"></span><span class="sig-name descname"><span class="n"><span class="pre">VRNA_ALN_LOWERCASE</span></span></span><a class="headerlink" href="#c.VRNA_ALN_LOWERCASE" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Convert to lowercase nucleotide letters. </p>
</dd></dl>

<dl class="c macro">
<dt class="sig sig-object c" id="c.VRNA_MEASURE_SHANNON_ENTROPY">
<span class="target" id="group__aln__utils_1ga1e659227c9fc077d29989f576f129000"></span><span class="sig-name descname"><span class="n"><span class="pre">VRNA_MEASURE_SHANNON_ENTROPY</span></span></span><a class="headerlink" href="#c.VRNA_MEASURE_SHANNON_ENTROPY" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Flag indicating Shannon Entropy measure. </p>
<p>Shannon Entropy is defined as <span class="math notranslate nohighlight">\( H = - \sum_c p_c \cdot \log_2 p_c \)</span></p>
</dd></dl>

</div>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric" id="breathe-section-title-typedefs">Typedefs</p>
<dl class="c type">
<dt class="sig sig-object c" id="c.vrna_pinfo_t">
<span class="target" id="group__aln__utils_1ga6660dfca23debee7306e0cd53341263f"></span><span class="k"><span class="pre">typedef</span></span><span class="w"> </span><span class="k"><span class="pre">struct</span></span><span class="w"> </span><a class="reference internal" href="#c.vrna_pinfo_s" title="vrna_pinfo_s"><span class="n"><span class="pre">vrna_pinfo_s</span></span></a><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">vrna_pinfo_t</span></span></span><a class="headerlink" href="#c.vrna_pinfo_t" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Typename for the base pair info repesenting data structure <a class="reference internal" href="#structvrna__pinfo__s"><span class="std std-ref">vrna_pinfo_s</span></a>. </p>
</dd></dl>

</div>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric" id="breathe-section-title-functions">Functions</p>
<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_mpi">
<span class="target" id="group__aln__utils_1gade5a1f2d16e7fd9a57b37d8514f08e8e"></span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_mpi</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_mpi" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Get the mean pairwise identity in steps from ?to?(ident) </p>
<p><dl>
<dt class="sig sig-object cpp">
<a class="reference internal" href="../wrappers.html#wrappers_1_wrappers000005"><span class="std std-ref"><span class="pre">SWIG</span> <span class="pre">Wrapper</span> <span class="pre">Notes:</span></span></a></dt>
<dd><p>This function is available as function <code class="docutils literal notranslate"><span class="pre">aln_mpi()</span></code>. See e.g.  <span><a class="reference internal" href="../api_python.html#RNA.aln_mpi" title="RNA.aln_mpi"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_mpi()</span></code></a> in the <a class="reference internal" href="../api_python.html"><span class="doc">Python API</span></a></span>. </p>
</dd></dl>

</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – Aligned sequences </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The mean pairwise identity</p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_pinfo">
<span class="target" id="group__aln__utils_1ga9545e9d5f82d141da39e3ba772516380"></span><a class="reference internal" href="#c.vrna_pinfo_t" title="vrna_pinfo_t"><span class="n"><span class="pre">vrna_pinfo_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_pinfo</span></span></span><span class="sig-paren">(</span><a class="reference internal" href="datastructures/fold_compound.html#c.vrna_fold_compound_t" title="vrna_fold_compound_t"><span class="n"><span class="pre">vrna_fold_compound_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">fc</span></span>, <span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">structure</span></span>, <span class="kt"><span class="pre">double</span></span><span class="w"> </span><span class="n"><span class="pre">threshold</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_pinfo" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Retrieve an array of <a class="reference internal" href="#group__aln__utils_1ga6660dfca23debee7306e0cd53341263f"><span class="std std-ref">vrna_pinfo_t</span></a> structures from precomputed pair probabilities. </p>
<p>This array of structures contains information about positionwise pair probabilies, base pair entropy and more</p>
<p><div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#group__aln__utils_1ga6660dfca23debee7306e0cd53341263f"><span class="std std-ref">vrna_pinfo_t</span></a>, and <a class="reference internal" href="../partfunc/global.html#group__part__func__global_1gaff6fe7a499a62d62d10cb850d5222f00"><span class="std std-ref">vrna_pf()</span></a></p>
</div>
</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fc</strong> – The <a class="reference internal" href="datastructures/fold_compound.html#group__fold__compound_1ga1b0cef17fd40466cef5968eaeeff6166"><span class="std std-ref">vrna_fold_compound_t</span></a> of type <a class="reference internal" href="datastructures/fold_compound.html#group__fold__compound_1gga01a4ff86fa71deaaa5d1abbd95a1447dab821ce46ea3cf665be97df22a76f5023"><span class="std std-ref">VRNA_FC_TYPE_COMPARATIVE</span></a> with precomputed partition function matrices </p></li>
<li><p><strong>structure</strong> – An optional structure in dot-bracket notation (Maybe NULL) </p></li>
<li><p><strong>threshold</strong> – Do not include results with pair probabilities below threshold </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <a class="reference internal" href="#group__aln__utils_1ga6660dfca23debee7306e0cd53341263f"><span class="std std-ref">vrna_pinfo_t</span></a> array </p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_pscore">
<span class="target" id="group__aln__utils_1ga697bf6f44356918cceeff4ed8ee885e1"></span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_pscore</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span>, <a class="reference internal" href="../grammar/model_details.html#c.vrna_md_t" title="vrna_md_t"><span class="n"><span class="pre">vrna_md_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">md</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_pscore" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p><dl>
<dt class="sig sig-object cpp">
<a class="reference internal" href="../wrappers.html#wrappers_1_wrappers000006"><span class="std std-ref"><span class="pre">SWIG</span> <span class="pre">Wrapper</span> <span class="pre">Notes:</span></span></a></dt>
<dd><p>This function is available as overloaded function <code class="docutils literal notranslate"><span class="pre">aln_pscore()</span></code> where the last parameter may be omitted, indicating <code class="docutils literal notranslate"><span class="pre">md</span></code> = <code class="docutils literal notranslate"><span class="pre">NULL</span></code>. See e.g.  <span><a class="reference internal" href="../api_python.html#RNA.aln_pscore" title="RNA.aln_pscore"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_pscore()</span></code></a> in the <a class="reference internal" href="../api_python.html"><span class="doc">Python API</span></a></span>. </p>
</dd></dl>

</p>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_pscore">
<span class="target" id="group__aln__utils_1gadd5e56bd49b93fa7859a9f802a11540f"></span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">vrna_pscore</span></span></span><span class="sig-paren">(</span><a class="reference internal" href="datastructures/fold_compound.html#c.vrna_fold_compound_t" title="vrna_fold_compound_t"><span class="n"><span class="pre">vrna_fold_compound_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">fc</span></span>, <span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="n"><span class="pre">i</span></span>, <span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="n"><span class="pre">j</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_pscore" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em></dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_pscore_freq">
<span class="target" id="group__aln__utils_1ga9071db7f59818374bc097b6be5fd3376"></span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">vrna_pscore_freq</span></span></span><span class="sig-paren">(</span><a class="reference internal" href="datastructures/fold_compound.html#c.vrna_fold_compound_t" title="vrna_fold_compound_t"><span class="n"><span class="pre">vrna_fold_compound_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">fc</span></span>, <span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">frequencies</span></span>, <span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="n"><span class="pre">pairs</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_pscore_freq" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em></dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_slice">
<span class="target" id="group__aln__utils_1gac801b0a72eac4ab9d5b2ef7533db9193"></span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_slice</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span>, <span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="n"><span class="pre">i</span></span>, <span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="n"><span class="pre">j</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_slice" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Slice out a subalignment from a larger alignment. </p>
<p><div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#group__aln__utils_1gac00dc80b8a324f151f92f11ab6850ecf"><span class="std std-ref">vrna_aln_free()</span></a></p>
</div>
</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The user is responsible to free the memory occupied by the returned subalignment</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input alignment </p></li>
<li><p><strong>i</strong> – The first column of the subalignment (1-based) </p></li>
<li><p><strong>j</strong> – The last column of the subalignment (1-based) </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The subalignment between column <span class="math notranslate nohighlight">\(i\)</span> and <span class="math notranslate nohighlight">\(j\)</span></p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_free">
<span class="target" id="group__aln__utils_1gac00dc80b8a324f151f92f11ab6850ecf"></span><span class="kt"><span class="pre">void</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_free</span></span></span><span class="sig-paren">(</span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_free" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Free memory occupied by a set of aligned sequences. </p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input alignment </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_uppercase">
<span class="target" id="group__aln__utils_1ga3a0476a71c4bae3d302790de58055704"></span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_uppercase</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_uppercase" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Create a copy of an alignment with only uppercase letters in the sequences. </p>
<p><div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#group__aln__utils_1ga08cdca592461436860daf9738279ce17"><span class="std std-ref">vrna_aln_copy</span></a></p>
</div>
</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input sequence alignment (last entry must be <em>NULL</em> terminated) </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A copy of the input alignment where lowercase sequence letters are replaced by uppercase letters </p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_toRNA">
<span class="target" id="group__aln__utils_1ga7886fc678553c06472a712a335cde9cf"></span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_toRNA</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_toRNA" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Create a copy of an alignment where DNA alphabet is replaced by RNA alphabet. </p>
<p><div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#group__aln__utils_1ga08cdca592461436860daf9738279ce17"><span class="std std-ref">vrna_aln_copy</span></a></p>
</div>
</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input sequence alignment (last entry must be <em>NULL</em> terminated) </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A copy of the input alignment where DNA alphabet is replaced by RNA alphabet (T -&gt; U) </p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_copy">
<span class="target" id="group__aln__utils_1ga08cdca592461436860daf9738279ce17"></span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_copy</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span>, <span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="n"><span class="pre">options</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_copy" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Make a copy of a multiple sequence alignment. </p>
<p>This function allows one to create a copy of a multiple sequence alignment. The <code class="docutils literal notranslate"><span class="pre">options</span></code> parameter additionally allows for sequence manipulation, such as converting DNA to RNA alphabet, and conversion to uppercase letters.</p>
<p><div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#group__aln__utils_1ga08cdca592461436860daf9738279ce17"><span class="std std-ref">vrna_aln_copy()</span></a>, <a class="reference internal" href="#group__aln__utils_1ga1a2aa60bcc51dc8e48c07bf146bd28c1"><span class="std std-ref">VRNA_ALN_RNA</span></a>, <a class="reference internal" href="#group__aln__utils_1ga0de72fc917d72acafe862750a3a5e0bc"><span class="std std-ref">VRNA_ALN_UPPERCASE</span></a>, <a class="reference internal" href="#group__aln__utils_1ga7437bcbc3142b266f1f3b086eb669092"><span class="std std-ref">VRNA_ALN_DEFAULT</span></a></p>
</div>
</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input sequence alignment (last entry must be <em>NULL</em> terminated) </p></li>
<li><p><strong>options</strong> – Option flags indicating whether the aligned sequences should be converted </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A (manipulated) copy of the input alignment </p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_conservation_struct">
<span class="target" id="group__aln__utils_1gab6f16a2ea93f3bfd4d089cc8d448bb16"></span><span class="kt"><span class="pre">float</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_conservation_struct</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span>, <span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">structure</span></span>, <span class="k"><span class="pre">const</span></span><span class="w"> </span><a class="reference internal" href="../grammar/model_details.html#c.vrna_md_t" title="vrna_md_t"><span class="n"><span class="pre">vrna_md_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">md</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_conservation_struct" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Compute base pair conservation of a consensus structure. </p>
<p>This function computes the base pair conservation (fraction of canonical base pairs) of a consensus structure given a multiple sequence alignment. The base pair types that are considered canonical may be specified using the <a class="reference internal" href="../grammar/model_details.html#structvrna__md__s_1ab4da594c638707e212f64aadb54a7454"><span class="std std-ref">vrna_md_t.pair</span></a> array. Passing <em>NULL</em> as parameter <code class="docutils literal notranslate"><span class="pre">md</span></code> results in default pairing rules, i.e. canonical Watson-Crick and GU Wobble pairs.</p>
<p><dl>
<dt class="sig sig-object cpp">
<a class="reference internal" href="../wrappers.html#wrappers_1_wrappers000001"><span class="std std-ref"><span class="pre">SWIG</span> <span class="pre">Wrapper</span> <span class="pre">Notes:</span></span></a></dt>
<dd><p>This function is available as overloaded function <code class="docutils literal notranslate"><span class="pre">aln_conservation_struct()</span></code> where the last parameter <code class="docutils literal notranslate"><span class="pre">md</span></code> may be omitted, indicating <code class="docutils literal notranslate"><span class="pre">md</span></code> = <code class="docutils literal notranslate"><span class="pre">NULL</span></code>. See, e.g.  <span><a class="reference internal" href="../api_python.html#RNA.aln_conservation_struct" title="RNA.aln_conservation_struct"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_conservation_struct()</span></code></a> in the <a class="reference internal" href="../api_python.html"><span class="doc">Python API</span></a></span>. </p>
</dd></dl>

</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input sequence alignment (last entry must be <em>NULL</em> terminated) </p></li>
<li><p><strong>structure</strong> – The consensus structure in dot-bracket notation </p></li>
<li><p><strong>md</strong> – Model details that specify compatible base pairs (Maybe <em>NULL</em>) </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 1-based vector of base pair conservations</p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_conservation_col">
<span class="target" id="group__aln__utils_1gaa12b481a7e7b965ef2eb1bcc4399e759"></span><span class="kt"><span class="pre">float</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_conservation_col</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span>, <span class="k"><span class="pre">const</span></span><span class="w"> </span><a class="reference internal" href="../grammar/model_details.html#c.vrna_md_t" title="vrna_md_t"><span class="n"><span class="pre">vrna_md_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">md_p</span></span>, <span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="n"><span class="pre">options</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_conservation_col" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Compute nucleotide conservation in an alignment. </p>
<p>This function computes the conservation of nucleotides in alignment columns. The simples measure is Shannon Entropy and can be selected by passing the <a class="reference internal" href="#group__aln__utils_1ga1e659227c9fc077d29989f576f129000"><span class="std std-ref">VRNA_MEASURE_SHANNON_ENTROPY</span></a> flag in the <code class="docutils literal notranslate"><span class="pre">options</span></code> parameter.</p>
<p><dl>
<dt class="sig sig-object cpp">
<a class="reference internal" href="../wrappers.html#wrappers_1_wrappers000002"><span class="std std-ref"><span class="pre">SWIG</span> <span class="pre">Wrapper</span> <span class="pre">Notes:</span></span></a></dt>
<dd><p>This function is available as overloaded function <code class="docutils literal notranslate"><span class="pre">aln_conservation_col()</span></code> where the last two parameters may be omitted, indicating <code class="docutils literal notranslate"><span class="pre">md</span></code> = <code class="docutils literal notranslate"><span class="pre">NULL</span></code>, and <code class="docutils literal notranslate"><span class="pre">options</span></code> = <a class="reference internal" href="#group__aln__utils_1ga1e659227c9fc077d29989f576f129000"><span class="std std-ref">VRNA_MEASURE_SHANNON_ENTROPY</span></a>, respectively. See e.g. <span><a class="reference internal" href="../api_python.html#RNA.aln_conservation_col" title="RNA.aln_conservation_col"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_conservation_col()</span></code></a> in the <a class="reference internal" href="../api_python.html"><span class="doc">Python API</span></a></span>. </p>
</dd></dl>

<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#group__aln__utils_1ga1e659227c9fc077d29989f576f129000"><span class="std std-ref">VRNA_MEASURE_SHANNON_ENTROPY</span></a></p>
</div>
</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently, only <a class="reference internal" href="#group__aln__utils_1ga1e659227c9fc077d29989f576f129000"><span class="std std-ref">VRNA_MEASURE_SHANNON_ENTROPY</span></a> is supported as conservation measure.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input sequence alignment (last entry must be <em>NULL</em> terminated) </p></li>
<li><p><strong>md</strong> – Model details that specify known nucleotides (Maybe <em>NULL</em>) </p></li>
<li><p><strong>options</strong> – A flag indicating which measure of conservation should be applied </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 1-based vector of column conservations</p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_consensus_sequence">
<span class="target" id="group__aln__utils_1gac6a33fa856d66dcb6f5d40ce0b61d7c7"></span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_consensus_sequence</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span>, <span class="k"><span class="pre">const</span></span><span class="w"> </span><a class="reference internal" href="../grammar/model_details.html#c.vrna_md_t" title="vrna_md_t"><span class="n"><span class="pre">vrna_md_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">md_p</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_consensus_sequence" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Compute the consensus sequence for a given multiple sequence alignment. </p>
<p><dl>
<dt class="sig sig-object cpp">
<a class="reference internal" href="../wrappers.html#wrappers_1_wrappers000003"><span class="std std-ref"><span class="pre">SWIG</span> <span class="pre">Wrapper</span> <span class="pre">Notes:</span></span></a></dt>
<dd><p>This function is available as overloaded function <code class="docutils literal notranslate"><span class="pre">aln_consensus_sequence()</span></code> where the last parameter may be omitted, indicating <code class="docutils literal notranslate"><span class="pre">md</span></code> = <code class="docutils literal notranslate"><span class="pre">NULL</span></code>. See e.g.  <span><a class="reference internal" href="../api_python.html#RNA.aln_consensus_sequence" title="RNA.aln_consensus_sequence"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_consensus_sequence()</span></code></a> in the <a class="reference internal" href="../api_python.html"><span class="doc">Python API</span></a></span>. </p>
</dd></dl>

</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input sequence alignment (last entry must be <em>NULL</em> terminated) </p></li>
<li><p><strong>md_p</strong> – Model details that specify known nucleotides (Maybe <em>NULL</em>) </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The consensus sequence of the alignment, i.e. the most frequent nucleotide for each alignment column</p>
</dd>
</dl>
</dd></dl>

<dl class="c function">
<dt class="sig sig-object c" id="c.vrna_aln_consensus_mis">
<span class="target" id="group__aln__utils_1gacd02c66d9e6f8dd5318f4fca0dc76cba"></span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="sig-name descname"><span class="n"><span class="pre">vrna_aln_consensus_mis</span></span></span><span class="sig-paren">(</span><span class="k"><span class="pre">const</span></span><span class="w"> </span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">alignment</span></span>, <span class="k"><span class="pre">const</span></span><span class="w"> </span><a class="reference internal" href="../grammar/model_details.html#c.vrna_md_t" title="vrna_md_t"><span class="n"><span class="pre">vrna_md_t</span></span></a><span class="w"> </span><span class="p"><span class="pre">*</span></span><span class="n"><span class="pre">md_p</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.vrna_aln_consensus_mis" title="Permalink to this definition"></a><br /></dt>
<dd><em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em><p>Compute the Most Informative Sequence (MIS) for a given multiple sequence alignment. </p>
<p>The most informative sequence (MIS) <span> <span id="id1">[<a class="reference internal" href="../bibliography.html#id39" title="Eva Freyhult, Vincent Moulton, and Paul Gardner. Predicting RNA structure using mutual information. Applied bioinformatics, 4(1):53–59, 2005. doi:10.2165/00822942-200504010-00006.">Freyhult <em>et al.</em>, 2005</a>]</span> </span> displays for each alignment column the nucleotides with frequency greater than the background frequency, projected into IUPAC notation. Columns where gaps are over-represented are in lower case.</p>
<p><dl>
<dt class="sig sig-object cpp">
<a class="reference internal" href="../wrappers.html#wrappers_1_wrappers000004"><span class="std std-ref"><span class="pre">SWIG</span> <span class="pre">Wrapper</span> <span class="pre">Notes:</span></span></a></dt>
<dd><p>This function is available as overloaded function <code class="docutils literal notranslate"><span class="pre">aln_consensus_mis()</span></code> where the last parameter may be omitted, indicating <code class="docutils literal notranslate"><span class="pre">md</span></code> = <code class="docutils literal notranslate"><span class="pre">NULL</span></code>. See e.g.  <span><a class="reference internal" href="../api_python.html#RNA.aln_consensus_mis" title="RNA.aln_consensus_mis"><code class="xref py py-func docutils literal notranslate"><span class="pre">RNA.aln_consensus_mis()</span></code></a> in the <a class="reference internal" href="../api_python.html"><span class="doc">Python API</span></a></span>. </p>
</dd></dl>

</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> – The input sequence alignment (last entry must be <em>NULL</em> terminated) </p></li>
<li><p><strong>md_p</strong> – Model details that specify known nucleotides (Maybe <em>NULL</em>) </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The most informative sequence for the alignment</p>
</dd>
</dl>
</dd></dl>

</div>
<dl class="c struct">
<dt class="sig sig-object c" id="c.vrna_pinfo_s">
<span class="target" id="structvrna__pinfo__s"></span><span class="k"><span class="pre">struct</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">vrna_pinfo_s</span></span></span><a class="headerlink" href="#c.vrna_pinfo_s" title="Permalink to this definition"></a><br /></dt>
<dd><div class="docutils container">
<em>#include &lt;ViennaRNA/utils/alignments.h&gt;</em></div>
<p>A base pair info structure. </p>
<p>For each base pair (i,j) with i,j in [0, n-1] the structure lists:<ul class="simple">
<li><p>its probability ‘p’</p></li>
<li><p>an entropy-like measure for its well-definedness ‘ent’</p></li>
<li><p>the frequency of each type of pair in ‘bp[]’<ul>
<li><p>’bp[0]’ contains the number of non-compatible sequences</p></li>
<li><p>’bp[1]’ the number of CG pairs, etc. </p></li>
</ul>
</p></li>
</ul>
</p>
<div class="breathe-sectiondef docutils container">
<p class="breathe-sectiondef-title rubric" id="breathe-section-title-public-members">Public Members</p>
<dl class="c var">
<dt class="sig sig-object c" id="c.vrna_pinfo_s.i">
<span class="target" id="structvrna__pinfo__s_1ab91db0a87ef8402dc151795ba5a64c6f"></span><span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">i</span></span></span><a class="headerlink" href="#c.vrna_pinfo_s.i" title="Permalink to this definition"></a><br /></dt>
<dd><p>nucleotide position i </p>
</dd></dl>

<dl class="c var">
<dt class="sig sig-object c" id="c.vrna_pinfo_s.j">
<span class="target" id="structvrna__pinfo__s_1a4142e38d6ba127acccdf680300a88e1f"></span><span class="kt"><span class="pre">unsigned</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">j</span></span></span><a class="headerlink" href="#c.vrna_pinfo_s.j" title="Permalink to this definition"></a><br /></dt>
<dd><p>nucleotide position j </p>
</dd></dl>

<dl class="c var">
<dt class="sig sig-object c" id="c.vrna_pinfo_s.p">
<span class="target" id="structvrna__pinfo__s_1a71990eae41fa100db8b557e4512ec281"></span><span class="kt"><span class="pre">float</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">p</span></span></span><a class="headerlink" href="#c.vrna_pinfo_s.p" title="Permalink to this definition"></a><br /></dt>
<dd><p>Probability. </p>
</dd></dl>

<dl class="c var">
<dt class="sig sig-object c" id="c.vrna_pinfo_s.ent">
<span class="target" id="structvrna__pinfo__s_1a5ab2a63f3da8e12c6d54d8e9dd9e3e73"></span><span class="kt"><span class="pre">float</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">ent</span></span></span><a class="headerlink" href="#c.vrna_pinfo_s.ent" title="Permalink to this definition"></a><br /></dt>
<dd><p>Pseudo entropy for <span class="math notranslate nohighlight">\( p(i,j) = S_i + S_j - p_ij*ln(p_ij) \)</span>. </p>
</dd></dl>

<dl class="c var">
<dt class="sig sig-object c" id="c.vrna_pinfo_s.bp">
<span class="target" id="structvrna__pinfo__s_1aa5feac5559b36dcd7cb38111c45d444d"></span><span class="kt"><span class="pre">short</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">bp</span></span></span><span class="p"><span class="pre">[</span></span><span class="m"><span class="pre">8</span></span><span class="p"><span class="pre">]</span></span><a class="headerlink" href="#c.vrna_pinfo_s.bp" title="Permalink to this definition"></a><br /></dt>
<dd><p>Frequencies of pair_types. </p>
</dd></dl>

<dl class="c var">
<dt class="sig sig-object c" id="c.vrna_pinfo_s.comp">
<span class="target" id="structvrna__pinfo__s_1a0720ae74ce53802a759ee9f98e9c8c43"></span><span class="kt"><span class="pre">char</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">comp</span></span></span><a class="headerlink" href="#c.vrna_pinfo_s.comp" title="Permalink to this definition"></a><br /></dt>
<dd><p>1 iff pair is in mfe structure </p>
</dd></dl>

</div>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="struct/deprecated.html" class="btn btn-neutral float-left" title="Deprecated Interface for Secondary Structure Utilities" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="msa_deprecated.html" class="btn btn-neutral float-right" title="Deprecated Interface for Multiple Sequence Alignment Utilities" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>