<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Multiple Sequence Alignments (MSA) &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/vrna_32.png"/>
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Command Files" href="commands.html" />
    <link rel="prev" title="RNA Structures" href="rna_structures.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manpages.html">Manpages</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">I/O Formats</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="rna_structures.html">RNA Structures</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Multiple Sequence Alignments (MSA)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#clustalw-format">ClustalW format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#stockholm-1-0-format">Stockholm 1.0 format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fasta-pearson-format">FASTA (Pearson) format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#maf-format">MAF format</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="commands.html">Command Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="parameters.html">Energy Parameters</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../io.html">I/O Formats</a> &raquo;</li>
      <li>Multiple Sequence Alignments (MSA)</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="multiple-sequence-alignments-msa">
<h1>Multiple Sequence Alignments (MSA)<a class="headerlink" href="#multiple-sequence-alignments-msa" title="Permalink to this headline"></a></h1>
<section id="clustalw-format">
<h2>ClustalW format<a class="headerlink" href="#clustalw-format" title="Permalink to this headline"></a></h2>
<p>The <em>ClustalW</em> format is a relatively simple text file containing a single
multiple sequence alignment of DNA, RNA, or protein sequences. It was first used
as an output format for the <em>clustalw</em> programs, but nowadays it may also be
generated by various other sequence alignment tools. The specification is straight
forward:</p>
<ul>
<li><p>The first line starts with the words:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CLUSTAL</span> <span class="n">W</span>
</pre></div>
</div>
<p>or:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CLUSTALW</span>
</pre></div>
</div>
</li>
<li><p>After the above header there is at least one empty line</p></li>
<li><p>Finally, one or more blocks of sequence data are following, where each
block is separated by at least one empty line.</p></li>
</ul>
<p>Each line in a blocks of sequence data consists of the sequence name followed by
the sequence symbols, separated by at least one whitespace character. Usually,
the length of a sequence in one block does not exceed 60 symbols. Optionally,
an additional whitespace separated cumulative residue count may follow the sequence
symbols. Optionally, a block may be followed by a line depicting the degree of
conservation of the respective alignment columns.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Sequence names and the sequences must not contain whitespace characters!
Allowed gap symbols are the hyphen (<code class="docutils literal notranslate"><span class="pre">-</span></code>), and dot (<code class="docutils literal notranslate"><span class="pre">.</span></code>).</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Please note that many programs that output this format tend to truncate
the sequence names to a limited number of characters, for instance the
first 15 characters. This can destroy the uniqueness of identifiers in
your MSA.</p>
</div>
<p>Here is an example alignment in ClustalW format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CLUSTAL</span> <span class="n">W</span> <span class="p">(</span><span class="mf">1.83</span><span class="p">)</span> <span class="n">multiple</span> <span class="n">sequence</span> <span class="n">alignment</span>


<span class="n">AL031296</span><span class="mf">.1</span><span class="o">/</span><span class="mi">85969</span><span class="o">-</span><span class="mi">86120</span>      <span class="n">CUGCCUCACAACGUUUGUGCCUCAGUUACCCGUAGAUGUAGUGAGGGUAACAAUACUUAC</span>
<span class="n">AANU01225121</span><span class="mf">.1</span><span class="o">/</span><span class="mi">438</span><span class="o">-</span><span class="mi">603</span>      <span class="n">CUGCCUCACAACAUUUGUGCCUCAGUUACUCAUAGAUGUAGUGAGGGUGACAAUACUUAC</span>
<span class="n">AAWR02037329</span><span class="mf">.1</span><span class="o">/</span><span class="mi">29294</span><span class="o">-</span><span class="mi">29150</span>  <span class="o">---</span><span class="n">CUCGACACCACU</span><span class="o">---</span><span class="n">GCCUCGGUUACCCAUCGGUGCAGUGCGGGUAGUAGUACCAAU</span>

<span class="n">AL031296</span><span class="mf">.1</span><span class="o">/</span><span class="mi">85969</span><span class="o">-</span><span class="mi">86120</span>      <span class="n">UCUCGUUGGUGAUAAGGAACAGCU</span>
<span class="n">AANU01225121</span><span class="mf">.1</span><span class="o">/</span><span class="mi">438</span><span class="o">-</span><span class="mi">603</span>      <span class="n">UCUCGUUGGUGAUAAGGAACAGCU</span>
<span class="n">AAWR02037329</span><span class="mf">.1</span><span class="o">/</span><span class="mi">29294</span><span class="o">-</span><span class="mi">29150</span>  <span class="n">GCUAAUUAGUUGUGAGGACCAACU</span> 
</pre></div>
</div>
</section>
<section id="stockholm-1-0-format">
<h2>Stockholm 1.0 format<a class="headerlink" href="#stockholm-1-0-format" title="Permalink to this headline"></a></h2>
<p>Here is an example alignment in Stockholm 1.0 format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># STOCKHOLM 1.0</span>

<span class="c1">#=GF AC   RF01293</span>
<span class="c1">#=GF ID   ACA59</span>
<span class="c1">#=GF DE   Small nucleolar RNA ACA59</span>
<span class="c1">#=GF AU   Wilkinson A</span>
<span class="c1">#=GF SE   Predicted; WAR; Wilkinson A</span>
<span class="c1">#=GF SS   Predicted; WAR; Wilkinson A</span>
<span class="c1">#=GF GA   43.00</span>
<span class="c1">#=GF TC   44.90</span>
<span class="c1">#=GF NC   40.30</span>
<span class="c1">#=GF TP   Gene; snRNA; snoRNA; HACA-box;</span>
<span class="c1">#=GF BM   cmbuild -F CM SEED</span>
<span class="c1">#=GF CB   cmcalibrate --mpi CM</span>
<span class="c1">#=GF SM   cmsearch --cpu 4 --verbose --nohmmonly -E 1000 -Z 549862.597050 CM SEQDB</span>
<span class="c1">#=GF DR   snoRNABase; ACA59;</span>
<span class="c1">#=GF DR   SO; 0001263; ncRNA_gene;</span>
<span class="c1">#=GF DR   GO; 0006396; RNA processing;</span>
<span class="c1">#=GF DR   GO; 0005730; nucleolus;</span>
<span class="c1">#=GF RN   [1]</span>
<span class="c1">#=GF RM   15199136</span>
<span class="c1">#=GF RT   Human box H/ACA pseudouridylation guide RNA machinery.</span>
<span class="c1">#=GF RA   Kiss AM, Jady BE, Bertrand E, Kiss T</span>
<span class="c1">#=GF RL   Mol Cell Biol. 2004;24:5797-5807.</span>
<span class="c1">#=GF WK   Small_nucleolar_RNA</span>
<span class="c1">#=GF SQ   3</span>


<span class="n">AL031296</span><span class="mf">.1</span><span class="o">/</span><span class="mi">85969</span><span class="o">-</span><span class="mi">86120</span>     <span class="n">CUGCCUCACAACGUUUGUGCCUCAGUUACCCGUAGAUGUAGUGAGGGUAACAAUACUUACUCUCGUUGGUGAUAAGGAACAGCU</span>
<span class="n">AANU01225121</span><span class="mf">.1</span><span class="o">/</span><span class="mi">438</span><span class="o">-</span><span class="mi">603</span>     <span class="n">CUGCCUCACAACAUUUGUGCCUCAGUUACUCAUAGAUGUAGUGAGGGUGACAAUACUUACUCUCGUUGGUGAUAAGGAACAGCU</span>
<span class="n">AAWR02037329</span><span class="mf">.1</span><span class="o">/</span><span class="mi">29294</span><span class="o">-</span><span class="mi">29150</span> <span class="o">---</span><span class="n">CUCGACACCACU</span><span class="o">---</span><span class="n">GCCUCGGUUACCCAUCGGUGCAGUGCGGGUAGUAGUACCAAUGCUAAUUAGUUGUGAGGACCAACU</span>
<span class="c1">#=GC SS_cons               -----((((,&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;___________&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;,,,,&lt;&lt;&lt;&lt;&lt;&lt;&lt;______&gt;&gt;&gt;&gt;&gt;&gt;&gt;,,,,,))))::::::::::::</span>
<span class="c1">#=GC RF                    CUGCcccaCAaCacuuguGCCUCaGUUACcCauagguGuAGUGaGgGuggcAaUACccaCcCucgUUgGuggUaAGGAaCAgCU</span>
<span class="o">//</span>
</pre></div>
</div>
<div class="admonition-see-also admonition">
<p class="admonition-title">See also…</p>
<p><a class="reference internal" href="rna_structures.html#wuss-notation"><span class="std std-ref">WUSS notation</span></a> for legal characters and
their interpretation in the consensus secondary structure line <code class="docutils literal notranslate"><span class="pre">SS_cons</span></code>.</p>
</div>
</section>
<section id="fasta-pearson-format">
<h2>FASTA (Pearson) format<a class="headerlink" href="#fasta-pearson-format" title="Permalink to this headline"></a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Sequence names must not contain whitespace characters. Otherwise, the parts after
the first whitespace will be dropped. The only allowed gap character is the hyphen
(<code class="docutils literal notranslate"><span class="pre">-</span></code>).</p>
</div>
<p>Here is an example alignment in FASTA format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">AL031296</span><span class="mf">.1</span><span class="o">/</span><span class="mi">85969</span><span class="o">-</span><span class="mi">86120</span>
<span class="n">CUGCCUCACAACGUUUGUGCCUCAGUUACCCGUAGAUGUAGUGAGGGUAACAAUACUUAC</span>
<span class="n">UCUCGUUGGUGAUAAGGAACAGCU</span>
<span class="o">&gt;</span><span class="n">AANU01225121</span><span class="mf">.1</span><span class="o">/</span><span class="mi">438</span><span class="o">-</span><span class="mi">603</span>
<span class="n">CUGCCUCACAACAUUUGUGCCUCAGUUACUCAUAGAUGUAGUGAGGGUGACAAUACUUAC</span>
<span class="n">UCUCGUUGGUGAUAAGGAACAGCU</span>
<span class="o">&gt;</span><span class="n">AAWR02037329</span><span class="mf">.1</span><span class="o">/</span><span class="mi">29294</span><span class="o">-</span><span class="mi">29150</span>
<span class="o">---</span><span class="n">CUCGACACCACU</span><span class="o">---</span><span class="n">GCCUCGGUUACCCAUCGGUGCAGUGCGGGUAGUAGUACCAAU</span>
<span class="n">GCUAAUUAGUUGUGAGGACCAACU</span>
</pre></div>
</div>
</section>
<section id="maf-format">
<h2>MAF format<a class="headerlink" href="#maf-format" title="Permalink to this headline"></a></h2>
<p>The multiple alignment format (MAF) is usually used to store multiple alignments on DNA level
between entire genomes. It consists of independent blocks of aligned sequences which are
annotated by their genomic location. Consequently, an MAF formatted MSA file may contain
multiple records. MAF files start with a line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">##maf</span>
</pre></div>
</div>
<p>which is optionally extended by whitespace delimited key=value pairs. Lines starting with
the character (<code class="docutils literal notranslate"><span class="pre">#</span></code>) are considered comments and usually ignored.</p>
<p>A MAF block starts with character (<code class="docutils literal notranslate"><span class="pre">a</span></code>) at the beginning of a line, optionally followed
by whitespace delimited <code class="docutils literal notranslate"><span class="pre">key=value</span></code> pairs. The next lines start with character (<code class="docutils literal notranslate"><span class="pre">s</span></code>) and
contain sequence information of the form:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">s</span> <span class="n">src</span> <span class="n">start</span> <span class="n">size</span> <span class="n">strand</span> <span class="n">srcSize</span> <span class="n">sequence</span>
</pre></div>
</div>
<p>where:</p>
<ul class="simple">
<li><p><em>src</em> is the name of the sequence source</p></li>
<li><p><em>start</em> is the start of the aligned region within the source (0-based)</p></li>
<li><p><em>size</em> is the length of the aligned region without gap characters</p></li>
<li><p><em>strand</em> is either (<code class="docutils literal notranslate"><span class="pre">+</span></code>) or (<code class="docutils literal notranslate"><span class="pre">-</span></code>), depicting the location of the aligned
region relative to the source</p></li>
<li><p><em>srcSize</em> is the size of the entire sequence source, e.g. the full chromosome</p></li>
<li><p><em>sequence</em> is the aligned sequence including gaps depicted by the hyphen (<code class="docutils literal notranslate"><span class="pre">-</span></code>)</p></li>
</ul>
<p>Here is an example alignment in MAF format (bluntly taken from the
<a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5">UCSC Genome browser website</a>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">##maf version=1 scoring=tba.v8 </span>
<span class="c1"># tba.v8 (((human chimp) baboon) (mouse rat)) </span>
<span class="c1"># multiz.v7</span>
<span class="c1"># maf_project.v5 _tba_right.maf3 mouse _tba_C</span>
<span class="c1"># single_cov2.v4 single_cov2 /dev/stdin</span>
                   
<span class="n">a</span> <span class="n">score</span><span class="o">=</span><span class="mf">23262.0</span>     
<span class="n">s</span> <span class="n">hg16</span><span class="o">.</span><span class="n">chr7</span>    <span class="mi">27578828</span> <span class="mi">38</span> <span class="o">+</span> <span class="mi">158545518</span> <span class="n">AAA</span><span class="o">-</span><span class="n">GGGAATGTTAACCAAATGA</span><span class="o">---</span><span class="n">ATTGTCTCTTACGGTG</span>
<span class="n">s</span> <span class="n">panTro1</span><span class="o">.</span><span class="n">chr6</span> <span class="mi">28741140</span> <span class="mi">38</span> <span class="o">+</span> <span class="mi">161576975</span> <span class="n">AAA</span><span class="o">-</span><span class="n">GGGAATGTTAACCAAATGA</span><span class="o">---</span><span class="n">ATTGTCTCTTACGGTG</span>
<span class="n">s</span> <span class="n">baboon</span>         <span class="mi">116834</span> <span class="mi">38</span> <span class="o">+</span>   <span class="mi">4622798</span> <span class="n">AAA</span><span class="o">-</span><span class="n">GGGAATGTTAACCAAATGA</span><span class="o">---</span><span class="n">GTTGTCTCTTATGGTG</span>
<span class="n">s</span> <span class="n">mm4</span><span class="o">.</span><span class="n">chr6</span>     <span class="mi">53215344</span> <span class="mi">38</span> <span class="o">+</span> <span class="mi">151104725</span> <span class="o">-</span><span class="n">AATGGGAATGTTAAGCAAACGA</span><span class="o">---</span><span class="n">ATTGTCTCTCAGTGTG</span>
<span class="n">s</span> <span class="n">rn3</span><span class="o">.</span><span class="n">chr4</span>     <span class="mi">81344243</span> <span class="mi">40</span> <span class="o">+</span> <span class="mi">187371129</span> <span class="o">-</span><span class="n">AA</span><span class="o">-</span><span class="n">GGGGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTG</span>
                   
<span class="n">a</span> <span class="n">score</span><span class="o">=</span><span class="mf">5062.0</span>                    
<span class="n">s</span> <span class="n">hg16</span><span class="o">.</span><span class="n">chr7</span>    <span class="mi">27699739</span> <span class="mi">6</span> <span class="o">+</span> <span class="mi">158545518</span> <span class="n">TAAAGA</span>
<span class="n">s</span> <span class="n">panTro1</span><span class="o">.</span><span class="n">chr6</span> <span class="mi">28862317</span> <span class="mi">6</span> <span class="o">+</span> <span class="mi">161576975</span> <span class="n">TAAAGA</span>
<span class="n">s</span> <span class="n">baboon</span>         <span class="mi">241163</span> <span class="mi">6</span> <span class="o">+</span>   <span class="mi">4622798</span> <span class="n">TAAAGA</span> 
<span class="n">s</span> <span class="n">mm4</span><span class="o">.</span><span class="n">chr6</span>     <span class="mi">53303881</span> <span class="mi">6</span> <span class="o">+</span> <span class="mi">151104725</span> <span class="n">TAAAGA</span>
<span class="n">s</span> <span class="n">rn3</span><span class="o">.</span><span class="n">chr4</span>     <span class="mi">81444246</span> <span class="mi">6</span> <span class="o">+</span> <span class="mi">187371129</span> <span class="n">taagga</span>

<span class="n">a</span> <span class="n">score</span><span class="o">=</span><span class="mf">6636.0</span>
<span class="n">s</span> <span class="n">hg16</span><span class="o">.</span><span class="n">chr7</span>    <span class="mi">27707221</span> <span class="mi">13</span> <span class="o">+</span> <span class="mi">158545518</span> <span class="n">gcagctgaaaaca</span>
<span class="n">s</span> <span class="n">panTro1</span><span class="o">.</span><span class="n">chr6</span> <span class="mi">28869787</span> <span class="mi">13</span> <span class="o">+</span> <span class="mi">161576975</span> <span class="n">gcagctgaaaaca</span>
<span class="n">s</span> <span class="n">baboon</span>         <span class="mi">249182</span> <span class="mi">13</span> <span class="o">+</span>   <span class="mi">4622798</span> <span class="n">gcagctgaaaaca</span>
<span class="n">s</span> <span class="n">mm4</span><span class="o">.</span><span class="n">chr6</span>     <span class="mi">53310102</span> <span class="mi">13</span> <span class="o">+</span> <span class="mi">151104725</span> <span class="n">ACAGCTGAAAATA</span>

</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="rna_structures.html" class="btn btn-neutral float-left" title="RNA Structures" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="commands.html" class="btn btn-neutral float-right" title="Command Files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>