<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>RNA Structures &mdash; ViennaRNA Package</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/vrna_32.png"/>
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Multiple Sequence Alignments (MSA)" href="msa.html" />
    <link rel="prev" title="I/O Formats" href="../io.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> ViennaRNA
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
    
              <p class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../configuration.html">Configuration</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Usage and Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../manpages.html">Manpages</a></li>
<li class="toctree-l1"><a class="reference internal" href="../using_rnalib.html">Using RNAlib</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">I/O Formats</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">RNA Structures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#dot-bracket-notation">Dot-Bracket Notation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pseudo-dot-bracket-notation">Pseudo Dot-Bracket Notation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#extended-dot-bracket-notation">Extended Dot-Bracket Notation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#wuss-notation">WUSS notation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#abstract-shapes">Abstract Shapes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tree-representations">Tree Representations</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="msa.html">Multiple Sequence Alignments (MSA)</a></li>
<li class="toctree-l2"><a class="reference internal" href="commands.html">Command Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="parameters.html">Energy Parameters</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">RNAlib API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Concepts and Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../wrappers.html">SWIG Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api_python.html">Python API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../news.html">What's new</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
</ul>


    
        <p class="caption">
            <span class="caption-text">
            
                Links
            
            </span>
        </p>
        <ul>
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA"><i class="fa fa-github fa-fw"></i> Source code</a></li>
                
            
                
                    <li class="toctree-l1"><a href="https://github.com/ViennaRNA/ViennaRNA/issues"><i class="fa fa-bug fa-fw"></i> Issue tracker</a></li>
                
            
        </ul>
    


        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ViennaRNA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../io.html">I/O Formats</a> &raquo;</li>
      <li>RNA Structures</li>

  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rna-structures">
<h1>RNA Structures<a class="headerlink" href="#rna-structures" title="Permalink to this headline"></a></h1>
<p>Here, we describe the different notations and representations of RNA secondary structures
used throughout our library and prediction tools.</p>
<div class="toctree-wrapper compound">
</div>
<section id="dot-bracket-notation">
<h2>Dot-Bracket Notation<a class="headerlink" href="#dot-bracket-notation" title="Permalink to this headline"></a></h2>
<p>The Dot-Bracket notation as introduced already in the early times of the ViennaRNA Package
denotes base pairs by matching pairs of parenthesis <code class="docutils literal notranslate"><span class="pre">()</span></code> and unpaired nucleotides by dots
<code class="docutils literal notranslate"><span class="pre">.</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is the standard representation of a secondary structure in our library.</p>
</div>
<p>Based on that notation, more elaborate representations have been developed to
include additional information, such as the loop context a nucleotide belongs
to and to annotated pseudo-knots.</p>
<p>Consider the following secondary structure in dot-bracket notation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(((</span><span class="o">..</span><span class="p">((((</span><span class="o">...</span><span class="p">))))</span><span class="o">.</span><span class="p">)))</span>
</pre></div>
</div>
<p>which, drawn as a secondary structure graph, looks like:</p>
<img alt="secondary structure layout iamge" class="align-center" src="../_images/bracket.png" />
<p>It is a stem-loop structure consisting of a an outer helix of 3 base pairs followed by
an interior loop of size 3, a second helix of length 4, and a hairpin loop of size 3.</p>
<section id="pseudo-dot-bracket-notation">
<h3>Pseudo Dot-Bracket Notation<a class="headerlink" href="#pseudo-dot-bracket-notation" title="Permalink to this headline"></a></h3>
<p>Base pair probabilities are sometimes summarized in <em>pseudo dot-bracket notation</em> with
the additional symbols <code class="docutils literal notranslate"><span class="pre">,</span></code>, <code class="docutils literal notranslate"><span class="pre">|</span></code>, <code class="docutils literal notranslate"><span class="pre">{</span></code>, <code class="docutils literal notranslate"><span class="pre">}</span></code>. Here, the usual <code class="docutils literal notranslate"><span class="pre">(</span></code>, <code class="docutils literal notranslate"><span class="pre">)</span></code>, <code class="docutils literal notranslate"><span class="pre">.</span></code>,
represent bases with a strong preference (more than 2/3) to pair upstream (with a partner
further 3’), pair down-stream, or do not pair, respectively. <code class="docutils literal notranslate"><span class="pre">{</span></code>, <code class="docutils literal notranslate"><span class="pre">}</span></code>, and <code class="docutils literal notranslate"><span class="pre">,</span></code> are
just the weaker version of the above and <code class="docutils literal notranslate"><span class="pre">|</span></code> represents a base that is mostly paired
but has pairing partners both upstream and downstream. In this case opening and closing
brackets do not need to match.</p>
</section>
<section id="extended-dot-bracket-notation">
<h3>Extended Dot-Bracket Notation<a class="headerlink" href="#extended-dot-bracket-notation" title="Permalink to this headline"></a></h3>
<p>A more generalized version of the original Dot-Bracket notation may use additional pairs
of brackets, such as <code class="docutils literal notranslate"><span class="pre">&lt;&gt;</span></code>, <code class="docutils literal notranslate"><span class="pre">{}</span></code>, and <code class="docutils literal notranslate"><span class="pre">[]</span></code>, and matching pairs of
uppercase/lowercase letters. This allows for anotating pseudo-knots, since different
pairs of brackets are not required to be nested.</p>
<p>The follwing annotations of a simple structure with two crossing helices of size 4 are equivalent:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;&lt;&lt;&lt;</span><span class="p">[[[[</span><span class="o">....&gt;&gt;&gt;&gt;</span><span class="p">]]]]</span>
<span class="p">((((</span><span class="n">AAAA</span><span class="o">....</span><span class="p">))))</span><span class="n">aaaa</span>
<span class="n">AAAA</span><span class="p">{{{{</span><span class="o">....</span><span class="n">aaaa</span><span class="p">}}}}</span>
</pre></div>
</div>
<div class="admonition-see-also admonition">
<p class="admonition-title">See also…</p>
<p><a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_pack" title="vrna_db_pack"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_pack()</span></code></a>, <a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_unpack" title="vrna_db_unpack"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_unpack()</span></code></a>, <a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_flatten" title="vrna_db_flatten"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_flatten()</span></code></a>, <a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_flatten_to" title="vrna_db_flatten_to"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_flatten_to()</span></code></a>,
<a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_from_ptable" title="vrna_db_from_ptable"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_from_ptable()</span></code></a>, <a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_from_plist" title="vrna_db_from_plist"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_from_plist()</span></code></a>, <a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_to_element_string" title="vrna_db_to_element_string"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_to_element_string()</span></code></a>,
<a class="reference internal" href="../utils/struct/dotbracket.html#c.vrna_db_pk_remove" title="vrna_db_pk_remove"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_pk_remove()</span></code></a></p>
</div>
</section>
</section>
<section id="wuss-notation">
<h2>WUSS notation<a class="headerlink" href="#wuss-notation" title="Permalink to this headline"></a></h2>
<p>The Washington University Secondary Structure (WUSS) notation is frequently used for
consensus secondary structures, e.g. in <a class="reference internal" href="msa.html#stockholm-1-0-format"><span class="std std-ref">Stockholm 1.0 format</span></a></p>
<p>This notation allows for a fine-grained annotation of base pairs and unpaired nucleotides,
including pseudo-knots.</p>
<div class="admonition-see-also admonition">
<p class="admonition-title">See also…</p>
<p>WUSS notation in the infernal user guide at <a class="reference external" href="http://eddylab.org/infernal/Userguide.pdf">http://eddylab.org/infernal/Userguide.pdf</a></p>
</div>
<p>Below, you’ll find a list of secondary structure elements and their corresponding WUSS annotation.</p>
<ul>
<li><p><strong>Base pairs</strong></p>
<p>Nested base pairs are annotated by matching pairs of the symbols <code class="docutils literal notranslate"><span class="pre">&lt;&gt;</span></code>, <code class="docutils literal notranslate"><span class="pre">()</span></code>, <code class="docutils literal notranslate"><span class="pre">{}</span></code>,
and <code class="docutils literal notranslate"><span class="pre">[]</span></code>. Each of the matching pairs  of parenthesis have their special meaning, however,
when used as input in our programs, e.g. structure constraint, these details are usually
ignored. Furthermore, base pairs that constitute as pseudo-knot are denoted by letters from
the latin alphabet and are, if not denoted otherwise, ignored entirely in our programs.</p>
</li>
<li><p><strong>Hairpin loops</strong></p>
<p>Unpaired nucleotides that constitute the hairpin loop are indicated by underscores, <code class="docutils literal notranslate"><span class="pre">_</span></code>.
Here is an example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;&lt;&lt;&lt;&lt;</span><span class="n">_____</span><span class="o">&gt;&gt;&gt;&gt;&gt;</span>
</pre></div>
</div>
</li>
<li><p><strong>Bulges and interior loops</strong></p>
<p>Residues that constitute a bulge or interior loop are denoted by dashes, <code class="docutils literal notranslate"><span class="pre">-</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(((</span><span class="o">--&lt;&lt;</span><span class="n">_____</span><span class="o">&gt;&gt;-</span><span class="p">)))</span>
</pre></div>
</div>
</li>
<li><p><strong>Multibranch loops</strong></p>
<p>Unpaired nucleotides in multibranch loops are indicated by commas <code class="docutils literal notranslate"><span class="pre">,</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(((,,</span><span class="o">&lt;&lt;</span><span class="n">_____</span><span class="o">&gt;&gt;</span><span class="p">,</span><span class="o">&lt;&lt;</span><span class="n">____</span><span class="o">&gt;&gt;</span><span class="p">)))</span>
</pre></div>
</div>
</li>
<li><p><strong>External residues</strong></p>
<p>Single stranded nucleotides in the exterior loop, i.e. not enclosed by any other pair are
denoted by colons, <code class="docutils literal notranslate"><span class="pre">:</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;&lt;&lt;</span><span class="n">____</span><span class="o">&gt;&gt;&gt;</span><span class="p">:::</span>
</pre></div>
</div>
</li>
<li><p><strong>Insertions</strong></p>
<p>In cases where an alignment represents the consensus with a known structure, insertions
relative to the known structure are denoted by periods, <code class="docutils literal notranslate"><span class="pre">.</span></code>. Regions where local
structural alignment was invoked, leaving regions of both target and query sequence
unaligned, are indicated by tildes, <code class="docutils literal notranslate"><span class="pre">~</span></code>.</p>
<p>These symbols only appear in alignments of a known (query) structure annotation to a target
sequence of unknown structure.</p>
</li>
<li><p><strong>Pseudo-knots</strong></p>
<p>The WUSS notation allows for annotation of pseudo-knots using pairs of upper-case/lower-case
letters. Our programs and library functions usually ignore pseudo-knots entirely treating
them as unpaired nucleotides, if not stated otherwise:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;&lt;&lt;</span><span class="n">_AAA___</span><span class="o">&gt;&gt;&gt;</span><span class="n">aaa</span>
</pre></div>
</div>
</li>
</ul>
<div class="admonition-see-also admonition">
<p class="admonition-title">See also…</p>
<p><a class="reference internal" href="../utils/struct/wuss.html#c.vrna_db_from_WUSS" title="vrna_db_from_WUSS"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_from_WUSS()</span></code></a></p>
</div>
</section>
<section id="abstract-shapes">
<h2>Abstract Shapes<a class="headerlink" href="#abstract-shapes" title="Permalink to this headline"></a></h2>
<p>Abstract Shapes, introduced by <span id="id1">Giegerich <em>et al.</em> [<a class="reference internal" href="../bibliography.html#id42" title="Robert Giegerich, Björn Voß, and Marc Rehmsmeier. Abstract shapes of RNA. Nucleic Acids Research, 32(16):4843–4851, 2004. doi:10.1093/nar/gkh779.">2004</a>]</span>, collapse the secondary structure while
retaining the nestedness of helices and hairpin loops.</p>
<p>The abstract shapes representation abstracts the structure from individual base pairs
and their corresponding location in the sequence, while retaining the inherent nestedness
of helices and hairpin loops.</p>
<p>Below is a description of what is included in the abstract shapes abstraction for each
respective level together with an example structure:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CGUCUUAAACUCAUCACCGUGUGGAGCUGCGACCCUUCCCUAGAUUCGAAGACGAG</span>
<span class="p">((((((</span><span class="o">...</span><span class="p">(((</span><span class="o">..</span><span class="p">(((</span><span class="o">...</span><span class="p">))))))</span><span class="o">...</span><span class="p">(((</span><span class="o">..</span><span class="p">((</span><span class="o">.....</span><span class="p">))</span><span class="o">..</span><span class="p">)))))))))</span><span class="o">..</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 71%" />
<col style="width: 18%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Shape Level</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Result</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>Most accurate - all loops and all unpaired</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">[_[_[]]_[_[]_]]_</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>Nesting pattern for all loop types and unpaired regions in external loop and multiloop</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">[[_[]][_[]_]]</span></code></p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>Nesting pattern for all loop types but no unpaired regions</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">[[[]][[]]]</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>Helix nesting pattern in external loop and multiloop</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">[[][[]]]</span></code></p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>Most abstract - helix nesting pattern and no unpaired regions</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">[[][]]</span></code></p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Our implementations also provide the special Shape Level 0, which does not
collapse any structural features but simply convert base pairs and unpaired
nucleotides into their corresponding set of symbols for abstract shapes.</p>
</div>
<div class="admonition-see-also admonition">
<p class="admonition-title">See also…</p>
<p><a class="reference internal" href="../utils/struct/abstract_shapes.html#c.vrna_abstract_shapes" title="vrna_abstract_shapes"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_abstract_shapes()</span></code></a>, <a class="reference internal" href="../utils/struct/abstract_shapes.html#c.vrna_abstract_shapes_pt" title="vrna_abstract_shapes_pt"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_abstract_shapes_pt()</span></code></a></p>
</div>
</section>
<section id="tree-representations">
<h2>Tree Representations<a class="headerlink" href="#tree-representations" title="Permalink to this headline"></a></h2>
<p>Secondary structures can be readily represented as trees, where internal
nodes represent base pairs, and leaves represent unpaired nucleotides.
The dot-bracket structure string already is a tree represented by a string
of parenthesis (base pairs) and dots for the leaf nodes (unpaired nucleotides).</p>
<p>Alternatively, one may find representations with two types of node labels,
<code class="docutils literal notranslate"><span class="pre">P</span></code> for paired and <code class="docutils literal notranslate"><span class="pre">U</span></code> for unpaired; a dot is then replaced by <code class="docutils literal notranslate"><span class="pre">(U)</span></code>, and
each closed bracket is assigned an additional identifier <code class="docutils literal notranslate"><span class="pre">P</span></code>.
We call this the expanded notation. In <span id="id2">Fontana <em>et al.</em> [<a class="reference internal" href="../bibliography.html#id25" title="W. Fontana, P.F. Stadler, E.G. Bornberg-Bauer, T. Griesmacher, I.L. Hofacker, M. Tacker, P. Tarazona, E.D. Weinberger, and P. Schuster. RNA folding and combinatory landscapes. Physical review E, 47(3):2083, 1993. doi:10.1103/PhysRevE.47.2083.">1993</a>]</span> a condensed
representation of the secondary structure is proposed, the so-called
homeomorphically irreducible tree (HIT) representation. Here a stack is
represented as a single pair of matching brackets labeled <code class="docutils literal notranslate"><span class="pre">P</span></code> and
weighted by the number of base pairs.  Correspondingly, a contiguous
strain of unpaired bases is shown as one pair of matching brackets
labeled <code class="docutils literal notranslate"><span class="pre">U</span></code> and weighted by its length.  Generally any string consisting
of matching brackets and identifiers is equivalent to a plane tree with
as many different types of nodes as there are identifiers.</p>
<p><span id="id3">Shapiro [<a class="reference internal" href="../bibliography.html#id27" title="B.A. Shapiro. An algorithm for comparing multiple RNA secondary structures. Computer applications in the biosciences: CABIOS, 4(3):387–393, 1988. doi:10.1093/bioinformatics/4.3.387.">1988</a>]</span> proposed a coarse grained representation which
does not retain the full information of the secondary structure. He
represents the different structure elements by single matching brackets
and labels them as</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">H</span></code>  (hairpin loop),</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">I</span></code>  (interior loop),</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">B</span></code>  (bulge),</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">M</span></code>  (multi-loop), and</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">S</span></code>  (stack).</p></li>
</ul>
<p>We extend his alphabet by an extra letter for external elements <code class="docutils literal notranslate"><span class="pre">E</span></code>.
Again these identifiers may be followed by a weight corresponding to the
number of unpaired bases or base pairs in the structure element.  All tree
representations (except for the dot-bracket form) can be encapsulated into
a virtual root (labeled <code class="docutils literal notranslate"><span class="pre">R</span></code>).</p>
<p>The following example illustrates the different linear tree representations
used by the package:</p>
<p>Consider the secondary structure represented by the dot-bracket string (full tree):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">.</span><span class="p">((</span><span class="o">..</span><span class="p">(((</span><span class="o">...</span><span class="p">)))</span><span class="o">..</span><span class="p">((</span><span class="o">..</span><span class="p">))))</span><span class="o">.</span>
</pre></div>
</div>
<p>which is the most convenient condensed notation used by our programs and
library functions.</p>
<p>Then, the following tree representations are equivalent:</p>
<ul>
<li><p>Expanded tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">((</span><span class="n">U</span><span class="p">)(((</span><span class="n">U</span><span class="p">)(</span><span class="n">U</span><span class="p">)((((</span><span class="n">U</span><span class="p">)(</span><span class="n">U</span><span class="p">)(</span><span class="n">U</span><span class="p">)</span><span class="n">P</span><span class="p">)</span><span class="n">P</span><span class="p">)</span><span class="n">P</span><span class="p">)(</span><span class="n">U</span><span class="p">)(</span><span class="n">U</span><span class="p">)(((</span><span class="n">U</span><span class="p">)(</span><span class="n">U</span><span class="p">)</span><span class="n">P</span><span class="p">)</span><span class="n">P</span><span class="p">)</span><span class="n">P</span><span class="p">)</span><span class="n">P</span><span class="p">)(</span><span class="n">U</span><span class="p">)</span><span class="n">R</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>HIT representation (<span id="id4">[<a class="reference internal" href="../bibliography.html#id25" title="W. Fontana, P.F. Stadler, E.G. Bornberg-Bauer, T. Griesmacher, I.L. Hofacker, M. Tacker, P. Tarazona, E.D. Weinberger, and P. Schuster. RNA folding and combinatory landscapes. Physical review E, 47(3):2083, 1993. doi:10.1103/PhysRevE.47.2083.">Fontana <em>et al.</em>, 1993</a>]</span>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">((</span><span class="n">U1</span><span class="p">)((</span><span class="n">U2</span><span class="p">)((</span><span class="n">U3</span><span class="p">)</span><span class="n">P3</span><span class="p">)(</span><span class="n">U2</span><span class="p">)((</span><span class="n">U2</span><span class="p">)</span><span class="n">P2</span><span class="p">)</span><span class="n">P2</span><span class="p">)(</span><span class="n">U1</span><span class="p">)</span><span class="n">R</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>Coarse Grained Tree Representation (<span id="id5">[<a class="reference internal" href="../bibliography.html#id27" title="B.A. Shapiro. An algorithm for comparing multiple RNA secondary structures. Computer applications in the biosciences: CABIOS, 4(3):387–393, 1988. doi:10.1093/bioinformatics/4.3.387.">Shapiro, 1988</a>]</span>):</p>
<ul>
<li><p>Short (with root node <code class="docutils literal notranslate"><span class="pre">R</span></code>, without stem nodes <code class="docutils literal notranslate"><span class="pre">S</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">((</span><span class="n">H</span><span class="p">)((</span><span class="n">H</span><span class="p">)</span><span class="n">M</span><span class="p">)</span><span class="n">R</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>Full (with root node <cite>R`</cite>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(((((</span><span class="n">H</span><span class="p">)</span><span class="n">S</span><span class="p">)((</span><span class="n">H</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">M</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">R</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>Extended (with root node <code class="docutils literal notranslate"><span class="pre">R</span></code>, with external nodes <code class="docutils literal notranslate"><span class="pre">E</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">((((((</span><span class="n">H</span><span class="p">)</span><span class="n">S</span><span class="p">)((</span><span class="n">H</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">M</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">E</span><span class="p">)</span><span class="n">R</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>Weighted (with root node <code class="docutils literal notranslate"><span class="pre">R</span></code>, with external nodes <code class="docutils literal notranslate"><span class="pre">E</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">((((((</span><span class="n">H3</span><span class="p">)</span><span class="n">S3</span><span class="p">)((</span><span class="n">H2</span><span class="p">)</span><span class="n">S2</span><span class="p">)</span><span class="n">M4</span><span class="p">)</span><span class="n">S2</span><span class="p">)</span><span class="n">E2</span><span class="p">)</span><span class="n">R</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
</li>
</ul>
<p>The Expanded tree is rather clumsy and mostly included for the sake of
completeness. The different versions of Coarse Grained Tree Representations
are variatios of Shapiro’s linear tree notation.</p>
<p>For the output of aligned structures from string editing, different
representations are needed, where we put the label on both sides.
The above examples for tree representations would then look like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">a</span><span class="p">)</span> <span class="p">(</span><span class="n">UU</span><span class="p">)(</span><span class="n">P</span><span class="p">(</span><span class="n">P</span><span class="p">(</span><span class="n">P</span><span class="p">(</span><span class="n">P</span><span class="p">(</span><span class="n">UU</span><span class="p">)(</span><span class="n">UU</span><span class="p">)(</span><span class="n">P</span><span class="p">(</span><span class="n">P</span><span class="p">(</span><span class="n">P</span><span class="p">(</span><span class="n">UU</span><span class="p">)(</span><span class="n">UU</span><span class="p">)(</span><span class="n">UU</span><span class="p">)</span><span class="n">P</span><span class="p">)</span><span class="n">P</span><span class="p">)</span><span class="n">P</span><span class="p">)(</span><span class="n">UU</span><span class="p">)(</span><span class="n">UU</span><span class="p">)(</span><span class="n">P</span><span class="p">(</span><span class="n">P</span><span class="p">(</span><span class="n">UU</span><span class="p">)(</span><span class="n">U</span><span class="o">...</span>
<span class="n">b</span><span class="p">)</span> <span class="p">(</span><span class="n">UU</span><span class="p">)(</span><span class="n">P2</span><span class="p">(</span><span class="n">P2</span><span class="p">(</span><span class="n">U2U2</span><span class="p">)(</span><span class="n">P2</span><span class="p">(</span><span class="n">U3U3</span><span class="p">)</span><span class="n">P3</span><span class="p">)(</span><span class="n">U2U2</span><span class="p">)(</span><span class="n">P2</span><span class="p">(</span><span class="n">U2U2</span><span class="p">)</span><span class="n">P2</span><span class="p">)</span><span class="n">P2</span><span class="p">)(</span><span class="n">UU</span><span class="p">)</span><span class="n">P2</span><span class="p">)(</span><span class="n">UU</span><span class="p">)</span>
<span class="n">c</span><span class="p">)</span> <span class="p">(</span><span class="n">B</span><span class="p">(</span><span class="n">M</span><span class="p">(</span><span class="n">HH</span><span class="p">)(</span><span class="n">HH</span><span class="p">)</span><span class="n">M</span><span class="p">)</span><span class="n">B</span><span class="p">)</span>
   <span class="p">(</span><span class="n">S</span><span class="p">(</span><span class="n">B</span><span class="p">(</span><span class="n">S</span><span class="p">(</span><span class="n">M</span><span class="p">(</span><span class="n">S</span><span class="p">(</span><span class="n">HH</span><span class="p">)</span><span class="n">S</span><span class="p">)(</span><span class="n">S</span><span class="p">(</span><span class="n">HH</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">M</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">B</span><span class="p">)</span><span class="n">S</span><span class="p">)</span>
   <span class="p">(</span><span class="n">E</span><span class="p">(</span><span class="n">S</span><span class="p">(</span><span class="n">B</span><span class="p">(</span><span class="n">S</span><span class="p">(</span><span class="n">M</span><span class="p">(</span><span class="n">S</span><span class="p">(</span><span class="n">HH</span><span class="p">)</span><span class="n">S</span><span class="p">)(</span><span class="n">S</span><span class="p">(</span><span class="n">HH</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">M</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">B</span><span class="p">)</span><span class="n">S</span><span class="p">)</span><span class="n">E</span><span class="p">)</span>
<span class="n">d</span><span class="p">)</span> <span class="p">(</span><span class="n">R</span><span class="p">(</span><span class="n">E2</span><span class="p">(</span><span class="n">S2</span><span class="p">(</span><span class="n">B1</span><span class="p">(</span><span class="n">S2</span><span class="p">(</span><span class="n">M4</span><span class="p">(</span><span class="n">S3</span><span class="p">(</span><span class="n">H3</span><span class="p">)</span><span class="n">S3</span><span class="p">)((</span><span class="n">H2</span><span class="p">)</span><span class="n">S2</span><span class="p">)</span><span class="n">M4</span><span class="p">)</span><span class="n">S2</span><span class="p">)</span><span class="n">B1</span><span class="p">)</span><span class="n">S2</span><span class="p">)</span><span class="n">E2</span><span class="p">)</span><span class="n">R</span><span class="p">)</span>
</pre></div>
</div>
<p>Aligned structures additionally contain the gap character <code class="docutils literal notranslate"><span class="pre">_</span></code>.</p>
<div class="admonition-see-also admonition">
<p class="admonition-title">See also…</p>
<p><a class="reference internal" href="../utils/struct/trees.html#c.vrna_db_to_tree_string" title="vrna_db_to_tree_string"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_db_to_tree_string()</span></code></a>, <a class="reference internal" href="../utils/struct/trees.html#c.vrna_tree_string_unweight" title="vrna_tree_string_unweight"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_tree_string_unweight()</span></code></a>,
<a class="reference internal" href="../utils/struct/trees.html#c.vrna_tree_string_to_db" title="vrna_tree_string_to_db"><code class="xref c c-func docutils literal notranslate"><span class="pre">vrna_tree_string_to_db()</span></code></a></p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../io.html" class="btn btn-neutral float-left" title="I/O Formats" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="msa.html" class="btn btn-neutral float-right" title="Multiple Sequence Alignments (MSA)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1994 - 2023, Ronny Lorenz, Ivo L. Hofacker, et al..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>